<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>BCR • immunarch</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><!-- docsearch --><script src="../../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><link href="../../extra.css" rel="stylesheet">
<meta property="og:title" content="BCR">
<meta property="og:description" content="immunarch">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-128418614-2"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-128418614-2');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">immunarch</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.7.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Start Here
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/v1_introduction.html">Installation &amp; troubleshooting</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">First steps</li>
    <li>
      <a href="../../articles/v2_data.html">Data loading</a>
    </li>
    <li>
      <a href="../../articles/web_only/load_mixcr.html">How-to: Loading MiXCR Data</a>
    </li>
    <li>
      <a href="../../articles/web_only/load_10x.html">How-to: Loading 10x Genomics Data</a>
    </li>
    <li>
      <a href="../../articles/web_only/v21_singlecell.html">How-to: Single-cell and paired chain data</a>
    </li>
    <li>
      <a href="../../articles/web_only/v3_basic_analysis.html">Basic statistics and clonality</a>
    </li>
    <li>
      <a href="../../articles/web_only/repFilter_v3.html">Data filtering</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Repertoire-level exploration and comparison</li>
    <li>
      <a href="../../articles/web_only/v4_overlap.html">Repertoire overlap and public clonotypes</a>
    </li>
    <li>
      <a href="../../articles/web_only/v5_gene_usage.html">Gene usage</a>
    </li>
    <li>
      <a href="../../articles/web_only/v6_diversity.html">Diversity estimation</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Clonotype-level exploration and comparison</li>
    <li>
      <a href="../../articles/web_only/v8_tracking.html">Track clonotypes across samples and time</a>
    </li>
    <li>
      <a href="../../articles/web_only/v11_db.html">Annotate clonotypes using immune receptor databases</a>
    </li>
    <li>
      <a href="../../articles/web_only/v9_kmers.html">Kmer and sequence motif analysis and visualisation</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Preparing to publication</li>
    <li>
      <a href="../../articles/web_only/v7_fixvis.html">Make your plots publication-ready with fixVis</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">BCR analysis</li>
    <li>
      <a href="../../articles/web_only/BCRpipeline.html">BCR pipeline</a>
    </li>
    <li>
      <a href="../../articles/web_only/clustering.html">Clustering</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
<li>
  <a href="https://github.com/immunomind/covid19" class="external-link">COVID-19</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://immunomind.io" class="external-link">ImmunoMind</a>
</li>
<li>
  <a href="http://twitter.com/immunomind" class="external-link">
    <span class="fa fa-lg fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/immunomind/immunarch" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>BCR</h1>
                        <h4 data-toc-skip class="author">
<b>ImmunoMind</b> – improving design of T-cell therapies using multi-omics and AI. Research and biopharma partnerships, more details: <a href="https://immunomind.io" class="external-link">immunomind.io</a>
</h4>
            
            <h4 data-toc-skip class="date"><a href="mailto:support@immunomind.io" class="email">support@immunomind.io</a></h4>
      
      <small class="dont-index">Source: <a href="https://github.com/immunomind/immunarch/blob/HEAD/vignettes/web_only/BCRpipeline.Rmd" class="external-link"><code>vignettes/web_only/BCRpipeline.Rmd</code></a></small>
      <div class="hidden name"><code>BCRpipeline.Rmd</code></div>

    </div>

    
    
<!--
%\VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{Gene usage analysis}
%\VignettePackage{immunarch}
-->
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>A <strong>B-cell receptor (BCR)</strong> consists of an immunoglobulin molecule and a СD79 molecule. BCR includes two identical heavy chains (generated by recombination of V, D, and J segments), and two identical light chains (generated by recombination of V and J segments). Rapid improvements in high-throughput sequencing provide an opportunity to study B-cell immunoglobulin receptors on the cell surface [1] via the process called <strong>BCR repertoire sequencing</strong>. This pipeline describes how to work with BCR repertoire data after preprocessing raw sequenced data with MiXCR (<a href="https://mixcr.readthedocs.io/en/master/" class="external-link">https://mixcr.readthedocs.io/en/master/</a>) or other analogical programs.</p>
<p>The pipeline involves five steps:</p>
<ol style="list-style-type: decimal">
<li>
<p><strong>Data loading.</strong></p>
<p>The step includes loading the data and checking whether there is enough information for the pipeline to work.</p>
</li>
<li>
<p><strong>Reconstructing clonal lineages.</strong></p>
<p>At this step, all BCR sequences are divided into clonal lineages — sets of sequences that presumably share a common ancestor.</p>
</li>
<li>
<p><strong>Building germline.</strong></p>
<p>At this step, the algorithm generates a sequence that represents the ancestral sequence for each BCR in a clonal lineage.</p>
</li>
<li>
<p><strong>Aligning sequences within clonal lineages.</strong></p>
<p>This step involves preparation for phylogenetic and somatic hypermutation analysis.</p>
</li>
<li>
<p><strong>Phylogenetic analysis.</strong></p>
<p>This final step provides phylogeny reconstruction and trunk length calculation (by running the PHYLIP package).</p>
</li>
</ol>
</div>
<div class="section level2">
<h2 id="data-loading">Data loading<a class="anchor" aria-label="anchor" href="#data-loading"></a>
</h2>
<p>BCR data are loaded with <code>repLoad</code> functions, like any other data.</p>
<p>Check that BCR data has the following:</p>
<ul>
<li>information about full sequence read of each BCR,</li>
<li>positions of CDR3 region start and end,</li>
<li>number of mutations per CDR3 region. In MiXCR output files, this information is located in the column ‘<code>refPoints</code>’</li>
</ul>
<p>Look at the sample BCR data implemented in Immunarch:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#load the package into the R environment:</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://immunarch.com/">immunarch</a></span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">bcrdata</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="reconstructing-clonal-lineages">Reconstructing clonal lineages<a class="anchor" aria-label="anchor" href="#reconstructing-clonal-lineages"></a>
</h2>
<p>B-cell <strong>clonal lineage</strong> represents a set of B cells that presumably have a common origin (arising from the same VDJ rearrangement event) and a common ancestor. BCR sequences are clustered according to their similarity, and sets of sequences in one cluster are named clonal lineages. Clustering involves two steps.</p>
<ul>
<li>The first step is calculating the distance between sequences.</li>
<li>The second step is clustering the sequences using the information about the distance between them. Check our tutorial for more information about clustering: <a href="https://www.notion.so/Clustering-1752b661f8794fb6aba19f25e9373488" class="external-link">https://www.notion.so/immunomind/Clustering-1752b661f8794fb6aba19f25e9373488</a>
</li>
</ul>
<p>An example of reconstructing clonal lineages using default Immunarch options:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#calulate distance matrix</span>
<span class="va">distBCR</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/seqDist.html">seqDist</a></span><span class="op">(</span><span class="va">bcrdata</span><span class="op">$</span><span class="va">data</span>  <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="../../reference/top.html">top</a></span><span class="op">(</span><span class="fl">500</span><span class="op">)</span><span class="op">)</span>

<span class="co">#find clusters</span>
<span class="va">bcrdata</span><span class="op">$</span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/seqCluster.html">seqCluster</a></span><span class="op">(</span><span class="va">bcrdata</span><span class="op">$</span><span class="va">data</span>  <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="../../reference/top.html">top</a></span><span class="op">(</span><span class="fl">500</span><span class="op">)</span>, <span class="va">distBCR</span>, .perc_similarity <span class="op">=</span> <span class="fl">0.9</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="building-germline">Building germline<a class="anchor" aria-label="anchor" href="#building-germline"></a>
</h2>
<p>Each clonal lineage has its own <strong>germline sequence</strong> that represents the ancestral sequence for each BCR in the clonal lineage. In other words, germline sequence is a sequence of B-cells immediately after VDJ recombination, before B-cell maturation and hypermutation process. Germline sequence is useful for assessing the degree of mutation and maturity of the repertoire.</p>
<p>In Immunarch, <code><a href="../../reference/repGermline.html">repGermline()</a></code> function generates germline for each sequence:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#generate germline</span>
<span class="va">bcrdata</span><span class="op">$</span><span class="va">data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
    <span class="fu"><a href="../../reference/repGermline.html">repGermline</a></span><span class="op">(</span>.threads <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<p>A germline is represented via sequences of V gene - N…N (CDR3 length) - J gene:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#germline example</span>
<span class="va">bcrdata</span><span class="op">$</span><span class="va">data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
     <span class="fu"><a href="../../reference/top.html">top</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
     <span class="fu"><a href="../../reference/repGermline.html">repGermline</a></span><span class="op">(</span>.threads <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">.</span><span class="op">$</span><span class="va">full_clones</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">.</span><span class="op">$</span><span class="va">Germline.sequence</span></code></pre></div>
<pre><code><span class="co">## [1] "CAGCTGCAGCTGCAGGAGTCGGGCCCAGGACTGGTGAAGCCTTCGGAGACCCTGTCCCTCACCTGCACTGTCTCTGGTGGCTCCATCAGCAGTAGTAGTTACTACTGGGGCTGGATCCGCCAGCCCCCAGGGAAGGGGCTGGAGTGGATTGGGAGTATCTATTATAGTGGGAGCACCTACTACAACCCGTCCCTCAAGAGTCGAGTCACCATATCCGTAGACACGTCCAAGAACCAGTTCTCCCTGAAGCTGAGCTCTGTGACCGCCGCAGACACGGCTGTGTATTACNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNGGCCAAGGAACCCTGGTCACCGTCTCCTCAG"</span></code></pre>
<p>Оptions:</p>
<p><code>.data</code> - The data to be processed. Can be <code>data.frame</code>, <code>data.table</code>, or a list of these objects.</p>
<p><code>.species</code> - Specifies species from which reference V and J are taken. Available species: “HomoSapiens” (default), “MusMusculus”, “BosTaurus”, “CamelusDromedarius”, “CanisLupusFamiliaris”, “DanioRerio”, “MacacaMulatta”, “MusMusculusDomesticus”, “MusMusculusCastaneus”, “MusMusculusMolossinus”, “MusMusculusMusculus”, “MusSpretus”, “OncorhynchusMykiss”, “OrnithorhynchusAnatinus”, “OryctolagusCuniculus”, “RattusNorvegicus”, “SusScrofa”.</p>
<p><code>.min_nuc_outside_cdr3</code> — filters out reads with low numbers of nucleotides from V or J gene outside of CDR3 region</p>
<p><code>.align_j_gene</code> - if the germline sequence does not assemble correctly in the region of the J gene, then set this parameter to True. This will slow down the algorithm, but the assembly of the germline sequence will be more accurate.</p>
<p><code>.ref_only_first</code></p>
<ul>
<li>if <code>TRUE</code>, only the first sequence from reference for each allele name will be used in the analysis;</li>
<li>if <code>FALSE</code>, all allele variants will be analyzed, and the output table will increase in size as a result.</li>
</ul>
<p>If <code>.ref_only_first = FALSE</code>, you will get several different germlines:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">bcrdata</span><span class="op">$</span><span class="va">data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
    <span class="fu"><a href="../../reference/top.html">top</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
    <span class="fu"><a href="../../reference/repGermline.html">repGermline</a></span><span class="op">(</span>.ref_only_first <span class="op">=</span> <span class="cn">F</span>, .threads <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">.</span><span class="op">$</span><span class="va">full_clones</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">.</span><span class="op">$</span><span class="va">Germline.sequence</span></code></pre></div>
<pre><code><span class="co">## [1] "CAGCTGCAGCTGCAGGAGTCGGGCCCAGGACTGGTGAAGCCTTCGGAGACCCTGTCCCTCACCTGCACTGTCTCTGGTGGCTCCATCAGCAGTAGTAGTTACTACTGGGGCTGGATCCGCCAGCCCCCAGGGAAGGGGCTGGAGTGGATTGGGAGTATCTATTATAGTGGGAGCACCTACTACAACCCGTCCCTCAAGAGTCGAGTCACCATATCCGTAGACACGTCCAAGAACCAGTTCTCCCTGAAGCTGAGCTCTGTGACCGCCGCAGACACGGCTGTGTATTACNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNGGCCAAGGAACCCTGGTCACCGTCTCCTCAG"</span></code></pre>
</div>
<div class="section level2">
<h2 id="aligning-sequences-within-a-clonal-lineage">Aligning sequences within a clonal lineage<a class="anchor" aria-label="anchor" href="#aligning-sequences-within-a-clonal-lineage"></a>
</h2>
<p>After building clonal lineage and germline, we can start analyzing the degree of mutation and maturity of each clonal lineage. This allows us to find cells with a large number of mutated clones. The phylogenetic analysis will find mutations that influence the affinity of BCRs. Aligning the sequence is the first step toward sequence phylogenetic analysis.</p>
<p>In the Immunarch package, the function <code>repAlignLineage</code> aligns sequences within clonal lineages. This function requires <code>Clustal W</code> app to be installed. The app could be downloaded here: <a href="http://www.clustal.org/download/current/%5C%5Cn" class="external-link">http://www.clustal.org/download/current/</a>, or installed via your system package manager (such as apt or dnf).</p>
<ul>
<li>For Ubuntu, check this guide: <a href="https://www.howtoinstall.me/ubuntu/18-04/clustalw/" class="external-link">https://www.howtoinstall.me/ubuntu/18-04/clustalw/</a>
</li>
</ul>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>sudo apt update</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>sudo apt install clustalw</span></code></pre></div>
<ul>
<li>For Windows, download and run <code>clustalw-x.x-win.msi</code>
</li>
</ul>
<p><code>repAlignLineage</code> usage example:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">bcrdata</span><span class="op">)</span>
<span class="va">bcr_data</span> <span class="op">&lt;-</span> <span class="va">bcrdata</span><span class="op">$</span><span class="va">data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="../../reference/top.html">top</a></span><span class="op">(</span><span class="fl">500</span><span class="op">)</span>
<span class="va">bcr_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="../../reference/seqCluster.html">seqCluster</a></span><span class="op">(</span><span class="fu"><a href="../../reference/seqDist.html">seqDist</a></span><span class="op">(</span><span class="va">bcr_data</span><span class="op">)</span>, .fixed_threshold <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="../../reference/repGermline.html">repGermline</a></span><span class="op">(</span>.threads <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="../../reference/repAlignLineage.html">repAlignLineage</a></span><span class="op">(</span>.min_lineage_sequences <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></code></pre></div>
<p>The function has several parameters:</p>
<ul>
<li>
<code>.min_lineage_sequences</code> — Filters clusters (clonal lineages) with the number of clonotypes lower than the threshold. Aligning clonal lineages with few sequences is of little use.</li>
</ul>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># take clusters that contain at least 1 sequence</span>
<span class="va">align_dt</span> <span class="op">&lt;-</span> <span class="va">bcr_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="../../reference/seqCluster.html">seqCluster</a></span><span class="op">(</span><span class="fu"><a href="../../reference/seqDist.html">seqDist</a></span><span class="op">(</span><span class="va">bcr_data</span><span class="op">)</span>, .fixed_threshold <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="../../reference/repGermline.html">repGermline</a></span><span class="op">(</span>.threads <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="../../reference/repAlignLineage.html">repAlignLineage</a></span><span class="op">(</span>.min_lineage_sequences <span class="op">=</span> <span class="fl">2</span>, .align_threads <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></code></pre></div>
<ul>
<li>
<code>.prepare_threads</code> — the number of threads to prepare results table. A high number can cause memory overload!</li>
<li>
<code>.align_threads</code> — the number of threads for lineage alignment.</li>
</ul>
<p>Requirements for the input table for <code><a href="../../reference/repAlignLineage.html">repAlignLineage()</a></code></p>
<ul>
<li>must have columns in the immunarch compatible format <code>immunarch_data_format</code>,</li>
<li>must contain the ‘Cluster’ column generated by seqCluster() function,</li>
<li>must contain the ‘Sequence.germline’ column generated by repGermline() function.</li>
</ul>
<p>Align sequences in a cluster can be visualized using standard functions:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># A name of the first cluster</span>
<span class="va">align_dt</span><span class="op">$</span><span class="va">full_clones</span><span class="op">$</span><span class="va">Cluster</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<pre><code><span class="co">## [1] "IGHV4-61/IGHJ4_length_45_cluster_1"</span></code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Alignment of sequences from the first cluster</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/image.html" class="external-link">image</a></span><span class="op">(</span><span class="va">align_dt</span><span class="op">$</span><span class="va">full_clones</span><span class="op">$</span><span class="va">Alignment</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, grid <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></code></pre></div>
<p><img src="BCRpipeline_files/figure-html/align%20visualisation-1.png" width="1152" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="phylogenetic-analysis">Phylogenetic analysis<a class="anchor" aria-label="anchor" href="#phylogenetic-analysis"></a>
</h2>
<p>For BCR phylogenetic analysis, we use the <strong>maximum parsimony method</strong>. The maximum parsimony method reconstructs a phylogenetic tree for the lineage by minimizing the total number of mutation events [2].</p>
<p>This method also enables the reconstruction of intermediate sequences that may have existed between BCR and germline sequence. These simulated sequences contain mutations that were ancestral to the clonotype group selected with the maximum parsimony analysis.</p>
<p>In Immunarch, function <code>repClonalFamily</code> builds a phylogeny of B-cells, generates a common ancestor, and calculates a trunk length. The mean trunk length represents the distance between the most recent common ancestor and germline sequence [3]. Mean trunk length serves as a measure of the maturity of a lineage. The trunk length of the lineage tree approximates the maturation state of the initiating B cell for each clone [4].</p>
<p>This function requires <code>PHYLIP</code> app to be installed. The app can be downloaded here: <a href="https://evolution.genetics.washington.edu/phylip/getme-new1.html" class="external-link">https://evolution.genetics.washington.edu/phylip/getme-new1.html</a>, or could be installed with your system package manager (such as apt or dnf).</p>
<ul>
<li>For Ubuntu, follow the installation guide (<a href="https://zoomadmin.com/HowToInstall/UbuntuPackage/phylip" class="external-link">https://zoomadmin.com/HowToInstall/UbuntuPackage/phylip</a> ):</li>
</ul>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>sudo apt<span class="sc">-</span>get update <span class="sc">-</span>y</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>sudo apt<span class="sc">-</span>get install <span class="sc">-</span>y phylip</span></code></pre></div>
<ul>
<li>For Windows:
<ol style="list-style-type: decimal">
<li>Download a Zip archive (<a href="https://evolution.genetics.washington.edu/phylip/install.html" class="external-link">https://evolution.genetics.washington.edu/phylip/install.html</a>).</li>
<li>Extract the archive into a folder.</li>
<li>Add the folder to the PATH (<a href="https://www.architectryan.com/2018/03/17/add-to-the-path-on-windows-10/" class="external-link">https://www.architectryan.com/2018/03/17/add-to-the-path-on-windows-10/</a>)</li>
</ol>
</li>
</ul>
<p>repClonalFamily usage example:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://ape-package.ird.fr/" class="external-link">ape</a></span><span class="op">)</span>
<span class="va">bcr</span> <span class="op">&lt;-</span> <span class="va">bcr_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="../../reference/seqCluster.html">seqCluster</a></span><span class="op">(</span><span class="fu"><a href="../../reference/seqDist.html">seqDist</a></span><span class="op">(</span><span class="va">bcr_data</span><span class="op">)</span>, .fixed_threshold <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="../../reference/repGermline.html">repGermline</a></span><span class="op">(</span>.threads <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="../../reference/repAlignLineage.html">repAlignLineage</a></span><span class="op">(</span>.min_lineage_sequences <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="../../reference/repClonalFamily.html">repClonalFamily</a></span><span class="op">(</span>.threads <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>

<span class="co">#plot visualization of the first tree</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">bcr</span><span class="op">$</span><span class="va">full_clones</span><span class="op">$</span><span class="va">Tree</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/ape/man/nodelabels.html" class="external-link">tiplabels</a></span><span class="op">(</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/ape/man/nodelabels.html" class="external-link">nodelabels</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="BCRpipeline_files/figure-html/example%2010-1.png" width="1152" style="display: block; margin: auto;"></p>
<p>We have found 4 clusters:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">bcr</span><span class="op">$</span><span class="va">full_clones</span><span class="op">$</span><span class="va">Cluster</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/duplicated.html" class="external-link">unique</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "IGHV4-61/IGHJ4_length_45_cluster_1" "IGHV4-61/IGHJ4_length_63_cluster_1"</span>
<span class="co">## [3] "IGHV5-51/IGHJ4_length_60_cluster_1" "IGHV5-51/IGHJ5_length_45_cluster_1"</span></code></pre>
<p>We have found mismatches between a germline and an ancestor sequence. Dots represent nucleotides matches between the sequences, letters represent mismatches between the sequences:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">bcr</span><span class="op">$</span><span class="va">full_clones</span><span class="op">$</span><span class="va">Germline.Output</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></code></pre></div>
<pre><code><span class="co">## [1] "....................................T....................................................A....................................C...G................................A.NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN.....A........................."</span></code></pre>
<p>We have calculated a trunk length for each cluster:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">bcr</span><span class="op">$</span><span class="va">full_clones</span><span class="op">$</span><span class="va">Trunk.Length</span></code></pre></div>
<pre><code><span class="co">## [1] 51 70 66 51</span></code></pre>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<ol style="list-style-type: decimal">
<li><a href="https://genomemedicine.biomedcentral.com/articles/10.1186/s13073-015-0243-2" class="external-link">https://genomemedicine.biomedcentral.com/articles/10.1186/s13073-015-0243-2</a></li>
<li><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4754972/" class="external-link">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4754972/</a></li>
<li><a href="https://www.frontiersin.org/articles/10.3389/fimmu.2020.01734/full" class="external-link">https://www.frontiersin.org/articles/10.3389/fimmu.2020.01734/full</a></li>
<li><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4754972/" class="external-link">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4754972/</a></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://www.linkedin.com/in/vdnaz" class="external-link">Vadim I. Nazarov</a>, <a href="https://www.linkedin.com/in/vasily-tsvetkov-227218ab" class="external-link">Vasily O. Tsvetkov</a>, <a href="https://www.linkedin.com/in/erumynskiy/" class="external-link">Eugene Rumynskiy</a>, <a href="https://www.linkedin.com/in/aleksandr-popov-634878105/" class="external-link">Aleksandr A. Popov</a>, <a href="https://www.linkedin.com/in/ivan-balashov/" class="external-link">Ivan Balashov</a>, <a href="https://www.linkedin.com/in/maria-volobueva-b0856a223/" class="external-link">Maria Samokhina</a>, <a href="https://immunomind.io" class="external-link">ImmunoMind</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '94207562f34b455c0796790d104e3549',
    indexName: 'immunarch',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
