<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>BCR • immunarch</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><!-- docsearch --><script src="../../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><link href="../../extra.css" rel="stylesheet">
<meta property="og:title" content="BCR">
<meta property="og:description" content="immunarch">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-128418614-2"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-128418614-2');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">immunarch</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Start Here
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/v1_introduction.html">Installation &amp; troubleshooting</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">First steps</li>
    <li>
      <a href="../../articles/v2_data.html">Data loading</a>
    </li>
    <li>
      <a href="../../articles/web_only/load_mixcr.html">How-to: Loading MiXCR Data</a>
    </li>
    <li>
      <a href="../../articles/web_only/load_10x.html">How-to: Loading 10x Genomics Data</a>
    </li>
    <li>
      <a href="../../articles/web_only/v21_singlecell.html">How-to: Single-cell and paired chain data</a>
    </li>
    <li>
      <a href="../../articles/web_only/v3_basic_analysis.html">Basic statistics and clonality</a>
    </li>
    <li>
      <a href="../../articles/web_only/repFilter_v3.html">Data filtering</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Repertoire-level exploration and comparison</li>
    <li>
      <a href="../../articles/web_only/v4_overlap.html">Repertoire overlap and public clonotypes</a>
    </li>
    <li>
      <a href="../../articles/web_only/v5_gene_usage.html">Gene usage</a>
    </li>
    <li>
      <a href="../../articles/web_only/v6_diversity.html">Diversity estimation</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Clonotype-level exploration and comparison</li>
    <li>
      <a href="../../articles/web_only/v8_tracking.html">Track clonotypes across samples and time</a>
    </li>
    <li>
      <a href="../../articles/web_only/v11_db.html">Annotate clonotypes using immune receptor databases</a>
    </li>
    <li>
      <a href="../../articles/web_only/v9_kmers.html">Kmer and sequence motif analysis and visualisation</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Preparing to publication</li>
    <li>
      <a href="../../articles/web_only/v7_fixvis.html">Make your plots publication-ready with fixVis</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">BCR analysis</li>
    <li>
      <a href="../../articles/web_only/BCRpipeline.html">BCR pipeline</a>
    </li>
    <li>
      <a href="../../articles/web_only/clustering.html">Clustering</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
<li>
  <a href="https://github.com/immunomind/covid19" class="external-link">COVID-19</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://immunomind.io" class="external-link">ImmunoMind</a>
</li>
<li>
  <a href="http://twitter.com/immunomind" class="external-link">
    <span class="fa fa-lg fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/immunomind/immunarch" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="BCRpipeline_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>BCR</h1>
                        <h4 data-toc-skip class="author">
<b>ImmunoMind</b> – improving design of T-cell therapies using multi-omics and AI. Research and biopharma partnerships, more details: <a href="https://immunomind.io" class="external-link">immunomind.io</a>
</h4>
            
            <h4 data-toc-skip class="date"><a href="mailto:support@immunomind.io" class="email">support@immunomind.io</a></h4>
      
      <small class="dont-index">Source: <a href="https://github.com/immunomind/immunarch/blob/HEAD/vignettes/web_only/BCRpipeline.Rmd" class="external-link"><code>vignettes/web_only/BCRpipeline.Rmd</code></a></small>
      <div class="hidden name"><code>BCRpipeline.Rmd</code></div>

    </div>

    
    
<!--
%\VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{Gene usage analysis}
%\VignettePackage{immunarch}
-->
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>A <strong>B-cell receptor (BCR)</strong> consists of an immunoglobulin molecule and a СD79 molecule. BCR includes two identical heavy chains (generated by recombination of V, D, and J segments), and two identical light chains (generated by recombination of V and J segments). Rapid improvements in high-throughput sequencing provide an opportunity to study B-cell immunoglobulin receptors on the cell surface [1] via the process called <strong>BCR repertoire sequencing</strong>. This pipeline describes how to work with BCR repertoire data after preprocessing raw sequenced data with MiXCR (<a href="https://docs.milaboratories.com/mixcr/about/" class="external-link">https://docs.milaboratories.com/mixcr/about/</a>) or any other programs.</p>
<p>The pipeline involves five steps:</p>
<ol style="list-style-type: decimal">
<li>
<p><strong>Data loading.</strong></p>
<p>The step includes loading the data and checking whether there is enough information for the pipeline to work.</p>
</li>
<li>
<p><strong>Reconstructing clonal lineages.</strong></p>
<p>At this step, all BCR sequences are divided into clonal lineages — sets of sequences that presumably share a common ancestor.</p>
</li>
<li>
<p><strong>Building germline.</strong></p>
<p>At this step, the algorithm generates a sequence that represents the ancestral sequence for each BCR in a clonal lineage.</p>
</li>
<li>
<p><strong>Aligning sequences within clonal lineages.</strong></p>
<p>This step involves preparation for phylogenetic and somatic hypermutation analysis.</p>
</li>
<li>
<p><strong>Phylogenetic analysis.</strong></p>
<p>This step provides phylogeny reconstruction and trunk length calculation (by running the PHYLIP package).</p>
</li>
<li>
<p><strong>Somatic hypermutation analysis.</strong></p>
<p>At this step, we compare clonotype and germline sequences to detect and count the number of mutations in clonotype sequence.</p>
</li>
</ol>
</div>
<div class="section level2">
<h2 id="data-loading">Data loading<a class="anchor" aria-label="anchor" href="#data-loading"></a>
</h2>
<p>BCR data are loaded with <code>repLoad</code> functions, like any other data.</p>
<p>Check that BCR data has the following:</p>
<ul>
<li>information about full sequence read of each BCR,</li>
<li>positions of CDR3 region start and end,</li>
<li>number of mutations per CDR3 region. In MiXCR output files, this information is located in the column ‘<code>refPoints</code>’</li>
</ul>
<p>Look at the sample BCR data implemented in Immunarch:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#load the package into the R environment:</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://immunarch.com/">immunarch</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">bcrdata</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="reconstructing-clonal-lineages">Reconstructing clonal lineages<a class="anchor" aria-label="anchor" href="#reconstructing-clonal-lineages"></a>
</h2>
<p>B-cell <strong>clonal lineage</strong> represents a set of B cells that presumably have a common origin (arising from the same VDJ rearrangement event) and a common ancestor. BCR sequences are clustered according to their similarity, and sets of sequences in one cluster are named clonal lineages. Clustering involves two steps.</p>
<ul>
<li>The first step is calculating the distance between sequences.</li>
<li>The second step is clustering the sequences using the information about the distance between them. Check our tutorial for more information about clustering: <a href="https://immunarch.com/articles/web_only/clustering.html">https://immunarch.com/articles/web_only/clustering.html</a>
</li>
</ul>
<p>An example of reconstructing clonal lineages using default Immunarch options:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#calulate distance matrix</span></span>
<span><span class="va">distBCR</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/seqDist.html">seqDist</a></span><span class="op">(</span><span class="va">bcrdata</span><span class="op">$</span><span class="va">data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="../../reference/top.html">top</a></span><span class="op">(</span><span class="fl">500</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#find clusters</span></span>
<span><span class="va">bcrdata</span><span class="op">$</span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/seqCluster.html">seqCluster</a></span><span class="op">(</span><span class="va">bcrdata</span><span class="op">$</span><span class="va">data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="../../reference/top.html">top</a></span><span class="op">(</span><span class="fl">500</span><span class="op">)</span>, <span class="va">distBCR</span>, .perc_similarity <span class="op">=</span> <span class="fl">0.9</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: There was 1 warning in `mutate()`.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> In argument: `length_value = map_chr(.y, ~ifelse(all(.x == .x[1]), yes =</span></span>
<span><span class="co">##   .x[1], no = glue("range_{min(.x)}:{max(.x)}")))`.</span></span>
<span><span class="co">## Caused by warning:</span></span>
<span><span class="co">## <span style="color: #BBBB00;">!</span> Automatic coercion from integer to character was deprecated in purrr 1.0.0.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Please use an explicit call to `as.character()` within `map_chr()` instead.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> The deprecated feature was likely used in the <span style="color: #0000BB;">immunarch</span> package.</span></span>
<span><span class="co">##   Please report the issue at <span style="color: #0000BB; font-style: italic;">&lt;https://github.com/immunomind/immunarch/issues&gt;</span>.</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="building-germline">Building germline<a class="anchor" aria-label="anchor" href="#building-germline"></a>
</h2>
<p>Each clonal lineage has its own <strong>germline sequence</strong> that represents the ancestral sequence for each BCR in the clonal lineage. In other words, germline sequence is a sequence of B-cells immediately after VDJ recombination, before B-cell maturation and hypermutation process. Germline sequence is useful for assessing the degree of mutation and maturity of the repertoire.</p>
<p>In Immunarch, <code><a href="../../reference/repGermline.html">repGermline()</a></code> function generates germline for each sequence:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#generate germline</span></span>
<span><span class="va">bcrdata</span><span class="op">$</span><span class="va">data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../../reference/repGermline.html">repGermline</a></span><span class="op">(</span>.threads <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p>A germline is represented via sequences of V gene - N…N (CDR3 length) - J gene:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#germline example</span></span>
<span><span class="va">bcrdata</span><span class="op">$</span><span class="va">data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>     <span class="fu"><a href="../../reference/top.html">top</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>     <span class="fu"><a href="../../reference/repGermline.html">repGermline</a></span><span class="op">(</span>.threads <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">.</span><span class="op">$</span><span class="va">full_clones</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">.</span><span class="op">$</span><span class="va">Germline.sequence</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "CAGCTGCAGCTGCAGGAGTCGGGCCCAGGACTGGTGAAGCCTTCGGAGACCCTGTCCCTCACCTGCACTGTCTCTGGTGGCTCCATCAGCAGTAGTAGTTACTACTGGGGCTGGATCCGCCAGCCCCCAGGGAAGGGGCTGGAGTGGATTGGGAGTATCTATTATAGTGGGAGCACCTACTACAACCCGTCCCTCAAGAGTCGAGTCACCATATCCGTAGACACGTCCAAGAACCAGTTCTCCCTGAAGCTGAGCTCTGTGACCGCCGCAGACACGGCTGTGTATTACNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNGGCCAAGGAACCCTGGTCACCGTCTCCTCAG"</span></span></code></pre>
<p>Оptions:</p>
<p><code>.data</code> - The data to be processed. Can be <code>data.frame</code>, <code>data.table</code>, or a list of these objects.</p>
<p><code>.species</code> - Specifies species from which reference V and J are taken. Available species: “HomoSapiens” (default), “MusMusculus”, “BosTaurus”, “CamelusDromedarius”, “CanisLupusFamiliaris”, “DanioRerio”, “MacacaMulatta”, “MusMusculusDomesticus”, “MusMusculusCastaneus”, “MusMusculusMolossinus”, “MusMusculusMusculus”, “MusSpretus”, “OncorhynchusMykiss”, “OrnithorhynchusAnatinus”, “OryctolagusCuniculus”, “RattusNorvegicus”, “SusScrofa”.</p>
<p><code>.min_nuc_outside_cdr3</code> - This parameter sets how many nucleotides should have V or J chain outside of CDR3 to be considered good for further alignment. Reads with too short chains are filtered out.</p>
<p><code>.threads</code> - The number of threads to use.</p>
</div>
<div class="section level2">
<h2 id="aligning-sequences-within-a-clonal-lineage">Aligning sequences within a clonal lineage<a class="anchor" aria-label="anchor" href="#aligning-sequences-within-a-clonal-lineage"></a>
</h2>
<p>After building clonal lineage and germline, we can start analyzing the degree of mutation and maturity of each clonal lineage. This allows us to find cells with a large number of mutated clones. The phylogenetic analysis will find mutations that influence the affinity of BCRs. Aligning the sequence is the first step toward sequence phylogenetic analysis.</p>
<p>In the Immunarch package, the function <code>repAlignLineage</code> aligns sequences within clonal lineages. This function requires <code>Clustal W</code> app to be installed. The app could be downloaded here: <a href="http://www.clustal.org/download/current/" class="external-link">http://www.clustal.org/download/current/</a>, or installed via your system package manager (such as apt or dnf).</p>
<ul>
<li>For Ubuntu, check this guide: <a href="https://www.howtoinstall.me/ubuntu/18-04/clustalw/" class="external-link">https://www.howtoinstall.me/ubuntu/18-04/clustalw/</a>
</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true"></a>sudo apt update</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true"></a>sudo apt install clustalw</span></code></pre></div>
<ul>
<li>For Windows, download and run <code>clustalw-x.x-win.msi</code>
</li>
</ul>
<p><code>repAlignLineage</code> usage example:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">bcrdata</span><span class="op">)</span></span>
<span><span class="va">bcr_data</span> <span class="op">&lt;-</span> <span class="va">bcrdata</span><span class="op">$</span><span class="va">data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="../../reference/top.html">top</a></span><span class="op">(</span><span class="fl">500</span><span class="op">)</span></span>
<span><span class="va">bcr_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/seqCluster.html">seqCluster</a></span><span class="op">(</span><span class="fu"><a href="../../reference/seqDist.html">seqDist</a></span><span class="op">(</span><span class="va">bcr_data</span><span class="op">)</span>, .fixed_threshold <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/repGermline.html">repGermline</a></span><span class="op">(</span>.threads <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/repAlignLineage.html">repAlignLineage</a></span><span class="op">(</span>.min_lineage_sequences <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p>The function has several parameters:</p>
<ul>
<li>
<code>.min_lineage_sequences</code> — Filters clusters (clonal lineages) with the number of clonotypes lower than the threshold. Aligning clonal lineages with few sequences is of little use.</li>
</ul>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># take clusters that contain at least 1 sequence</span></span>
<span><span class="va">bcr_data</span> <span class="op">&lt;-</span> <span class="va">bcrdata</span><span class="op">$</span><span class="va">data</span></span>
<span><span class="va">align_dt</span> <span class="op">&lt;-</span> <span class="va">bcr_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/seqCluster.html">seqCluster</a></span><span class="op">(</span><span class="fu"><a href="../../reference/seqDist.html">seqDist</a></span><span class="op">(</span><span class="va">bcr_data</span>, .col <span class="op">=</span> <span class="st">'CDR3.nt'</span>, .group_by_seqLength <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>             .perc_similarity <span class="op">=</span> <span class="fl">0.6</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/repGermline.html">repGermline</a></span><span class="op">(</span>.threads <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/repAlignLineage.html">repAlignLineage</a></span><span class="op">(</span>.min_lineage_sequences <span class="op">=</span> <span class="fl">6</span>, .align_threads <span class="op">=</span> <span class="fl">2</span>, .nofail <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning in merge_reference_sequences(., reference, "V", species, sample_name): Genes or alleles IGHV4-55 from sample full_clones not found in the reference and will be dropped!</span></span>
<span><span class="co">## Probably, species argument is wrong (current value: HomoSapiens) or the data contains non-BCR genes.</span></span></code></pre>
<ul>
<li>
<code>.prepare_threads</code> — the number of threads to prepare results table. A high number can cause memory overload!</li>
<li>
<code>.align_threads</code> — the number of threads for lineage alignment.</li>
</ul>
<p>Requirements for the input table for <code><a href="../../reference/repAlignLineage.html">repAlignLineage()</a></code></p>
<ul>
<li>must have columns in the immunarch compatible format <code>immunarch_data_format</code>,</li>
<li>must contain the ‘Cluster’ column generated by seqCluster() function,</li>
<li>must contain the ‘Sequence.germline’ column generated by repGermline() function.</li>
</ul>
<p>Align sequences in a cluster can be visualized using standard functions:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># A name of the first cluster</span></span>
<span><span class="va">align_dt</span><span class="op">$</span><span class="va">full_clones</span><span class="op">$</span><span class="va">Cluster</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "IGHV1-69/IGHJ6_length_63_cluster_1"</span></span></code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Alignment of sequences from the first cluster</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/image.html" class="external-link">image</a></span><span class="op">(</span><span class="va">align_dt</span><span class="op">$</span><span class="va">full_clones</span><span class="op">$</span><span class="va">Alignment</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, grid <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="BCRpipeline_files/figure-html/align%20visualisation-1.png" width="1152" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="phylogenetic-analysis">Phylogenetic analysis<a class="anchor" aria-label="anchor" href="#phylogenetic-analysis"></a>
</h2>
<p>For BCR phylogenetic analysis, we use the <strong>maximum parsimony method</strong>. The maximum parsimony method reconstructs a phylogenetic tree for the lineage by minimizing the total number of mutation events [2].</p>
<p>This method also enables the reconstruction of intermediate sequences that may have existed between BCR and germline sequence. These simulated sequences contain mutations that were ancestral to the clonotype group selected with the maximum parsimony analysis.</p>
<p>In Immunarch, function <code>repClonalFamily</code> builds a phylogeny of B-cells, generates a common ancestor, and calculates a trunk length. The mean trunk length represents the distance between the most recent common ancestor and germline sequence [3]. Mean trunk length serves as a measure of the maturity of a lineage. The trunk length of the lineage tree approximates the maturation state of the initiating B cell for each clone [4].</p>
<p>This function requires <code>PHYLIP</code> app to be installed. The app can be downloaded here: <a href="https://evolution.genetics.washington.edu/phylip/getme-new1.html" class="external-link">https://evolution.genetics.washington.edu/phylip/getme-new1.html</a>, or could be installed with your system package manager (such as apt or dnf).</p>
<ul>
<li>For Ubuntu, follow the installation guide (<a href="https://zoomadmin.com/HowToInstall/UbuntuPackage/phylip" class="external-link">https://zoomadmin.com/HowToInstall/UbuntuPackage/phylip</a> ):</li>
</ul>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true"></a>sudo apt<span class="op">-</span>get update <span class="op">-</span>y</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true"></a>sudo apt<span class="op">-</span>get install <span class="op">-</span>y phylip</span></code></pre></div>
<ul>
<li>For Windows:
<ol style="list-style-type: decimal">
<li>Download a Zip archive (<a href="https://evolution.genetics.washington.edu/phylip/install.html" class="external-link">https://evolution.genetics.washington.edu/phylip/install.html</a>).</li>
<li>Extract the archive into a folder.</li>
<li>Add the folder to the PATH (<a href="https://www.architectryan.com/2018/03/17/add-to-the-path-on-windows-10/" class="external-link">https://www.architectryan.com/2018/03/17/add-to-the-path-on-windows-10/</a>)</li>
</ol>
</li>
</ul>
<p>repClonalFamily usage example:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bcr</span> <span class="op">&lt;-</span> <span class="va">align_dt</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/repClonalFamily.html">repClonalFamily</a></span><span class="op">(</span>.threads <span class="op">=</span> <span class="fl">2</span>, .nofail <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#plot visualization of the first tree</span></span>
<span><span class="fu"><a href="../../reference/vis.html">vis</a></span><span class="op">(</span><span class="va">bcr</span><span class="op">[[</span><span class="st">"full_clones"</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"TreeStats"</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="BCRpipeline_files/figure-html/example%2010-1.png" width="1152" style="display: block; margin: auto;"> For each cluster tree is represented as table (The default number of clones for CommonAncestor, Germline, Presumable is 1):</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#example for the first tree</span></span>
<span><span class="va">bcr</span><span class="op">[[</span><span class="st">"full_clones"</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"TreeStats"</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p>You can recolor leaves. For example, we recolor leaves where number of AA mutations is not 0:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#take sequence where number of AA mutations is not 0</span></span>
<span><span class="va">f</span> <span class="op">&lt;-</span> <span class="va">bcr</span><span class="op">[[</span><span class="st">"full_clones"</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"TreeStats"</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#rename these leaves</span></span>
<span><span class="va">f</span><span class="op">[</span><span class="va">f</span><span class="op">$</span><span class="va">DistanceAA</span> <span class="op">!=</span> <span class="fl">0</span>, <span class="op">]</span><span class="op">[</span><span class="st">'Type'</span><span class="op">]</span> <span class="op">=</span> <span class="st">'mutationAA'</span></span>
<span><span class="co">#new tree</span></span>
<span><span class="fu"><a href="../../reference/vis.html">vis</a></span><span class="op">(</span><span class="va">f</span><span class="op">)</span></span></code></pre></div>
<p><img src="BCRpipeline_files/figure-html/example%2010.3-1.png" width="1152" style="display: block; margin: auto;"></p>
<p>Another way to recolor leaves is to use <code>.vis_groups</code> parameter for repClonalFamily. It allows to assign group names for specific clone IDs, or lists of clone IDs:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#get all clone IDs from align_dt</span></span>
<span><span class="va">clone_ids</span> <span class="op">&lt;-</span> <span class="fu">unnest</span><span class="op">(</span><span class="va">align_dt</span><span class="op">[[</span><span class="st">"full_clones"</span><span class="op">]</span><span class="op">]</span>, <span class="st">"Sequences"</span><span class="op">)</span><span class="op">[[</span><span class="st">"Clone.ID"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#run repClonalFamily with assigning some of these clones to differently named and colored groups</span></span>
<span><span class="va">bcr_with_groups</span> <span class="op">&lt;-</span> <span class="va">align_dt</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/repClonalFamily.html">repClonalFamily</a></span><span class="op">(</span>.vis_groups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>                    Group1 <span class="op">=</span> <span class="va">clone_ids</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>                    Group2 <span class="op">=</span> <span class="va">clone_ids</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span>,</span>
<span>                    Group3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">clone_ids</span><span class="op">[</span><span class="fl">5</span><span class="op">]</span>, <span class="va">clone_ids</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                    Group4 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">clone_ids</span><span class="op">[</span><span class="fl">7</span><span class="op">]</span>, <span class="va">clone_ids</span><span class="op">[</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span></span>
<span>                   <span class="op">)</span>, .threads <span class="op">=</span> <span class="fl">2</span>, .nofail <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>                 <span class="op">)</span></span>
<span><span class="co">#display the first tree from repClonalFamily results</span></span>
<span><span class="fu"><a href="../../reference/vis.html">vis</a></span><span class="op">(</span><span class="va">bcr_with_groups</span><span class="op">[[</span><span class="st">"full_clones"</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"TreeStats"</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="BCRpipeline_files/figure-html/example%2010.4-1.png" width="1152" style="display: block; margin: auto;"></p>
<p>We have found 4 clusters:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bcr</span><span class="op">$</span><span class="va">full_clones</span><span class="op">$</span><span class="va">Cluster</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "IGHV1-69/IGHJ6_length_63_cluster_1" "IGHV1-69/IGHJ6_length_75_cluster_2"</span></span>
<span><span class="co">## [3] "IGHV3-23/IGHJ4_length_45_cluster_1" "IGHV3-23/IGHJ4_length_48_cluster_1"</span></span></code></pre>
<p>We have found mismatches between a germline and an ancestor sequence. Dots represent nucleotides matches between the sequences, letters represent mismatches between the sequences:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># the example of common ancestor sequence</span></span>
<span><span class="va">bcr</span><span class="op">$</span><span class="va">full_clones</span><span class="op">$</span><span class="va">Common.Ancestor</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "CAGGTGCAGCTGGTGCAGTCTGGGGCTGAGGTGAAGAAGCCTGGGTCCTCGGTGAAGGTCTCCTGCAAGGCTTCTGGAGGCACCTTCAGCAGCTATGCTATCAGCTGGGTGCGACAGGCCCCTGGACAAGGGCTTGAGTGGATGGGAGGGATCATCCCTATCTTTGGTACAGCAAACTACGCACAGAAGTTCCAGGGCAGAGTCACGATTACCGCGGACAAATCCACGAGCACAGCCTACATGGAGCTGAGCAGCCTGAGATCTGAGGACACGGCCGTGTATTACTGTGCGAGAGA-----TTDKGGCAGTGGTTATWCSMWKAACTACTACTACTACGGTATGGACGTCTGGGGCCAAGGGACCACGGTCACCGTCTCCTCAG"</span></span></code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># the example of mismatches between a germline and an ancestor sequence</span></span>
<span><span class="va">bcr</span><span class="op">$</span><span class="va">full_clones</span><span class="op">$</span><span class="va">Germline.Output</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "GAGGTCCAGCTGGTACAGTCTGGGGCTGAGGTGAAGAAGCCTGGGGCTACAGTGAAAATCTCCTGCAAGGTTTCTGGATACACCTTCACCGACTACTACATGCACTGGGTGCAACAGGCCCCTGGAAAAGGGCTTGAGTGGATGGGACTTGTTGATCCTGAAGATGGTGAAACAATATACGCAGAGAAGTTCCAGGGCAGAGTCACCATAACCGCGGACACGTCTACAGACACAGCCTACATGGAGCTGAGCAGCCTGAGATCTGAGGACACGGCCGTGTATTACNNNNNNNNNNN-----NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNGGGCAAGGGACCACGGTCACCGTCTCCTCAG"</span></span></code></pre>
<p>We have calculated a trunk length for each cluster:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bcr</span><span class="op">$</span><span class="va">full_clones</span><span class="op">[</span> , <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Cluster'</span>, <span class="st">'Trunk.Length'</span><span class="op">)</span> <span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##                              Cluster Trunk.Length</span></span>
<span><span class="co">## 1 IGHV1-69/IGHJ6_length_63_cluster_1           52</span></span>
<span><span class="co">## 2 IGHV1-69/IGHJ6_length_75_cluster_2           52</span></span>
<span><span class="co">## 3 IGHV3-23/IGHJ4_length_45_cluster_1            2</span></span>
<span><span class="co">## 4 IGHV3-23/IGHJ4_length_48_cluster_1            1</span></span></code></pre>
<p>Also trunk length specified in “TreeStats” table in column “DistanceNT”.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#example fot first tree</span></span>
<span><span class="va">bcr</span><span class="op">[[</span><span class="st">"full_clones"</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"TreeStats"</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span>, <span class="op">]</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="somatic-hypermutation-analysis">Somatic hypermutation analysis<a class="anchor" aria-label="anchor" href="#somatic-hypermutation-analysis"></a>
</h2>
<p>The rate of somatic hypermutation allows us to estimate repertoire maturation and detect the type of mutation contributing to the emergence of high affinity antibodies. This makes somatic hypermutation analysis a valuable asset for B-cell repertoire analysis.</p>
<p>In Immunarch, <code><a href="../../reference/repSomaticHypermutation.html">repSomaticHypermutation()</a></code> function is designed for hypermutation analysis:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bcr_data</span> <span class="op">&lt;-</span> <span class="va">bcrdata</span><span class="op">$</span><span class="va">data</span></span>
<span></span>
<span><span class="va">shm_data</span> <span class="op">&lt;-</span> <span class="va">bcr</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="../../reference/repSomaticHypermutation.html">repSomaticHypermutation</a></span><span class="op">(</span>.threads <span class="op">=</span> <span class="fl">2</span>, .nofail <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>The function repSomaticHypermutation() takes V and J germline sequences and V and J clonotype sequences as an input. Then the function aligns germline and clonotype sequences to detect and calculate occurring mutations.</p>
<p>Examples of germline and clonotype sequences:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">full_clones</span> <span class="op">&lt;-</span> <span class="va">shm_data</span><span class="op">$</span><span class="va">full_clones</span></span>
<span><span class="va">v_length</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nchar.html" class="external-link">nchar</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">full_clones</span><span class="op">[</span><span class="fl">1</span>, <span class="st">"FR1.nt"</span><span class="op">]</span>, <span class="va">full_clones</span><span class="op">[</span><span class="fl">1</span>, <span class="st">"CDR1.nt"</span><span class="op">]</span>,</span>
<span> <span class="va">full_clones</span><span class="op">[</span><span class="fl">1</span>, <span class="st">"FR2.nt"</span><span class="op">]</span>, <span class="va">full_clones</span><span class="op">[</span><span class="fl">1</span>, <span class="st">"CDR2.nt"</span><span class="op">]</span>, <span class="va">full_clones</span><span class="op">[</span><span class="fl">1</span>, <span class="st">"FR3.nt"</span><span class="op">]</span>, collapse<span class="op">=</span><span class="st">""</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">j_length</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nchar.html" class="external-link">nchar</a></span><span class="op">(</span><span class="va">full_clones</span><span class="op">[</span><span class="fl">1</span>, <span class="st">"FR4.nt"</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">seq_length</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nchar.html" class="external-link">nchar</a></span><span class="op">(</span><span class="va">full_clones</span><span class="op">[</span><span class="fl">1</span>, <span class="st">"Sequence"</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># the example of germline V sequence</span></span>
<span><span class="va">full_clones</span><span class="op">$</span><span class="va">Germline.Input</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/substr.html" class="external-link">substr</a></span><span class="op">(</span><span class="fl">1</span>, <span class="va">v_length</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "GAGGTCCAGCTGGTACAGTCTGGGGCTGAGGTGAAGAAGCCTGGGGCTACAGTGAAAATCTCCTGCAAGGTTTCTGGATACACCTTCACCGACTACTACATGCACTGGGTGCAACAGGCCCCTGGAAAAGGGCTTGAGTGGATGGGACTTGTTGATCCTGAAGATGGTGAAACAATATACGCAGAGAAGTTCCAGGGCAGAGTCACCATAACCGCGGACACGTCTACAGACACAGCCTACATGGAGCTGAGCAGCCTGAGATCTGAGGACACGGCCGTGTATTACNNNN"</span></span></code></pre>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># the example of germline J sequence</span></span>
<span><span class="va">full_clones</span><span class="op">$</span><span class="va">Germline.Input</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/substr.html" class="external-link">substr</a></span><span class="op">(</span><span class="va">seq_length</span> <span class="op">-</span> <span class="va">j_length</span>, <span class="va">seq_length</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "NGGGCAAGGGACCACGGTCACCGTCTCCTCAG"</span></span></code></pre>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># the example of clonotype V sequence</span></span>
<span><span class="va">full_clones</span><span class="op">$</span><span class="va">Sequence</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/substr.html" class="external-link">substr</a></span><span class="op">(</span><span class="fl">1</span>, <span class="va">v_length</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "CAGGTGCAGCTGGTGCAGTCTGGGGCTGAGGTGAAGAAGCCTGGGTCCTCGGTGAAGGTCTCCTGCAAGGCTTCTGGAGGCACCTTCAGCAGCTATGCTATCAGCTGGGTGCGACAGGCCCCTGGACAAGGGCTTGAGTGGATGGGAGGGATCATCCCTATCTTTGGTACAGCAAACTACGCACAGAAGTTCCAGGGCAGAGTCACGATTACCGCGGACGAATCCACGAGCACAGCCTACATGGAGCTGAGCAGCCTGAGATCTGAGGACACGGCCGTGTATTACTGTG"</span></span></code></pre>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># the example of clonotype J sequence</span></span>
<span><span class="va">full_clones</span><span class="op">$</span><span class="va">Sequence</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/substr.html" class="external-link">substr</a></span><span class="op">(</span><span class="va">seq_length</span> <span class="op">-</span> <span class="va">j_length</span>, <span class="va">seq_length</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "GGGCCAAGGGACCACGGTCACCGTCTCCTCAG"</span></span></code></pre>
<p>Example: aligning germline and clonotype V sequences:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/image.html" class="external-link">image</a></span><span class="op">(</span><span class="va">shm_data</span><span class="op">$</span><span class="va">full_clones</span><span class="op">$</span><span class="va">Germline.Alignment.V</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span>, grid <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="BCRpipeline_files/figure-html/example%2016-1.png" width="1152" style="display: block; margin: auto;"></p>
<p>Example: aligning germline and clonotype J sequences:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/image.html" class="external-link">image</a></span><span class="op">(</span><span class="va">shm_data</span><span class="op">$</span><span class="va">full_clones</span><span class="op">$</span><span class="va">Germline.Alignment.J</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span>, grid <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="BCRpipeline_files/figure-html/example%2017-1.png" width="1152" style="display: block; margin: auto;"></p>
<p>The number of mutations for each clonotype sequence:</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Clone.ID'</span>, <span class="st">'Substitutions'</span>, <span class="st">'Insertions'</span>, <span class="st">'Deletions'</span>, <span class="st">'Mutations'</span><span class="op">)</span></span>
<span><span class="va">shm_data</span><span class="op">$</span><span class="va">full_clones</span><span class="op">[</span> , <span class="va">cols</span> <span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##    Clone.ID Substitutions Insertions Deletions Mutations</span></span>
<span><span class="co">## 1       851            53          0         0        53</span></span>
<span><span class="co">## 2       921            53          0         0        53</span></span>
<span><span class="co">## 3       214            53          0         0        53</span></span>
<span><span class="co">## 4       159            52          0         0        52</span></span>
<span><span class="co">## 5       222            53          0         0        53</span></span>
<span><span class="co">## 6       270            53          0         0        53</span></span>
<span><span class="co">## 7       250            53          0         0        53</span></span>
<span><span class="co">## 8       330            53          0         0        53</span></span>
<span><span class="co">## 9       966            52          0         0        52</span></span>
<span><span class="co">## 10      672            52          0         0        52</span></span>
<span><span class="co">## 11      401            53          0         0        53</span></span>
<span><span class="co">## 12      386            53          0         0        53</span></span>
<span><span class="co">## 13      736            53          0         0        53</span></span>
<span><span class="co">## 14      295            53          0         0        53</span></span>
<span><span class="co">## 15      202            53          0         0        53</span></span>
<span><span class="co">## 16      716            52          0         0        52</span></span>
<span><span class="co">## 17      827             1          0         0         1</span></span>
<span><span class="co">## 18      591            16          0         0        16</span></span>
<span><span class="co">## 19      433             1          0         0         1</span></span>
<span><span class="co">## 20      163            22          0         0        22</span></span>
<span><span class="co">## 21      914            24          0         0        24</span></span>
<span><span class="co">## 22      524            12          0         0        12</span></span>
<span><span class="co">## 23       24             6          0         0         6</span></span>
<span><span class="co">## 24      342             6          0         0         6</span></span>
<span><span class="co">## 25      940            19          0         0        19</span></span>
<span><span class="co">## 26      238             6          0         0         6</span></span>
<span><span class="co">## 27      428            15          0         0        15</span></span>
<span><span class="co">## 28      484            13          0         0        13</span></span>
<span><span class="co">## 29      932             1          0         0         1</span></span>
<span><span class="co">## 30      517             4          0         0         4</span></span>
<span><span class="co">## 31      522             4          0         0         4</span></span>
<span><span class="co">## 32      933            12          0         0        12</span></span></code></pre>
<p>Then you could easily estimate the mutation rate:</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># estimate mutation rate</span></span>
<span><span class="va">shm_data</span><span class="op">$</span><span class="va">full_clones</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>Mutation.Rate <span class="op">=</span> <span class="va">Mutations</span> <span class="op">/</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nchar.html" class="external-link">nchar</a></span><span class="op">(</span><span class="va">Sequence</span><span class="op">)</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/nchar.html" class="external-link">nchar</a></span><span class="op">(</span><span class="va">CDR3.nt</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">Clone.ID</span>, <span class="va">Mutation.Rate</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    Clone.ID Mutation.Rate</span></span>
<span><span class="co">## 1       851   0.167721519</span></span>
<span><span class="co">## 2       921   0.167721519</span></span>
<span><span class="co">## 3       214   0.167721519</span></span>
<span><span class="co">## 4       159   0.164556962</span></span>
<span><span class="co">## 5       222   0.167721519</span></span>
<span><span class="co">## 6       270   0.167721519</span></span>
<span><span class="co">## 7       250   0.167721519</span></span>
<span><span class="co">## 8       330   0.167721519</span></span>
<span><span class="co">## 9       966   0.164556962</span></span>
<span><span class="co">## 10      672   0.164556962</span></span>
<span><span class="co">## 11      401   0.167721519</span></span>
<span><span class="co">## 12      386   0.167721519</span></span>
<span><span class="co">## 13      736   0.167721519</span></span>
<span><span class="co">## 14      295   0.167721519</span></span>
<span><span class="co">## 15      202   0.167721519</span></span>
<span><span class="co">## 16      716   0.164556962</span></span>
<span><span class="co">## 17      827   0.003164557</span></span>
<span><span class="co">## 18      591   0.050632911</span></span>
<span><span class="co">## 19      433   0.003164557</span></span>
<span><span class="co">## 20      163   0.069620253</span></span>
<span><span class="co">## 21      914   0.075949367</span></span>
<span><span class="co">## 22      524   0.037974684</span></span>
<span><span class="co">## 23       24   0.018987342</span></span>
<span><span class="co">## 24      342   0.018987342</span></span>
<span><span class="co">## 25      940   0.060126582</span></span>
<span><span class="co">## 26      238   0.018987342</span></span>
<span><span class="co">## 27      428   0.047468354</span></span>
<span><span class="co">## 28      484   0.041139241</span></span>
<span><span class="co">## 29      932   0.003164557</span></span>
<span><span class="co">## 30      517   0.012658228</span></span>
<span><span class="co">## 31      522   0.012658228</span></span>
<span><span class="co">## 32      933   0.037974684</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<ol style="list-style-type: decimal">
<li><a href="https://genomemedicine.biomedcentral.com/articles/10.1186/s13073-015-0243-2" class="external-link">https://genomemedicine.biomedcentral.com/articles/10.1186/s13073-015-0243-2</a></li>
<li><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4754972/" class="external-link">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4754972/</a></li>
<li><a href="https://www.frontiersin.org/articles/10.3389/fimmu.2020.01734/full" class="external-link">https://www.frontiersin.org/articles/10.3389/fimmu.2020.01734/full</a></li>
<li><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4754972/" class="external-link">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4754972/</a></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://www.linkedin.com/in/vdnaz" class="external-link">Vadim I. Nazarov</a>, <a href="https://www.linkedin.com/in/vasily-tsvetkov-227218ab" class="external-link">Vasily O. Tsvetkov</a>, <a href="https://www.linkedin.com/in/sergey-f-709b0415b/" class="external-link">Siarhei Fiadziushchanka</a>, <a href="https://www.linkedin.com/in/erumynskiy/" class="external-link">Eugene Rumynskiy</a>, <a href="https://www.linkedin.com/in/aleksandr-popov-634878105/" class="external-link">Aleksandr A. Popov</a>, <a href="https://www.linkedin.com/in/ivan-balashov/" class="external-link">Ivan Balashov</a>, <a href="https://www.linkedin.com/in/maria-volobueva-b0856a223/" class="external-link">Maria Samokhina</a>, <a href="https://immunomind.io" class="external-link">ImmunoMind</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '94207562f34b455c0796790d104e3549',
    indexName: 'immunarch',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
