<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Loading MiXCR Data • immunarch</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><!-- docsearch --><script src="../../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><link href="../../extra.css" rel="stylesheet">
<meta property="og:title" content="Loading MiXCR Data">
<meta property="og:description" content="immunarch">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-128418614-2"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-128418614-2');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">immunarch</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.7.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Start Here
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/v1_introduction.html">Installation &amp; troubleshooting</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">First steps</li>
    <li>
      <a href="../../articles/v2_data.html">Data loading</a>
    </li>
    <li>
      <a href="../../articles/web_only/load_mixcr.html">How-to: Loading MiXCR Data</a>
    </li>
    <li>
      <a href="../../articles/web_only/load_10x.html">How-to: Loading 10x Genomics Data</a>
    </li>
    <li>
      <a href="../../articles/web_only/v21_singlecell.html">How-to: Single-cell and paired chain data</a>
    </li>
    <li>
      <a href="../../articles/web_only/v3_basic_analysis.html">Basic statistics and clonality</a>
    </li>
    <li>
      <a href="../../articles/web_only/repFilter_v3.html">Data filtering</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Repertoire-level exploration and comparison</li>
    <li>
      <a href="../../articles/web_only/v4_overlap.html">Repertoire overlap and public clonotypes</a>
    </li>
    <li>
      <a href="../../articles/web_only/v5_gene_usage.html">Gene usage</a>
    </li>
    <li>
      <a href="../../articles/web_only/v6_diversity.html">Diversity estimation</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Clonotype-level exploration and comparison</li>
    <li>
      <a href="../../articles/web_only/v8_tracking.html">Track clonotypes across samples and time</a>
    </li>
    <li>
      <a href="../../articles/web_only/v11_db.html">Annotate clonotypes using immune receptor databases</a>
    </li>
    <li>
      <a href="../../articles/web_only/v9_kmers.html">Kmer and sequence motif analysis and visualisation</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Preparing to publication</li>
    <li>
      <a href="../../articles/web_only/v7_fixvis.html">Make your plots publication-ready with fixVis</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">BCR analysis</li>
    <li>
      <a href="../../articles/web_only/BCRpipeline.html">BCR pipeline</a>
    </li>
    <li>
      <a href="../../articles/web_only/clustering.html">Clustering</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
<li>
  <a href="https://github.com/immunomind/covid19" class="external-link">COVID-19</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://immunomind.io" class="external-link">ImmunoMind</a>
</li>
<li>
  <a href="http://twitter.com/immunomind" class="external-link">
    <span class="fa fa-lg fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/immunomind/immunarch" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Loading MiXCR Data</h1>
                        <h4 data-toc-skip class="author">
<b>ImmunoMind</b> – improving design of T-cell therapies using multi-omics and AI. Research and biopharma partnerships, more details: <a href="https://immunomind.io" class="external-link">immunomind.io</a>
</h4>
            
            <h4 data-toc-skip class="date"><a href="mailto:support@immunomind.io" class="email">support@immunomind.io</a></h4>
      
      <small class="dont-index">Source: <a href="https://github.com/immunomind/immunarch/blob/HEAD/vignettes/web_only/load_mixcr.Rmd" class="external-link"><code>vignettes/web_only/load_mixcr.Rmd</code></a></small>
      <div class="hidden name"><code>load_mixcr.Rmd</code></div>

    </div>

    
    
<!--
%\VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{Loading MiXCR Data}
%\VignettePackage{immunarch}
-->
<div class="section level2">
<h2 id="intro-to-mixcr">Intro to MiXCR<a class="anchor" aria-label="anchor" href="#intro-to-mixcr"></a>
</h2>
<p>MiXCR is a universal software for fast and accurate extraction of T- and B- cell receptor repertoires from any type of sequencing data. It handles paired- and single-end reads, considers sequence quality, corrects PCR errors and identifies germline hypermutations. The software supports both partial- and full-length profiling and employs all available RNA or DNA information, including sequences upstream of V and downstream of J gene segments.</p>
<p>Some of its features include:</p>
<ul>
<li>Extracting both T- and B-cell receptor repertoires</li>
<li>Extracting repertoire data even from regular RNA-Seq</li>
<li>Successfully analysing full-length antibody data</li>
</ul>
<p>Find more info about MiXCR at <a href="https://milaboratory.com/software/mixcr/" class="external-link">here</a></p>
</div>
<div class="section level2">
<h2 id="prepare-mixcr-data">Prepare MiXCR Data<a class="anchor" aria-label="anchor" href="#prepare-mixcr-data"></a>
</h2>
<p>Follow these instructions <a href="https://mixcr.readthedocs.io/en/master/install.html" class="external-link">here</a> to install MiXCR and get started with processing data. <strong>Important Note:</strong> Currently supports MiXCR version 3 and above.</p>
<p>MiXCR supports the following formats of sequencing data: fasta, fastq, fastq.gz, paired-end fastq and fastq.gz. In this tutorial we use the real IGH data from <a href="https://github.com/milaboratory/mixcr/blob/develop/doc/paper/paperAttachments.md" class="external-link">here</a>.</p>
<p>You can choose to use the <code>analyze amplicon</code> method to process in one go:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> mixcr analyze amplicon <span class="sc">--</span>species hs \</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>starting<span class="sc">-</span>material dna \</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span><span class="dv">5</span><span class="sc">-</span>end v<span class="sc">-</span>primers \</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span><span class="dv">3</span><span class="sc">-</span>end j<span class="sc">-</span>primers \</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>adapters adapters<span class="sc">-</span>present \</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>receptor<span class="sc">-</span>type IGH \</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>        input_R1.fastq input_R2.fastq analysis</span></code></pre></div>
<p>or execute each step <code>align</code>, <code>assemble</code>, and <code>exportClones</code> individually.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> mixcr align <span class="sc">-</span>s hs <span class="sc">-</span>OvParameters.geneFeatureToAlign<span class="ot">=</span>VTranscript \</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>  <span class="sc">--</span>report analysis.report input_R1.fastq input_R2.fastq analysis.vdjca</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>Analysis Date<span class="sc">:</span> Mon Aug <span class="dv">25</span> <span class="dv">15</span><span class="sc">:</span><span class="dv">22</span><span class="sc">:</span><span class="dv">39</span> MSK <span class="dv">2014</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>Input <span class="fu">file</span>(s)<span class="sc">:</span> input_r1.fastq,input_r2.fastq</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>Output file<span class="sc">:</span> alignments.vdjca</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>Command line arguments<span class="sc">:</span> align <span class="sc">--</span>report alignmentReport.log input_r1.fastq input_r2.fastq alignments.vdjca</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>Total sequencing reads<span class="sc">:</span> <span class="dv">323248</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>Successfully aligned reads<span class="sc">:</span> <span class="dv">210360</span></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>Successfully aligned, percent<span class="sc">:</span> <span class="fl">65.08</span>%</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>Alignment failed because of absence of V hits<span class="sc">:</span> <span class="fl">4.26</span>%</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>Alignment failed because of absence of J hits<span class="sc">:</span> <span class="fl">30.19</span>%</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>Alignment failed because of low total score<span class="sc">:</span> <span class="fl">0.48</span>%</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="prepare-files-to-load">Prepare files to load<a class="anchor" aria-label="anchor" href="#prepare-files-to-load"></a>
</h2>
<p>After you run these commands, you will generate these files with detailed information about calculated clonotypes:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>.</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>├── analysis.clonotypes.<span class="sc">&lt;</span>chains<span class="sc">&gt;</span>.txt <span class="ot">&lt;-</span><span class="sc">-</span> This contains the count data we want<span class="sc">!</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>├── analysis.clna <span class="ot">&lt;-</span> Build clonotypes correct PCR and sequencing errors</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>├── analysis.vdjca <span class="ot">&lt;-</span> Align raw sequences to reference sequences of <span class="fu">segments</span> (V, D, J) of IGH gene</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>├── analysis.report <span class="ot">&lt;-</span> Information on the run</span></code></pre></div>
<p>Create a new folder that only contains the specified clonotype text files from your runs, and create a metadata.txt file in the following format.</p>
<p>The metadata file “metadata.txt” has to be tab delimited file with first column named “Sample” and any number of additional columns with arbitrary names. The first column should contain base names of files without extensions in your folder.</p>
<table class="table">
<thead><tr class="header">
<th align="center"><strong>Sample</strong></th>
<th align="center"><strong>Sex</strong></th>
<th align="center"><strong>Age</strong></th>
<th align="center"><strong>Status</strong></th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center">immunoseq_1</td>
<td align="center">M</td>
<td align="center">1</td>
<td align="center">C</td>
</tr>
<tr class="even">
<td align="center">immunoseq_2</td>
<td align="center">M</td>
<td align="center">2</td>
<td align="center">C</td>
</tr>
<tr class="odd">
<td align="center">immunoseq_3</td>
<td align="center">F</td>
<td align="center">3</td>
<td align="center">A</td>
</tr>
</tbody>
</table>
<p>The next section will explain how to load a single file or multiple samples in a folder.</p>
</div>
<div class="section level2">
<h2 id="load-into-immunarch">Load into Immunarch<a class="anchor" aria-label="anchor" href="#load-into-immunarch"></a>
</h2>
<p>In your R environment, run the commands below. The output should be similar. You can run it on the <a href="#loading-a-folder">entire folder</a> or a <a href="#loading-a-single-file">single file</a>. <code>repLoad</code> will ignore the file formats that are unsupported.</p>
<div class="section level3">
<h3 id="loading-a-single-file">Loading a single file<a class="anchor" aria-label="anchor" href="#loading-a-single-file"></a>
</h3>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># 1.1) Load the package into R:</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">library</span>(immunarch)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="co"># 1.2) Replace with the path to your clonotypes file</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> file_path <span class="ot">=</span> <span class="st">"path/to/your/mixcr/data/analysis.clonotypes.IGH.txt"</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="co"># 1.3) Load MiXCR data with repLoad</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> immdata_mixcr <span class="ot">&lt;-</span> <span class="fu">repLoad</span>(file_path)</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a><span class="sc">==</span> Step <span class="dv">1</span><span class="sc">/</span><span class="dv">3</span><span class="sc">:</span> loading repertoire files... <span class="sc">==</span></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>Processing <span class="st">"&lt;initial&gt;"</span> ...</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>  <span class="sc">--</span> Parsing <span class="st">"/path/to/your/mixcr/data/analysis.clonotypes.IGH.txt"</span> <span class="sc">--</span> mixcr</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a><span class="sc">==</span> Step <span class="dv">2</span><span class="sc">/</span><span class="dv">3</span><span class="sc">:</span> checking metadata files and merging... <span class="sc">==</span></span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a>Processing <span class="st">"&lt;initial&gt;"</span> ...</span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a>  <span class="sc">--</span> Metadata file not found; creating a dummy metadata...</span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a><span class="sc">==</span> Step <span class="dv">3</span><span class="sc">/</span><span class="dv">3</span><span class="sc">:</span> splitting data by barcodes and chains... <span class="sc">==</span></span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a>Done<span class="sc">!</span></span></code></pre></div>
<p>Congrats! Now your data is ready for exploration. Follow the steps here to learn more about how to explore your dataset.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>r<span class="sc">$</span><span class="er">&gt;</span> immdata_mixcr</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>data</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>data<span class="sc">$</span>analysis.clonotypes.IGH</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="co"># A tibble: 33,812 x 15</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>   Clones Proportion CDR3.nt            CDR3.aa    V.name    D.name    J.name V.end D.start D.end J.start VJ.ins VD.ins DJ.ins Sequence</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>    <span class="sc">&lt;</span>dbl<span class="sc">&gt;</span>      <span class="er">&lt;</span>dbl<span class="sc">&gt;</span> <span class="er">&lt;</span>chr<span class="sc">&gt;</span>              <span class="er">&lt;</span>chr<span class="sc">&gt;</span>      <span class="er">&lt;</span>chr<span class="sc">&gt;</span>     <span class="er">&lt;</span>chr<span class="sc">&gt;</span>     <span class="er">&lt;</span>chr<span class="sc">&gt;</span>  <span class="er">&lt;</span>int<span class="sc">&gt;</span>   <span class="er">&lt;</span>int<span class="sc">&gt;</span> <span class="er">&lt;</span>int<span class="sc">&gt;</span>   <span class="er">&lt;</span>int<span class="sc">&gt;</span>  <span class="er">&lt;</span>int<span class="sc">&gt;</span>  <span class="er">&lt;</span>int<span class="sc">&gt;</span>  <span class="er">&lt;</span>int<span class="sc">&gt;</span> <span class="er">&lt;</span>chr<span class="sc">&gt;</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a> <span class="dv">1</span>    <span class="dv">230</span>    <span class="fl">0.00284</span> TGTGTGAGACATAAACC… CVRHKPMVQ… IGHV4<span class="dv">-39</span>  IGHD3<span class="sc">-</span>10… IGHJ6     <span class="dv">12</span>      <span class="cn">NA</span>     <span class="dv">5</span>      <span class="dv">36</span>      <span class="dv">9</span>      <span class="dv">3</span>      <span class="dv">6</span> TGTGTGAGACATAAACC…</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a> <span class="dv">2</span>    <span class="dv">201</span>    <span class="fl">0.00248</span> TGTGCGATTTGGGATGT… CAIWDVGLR… IGHV4<span class="dv">-34</span>  IGHD2<span class="dv">-21</span>  IGHJ4…     <span class="dv">7</span>      <span class="cn">NA</span>     <span class="dv">5</span>      <span class="dv">29</span>     <span class="dv">10</span>      <span class="dv">7</span>      <span class="dv">3</span> TGTGCGATTTGGGATGT…</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a> <span class="dv">3</span>    <span class="dv">179</span>    <span class="fl">0.00221</span> TGTGCGAGAGATCATGC… CARDHAGFG… IGHV1<span class="sc">-</span>69… IGHD3<span class="dv">-10</span>  IGHJ6     <span class="dv">13</span>      <span class="cn">NA</span>     <span class="dv">4</span>      <span class="dv">40</span>     <span class="dv">18</span>      <span class="dv">5</span>     <span class="dv">13</span> TGTGCGAGAGATCATGC…</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a> <span class="dv">4</span>     <span class="dv">99</span>    <span class="fl">0.00122</span> TGTGCGAGATGGGGATA… CARWGYCIN… IGHV4<span class="dv">-39</span>  IGHD2<span class="dv">-8</span>   IGHJ6      <span class="dv">9</span>      <span class="cn">NA</span>     <span class="dv">6</span>      <span class="dv">64</span>     <span class="dv">23</span>      <span class="dv">2</span>     <span class="dv">21</span> TGTGCGAGATGGGGATA…</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a> <span class="dv">5</span>     <span class="dv">97</span>    <span class="fl">0.00120</span> TGTGCGAGAGGCCCCAC… CARGPTSSE… IGHV4<span class="dv">-34</span>  IGHD3<span class="sc">-</span>22… IGHJ6     <span class="dv">13</span>      <span class="cn">NA</span>     <span class="dv">6</span>      <span class="dv">52</span>     <span class="dv">26</span>     <span class="dv">24</span>      <span class="dv">2</span> TGTGCGAGAGGCCCCAC…</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a> <span class="dv">6</span>     <span class="dv">97</span>    <span class="fl">0.00120</span> TGTGCGCACCACTATAC… CAHHYTSDY… IGHV2<span class="dv">-5</span>   IGHD1<span class="dv">-26</span>  IGHJ5      <span class="dv">9</span>      <span class="cn">NA</span>     <span class="dv">2</span>      <span class="dv">39</span>     <span class="dv">19</span>     <span class="cn">NA</span>     <span class="dv">20</span> TGTGCGCACCACTATAC…</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a> <span class="dv">7</span>     <span class="dv">92</span>    <span class="fl">0.00114</span> TGTGCGAGAGGCCCTCC… CARGPPSMG… IGHV4<span class="dv">-34</span>  IGHD5<span class="sc">-</span>24… IGHJ4     <span class="dv">13</span>      <span class="cn">NA</span>     <span class="dv">3</span>      <span class="dv">38</span>     <span class="dv">11</span>      <span class="dv">6</span>      <span class="dv">5</span> TGTGCGAGAGGCCCTCC…</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a> <span class="dv">8</span>     <span class="dv">84</span>    <span class="fl">0.00104</span> TGTGCGAGGTGGCTTGG… CARWLGEDI… IGHV4<span class="dv">-39</span>  IGHD3<span class="sc">-</span>16… IGHJ4…     <span class="dv">8</span>      <span class="cn">NA</span>     <span class="dv">6</span>      <span class="dv">32</span>     <span class="dv">13</span>      <span class="dv">4</span>      <span class="dv">9</span> TGTGCGAGGTGGCTTGG…</span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a> <span class="dv">9</span>     <span class="dv">83</span>    <span class="fl">0.00103</span> TGTGCGAGAGGCCGCAG… CARGRSGDP… IGHV4<span class="dv">-34</span>  IGHD2<span class="dv">-2</span>,… IGHJ5     <span class="dv">13</span>      <span class="cn">NA</span>     <span class="dv">4</span>      <span class="dv">50</span>     <span class="dv">18</span>     <span class="dv">13</span>      <span class="dv">5</span> TGTGCGAGAGGCCGCAG…</span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a><span class="dv">10</span>     <span class="dv">81</span>    <span class="fl">0.00100</span> TGTGTGAGTCACCTCCT… CVSHLLDTS… IGHV1<span class="dv">-2</span>   IGHD2<span class="sc">-</span>21… IGHJ4…     <span class="dv">8</span>      <span class="cn">NA</span>     <span class="dv">3</span>      <span class="dv">40</span>     <span class="dv">20</span>     <span class="dv">14</span>      <span class="dv">6</span> TGTGTGAGTCACCTCCT…</span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a><span class="co"># … with 33,802 more rows</span></span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>meta</span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a><span class="co"># A tibble: 1 x 1</span></span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a>  Sample</span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a>  <span class="sc">&lt;</span>chr<span class="sc">&gt;</span></span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span> analysis.clonotypes.IGH</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="loading-a-folder">Loading a folder<a class="anchor" aria-label="anchor" href="#loading-a-folder"></a>
</h3>
<p>In this tutorial we use three identical samples just to demonstrate the output, but you should put all your output <code>.txt</code> clonotype files in this folder, along with your <code>metadata.txt</code> file.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># 1.1) Load the package into R:</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">library</span>(immunarch)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="co"># 1.2) Replace with the path to the folder with your processed MiXCR data.</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> file_path <span class="ot">=</span> <span class="st">"/path/to/your/mixcr/data/"</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a><span class="co"># 1.3) Load MiXCR data with repLoad</span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> immdata_mixcr <span class="ot">&lt;-</span> <span class="fu">repLoad</span>(file_path)</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a><span class="sc">==</span> Step <span class="dv">1</span><span class="sc">/</span><span class="dv">3</span><span class="sc">:</span> loading repertoire files... <span class="sc">==</span></span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>Processing <span class="st">"/path/to/your/mixcr/data/"</span> ...</span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>  <span class="sc">--</span> Parsing <span class="st">"/path/to/your/mixcr/data/analysis.clonotypes.IGH_1.txt"</span> <span class="sc">--</span> mixcr</span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>  <span class="sc">--</span> Parsing <span class="st">"/path/to/your/mixcr/data/analysis.clonotypes.IGH_2.txt"</span> <span class="sc">--</span> mixcr</span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>  <span class="sc">--</span> Parsing <span class="st">"/path/to/your/mixcr/data/analysis.clonotypes.IGH_3.txt"</span> <span class="sc">--</span> mixcr</span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a>  <span class="sc">--</span> Parsing <span class="st">"/path/to/your/mixcr/data/metadata.txt"</span> <span class="sc">--</span> metadata</span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a><span class="sc">==</span> Step <span class="dv">2</span><span class="sc">/</span><span class="dv">3</span><span class="sc">:</span> checking metadata files and merging files... <span class="sc">==</span></span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a>Processing <span class="st">"/path/to/your/mixcr/data/"</span> ...</span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a>  <span class="sc">--</span> Everything is OK<span class="sc">!</span></span>
<span id="cb6-22"><a href="#cb6-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-23"><a href="#cb6-23" aria-hidden="true" tabindex="-1"></a><span class="er">==</span> Step <span class="dv">3</span><span class="sc">/</span><span class="dv">3</span><span class="sc">:</span> processing paired chain data... <span class="sc">==</span></span>
<span id="cb6-24"><a href="#cb6-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-25"><a href="#cb6-25" aria-hidden="true" tabindex="-1"></a>Done<span class="sc">!</span></span></code></pre></div>
<p>Now let’s take a look at the data! Your output should look something like that.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>r<span class="sc">$</span><span class="er">&gt;</span> immdata_mixcr</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>data</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>data<span class="sc">$</span>analysis.clonotypes.IGH_1</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="co"># A tibble: 32,744 x 15</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>   Clones Proportion CDR3.nt                 CDR3.aa     V.name    D.name      J.name  V.end D.start D.end J.start VJ.ins VD.ins DJ.ins Sequence</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>    <span class="sc">&lt;</span>dbl<span class="sc">&gt;</span>      <span class="er">&lt;</span>dbl<span class="sc">&gt;</span> <span class="er">&lt;</span>chr<span class="sc">&gt;</span>                   <span class="er">&lt;</span>chr<span class="sc">&gt;</span>       <span class="er">&lt;</span>chr<span class="sc">&gt;</span>     <span class="er">&lt;</span>chr<span class="sc">&gt;</span>       <span class="er">&lt;</span>chr<span class="sc">&gt;</span>   <span class="er">&lt;</span>int<span class="sc">&gt;</span>   <span class="er">&lt;</span>int<span class="sc">&gt;</span> <span class="er">&lt;</span>int<span class="sc">&gt;</span>   <span class="er">&lt;</span>int<span class="sc">&gt;</span>  <span class="er">&lt;</span>int<span class="sc">&gt;</span>  <span class="er">&lt;</span>int<span class="sc">&gt;</span>  <span class="er">&lt;</span>int<span class="sc">&gt;</span> <span class="er">&lt;</span>chr<span class="sc">&gt;</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a> <span class="dv">1</span>    <span class="dv">230</span>    <span class="fl">0.00284</span> TGTGTGAGACATAAACCTATGG… CVRHKPMVQG… IGHV4<span class="dv">-39</span>  IGHD3<span class="dv">-10</span>, … IGHJ6      <span class="dv">12</span>      <span class="cn">NA</span>     <span class="dv">5</span>      <span class="dv">36</span>      <span class="dv">9</span>      <span class="dv">3</span>      <span class="dv">6</span> TGTGTGAGACATAAACCTATG…</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a> <span class="dv">2</span>    <span class="dv">201</span>    <span class="fl">0.00248</span> TGTGCGATTTGGGATGTGGGAC… CAIWDVGLRH… IGHV4<span class="dv">-34</span>  IGHD2<span class="dv">-21</span>    IGHJ4,…     <span class="dv">7</span>      <span class="cn">NA</span>     <span class="dv">5</span>      <span class="dv">29</span>     <span class="dv">10</span>      <span class="dv">7</span>      <span class="dv">3</span> TGTGCGATTTGGGATGTGGGA…</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a> <span class="dv">3</span>    <span class="dv">179</span>    <span class="fl">0.00221</span> TGTGCGAGAGATCATGCGGGGT… CARDHAGFGK… IGHV1<span class="sc">-</span>69… IGHD3<span class="dv">-10</span>    IGHJ6      <span class="dv">13</span>      <span class="cn">NA</span>     <span class="dv">4</span>      <span class="dv">40</span>     <span class="dv">18</span>      <span class="dv">5</span>     <span class="dv">13</span> TGTGCGAGAGATCATGCGGGG…</span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a> <span class="dv">4</span>     <span class="dv">99</span>    <span class="fl">0.00122</span> TGTGCGAGATGGGGATATTGTA… CARWGYCING… IGHV4<span class="dv">-39</span>  IGHD2<span class="dv">-8</span>     IGHJ6       <span class="dv">9</span>      <span class="cn">NA</span>     <span class="dv">6</span>      <span class="dv">64</span>     <span class="dv">23</span>      <span class="dv">2</span>     <span class="dv">21</span> TGTGCGAGATGGGGATATTGT…</span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a> <span class="dv">5</span>     <span class="dv">97</span>    <span class="fl">0.00120</span> TGTGCGAGAGGCCCCACGAGCA… CARGPTSSEW… IGHV4<span class="dv">-34</span>  IGHD3<span class="dv">-22</span>, … IGHJ6      <span class="dv">13</span>      <span class="cn">NA</span>     <span class="dv">6</span>      <span class="dv">52</span>     <span class="dv">26</span>     <span class="dv">24</span>      <span class="dv">2</span> TGTGCGAGAGGCCCCACGAGC…</span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a> <span class="dv">6</span>     <span class="dv">97</span>    <span class="fl">0.00120</span> TGTGCGCACCACTATACCAGCG… CAHHYTSDYY… IGHV2<span class="dv">-5</span>   IGHD1<span class="dv">-26</span>    IGHJ5       <span class="dv">9</span>      <span class="cn">NA</span>     <span class="dv">2</span>      <span class="dv">39</span>     <span class="dv">19</span>     <span class="cn">NA</span>     <span class="dv">20</span> TGTGCGCACCACTATACCAGC…</span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a> <span class="dv">7</span>     <span class="dv">92</span>    <span class="fl">0.00114</span> TGTGCGAGAGGCCCTCCGTCGA… CARGPPSMGT… IGHV4<span class="dv">-34</span>  IGHD5<span class="dv">-24</span>, … IGHJ4      <span class="dv">13</span>      <span class="cn">NA</span>     <span class="dv">3</span>      <span class="dv">38</span>     <span class="dv">11</span>      <span class="dv">6</span>      <span class="dv">5</span> TGTGCGAGAGGCCCTCCGTCG…</span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a> <span class="dv">8</span>     <span class="dv">84</span>    <span class="fl">0.00104</span> TGTGCGAGGTGGCTTGGGGAAG… CARWLGEDIR… IGHV4<span class="dv">-39</span>  IGHD3<span class="dv">-16</span>, … IGHJ4,…     <span class="dv">8</span>      <span class="cn">NA</span>     <span class="dv">6</span>      <span class="dv">32</span>     <span class="dv">13</span>      <span class="dv">4</span>      <span class="dv">9</span> TGTGCGAGGTGGCTTGGGGAA…</span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a> <span class="dv">9</span>     <span class="dv">83</span>    <span class="fl">0.00103</span> TGTGCGAGAGGCCGCAGCGGCG… CARGRSGDPY… IGHV4<span class="dv">-34</span>  IGHD2<span class="dv">-2</span>, I… IGHJ5      <span class="dv">13</span>      <span class="cn">NA</span>     <span class="dv">4</span>      <span class="dv">50</span>     <span class="dv">18</span>     <span class="dv">13</span>      <span class="dv">5</span> TGTGCGAGAGGCCGCAGCGGC…</span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a><span class="dv">10</span>     <span class="dv">81</span>    <span class="fl">0.00100</span> TGTGTGAGTCACCTCCTCGACA… CVSHLLDTSD… IGHV1<span class="dv">-2</span>   IGHD2<span class="dv">-21</span>, … IGHJ4,…     <span class="dv">8</span>      <span class="cn">NA</span>     <span class="dv">3</span>      <span class="dv">40</span>     <span class="dv">20</span>     <span class="dv">14</span>      <span class="dv">6</span> TGTGTGAGTCACCTCCTCGAC…</span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a><span class="co"># … with 32,734 more rows</span></span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>data<span class="sc">$</span>analysis.clonotypes.IGH_2</span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true" tabindex="-1"></a><span class="co"># A tibble: 32,744 x 15</span></span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true" tabindex="-1"></a>   Clones Proportion CDR3.nt                 CDR3.aa     V.name    D.name      J.name  V.end D.start D.end J.start VJ.ins VD.ins DJ.ins Sequence</span>
<span id="cb7-22"><a href="#cb7-22" aria-hidden="true" tabindex="-1"></a>    <span class="sc">&lt;</span>dbl<span class="sc">&gt;</span>      <span class="er">&lt;</span>dbl<span class="sc">&gt;</span> <span class="er">&lt;</span>chr<span class="sc">&gt;</span>                   <span class="er">&lt;</span>chr<span class="sc">&gt;</span>       <span class="er">&lt;</span>chr<span class="sc">&gt;</span>     <span class="er">&lt;</span>chr<span class="sc">&gt;</span>       <span class="er">&lt;</span>chr<span class="sc">&gt;</span>   <span class="er">&lt;</span>int<span class="sc">&gt;</span>   <span class="er">&lt;</span>int<span class="sc">&gt;</span> <span class="er">&lt;</span>int<span class="sc">&gt;</span>   <span class="er">&lt;</span>int<span class="sc">&gt;</span>  <span class="er">&lt;</span>int<span class="sc">&gt;</span>  <span class="er">&lt;</span>int<span class="sc">&gt;</span>  <span class="er">&lt;</span>int<span class="sc">&gt;</span> <span class="er">&lt;</span>chr<span class="sc">&gt;</span></span>
<span id="cb7-23"><a href="#cb7-23" aria-hidden="true" tabindex="-1"></a> <span class="dv">1</span>    <span class="dv">230</span>    <span class="fl">0.00284</span> TGTGTGAGACATAAACCTATGG… CVRHKPMVQG… IGHV4<span class="dv">-39</span>  IGHD3<span class="dv">-10</span>, … IGHJ6      <span class="dv">12</span>      <span class="cn">NA</span>     <span class="dv">5</span>      <span class="dv">36</span>      <span class="dv">9</span>      <span class="dv">3</span>      <span class="dv">6</span> TGTGTGAGACATAAACCTATG…</span>
<span id="cb7-24"><a href="#cb7-24" aria-hidden="true" tabindex="-1"></a> <span class="dv">2</span>    <span class="dv">201</span>    <span class="fl">0.00248</span> TGTGCGATTTGGGATGTGGGAC… CAIWDVGLRH… IGHV4<span class="dv">-34</span>  IGHD2<span class="dv">-21</span>    IGHJ4,…     <span class="dv">7</span>      <span class="cn">NA</span>     <span class="dv">5</span>      <span class="dv">29</span>     <span class="dv">10</span>      <span class="dv">7</span>      <span class="dv">3</span> TGTGCGATTTGGGATGTGGGA…</span>
<span id="cb7-25"><a href="#cb7-25" aria-hidden="true" tabindex="-1"></a> <span class="dv">3</span>    <span class="dv">179</span>    <span class="fl">0.00221</span> TGTGCGAGAGATCATGCGGGGT… CARDHAGFGK… IGHV1<span class="sc">-</span>69… IGHD3<span class="dv">-10</span>    IGHJ6      <span class="dv">13</span>      <span class="cn">NA</span>     <span class="dv">4</span>      <span class="dv">40</span>     <span class="dv">18</span>      <span class="dv">5</span>     <span class="dv">13</span> TGTGCGAGAGATCATGCGGGG…</span>
<span id="cb7-26"><a href="#cb7-26" aria-hidden="true" tabindex="-1"></a> <span class="dv">4</span>     <span class="dv">99</span>    <span class="fl">0.00122</span> TGTGCGAGATGGGGATATTGTA… CARWGYCING… IGHV4<span class="dv">-39</span>  IGHD2<span class="dv">-8</span>     IGHJ6       <span class="dv">9</span>      <span class="cn">NA</span>     <span class="dv">6</span>      <span class="dv">64</span>     <span class="dv">23</span>      <span class="dv">2</span>     <span class="dv">21</span> TGTGCGAGATGGGGATATTGT…</span>
<span id="cb7-27"><a href="#cb7-27" aria-hidden="true" tabindex="-1"></a> <span class="dv">5</span>     <span class="dv">97</span>    <span class="fl">0.00120</span> TGTGCGAGAGGCCCCACGAGCA… CARGPTSSEW… IGHV4<span class="dv">-34</span>  IGHD3<span class="dv">-22</span>, … IGHJ6      <span class="dv">13</span>      <span class="cn">NA</span>     <span class="dv">6</span>      <span class="dv">52</span>     <span class="dv">26</span>     <span class="dv">24</span>      <span class="dv">2</span> TGTGCGAGAGGCCCCACGAGC…</span>
<span id="cb7-28"><a href="#cb7-28" aria-hidden="true" tabindex="-1"></a> <span class="dv">6</span>     <span class="dv">97</span>    <span class="fl">0.00120</span> TGTGCGCACCACTATACCAGCG… CAHHYTSDYY… IGHV2<span class="dv">-5</span>   IGHD1<span class="dv">-26</span>    IGHJ5       <span class="dv">9</span>      <span class="cn">NA</span>     <span class="dv">2</span>      <span class="dv">39</span>     <span class="dv">19</span>     <span class="cn">NA</span>     <span class="dv">20</span> TGTGCGCACCACTATACCAGC…</span>
<span id="cb7-29"><a href="#cb7-29" aria-hidden="true" tabindex="-1"></a> <span class="dv">7</span>     <span class="dv">92</span>    <span class="fl">0.00114</span> TGTGCGAGAGGCCCTCCGTCGA… CARGPPSMGT… IGHV4<span class="dv">-34</span>  IGHD5<span class="dv">-24</span>, … IGHJ4      <span class="dv">13</span>      <span class="cn">NA</span>     <span class="dv">3</span>      <span class="dv">38</span>     <span class="dv">11</span>      <span class="dv">6</span>      <span class="dv">5</span> TGTGCGAGAGGCCCTCCGTCG…</span>
<span id="cb7-30"><a href="#cb7-30" aria-hidden="true" tabindex="-1"></a> <span class="dv">8</span>     <span class="dv">84</span>    <span class="fl">0.00104</span> TGTGCGAGGTGGCTTGGGGAAG… CARWLGEDIR… IGHV4<span class="dv">-39</span>  IGHD3<span class="dv">-16</span>, … IGHJ4,…     <span class="dv">8</span>      <span class="cn">NA</span>     <span class="dv">6</span>      <span class="dv">32</span>     <span class="dv">13</span>      <span class="dv">4</span>      <span class="dv">9</span> TGTGCGAGGTGGCTTGGGGAA…</span>
<span id="cb7-31"><a href="#cb7-31" aria-hidden="true" tabindex="-1"></a> <span class="dv">9</span>     <span class="dv">83</span>    <span class="fl">0.00103</span> TGTGCGAGAGGCCGCAGCGGCG… CARGRSGDPY… IGHV4<span class="dv">-34</span>  IGHD2<span class="dv">-2</span>, I… IGHJ5      <span class="dv">13</span>      <span class="cn">NA</span>     <span class="dv">4</span>      <span class="dv">50</span>     <span class="dv">18</span>     <span class="dv">13</span>      <span class="dv">5</span> TGTGCGAGAGGCCGCAGCGGC…</span>
<span id="cb7-32"><a href="#cb7-32" aria-hidden="true" tabindex="-1"></a><span class="dv">10</span>     <span class="dv">81</span>    <span class="fl">0.00100</span> TGTGTGAGTCACCTCCTCGACA… CVSHLLDTSD… IGHV1<span class="dv">-2</span>   IGHD2<span class="dv">-21</span>, … IGHJ4,…     <span class="dv">8</span>      <span class="cn">NA</span>     <span class="dv">3</span>      <span class="dv">40</span>     <span class="dv">20</span>     <span class="dv">14</span>      <span class="dv">6</span> TGTGTGAGTCACCTCCTCGAC…</span>
<span id="cb7-33"><a href="#cb7-33" aria-hidden="true" tabindex="-1"></a><span class="co"># … with 32,734 more rows</span></span>
<span id="cb7-34"><a href="#cb7-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-35"><a href="#cb7-35" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>data<span class="sc">$</span>analysis.clonotypes.IGH_3</span>
<span id="cb7-36"><a href="#cb7-36" aria-hidden="true" tabindex="-1"></a><span class="co"># A tibble: 32,744 x 15</span></span>
<span id="cb7-37"><a href="#cb7-37" aria-hidden="true" tabindex="-1"></a>   Clones Proportion CDR3.nt                 CDR3.aa     V.name    D.name      J.name  V.end D.start D.end J.start VJ.ins VD.ins DJ.ins Sequence</span>
<span id="cb7-38"><a href="#cb7-38" aria-hidden="true" tabindex="-1"></a>    <span class="sc">&lt;</span>dbl<span class="sc">&gt;</span>      <span class="er">&lt;</span>dbl<span class="sc">&gt;</span> <span class="er">&lt;</span>chr<span class="sc">&gt;</span>                   <span class="er">&lt;</span>chr<span class="sc">&gt;</span>       <span class="er">&lt;</span>chr<span class="sc">&gt;</span>     <span class="er">&lt;</span>chr<span class="sc">&gt;</span>       <span class="er">&lt;</span>chr<span class="sc">&gt;</span>   <span class="er">&lt;</span>int<span class="sc">&gt;</span>   <span class="er">&lt;</span>int<span class="sc">&gt;</span> <span class="er">&lt;</span>int<span class="sc">&gt;</span>   <span class="er">&lt;</span>int<span class="sc">&gt;</span>  <span class="er">&lt;</span>int<span class="sc">&gt;</span>  <span class="er">&lt;</span>int<span class="sc">&gt;</span>  <span class="er">&lt;</span>int<span class="sc">&gt;</span> <span class="er">&lt;</span>chr<span class="sc">&gt;</span></span>
<span id="cb7-39"><a href="#cb7-39" aria-hidden="true" tabindex="-1"></a> <span class="dv">1</span>    <span class="dv">230</span>    <span class="fl">0.00284</span> TGTGTGAGACATAAACCTATGG… CVRHKPMVQG… IGHV4<span class="dv">-39</span>  IGHD3<span class="dv">-10</span>, … IGHJ6      <span class="dv">12</span>      <span class="cn">NA</span>     <span class="dv">5</span>      <span class="dv">36</span>      <span class="dv">9</span>      <span class="dv">3</span>      <span class="dv">6</span> TGTGTGAGACATAAACCTATG…</span>
<span id="cb7-40"><a href="#cb7-40" aria-hidden="true" tabindex="-1"></a> <span class="dv">2</span>    <span class="dv">201</span>    <span class="fl">0.00248</span> TGTGCGATTTGGGATGTGGGAC… CAIWDVGLRH… IGHV4<span class="dv">-34</span>  IGHD2<span class="dv">-21</span>    IGHJ4,…     <span class="dv">7</span>      <span class="cn">NA</span>     <span class="dv">5</span>      <span class="dv">29</span>     <span class="dv">10</span>      <span class="dv">7</span>      <span class="dv">3</span> TGTGCGATTTGGGATGTGGGA…</span>
<span id="cb7-41"><a href="#cb7-41" aria-hidden="true" tabindex="-1"></a> <span class="dv">3</span>    <span class="dv">179</span>    <span class="fl">0.00221</span> TGTGCGAGAGATCATGCGGGGT… CARDHAGFGK… IGHV1<span class="sc">-</span>69… IGHD3<span class="dv">-10</span>    IGHJ6      <span class="dv">13</span>      <span class="cn">NA</span>     <span class="dv">4</span>      <span class="dv">40</span>     <span class="dv">18</span>      <span class="dv">5</span>     <span class="dv">13</span> TGTGCGAGAGATCATGCGGGG…</span>
<span id="cb7-42"><a href="#cb7-42" aria-hidden="true" tabindex="-1"></a> <span class="dv">4</span>     <span class="dv">99</span>    <span class="fl">0.00122</span> TGTGCGAGATGGGGATATTGTA… CARWGYCING… IGHV4<span class="dv">-39</span>  IGHD2<span class="dv">-8</span>     IGHJ6       <span class="dv">9</span>      <span class="cn">NA</span>     <span class="dv">6</span>      <span class="dv">64</span>     <span class="dv">23</span>      <span class="dv">2</span>     <span class="dv">21</span> TGTGCGAGATGGGGATATTGT…</span>
<span id="cb7-43"><a href="#cb7-43" aria-hidden="true" tabindex="-1"></a> <span class="dv">5</span>     <span class="dv">97</span>    <span class="fl">0.00120</span> TGTGCGAGAGGCCCCACGAGCA… CARGPTSSEW… IGHV4<span class="dv">-34</span>  IGHD3<span class="dv">-22</span>, … IGHJ6      <span class="dv">13</span>      <span class="cn">NA</span>     <span class="dv">6</span>      <span class="dv">52</span>     <span class="dv">26</span>     <span class="dv">24</span>      <span class="dv">2</span> TGTGCGAGAGGCCCCACGAGC…</span>
<span id="cb7-44"><a href="#cb7-44" aria-hidden="true" tabindex="-1"></a> <span class="dv">6</span>     <span class="dv">97</span>    <span class="fl">0.00120</span> TGTGCGCACCACTATACCAGCG… CAHHYTSDYY… IGHV2<span class="dv">-5</span>   IGHD1<span class="dv">-26</span>    IGHJ5       <span class="dv">9</span>      <span class="cn">NA</span>     <span class="dv">2</span>      <span class="dv">39</span>     <span class="dv">19</span>     <span class="cn">NA</span>     <span class="dv">20</span> TGTGCGCACCACTATACCAGC…</span>
<span id="cb7-45"><a href="#cb7-45" aria-hidden="true" tabindex="-1"></a> <span class="dv">7</span>     <span class="dv">92</span>    <span class="fl">0.00114</span> TGTGCGAGAGGCCCTCCGTCGA… CARGPPSMGT… IGHV4<span class="dv">-34</span>  IGHD5<span class="dv">-24</span>, … IGHJ4      <span class="dv">13</span>      <span class="cn">NA</span>     <span class="dv">3</span>      <span class="dv">38</span>     <span class="dv">11</span>      <span class="dv">6</span>      <span class="dv">5</span> TGTGCGAGAGGCCCTCCGTCG…</span>
<span id="cb7-46"><a href="#cb7-46" aria-hidden="true" tabindex="-1"></a> <span class="dv">8</span>     <span class="dv">84</span>    <span class="fl">0.00104</span> TGTGCGAGGTGGCTTGGGGAAG… CARWLGEDIR… IGHV4<span class="dv">-39</span>  IGHD3<span class="dv">-16</span>, … IGHJ4,…     <span class="dv">8</span>      <span class="cn">NA</span>     <span class="dv">6</span>      <span class="dv">32</span>     <span class="dv">13</span>      <span class="dv">4</span>      <span class="dv">9</span> TGTGCGAGGTGGCTTGGGGAA…</span>
<span id="cb7-47"><a href="#cb7-47" aria-hidden="true" tabindex="-1"></a> <span class="dv">9</span>     <span class="dv">83</span>    <span class="fl">0.00103</span> TGTGCGAGAGGCCGCAGCGGCG… CARGRSGDPY… IGHV4<span class="dv">-34</span>  IGHD2<span class="dv">-2</span>, I… IGHJ5      <span class="dv">13</span>      <span class="cn">NA</span>     <span class="dv">4</span>      <span class="dv">50</span>     <span class="dv">18</span>     <span class="dv">13</span>      <span class="dv">5</span> TGTGCGAGAGGCCGCAGCGGC…</span>
<span id="cb7-48"><a href="#cb7-48" aria-hidden="true" tabindex="-1"></a><span class="dv">10</span>     <span class="dv">81</span>    <span class="fl">0.00100</span> TGTGTGAGTCACCTCCTCGACA… CVSHLLDTSD… IGHV1<span class="dv">-2</span>   IGHD2<span class="dv">-21</span>, … IGHJ4,…     <span class="dv">8</span>      <span class="cn">NA</span>     <span class="dv">3</span>      <span class="dv">40</span>     <span class="dv">20</span>     <span class="dv">14</span>      <span class="dv">6</span> TGTGTGAGTCACCTCCTCGAC…</span>
<span id="cb7-49"><a href="#cb7-49" aria-hidden="true" tabindex="-1"></a><span class="co"># … with 32,734 more rows</span></span>
<span id="cb7-50"><a href="#cb7-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-51"><a href="#cb7-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-52"><a href="#cb7-52" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>meta</span>
<span id="cb7-53"><a href="#cb7-53" aria-hidden="true" tabindex="-1"></a><span class="co"># A tibble: 3 x 4</span></span>
<span id="cb7-54"><a href="#cb7-54" aria-hidden="true" tabindex="-1"></a>  Sample                    Sex     Age Status</span>
<span id="cb7-55"><a href="#cb7-55" aria-hidden="true" tabindex="-1"></a>  <span class="sc">&lt;</span>chr<span class="sc">&gt;</span>                     <span class="er">&lt;</span>chr<span class="sc">&gt;</span> <span class="er">&lt;</span>dbl<span class="sc">&gt;</span> <span class="er">&lt;</span>chr<span class="sc">&gt;</span></span>
<span id="cb7-56"><a href="#cb7-56" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span> analysis.clonotypes.IGH_1 M         <span class="dv">1</span> C</span>
<span id="cb7-57"><a href="#cb7-57" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span> analysis.clonotypes.IGH_2 M         <span class="dv">2</span> C</span>
<span id="cb7-58"><a href="#cb7-58" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span> analysis.clonotypes.IGH_3 F         <span class="dv">3</span> A</span></code></pre></div>
<p>Congrats! Now your data is ready for exploration. Follow the steps <a href="https://immunarch.com/articles/v3_basic_analysis.html">here</a> to learn more about how to explore your dataset.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://www.linkedin.com/in/vdnaz" class="external-link">Vadim I. Nazarov</a>, <a href="https://www.linkedin.com/in/vasily-tsvetkov-227218ab" class="external-link">Vasily O. Tsvetkov</a>, <a href="https://www.linkedin.com/in/erumynskiy/" class="external-link">Eugene Rumynskiy</a>, <a href="https://www.linkedin.com/in/aleksandr-popov-634878105/" class="external-link">Aleksandr A. Popov</a>, <a href="https://www.linkedin.com/in/ivan-balashov/" class="external-link">Ivan Balashov</a>, <a href="https://www.linkedin.com/in/maria-volobueva-b0856a223/" class="external-link">Maria Samokhina</a>, <a href="https://immunomind.io" class="external-link">ImmunoMind</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '94207562f34b455c0796790d104e3549',
    indexName: 'immunarch',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
