<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Loading 10x Genomics Data • immunarch</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><!-- docsearch --><script src="../../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><link href="../../extra.css" rel="stylesheet">
<meta property="og:title" content="Loading 10x Genomics Data">
<meta property="og:description" content="immunarch">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-128418614-2"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-128418614-2');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">immunarch</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.7.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Start Here
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/v1_introduction.html">Installation &amp; troubleshooting</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">First steps</li>
    <li>
      <a href="../../articles/v2_data.html">Data loading</a>
    </li>
    <li>
      <a href="../../articles/web_only/load_mixcr.html">How-to: Loading MiXCR Data</a>
    </li>
    <li>
      <a href="../../articles/web_only/load_10x.html">How-to: Loading 10x Genomics Data</a>
    </li>
    <li>
      <a href="../../articles/web_only/v21_singlecell.html">How-to: Single-cell and paired chain data</a>
    </li>
    <li>
      <a href="../../articles/web_only/v3_basic_analysis.html">Basic statistics and clonality</a>
    </li>
    <li>
      <a href="../../articles/web_only/repFilter_v3.html">Data filtering</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Repertoire-level exploration and comparison</li>
    <li>
      <a href="../../articles/web_only/v4_overlap.html">Repertoire overlap and public clonotypes</a>
    </li>
    <li>
      <a href="../../articles/web_only/v5_gene_usage.html">Gene usage</a>
    </li>
    <li>
      <a href="../../articles/web_only/v6_diversity.html">Diversity estimation</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Clonotype-level exploration and comparison</li>
    <li>
      <a href="../../articles/web_only/v8_tracking.html">Track clonotypes across samples and time</a>
    </li>
    <li>
      <a href="../../articles/web_only/v11_db.html">Annotate clonotypes using immune receptor databases</a>
    </li>
    <li>
      <a href="../../articles/web_only/v9_kmers.html">Kmer and sequence motif analysis and visualisation</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Preparing to publication</li>
    <li>
      <a href="../../articles/web_only/v7_fixvis.html">Make your plots publication-ready with fixVis</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">BCR analysis</li>
    <li>
      <a href="../../articles/web_only/BCRpipeline.html">BCR pipeline</a>
    </li>
    <li>
      <a href="../../articles/web_only/clustering.html">Clustering</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
<li>
  <a href="https://github.com/immunomind/covid19" class="external-link">COVID-19</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://immunomind.io" class="external-link">ImmunoMind</a>
</li>
<li>
  <a href="http://twitter.com/immunomind" class="external-link">
    <span class="fa fa-lg fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/immunomind/immunarch" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="load_10x_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Loading 10x Genomics Data</h1>
                        <h4 data-toc-skip class="author">
<b>ImmunoMind</b> – improving design of T-cell therapies using multi-omics and AI. Research and biopharma partnerships, more details: <a href="https://immunomind.io" class="external-link">immunomind.io</a>
</h4>
            
            <h4 data-toc-skip class="date"><a href="mailto:support@immunomind.io" class="email">support@immunomind.io</a></h4>
      
      <small class="dont-index">Source: <a href="https://github.com/immunomind/immunarch/blob/HEAD/vignettes/web_only/load_10x.Rmd" class="external-link"><code>vignettes/web_only/load_10x.Rmd</code></a></small>
      <div class="hidden name"><code>load_10x.Rmd</code></div>

    </div>

    
    
<!--
%\VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{Loading 10x Genomics Data}
%\VignettePackage{immunarch}
-->
<div class="section level2">
<h2 id="intro-to-10x-genomics">Intro to 10x Genomics<a class="anchor" aria-label="anchor" href="#intro-to-10x-genomics"></a>
</h2>
<p>10x Genomics has various pipelines for single cell and spatial views of biological systems, including single cell immune profiling. The 10x Genomics Chromium Single Cell Immune Profiling Solution enables simultaneous analysis of the following:</p>
<ul>
<li>V(D)J transcripts and clonotypes for T and B cells.</li>
<li>5’ gene expression.</li>
<li>Cell surface proteins/antigen specificity (feature barcodes) at single-cell resolution for the same set of cells.</li>
</ul>
<p>Their end-to-end pipeline also includes the <a href="https://support.10xgenomics.com/single-cell-vdj/software/pipelines/latest/what-is-cell-ranger" class="external-link">Cell Ranger</a> software, which include the following pipelines for Immune profiling analysis:</p>
<p><code>cellranger mkfastq</code> demultiplexes raw base call (BCL) files generated by Illumina sequencers into FASTQ files. It is a wrapper around Illumina’s bcl2fastq, with additional useful features that are specific to 10x libraries and a simplified sample sheet format.</p>
<p><code>cellranger vdj</code> takes FASTQ files from cellranger mkfastq for V(D)J libraries and performs sequence assembly and paired clonotype calling. It uses the Chromium cellular barcodes and UMIs to assemble V(D)J transcripts per cell. Clonotypes and CDR3 sequences are output as a .vloupe file which can be loaded into Loupe V(D)J Browser.</p>
<p><code>cellranger count</code> takes FASTQ files for 5’ Gene Expression and/or Feature Barcode (cell surface protein or antigen) libraries and performs alignment, filtering, barcode counting, and UMI counting. It uses the Chromium cellular barcodes to generate feature-barcode matrices, determine clusters, and perform gene expression analysis. The cellranger count pipeline outputs a .cloupe file which can be loaded into Loupe Browser for interactive visualization, clustering, and differential expression analysis.</p>
</div>
<div class="section level2">
<h2 id="cell-ranger-data-pipeline">Cell Ranger Data Pipeline<a class="anchor" aria-label="anchor" href="#cell-ranger-data-pipeline"></a>
</h2>
<p>Follow the instructions <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger" class="external-link">here</a> to use Cellranger on your data. Note that Cellranger is currently only supported by Linux operating systems. The <code>cellranger count</code> and <code>cellranger vdj</code> methods will be particularly useful.</p>
<p>You can find sample data of full runs to download <a href="https://www.10xgenomics.com/resources/datasets?menu%5Bproducts.name%5D=Single+Cell+Immune+Profiling" class="external-link">here</a>. In this tutorial, we use <a href="https://support.10xgenomics.com/single-cell-vdj/datasets/3.0.0/vdj_v1_mm_c57bl6_pbmc_t" class="external-link">this</a> single cell mouse data. If you are using <code>immunarch</code> you can download only the .csv files.</p>
</div>
<div class="section level2">
<h2 id="prepare-10x-data">Prepare 10x Data<a class="anchor" aria-label="anchor" href="#prepare-10x-data"></a>
</h2>
<p>Upon processing the data, you will have a lot of output files. You should use the <code>filtered contigs</code> csv files because they contain barcode information.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true"></a>.</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true"></a>├── vdj_v1_mm_c57bl6_pbmc_t_filtered_contig_annotations.csv &lt;-<span class="op">-</span><span class="st"> </span>This contains the count data we want<span class="op">!</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true"></a>├── vdj_v1_mm_c57bl6_pbmc_t_consensus_annotations.csv</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true"></a>├── vdj_v1_mm_c57bl6_pbmc_t_clonotypes.csv</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true"></a>├── vdj_v1_mm_c57bl6_pbmc_t_all_contig_annotations.csv</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true"></a>├── vdj_v1_mm_c57bl6_pbmc_t_matrix.h5</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true"></a>├── vdj_v1_mm_c57bl6_pbmc_t_bam.bam.bai</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true"></a>├── vdj_v1_mm_c57bl6_pbmc_t_molecule_info.h5</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true"></a>├── vdj_v1_mm_c57bl6_pbmc_t_raw_feature_bc_matrix.tar.gz</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true"></a>├── vdj_v1_mm_c57bl6_pbmc_t_analysis.tar.gz</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="load-into-immunarch">Load into Immunarch<a class="anchor" aria-label="anchor" href="#load-into-immunarch"></a>
</h2>
<p>Run the code below in your R environment to load the data into Immunarch’s format. You can run it on the entire folder with the Cellranger output files. <code>repLoad</code> will ignore the file formats that are unsupported.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true"></a><span class="co"># 1.1) Load the package into R:</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true"></a><span class="op">&gt;</span><span class="st"> </span><span class="kw">library</span>(immunarch)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true"></a><span class="co"># 1.2) Replace with the path to your processed 10x data or to the clonotypes file</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true"></a><span class="op">&gt;</span><span class="st"> </span>file_path =<span class="st"> "~/path/to/your/cellranger/data/"</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true"></a></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true"></a><span class="co"># 1.3) Load 10x data with repLoad</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true"></a><span class="op">&gt;</span><span class="st"> </span>immdata_10x &lt;-<span class="st"> </span><span class="kw">repLoad</span>(file_path)</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true"></a></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true"></a><span class="op">==</span><span class="st"> </span>Step <span class="dv">1</span><span class="op">/</span><span class="dv">3</span><span class="op">:</span><span class="st"> </span>loading repertoire files... <span class="op">==</span></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true"></a></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true"></a>Processing <span class="st">"/filepath/C57BL_mice_igenrichment"</span> ...</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true"></a>  <span class="op">--</span><span class="st"> </span>Parsing <span class="st">"/filepath/vdj_v1_mm_c57bl6_pbmc_t_all_contig_annotations.csv"</span> <span class="op">--</span><span class="st"> </span>10x (filt.contigs)</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true"></a>  [<span class="op">!</span>] Removed <span class="dv">2917</span> clonotypes with no nucleotide and amino acid CDR3 sequence.</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true"></a>  <span class="op">--</span><span class="st"> </span>Parsing <span class="st">"/filepath/vdj_v1_mm_c57bl6_pbmc_t_clonotypes.csv"</span> <span class="op">--</span><span class="st"> </span>unsupported format, skipping</span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true"></a>  <span class="op">--</span><span class="st"> </span>Parsing <span class="st">"/filepath/vdj_v1_mm_c57bl6_pbmc_t_consensus_annotations.csv"</span> <span class="op">--</span><span class="st"> </span>10x (consensus)</span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true"></a>  <span class="op">--</span><span class="st"> </span>Parsing <span class="st">"/filepath/vdj_v1_mm_c57bl6_pbmc_t_filtered_contig_annotations.csv"</span> <span class="op">--</span><span class="st"> </span>10x (filt.contigs)</span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true"></a>  [<span class="op">!</span>] Removed <span class="dv">1198</span> clonotypes with no nucleotide and amino acid CDR3 sequence.</span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true"></a></span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true"></a><span class="op">==</span><span class="st"> </span>Step <span class="dv">2</span><span class="op">/</span><span class="dv">3</span><span class="op">:</span><span class="st"> </span>checking metadata files and merging... <span class="op">==</span></span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true"></a></span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true"></a>Processing <span class="st">"&lt;initial&gt;"</span> ...</span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true"></a>  <span class="op">--</span><span class="st"> </span>Metadata file not found; creating a dummy metadata...</span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true"></a></span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true"></a><span class="op">==</span><span class="st"> </span>Step <span class="dv">3</span><span class="op">/</span><span class="dv">3</span><span class="op">:</span><span class="st"> </span>splitting data by barcodes and chains... <span class="op">==</span></span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true"></a></span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true"></a>Done<span class="op">!</span></span></code></pre></div>
<p>Now let’s take a look at the data! Your output should look something like below.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true"></a><span class="op">&gt;</span><span class="st"> </span>immdata_10x</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true"></a><span class="op">$</span>data<span class="op">$</span>vdj_v1_mm_c57bl6_splenocytes_t_consensus_annotations_TRA</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true"></a><span class="co"># A tibble: 710 x 17</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true"></a>   Clones Proportion CDR3.nt      CDR3.aa  V.name  D.name J.name V.end D.start D.end J.start VJ.ins VD.ins DJ.ins chain ClonotypeID  ConsensusID</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true"></a>    <span class="op">&lt;</span>dbl<span class="op">&gt;</span><span class="st">      </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span><span class="st"> </span><span class="er">&lt;</span>chr<span class="op">&gt;</span><span class="st">        </span><span class="er">&lt;</span>chr<span class="op">&gt;</span><span class="st">    </span><span class="er">&lt;</span>chr<span class="op">&gt;</span><span class="st">   </span><span class="er">&lt;</span>chr<span class="op">&gt;</span><span class="st">  </span><span class="er">&lt;</span>chr<span class="op">&gt;</span><span class="st">  </span><span class="er">&lt;</span>int<span class="op">&gt;</span><span class="st">   </span><span class="er">&lt;</span>int<span class="op">&gt;</span><span class="st"> </span><span class="er">&lt;</span>int<span class="op">&gt;</span><span class="st">   </span><span class="er">&lt;</span>int<span class="op">&gt;</span><span class="st">  </span><span class="er">&lt;</span>int<span class="op">&gt;</span><span class="st">  </span><span class="er">&lt;</span>int<span class="op">&gt;</span><span class="st">  </span><span class="er">&lt;</span>int<span class="op">&gt;</span><span class="st"> </span><span class="er">&lt;</span>chr<span class="op">&gt;</span><span class="st"> </span><span class="er">&lt;</span>chr<span class="op">&gt;</span><span class="st">        </span><span class="er">&lt;</span>chr<span class="op">&gt;</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true"></a><span class="st"> </span><span class="dv">1</span>     <span class="dv">55</span>    <span class="fl">0.00414</span> TGTGCTATGGC… CAMATGG… TRAV13… None   TRAJ56    <span class="ot">NA</span>      <span class="ot">NA</span>    <span class="ot">NA</span>      <span class="ot">NA</span>     <span class="ot">NA</span>     <span class="ot">NA</span>     <span class="ot">NA</span> TRA   clonotype306 clonotype30…</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true"></a> <span class="dv">2</span>     <span class="dv">55</span>    <span class="fl">0.00414</span> TGTGCAGCTAG… CAASGNT… TRAV7<span class="dv">-4</span> None   TRAJ27    <span class="ot">NA</span>      <span class="ot">NA</span>    <span class="ot">NA</span>      <span class="ot">NA</span>     <span class="ot">NA</span>     <span class="ot">NA</span>     <span class="ot">NA</span> TRA   clonotype338 clonotype33…</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true"></a> <span class="dv">3</span>     <span class="dv">53</span>    <span class="fl">0.00399</span> TGTGCAGCAAG… CAARDSG… TRAV14… None   TRAJ11    <span class="ot">NA</span>      <span class="ot">NA</span>    <span class="ot">NA</span>      <span class="ot">NA</span>     <span class="ot">NA</span>     <span class="ot">NA</span>     <span class="ot">NA</span> TRA   clonotype617 clonotype61…</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true"></a> <span class="dv">4</span>     <span class="dv">45</span>    <span class="fl">0.00339</span> TGCGCAGTCAG… CAVSNNT… TRAV3<span class="dv">-3</span> None   TRAJ27    <span class="ot">NA</span>      <span class="ot">NA</span>    <span class="ot">NA</span>      <span class="ot">NA</span>     <span class="ot">NA</span>     <span class="ot">NA</span>     <span class="ot">NA</span> TRA   clonotype435 clonotype43…</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true"></a> <span class="dv">5</span>     <span class="dv">43</span>    <span class="fl">0.00324</span> TGTGCAGTCAG… CAVSNMG… TRAV7D… None   TRAJ9     <span class="ot">NA</span>      <span class="ot">NA</span>    <span class="ot">NA</span>      <span class="ot">NA</span>     <span class="ot">NA</span>     <span class="ot">NA</span>     <span class="ot">NA</span> TRA   clonotype401 clonotype40…</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true"></a> <span class="dv">6</span>     <span class="dv">42</span>    <span class="fl">0.00316</span> TGTGCAGCAAG… CAASPNY… TRAV14… None   TRAJ21    <span class="ot">NA</span>      <span class="ot">NA</span>    <span class="ot">NA</span>      <span class="ot">NA</span>     <span class="ot">NA</span>     <span class="ot">NA</span>     <span class="ot">NA</span> TRA   clonotype5   clonotype5_…</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true"></a> <span class="dv">7</span>     <span class="dv">37</span>    <span class="fl">0.00279</span> TGTGCAGTGAG… CAVSSGG… TRAV7D… None   TRAJ6     <span class="ot">NA</span>      <span class="ot">NA</span>    <span class="ot">NA</span>      <span class="ot">NA</span>     <span class="ot">NA</span>     <span class="ot">NA</span>     <span class="ot">NA</span> TRA   clonotype453 clonotype45…</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true"></a> <span class="dv">8</span>     <span class="dv">35</span>    <span class="fl">0.00264</span> TGTGCAGCAAG… CAASATS… TRAV14… None   TRAJ22    <span class="ot">NA</span>      <span class="ot">NA</span>    <span class="ot">NA</span>      <span class="ot">NA</span>     <span class="ot">NA</span>     <span class="ot">NA</span>     <span class="ot">NA</span> TRA   clonotype809 clonotype80…</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true"></a> <span class="dv">9</span>     <span class="dv">32</span>    <span class="fl">0.00241</span> TGTGCAGCAAG… CAASPNY… TRAV14… None   TRAJ21    <span class="ot">NA</span>      <span class="ot">NA</span>    <span class="ot">NA</span>      <span class="ot">NA</span>     <span class="ot">NA</span>     <span class="ot">NA</span>     <span class="ot">NA</span> TRA   clonotype150 clonotype15…</span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true"></a><span class="dv">10</span>     <span class="dv">32</span>    <span class="fl">0.00241</span> TGTGCTCTGGG… CALGDEA… TRAV6<span class="op">-</span>… None   TRAJ30    <span class="ot">NA</span>      <span class="ot">NA</span>    <span class="ot">NA</span>      <span class="ot">NA</span>     <span class="ot">NA</span>     <span class="ot">NA</span>     <span class="ot">NA</span> TRA   clonotype393 clonotype39…</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true"></a><span class="co"># … with 700 more rows</span></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true"></a></span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true"></a><span class="op">$</span>meta</span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true"></a>                                                       Sample Chain                                                Source</span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true"></a><span class="dv">1</span> vdj_v1_mm_c57bl6_splenocytes_t_all_contig_annotations_Multi Multi vdj_v1_mm_c57bl6_splenocytes_t_all_contig_annotations</span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true"></a><span class="dv">2</span>   vdj_v1_mm_c57bl6_splenocytes_t_all_contig_annotations_TRA   TRA vdj_v1_mm_c57bl6_splenocytes_t_all_contig_annotations</span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true"></a><span class="dv">3</span>   vdj_v1_mm_c57bl6_splenocytes_t_all_contig_annotations_TRB   TRB vdj_v1_mm_c57bl6_splenocytes_t_all_contig_annotations</span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true"></a><span class="dv">5</span>    vdj_v1_mm_c57bl6_splenocytes_t_consensus_annotations_TRA   TRA  vdj_v1_mm_c57bl6_splenocytes_t_consensus_annotations</span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true"></a><span class="dv">6</span>    vdj_v1_mm_c57bl6_splenocytes_t_consensus_annotations_TRB   TRB  vdj_v1_mm_c57bl6_splenocytes_t_consensus_annotations</span></code></pre></div>
<p>Congrats! Now your data is ready for exploration. Follow the steps <a href="https://immunarch.com/articles/v3_basic_analysis.html">here</a> to learn more about how to explore your dataset.</p>
<div class="section level3">
<h3 id="note-on-barcodes">Note on barcodes<a class="anchor" aria-label="anchor" href="#note-on-barcodes"></a>
</h3>
<p>Another important note is that some of the contigs files lack a column for barcodes – a unique identifier of any cell.</p>
<p>These files can be useful for analysis of single chain data (only alpha, or beta TCRs), but in order to analyze paired-chain data and fully utilize the full power of single-cell technologies, you should upload the file with barcodes to the Immunarch.</p>
</div>
</div>
<div class="section level2">
<h2 id="paired-chain-data">Paired-chain data<a class="anchor" aria-label="anchor" href="#paired-chain-data"></a>
</h2>
<p>Do you have single-cell data? Immunarch can now parse both single-chain and paired-chain single-cell data. Try out these features <a href="https://immunarch.com/articles/web_only/v21_singlecell.html">here</a>.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://www.linkedin.com/in/vdnaz" class="external-link">Vadim I. Nazarov</a>, <a href="https://www.linkedin.com/in/vasily-tsvetkov-227218ab" class="external-link">Vasily O. Tsvetkov</a>, <a href="https://www.linkedin.com/in/erumynskiy/" class="external-link">Eugene Rumynskiy</a>, <a href="https://www.linkedin.com/in/aleksandr-popov-634878105/" class="external-link">Aleksandr A. Popov</a>, <a href="https://www.linkedin.com/in/ivan-balashov/" class="external-link">Ivan Balashov</a>, <a href="https://www.linkedin.com/in/maria-volobueva-b0856a223/" class="external-link">Maria Samokhina</a>, <a href="https://immunomind.io" class="external-link">ImmunoMind</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '94207562f34b455c0796790d104e3549',
    indexName: 'immunarch',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
