<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Annotate clonotypes using immune receptor databases • immunarch</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><!-- docsearch --><script src="../../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><link href="../../extra.css" rel="stylesheet">
<meta property="og:title" content="Annotate clonotypes using immune receptor databases">
<meta property="og:description" content="immunarch">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-128418614-2"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-128418614-2');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">immunarch</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.9.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Start Here
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/v1_introduction.html">Installation &amp; troubleshooting</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">First steps</li>
    <li>
      <a href="../../articles/v2_data.html">Data loading</a>
    </li>
    <li>
      <a href="../../articles/web_only/load_mixcr.html">How-to: Loading MiXCR Data</a>
    </li>
    <li>
      <a href="../../articles/web_only/load_10x.html">How-to: Loading 10x Genomics Data</a>
    </li>
    <li>
      <a href="../../articles/web_only/v21_singlecell.html">How-to: Single-cell and paired chain data</a>
    </li>
    <li>
      <a href="../../articles/web_only/v3_basic_analysis.html">Basic statistics and clonality</a>
    </li>
    <li>
      <a href="../../articles/web_only/repFilter_v3.html">Data filtering</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Repertoire-level exploration and comparison</li>
    <li>
      <a href="../../articles/web_only/v4_overlap.html">Repertoire overlap and public clonotypes</a>
    </li>
    <li>
      <a href="../../articles/web_only/v5_gene_usage.html">Gene usage</a>
    </li>
    <li>
      <a href="../../articles/web_only/v6_diversity.html">Diversity estimation</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Clonotype-level exploration and comparison</li>
    <li>
      <a href="../../articles/web_only/v8_tracking.html">Track clonotypes across samples and time</a>
    </li>
    <li>
      <a href="../../articles/web_only/v11_db.html">Annotate clonotypes using immune receptor databases</a>
    </li>
    <li>
      <a href="../../articles/web_only/v9_kmers.html">Kmer and sequence motif analysis and visualisation</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Preparing to publication</li>
    <li>
      <a href="../../articles/web_only/v7_fixvis.html">Make your plots publication-ready with fixVis</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">BCR analysis</li>
    <li>
      <a href="../../articles/web_only/BCRpipeline.html">BCR pipeline</a>
    </li>
    <li>
      <a href="../../articles/web_only/clustering.html">Clustering</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
<li>
  <a href="https://github.com/immunomind/covid19" class="external-link">COVID-19</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://immunomind.io" class="external-link">ImmunoMind</a>
</li>
<li>
  <a href="http://twitter.com/immunomind" class="external-link">
    <span class="fa fa-lg fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/immunomind/immunarch" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="v11_db_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Annotate clonotypes using immune receptor databases</h1>
                        <h4 data-toc-skip class="author">
<b>ImmunoMind</b> – improving design of T-cell therapies using multi-omics and AI. Research and biopharma partnerships, more details: <a href="https://immunomind.io" class="external-link">immunomind.io</a>
</h4>
            
            <h4 data-toc-skip class="date"><a href="mailto:support@immunomind.io" class="email">support@immunomind.io</a></h4>
      
      <small class="dont-index">Source: <a href="https://github.com/immunomind/immunarch/blob/HEAD/vignettes/web_only/v11_db.Rmd" class="external-link"><code>vignettes/web_only/v11_db.Rmd</code></a></small>
      <div class="hidden name"><code>v11_db.Rmd</code></div>

    </div>

    
    
<!--
%\VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{Annotate clonotypes using immune receptor databases}
%\VignettePackage{immunarch}
-->
<div class="section level2">
<h2 id="introduction-to-immune-receptor-databases">Introduction to immune receptor databases<a class="anchor" aria-label="anchor" href="#introduction-to-immune-receptor-databases"></a>
</h2>
<p>Databases with aggregated information about immune receptor specificity provide a straightforward way to annotate your data and find condition-associated receptors. <code>immunarch</code> supports the tools to annotate your data using the most popular AIRR databases - VDJDB, McPAS-TCR and PIRD TBAdb.</p>
<p>Database annotation is a two-step process. First, you need to download database files - either the full database or filtered data obtained from the web interface of the database. After that, you can use <code>immunarch</code> functions to annotate your data and visualise the results. Below you can find a guide for annotation covering both steps.</p>
</div>
<div class="section level2">
<h2 id="downloading-databases">Downloading databases<a class="anchor" aria-label="anchor" href="#downloading-databases"></a>
</h2>
<div class="section level3">
<h3 id="vdjdb">VDJDB<a class="anchor" aria-label="anchor" href="#vdjdb"></a>
</h3>
<p>VDJDB is a curated database of T-cell receptor sequences of known antigen specificity. The database is GitHub-based and available here: <a href="https://github.com/antigenomics/vdjdb-db" class="external-link uri">https://github.com/antigenomics/vdjdb-db</a></p>
<p>Citation: <em>Shugay M et al. VDJdb: a curated database of T-cell receptor sequences with known antigen specificity. Nucleic Acids Research 2017</em></p>
<div class="section level4">
<h4 id="how-to-filter-and-download-data">How to filter and download data<a class="anchor" aria-label="anchor" href="#how-to-filter-and-download-data"></a>
</h4>
<p>It can be useful to filter out immune receptors that are not relevant from the database before working with it. For instance, if you analyse human T-cell beta repertoires, it is not necessary to keep immune receptors from other species, as well as non-TRB data. Use the web interface to VDJDB located at <a href="https://vdjdb.cdr3.net/search" class="external-link uri">https://vdjdb.cdr3.net/search</a> to filter out data. Having filtered the data and pressed the “Refresh table” button, locate the “Export” button and select the “TSV” label inside. It will start the downloading of the filtered database file with a name like “SearchTable-2019-10-17 12_36_11.989.tsv”, which can be used for annotation with <code>immunarch</code>.</p>
</div>
<div class="section level4">
<h4 id="how-to-download-full-vdjdb">How to download full VDJDB<a class="anchor" aria-label="anchor" href="#how-to-download-full-vdjdb"></a>
</h4>
<p>You can use the previous method to download the full database if you set all check marks in the “General” section of the “CDR3” tab. However, if you want to download the raw database files, here is the step by step guide to the rather complicated process of VDJDB downloading and unpacking.</p>
<ol style="list-style-type: decimal">
<li><p>First, you need to install JDK 8 - Java Development Kit. If you already have it, skip this step. If you don’t, just search for the proper installation instructions for your system.</p></li>
<li><p>Second, you need to install Groovy - a language that is used for processing VDJDB. Go to <a href="https://groovy.apache.org/download.html#distro" class="external-link">this link</a> and download the distribution or windows installer depending on your system. For Windows users the best way is to download the Windows installer. For Linux users the easiest way is to use OS package manager such as apt, dpkg, pacman, etc. For Mac users the most seamless way is to use <a href="https://brew.sh" class="external-link">Homebrew</a>.</p></li>
<li><p>Download the VDJDB repository from GitHub via this link: <a href="https://github.com/antigenomics/vdjdb-db/archive/master.zip" class="external-link uri">https://github.com/antigenomics/vdjdb-db/archive/master.zip</a></p></li>
<li><p>Unzip the archive and go to the unpacked “vdjdb-db-master” folder.</p></li>
<li><p>Go to the “src” folder.</p></li>
<li><p>Open your Terminal or Console and execute the following command: <code>groovy -cp . BuildDatabase.groovy --no2fix</code>.</p></li>
<li><p>After some processing, the database files will be available at the “database” folder inside the “vdjdb-db-master” folder. You will need to provide paths to this files for the <code>immunarch</code> annotation functions.</p></li>
</ol>
</div>
</div>
<div class="section level3">
<h3 id="mcpas-tcr">McPAS-TCR<a class="anchor" aria-label="anchor" href="#mcpas-tcr"></a>
</h3>
<p>McPAS-TCR is a manually curated catalogue of pathology associated T-cell receptor sequences. The database is available at <a href="http://friedmanlab.weizmann.ac.il/McPAS-TCR/" class="external-link uri">http://friedmanlab.weizmann.ac.il/McPAS-TCR/</a></p>
<p>Citation: <em>Tickotsky N, Sagiv T, Prilusky J, Shifrut E, Friedman N (2017). McPAS-TCR: A manually-curated catalogue of pathology-associated T cell receptor sequences. Bioinformatics 33:2924-2929</em></p>
<div class="section level4">
<h4 id="how-to-filter-and-download-data-1">How to filter and download data<a class="anchor" aria-label="anchor" href="#how-to-filter-and-download-data-1"></a>
</h4>
<p>The filtering feature of the database’s web interface is located in the “Search Database” tab. After processing the data, press the “Download .csv” button. The downloaded file named “McPAS-TCR_search.csv” can be used for annotation with <code>immunarch</code>.</p>
</div>
<div class="section level4">
<h4 id="how-to-download-full-mcpas-tcr">How to download full McPAS-TCR<a class="anchor" aria-label="anchor" href="#how-to-download-full-mcpas-tcr"></a>
</h4>
<p>To download McPAS-TCR you just need to go to <a href="http://friedmanlab.weizmann.ac.il/McPAS-TCR/" class="external-link uri">http://friedmanlab.weizmann.ac.il/McPAS-TCR/</a> and press the “Download the complete database” button there. Note that sometimes you need to press it twice or press it in a new browser tab to start the downloading process.</p>
</div>
</div>
<div class="section level3">
<h3 id="tbadb-from-pird">TBAdb from PIRD<a class="anchor" aria-label="anchor" href="#tbadb-from-pird"></a>
</h3>
<p>TBAdb is a manually curated database of T-cell receptor (TCR) and B-cell receptor (BCR) targeting specific antigen or diseases. The database contains three parts, namely TCR-AB, TCR-GD and BCR. These three parts are aimed at collecting sequences and specificity information of TCRA and TCRB, TCR- gamma and TCR-delta and BCR separately. The database is referenced in this paper: <a href="https://doi.org/10.1093/bioinformatics/btz614" class="external-link uri">https://doi.org/10.1093/bioinformatics/btz614</a></p>
<p>Currently there is no direct way to download TBAdb.</p>
<p>Citation: <em>ZHANG W, Wang L, Liu K, Wei X, Yang K, Du W, Wang S, Guo N, Ma C, Luo L, et al. PIRD: Pan immune repertoire database. Bioinformatics(2019)</em></p>
</div>
</div>
<div class="section level2">
<h2 id="annotation-of-the-clonotypes">Annotation of the clonotypes<a class="anchor" aria-label="anchor" href="#annotation-of-the-clonotypes"></a>
</h2>
<p>After downloading the database, we can proceed to the annotation part with R. To demonstrate the applicability of R and <code>immunarch</code>, we will use a common task of annotation of repertoires with Cytomegalovirus (CMV) infection.</p>
<div class="section level3">
<h3 id="preprocessing-databases-with-r">Preprocessing databases with R<a class="anchor" aria-label="anchor" href="#preprocessing-databases-with-r"></a>
</h3>
<p>As a start, we need to load databases into R and filter out non-human, non-TRB and non-CMV data from the input database. With databases, we follow the same philosophy as with <code>repLoad</code> and <code>vis</code> functions: the function <code>dbLoad</code> provides a single interface to loading and basic quering of all supported databases.</p>
<p>For demonstration purposes, we will process each of the supported databases below.</p>
<div class="section level4">
<h4 id="vdjdb-1">VDJDB<a class="anchor" aria-label="anchor" href="#vdjdb-1"></a>
</h4>
<p>Download the VDJDB database following the instructions above. In the examples, we use URLs to snippets of databases as file paths. In your own code you need to provide paths to your local database files, e.g., “/Users/yourname/Downloads/vdjdb-db-master/vdjdb.slim.txt”. Do not use the links below since they are only for testing purposes and do not contain the actual databases!</p>
<p>Note that VDJDB data obtained from the web interface differs from VDJDB obtained from raw files. Check the next section for working with VDJDB search tables.</p>
<p>The most basic way to load VDJDB to R:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vdjdb</span> <span class="op">=</span> <span class="fu"><a href="../../reference/dbLoad.html">dbLoad</a></span><span class="op">(</span><span class="st">"https://gitlab.com/immunomind/immunarch/raw/dev-0.5.0/private/vdjdb.slim.txt.gz"</span>, <span class="st">"vdjdb"</span><span class="op">)</span></span>
<span><span class="va">vdjdb</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 61,049 × 19</span></span></span>
<span><span class="co">##    gene  cdr3        species antig…¹ antig…² antig…³ compl…⁴ v.segm j.segm v.end</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> TRB   CASSTSRLSN… Macaca… STPESA… Tat     SIV     0   <span style="color: #949494;"> </span>   TRBV1… TRBJ1…     4</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> TRB   CASSQDRGPA… HomoSa… RLRAEA… EBNA3A  EBV     1.93<span style="color: #949494;">e</span> 4 TRBV4… TRBJ2…     5</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> TRB   CASSMSRSSN… Macaca… TTPESA… Tat     SIV     0   <span style="color: #949494;"> </span>   TRBV14 TRBJ1…     4</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> TRA   CASNTGTASK… HomoSa… GILGFV… M       Influe… 0   <span style="color: #949494;"> </span>   TRAV24 TRAJ44     2</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> TRB   CASSLGSQNT… MusMus… HGIRNA… M45     MCMV    2.24<span style="color: #949494;">e</span>24 TRBV1… TRBJ2…     5</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> TRB   CSASILGLAG… HomoSa… KLGGAL… IE1     CMV     8.58<span style="color: #949494;">e</span> 3 TRBV2… TRBJ2…     3</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> TRA   CAVLLEYGNK… HomoSa… GILGFV… M       Influe… 0   <span style="color: #949494;"> </span>   TRAV1… TRAJ47     3</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> TRB   CASSYFSATN… HomoSa… KLGGAL… IE1     CMV     3.44<span style="color: #949494;">e</span> 3 TRBV6… TRBJ2…     5</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> TRB   CASTGDSNER… MusMus… SSYRRP… PB1     Influe… 2.28<span style="color: #949494;">e</span> 4 TRBV1… TRBJ1…     3</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> TRB   CASSAFPCRE… HomoSa… NLVPMV… pp65    CMV     0   <span style="color: #949494;"> </span>   TRBV6… TRBJ2…     4</span></span>
<span><span class="co">## <span style="color: #949494;"># … with 61,039 more rows, 9 more variables: j.start &lt;dbl&gt;, mhc.a &lt;chr&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   mhc.b &lt;chr&gt;, mhc.class &lt;chr&gt;, reference.id &lt;chr&gt;, vdjdb.score &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   Species &lt;chr&gt;, Chain &lt;chr&gt;, Pathology &lt;chr&gt;, and abbreviated variable names</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   ¹​antigen.epitope, ²​antigen.gene, ³​antigen.species, ⁴​complex.id</span></span></span></code></pre>
<p>To load VDJDB and filter out information you need to provide <code>.species</code>, <code>.chain</code> and <code>.pathology</code> arguments:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vdjdb</span> <span class="op">=</span> <span class="fu"><a href="../../reference/dbLoad.html">dbLoad</a></span><span class="op">(</span><span class="st">"https://gitlab.com/immunomind/immunarch/raw/dev-0.5.0/private/vdjdb.slim.txt.gz"</span>, <span class="st">"vdjdb"</span>, .species <span class="op">=</span> <span class="st">"HomoSapiens"</span>, .chain <span class="op">=</span> <span class="st">"TRB"</span>, .pathology <span class="op">=</span> <span class="st">"CMV"</span><span class="op">)</span></span>
<span><span class="va">vdjdb</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 18,039 × 19</span></span></span>
<span><span class="co">##    gene  cdr3        species antig…¹ antig…² antig…³ compl…⁴ v.segm j.segm v.end</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> TRB   CSASILGLAG… HomoSa… KLGGAL… IE1     CMV        <span style="text-decoration: underline;">8</span>584 TRBV2… TRBJ2…     3</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> TRB   CASSYFSATN… HomoSa… KLGGAL… IE1     CMV        <span style="text-decoration: underline;">3</span>445 TRBV6… TRBJ2…     5</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> TRB   CASSAFPCRE… HomoSa… NLVPMV… pp65    CMV           0 TRBV6… TRBJ2…     4</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> TRB   CASSLWTTNY… HomoSa… KLGGAL… IE1     CMV       <span style="text-decoration: underline;">19</span>396 TRBV1… TRBJ1…     5</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> TRB   CASSLTTESG… HomoSa… NLVPMV… pp65    CMV           0 TRBV7… TRBJ2…     5</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> TRB   CASTAKQDFQ… HomoSa… KLGGAL… IE1     CMV       <span style="text-decoration: underline;">10</span>972 TRBV1… TRBJ2…     3</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> TRB   CASSGAGAGY… HomoSa… KLGGAL… IE1     CMV        <span style="text-decoration: underline;">6</span>231 TRBV6… TRBJ2…     4</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> TRB   CASSLIIGVS… HomoSa… KLGGAL… IE1     CMV       <span style="text-decoration: underline;">12</span>587 TRBV1… TRBJ1…     5</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> TRB   CATSSSGVQE… HomoSa… KLGGAL… IE1     CMV       <span style="text-decoration: underline;">13</span>267 TRBV15 TRBJ2…     4</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> TRB   CASSLGTLEE… HomoSa… NLVPMV… pp65    CMV           0 TRBV6… TRBJ2…     4</span></span>
<span><span class="co">## <span style="color: #949494;"># … with 18,029 more rows, 9 more variables: j.start &lt;dbl&gt;, mhc.a &lt;chr&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   mhc.b &lt;chr&gt;, mhc.class &lt;chr&gt;, reference.id &lt;chr&gt;, vdjdb.score &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   Species &lt;chr&gt;, Chain &lt;chr&gt;, Pathology &lt;chr&gt;, and abbreviated variable names</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   ¹​antigen.epitope, ²​antigen.gene, ³​antigen.species, ⁴​complex.id</span></span></span></code></pre>
</div>
<div class="section level4">
<h4 id="vdjdb-search-tables">VDJDB search tables<a class="anchor" aria-label="anchor" href="#vdjdb-search-tables"></a>
</h4>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vdjdb_st</span> <span class="op">=</span> <span class="fu"><a href="../../reference/dbLoad.html">dbLoad</a></span><span class="op">(</span><span class="st">"https://gitlab.com/immunomind/immunarch/raw/dev-0.5.0/private/SearchTable-2019-10-17%2012_36_11.989.tsv.gz"</span>, <span class="st">"vdjdb-search"</span>, .species <span class="op">=</span> <span class="st">"HomoSapiens"</span>, .chain <span class="op">=</span> <span class="st">"TRB"</span>, .pathology <span class="op">=</span> <span class="st">"CMV"</span><span class="op">)</span></span>
<span><span class="va">vdjdb_st</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 4,999 × 19</span></span></span>
<span><span class="co">##    complex.id Gene  CDR3     V     J     Species `MHC A` `MHC B` MHC c…¹ Epitope</span></span>
<span><span class="co">##         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span>          0 TRB   CASSVDG… TRBV… TRBJ… HomoSa… HLA-A*… B2M     MHCI    YILEET…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span>          0 TRB   CAWSWGD… TRBV… TRBJ… HomoSa… HLA-A*… B2M     MHCI    YILEET…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span>          0 TRB   CASSLVG… TRBV… TRBJ… HomoSa… HLA-A*… B2M     MHCI    YILEET…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span>          0 TRB   CASSLLM… TRBV… TRBJ… HomoSa… HLA-B*… B2M     MHCI    ELKRKM…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span>          0 TRB   CASSYRA… TRBV… TRBJ… HomoSa… HLA-B*… B2M     MHCI    ELKRKM…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span>          0 TRB   CASSQAV… TRBV… TRBJ… HomoSa… HLA-B*… B2M     MHCI    ELKRKM…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span>          0 TRB   CASVTGS… TRBV… TRBJ… HomoSa… HLA-B*… B2M     MHCI    ELKRKM…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span>          0 TRB   CASSDGT… TRBV… TRBJ… HomoSa… HLA-B*… B2M     MHCI    ELKRKM…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span>          0 TRB   CASSLAR… TRBV… TRBJ… HomoSa… HLA-B*… B2M     MHCI    CVETMC…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span>          0 TRB   CASGGAD… TRBV… TRBJ… HomoSa… HLA-B*… B2M     MHCI    CVETMC…</span></span>
<span><span class="co">## <span style="color: #949494;"># … with 4,989 more rows, 9 more variables: `Epitope gene` &lt;chr&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   `Epitope species` &lt;chr&gt;, Reference &lt;chr&gt;, Method &lt;chr&gt;, Meta &lt;chr&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   CDR3fix &lt;chr&gt;, Score &lt;dbl&gt;, Chain &lt;chr&gt;, Pathology &lt;chr&gt;, and abbreviated</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   variable name ¹​`MHC class`</span></span></span></code></pre>
</div>
<div class="section level4">
<h4 id="mcpas-tcr-1">McPAS-TCR<a class="anchor" aria-label="anchor" href="#mcpas-tcr-1"></a>
</h4>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mcpas</span> <span class="op">=</span> <span class="fu"><a href="../../reference/dbLoad.html">dbLoad</a></span><span class="op">(</span><span class="st">"https://gitlab.com/immunomind/immunarch/raw/dev-0.5.0/private/McPAS-TCR.csv.gz"</span>, <span class="st">"mcpas"</span>, .species <span class="op">=</span> <span class="st">"Human"</span>, .chain <span class="op">=</span> <span class="st">"TRB"</span>, .pathology <span class="op">=</span> <span class="st">"Cytomegalovirus (CMV)"</span><span class="op">)</span></span>
<span><span class="va">mcpas</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 2,723 × 29</span></span></span>
<span><span class="co">##    CDR3.…¹ CDR3.…² Species Categ…³ Patho…⁴ Patho…⁵ Addit…⁶ Antig…⁷ NGS   Antig…⁸</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> <span style="color: #BB0000;">NA</span>      CASLAP… Human   Pathog… Cytome… D003586 <span style="color: #BB0000;">NA</span>            1 No    pp65   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> <span style="color: #BB0000;">NA</span>      CASLQA… Human   Pathog… Cytome… D003586 <span style="color: #BB0000;">NA</span>            1 No    pp65   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> <span style="color: #BB0000;">NA</span>      CASLSG… Human   Pathog… Cytome… D003586 <span style="color: #BB0000;">NA</span>            1 No    pp65   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> <span style="color: #BB0000;">NA</span>      CASLVA… Human   Pathog… Cytome… D003586 <span style="color: #BB0000;">NA</span>            1 No    pp65   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> <span style="color: #BB0000;">NA</span>      CASSHR… Human   Pathog… Cytome… D003586 CMV           1 No    pp65   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> <span style="color: #BB0000;">NA</span>      CASSSA… Human   Pathog… Cytome… D003586 CMV           1 No    pp65   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> <span style="color: #BB0000;">NA</span>      CATSDP… Human   Pathog… Cytome… D003586 CMV           1 No    pp65   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> CARNTG… CACSLR… Human   Pathog… Cytome… D003586 CMV           1 No    pp65   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> CAGNTG… CASSAW… Human   Pathog… Cytome… D003586 Reumat…       1 No    pp65   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> CAYPYN… CASSEL… Human   Pathog… Cytome… D003586 CMV           1 No    pp65   </span></span>
<span><span class="co">## <span style="color: #949494;"># … with 2,713 more rows, 19 more variables: Protein.ID &lt;chr&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   Epitope.peptide &lt;chr&gt;, Epitope.ID &lt;chr&gt;, MHC &lt;chr&gt;, Tissue &lt;chr&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   T.Cell.Type &lt;chr&gt;, T.cell.characteristics &lt;chr&gt;, CDR3.alpha.nt &lt;chr&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   TRAV &lt;chr&gt;, TRAJ &lt;chr&gt;, TRBV &lt;chr&gt;, TRBD &lt;chr&gt;, TRBJ &lt;chr&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   Reconstructed.J.annotation &lt;chr&gt;, CDR3.beta.nt &lt;chr&gt;, Mouse.strain &lt;chr&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   PubMed.ID &lt;dbl&gt;, Remarks &lt;chr&gt;, Chain &lt;chr&gt;, and abbreviated variable names</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   ¹​CDR3.alpha.aa, ²​CDR3.beta.aa, ³​Category, ⁴​Pathology, ⁵​Pathology.Mesh.ID, …</span></span></span></code></pre>
</div>
<div class="section level4">
<h4 id="tbadb">TBAdb<a class="anchor" aria-label="anchor" href="#tbadb"></a>
</h4>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tbadb</span> <span class="op">=</span> <span class="fu"><a href="../../reference/dbLoad.html">dbLoad</a></span><span class="op">(</span><span class="st">"https://gitlab.com/immunomind/immunarch/raw/dev-0.5.0/private/TBAdb.xlsx"</span>, <span class="st">"tbadb"</span>, .species <span class="op">=</span> <span class="st">"Homo Sapiens"</span>, .chain <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TRB"</span>, <span class="st">"TRA-TRB"</span><span class="op">)</span>, .pathology <span class="op">=</span> <span class="st">"CMV"</span><span class="op">)</span></span>
<span><span class="va">tbadb</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="repertoire-annotation">Repertoire annotation<a class="anchor" aria-label="anchor" href="#repertoire-annotation"></a>
</h3>
<p>The key <code>immunarch</code> function for annotation is <code>dbAnnotate</code>. As an input it requires repertoires to search in, a database to look up, and columns of interest such as CDR3 amino acid sequence or V gene segment names columns. If you want to try it on the test data packaged with <code>immunarch</code>, execute the following line of code before proceeding further:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">immdata</span><span class="op">)</span></span></code></pre></div>
<p>Just in a single line of code you are able to find all clonotypes with matching CDR3 amino acid sequences in the input data and VDJDB database:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/dbAnnotate.html">dbAnnotate</a></span><span class="op">(</span><span class="va">immdata</span><span class="op">$</span><span class="va">data</span>, <span class="va">vdjdb</span>, <span class="st">"CDR3.aa"</span>, <span class="st">"cdr3"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##            CDR3.aa Samples A2-i129 A2-i131 A2-i133 A2-i132 A4-i191 A4-i192 MS1</span></span>
<span><span class="co">##   1:   CASSLGETQYF      11       6       4       2       8       4       0   1</span></span>
<span><span class="co">##   2:   CASSFQETQYF       9       3       2       2       4       2       0   1</span></span>
<span><span class="co">##   3:    CASSQETQYF       9       5       2       1       2       3       2   0</span></span>
<span><span class="co">##   4:   CASSSSYEQYF       9       1       0       0       1       2       2   1</span></span>
<span><span class="co">##   5:  CASSLEGYEQYF       8       0       0       1       1       3       0   1</span></span>
<span><span class="co">##  ---                                                                          </span></span>
<span><span class="co">## 579:  CSVGTGTYEQYF       1       0       0       0       0       1       0   0</span></span>
<span><span class="co">## 580: CSVQGGAYNEQFF       1       0       1       0       0       0       0   0</span></span>
<span><span class="co">## 581: CSVQGGSYNEQFF       1       0       1       0       0       0       0   0</span></span>
<span><span class="co">## 582:  CSVVATNEKLFF       1       0       0       1       0       0       0   0</span></span>
<span><span class="co">## 583: CSVVGTGNTEAFF       1       0       0       0       0       0       0   0</span></span>
<span><span class="co">##      MS2 MS3 MS4 MS5 MS6</span></span>
<span><span class="co">##   1:   3   1   2   5   1</span></span>
<span><span class="co">##   2:   1   0   4   0   2</span></span>
<span><span class="co">##   3:   1   0   0   4   1</span></span>
<span><span class="co">##   4:   1   0   1   1   3</span></span>
<span><span class="co">##   5:   0   1   1   1   1</span></span>
<span><span class="co">##  ---                    </span></span>
<span><span class="co">## 579:   0   0   0   0   0</span></span>
<span><span class="co">## 580:   0   0   0   0   0</span></span>
<span><span class="co">## 581:   0   0   0   0   0</span></span>
<span><span class="co">## 582:   0   0   0   0   0</span></span>
<span><span class="co">## 583:   0   0   0   1   0</span></span></code></pre>
<p>The “Samples” column specifies the number of samples in which the clonotype found. Other numbers in columns correspond to the clonal count of the clonotype in a specific input sample.</p>
<p>In the next example we will search the McPAS-TCR database for condition-associated sequences using both CDR3 amino acid sequences and V gene segments:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/dbAnnotate.html">dbAnnotate</a></span><span class="op">(</span><span class="va">immdata</span><span class="op">$</span><span class="va">data</span>, <span class="va">mcpas</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CDR3.aa"</span>, <span class="st">"V.name"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CDR3.beta.aa"</span>, <span class="st">"TRBV"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                CDR3.aa   V.name Samples A2-i129 A2-i131 A2-i133 A2-i132 A4-i191</span></span>
<span><span class="co">##    1:      CAISESYEQYF TRBV10-3       3       0       1       0       0       0</span></span>
<span><span class="co">##    2: CASSLAPGATNEKLFF  TRBV7-6       3       0       0       0       0       0</span></span>
<span><span class="co">##    3:     CASSLGENIQYF   TRBV13       3       0       0       0       0       1</span></span>
<span><span class="co">##    4:     CASSLGRETQYF   TRBV28       3       0       0       0       0       0</span></span>
<span><span class="co">##    5:   CSVGTGGTNEKLFF TRBV29-1       3       0       0       0       0       0</span></span>
<span><span class="co">##   ---                                                                          </span></span>
<span><span class="co">## 2123:           KNPTAF   TRBV19       1       0       0       0       0       0</span></span>
<span><span class="co">## 2124:       LLGGQETQYF  TRBV7-4       1       0       0       0       0       0</span></span>
<span><span class="co">## 2125:     WASSFQGFTEAF   TRBV28       1       0       0       0       0       0</span></span>
<span><span class="co">## 2126:    WASSQALPYEQYF TRBV12-4       1       0       0       0       0       0</span></span>
<span><span class="co">## 2127:  WASSQQTGTIGGYTF  TRBV6-5       1       0       0       0       0       0</span></span>
<span><span class="co">##       A4-i192 MS1 MS2 MS3 MS4 MS5 MS6</span></span>
<span><span class="co">##    1:       1   0   0   0   0   0   0</span></span>
<span><span class="co">##    2:       0   0   1   0   0   1   0</span></span>
<span><span class="co">##    3:       1   0   0   0   0   0   0</span></span>
<span><span class="co">##    4:       0   0   0   0   0  75   1</span></span>
<span><span class="co">##    5:       0   0   1   0   0   0   1</span></span>
<span><span class="co">##   ---                                </span></span>
<span><span class="co">## 2123:       0   0   0   0   0   0   0</span></span>
<span><span class="co">## 2124:       0   0   0   0   0   0   0</span></span>
<span><span class="co">## 2125:       0   0   0   0   0   0   0</span></span>
<span><span class="co">## 2126:       0   0   0   0   0   0   0</span></span>
<span><span class="co">## 2127:       0   0   0   0   0   0   0</span></span></code></pre>
<p>If you seek to search a database for a specific set of sequences, create a data frame containing them and use it as a database file:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">local_db</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>Seq <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CASSDSSGGANEQFF"</span>, <span class="st">"CSARLAGGQETQYF"</span><span class="op">)</span>, V <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TRBV6-4"</span>, <span class="st">"TRBV20-1"</span><span class="op">)</span>, stringsAsFactors <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../../reference/dbAnnotate.html">dbAnnotate</a></span><span class="op">(</span><span class="va">immdata</span><span class="op">$</span><span class="va">data</span>, <span class="va">local_db</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CDR3.aa"</span>, <span class="st">"V.name"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Seq"</span>, <span class="st">"V"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##            CDR3.aa   V.name Samples A2-i129 A2-i131 A2-i133 A2-i132 A4-i191</span></span>
<span><span class="co">## 1: CASSDSSGGANEQFF  TRBV6-4       7       1       1       2       0       3</span></span>
<span><span class="co">## 2:  CSARLAGGQETQYF TRBV20-1       6       1       3       0       1       0</span></span>
<span><span class="co">##    A4-i192 MS1 MS2 MS3 MS4 MS5 MS6</span></span>
<span><span class="co">## 1:       0   0   0   2   0   0  12</span></span>
<span><span class="co">## 2:       0   0   0   1   0   0   1</span></span></code></pre>
<div class="section level4">
<h4 id="visualisation">Visualisation<a class="anchor" aria-label="anchor" href="#visualisation"></a>
</h4>
<p>Visualisation with the <code><a href="../../reference/vis.html">vis()</a></code> function will be supported in the next major release of <code>immunarch</code>. You can use <code>ggplot2</code> to visualise distributions of found clonotypes.</p>
</div>
</div>
<div class="section level3">
<h3 id="advanced-filtering">Advanced filtering<a class="anchor" aria-label="anchor" href="#advanced-filtering"></a>
</h3>
<p><code>immunarch</code> provides a very basic query interface that permits filtering by species types, chain types and pathology types only. To perform advanced filtering such as filtering by antigen epitope, you need to use R. In the most cases, filtering with the <code>dplyr</code> package is the most seamless way. Here is an example on how to use <code>dplyr</code> to filter out specific antigen epitopes from VDJDB:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load the dplyr library</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load the database with immunarch</span></span>
<span><span class="va">vdjdb</span> <span class="op">=</span> <span class="fu"><a href="../../reference/dbLoad.html">dbLoad</a></span><span class="op">(</span><span class="st">"https://gitlab.com/immunomind/immunarch/raw/dev-0.5.0/private/vdjdb.slim.txt.gz"</span>, <span class="st">"vdjdb"</span>, .species <span class="op">=</span> <span class="st">"HomoSapiens"</span>, .chain <span class="op">=</span> <span class="st">"TRB"</span>, .pathology <span class="op">=</span> <span class="st">"CMV"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Check which antigen epitopes are presented in the database</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">vdjdb</span><span class="op">$</span><span class="va">antigen.epitope</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## ARNLVPMVATVQGQN       AYAQKIFKI    CPSQEPMSIYVY       CVETMCNEY       DEEDAIAAY </span></span>
<span><span class="co">##               3              39               2               2               2 </span></span>
<span><span class="co">## EDVPSGKLFMHVTLG      EFFWDANDIY       ELKRKMIYM       ELRRKMMYM        FPTKDVAL </span></span>
<span><span class="co">##               1               1               5              10              10 </span></span>
<span><span class="co">##       IPSINVHHY       KLGGALQAK LSEFCRVLCCYVLEE       MLNIPSINV        NEGVKAAW </span></span>
<span><span class="co">##              93           12667               2              73              49 </span></span>
<span><span class="co">##       NLVPMVATV       QIKVRVDMV       QIKVRVKMV       QYDPVAALF      RPHERNGFTV </span></span>
<span><span class="co">##            4496              15              24              39               4 </span></span>
<span><span class="co">##     RPHERNGFTVL      TPRVTGGGAM       VLEETSVML       VMAPRTLIL       VTEHDTLLY </span></span>
<span><span class="co">##              22             207              14               1             202 </span></span>
<span><span class="co">##       YILEETSVM     YSEHPTFTSQY </span></span>
<span><span class="co">##               3              53</span></span></code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Filter out all non NLVPMVATV epitopes</span></span>
<span><span class="va">vdjdb</span> <span class="op">=</span> <span class="va">vdjdb</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">antigen.epitope</span> <span class="op">==</span> <span class="st">"NLVPMVATV"</span><span class="op">)</span></span>
<span><span class="va">vdjdb</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 4,496 × 19</span></span></span>
<span><span class="co">##    gene  cdr3        species antig…¹ antig…² antig…³ compl…⁴ v.segm j.segm v.end</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> TRB   CASSAFPCRE… HomoSa… NLVPMV… pp65    CMV           0 TRBV6… TRBJ2…     4</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> TRB   CASSLTTESG… HomoSa… NLVPMV… pp65    CMV           0 TRBV7… TRBJ2…     5</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> TRB   CASSLGTLEE… HomoSa… NLVPMV… pp65    CMV           0 TRBV6… TRBJ2…     4</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> TRB   CASSLDSLNT… HomoSa… NLVPMV… pp65    CMV           0 TRBV5… TRBJ1…     5</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> TRB   CSADGLPISS… HomoSa… NLVPMV… pp65    CMV           0 TRBV2… TRBJ2…     2</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> TRB   CASSFRQGAF… HomoSa… NLVPMV… pp65    CMV           0 TRBV7… TRBJ2…     4</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> TRB   CASSFGPRAG… HomoSa… NLVPMV… pp65    CMV           0 TRBV7… TRBJ2…     4</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> TRB   CASSYGTGKD… HomoSa… NLVPMV… pp65    CMV           0 TRBV7… TRBJ2…     4</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> TRB   CSVEAYATDY… HomoSa… NLVPMV… pp65    CMV           0 TRBV2… TRBJ1…     4</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> TRB   CASSSGLISF… HomoSa… NLVPMV… pp65    CMV           0 TRBV5… TRBJ2…     4</span></span>
<span><span class="co">## <span style="color: #949494;"># … with 4,486 more rows, 9 more variables: j.start &lt;dbl&gt;, mhc.a &lt;chr&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   mhc.b &lt;chr&gt;, mhc.class &lt;chr&gt;, reference.id &lt;chr&gt;, vdjdb.score &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   Species &lt;chr&gt;, Chain &lt;chr&gt;, Pathology &lt;chr&gt;, and abbreviated variable names</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   ¹​antigen.epitope, ²​antigen.gene, ³​antigen.species, ⁴​complex.id</span></span></span></code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Check if everything is OK and there is no other epitopes</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">vdjdb</span><span class="op">$</span><span class="va">antigen.epitope</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## NLVPMVATV </span></span>
<span><span class="co">##      4496</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="get-in-contact-with-us">Get in contact with us<a class="anchor" aria-label="anchor" href="#get-in-contact-with-us"></a>
</h2>
<p>Cannot find an important feature? Have a question or found a bug? Contact us at <a href="mailto:support@immunomind.io" class="email">support@immunomind.io</a></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://www.linkedin.com/in/vdnaz" class="external-link">Vadim I. Nazarov</a>, <a href="https://www.linkedin.com/in/vasily-tsvetkov-227218ab" class="external-link">Vasily O. Tsvetkov</a>, <a href="https://www.linkedin.com/in/erumynskiy/" class="external-link">Eugene Rumynskiy</a>, <a href="https://www.linkedin.com/in/aleksandr-popov-634878105/" class="external-link">Aleksandr A. Popov</a>, <a href="https://www.linkedin.com/in/ivan-balashov/" class="external-link">Ivan Balashov</a>, <a href="https://www.linkedin.com/in/maria-volobueva-b0856a223/" class="external-link">Maria Samokhina</a>, <a href="https://immunomind.io" class="external-link">ImmunoMind</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '94207562f34b455c0796790d104e3549',
    indexName: 'immunarch',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
