<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Single-cell and paired chain data: scTCRseq and scBCRseq exploration • immunarch</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><!-- docsearch --><script src="../../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><link href="../../extra.css" rel="stylesheet">
<meta property="og:title" content="Single-cell and paired chain data: scTCRseq and scBCRseq exploration">
<meta property="og:description" content="immunarch">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-128418614-2"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-128418614-2');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">immunarch</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.6.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Start Here
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/v1_introduction.html">Installation &amp; troubleshooting</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">First steps</li>
    <li>
      <a href="../../articles/v2_data.html">Data loading</a>
    </li>
    <li>
      <a href="../../articles/web_only/load_mixcr.html">How-to: Loading MiXCR Data</a>
    </li>
    <li>
      <a href="../../articles/web_only/load_10x.html">How-to: Loading 10x Genomics Data</a>
    </li>
    <li>
      <a href="../../articles/web_only/v21_singlecell.html">How-to: Single-cell and paired chain data</a>
    </li>
    <li>
      <a href="../../articles/web_only/v3_basic_analysis.html">Basic statistics and clonality</a>
    </li>
    <li>
      <a href="../../articles/web_only/repFilter_v3.html">Data filtering</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Repertoire-level exploration and comparison</li>
    <li>
      <a href="../../articles/web_only/v4_overlap.html">Repertoire overlap and public clonotypes</a>
    </li>
    <li>
      <a href="../../articles/web_only/v5_gene_usage.html">Gene usage</a>
    </li>
    <li>
      <a href="../../articles/web_only/v6_diversity.html">Diversity estimation</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Clonotype-level exploration and comparison</li>
    <li>
      <a href="../../articles/web_only/v8_tracking.html">Track clonotypes across samples and time</a>
    </li>
    <li>
      <a href="../../articles/web_only/v11_db.html">Annotate clonotypes using immune receptor databases</a>
    </li>
    <li>
      <a href="../../articles/web_only/v9_kmers.html">Kmer and sequence motif analysis and visualisation</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Preparing to publication</li>
    <li>
      <a href="../../articles/web_only/v7_fixvis.html">Make your plots publication-ready with fixVis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
<li>
  <a href="https://github.com/immunomind/covid19">COVID-19</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://immunomind.io">ImmunoMind</a>
</li>
<li>
  <a href="http://twitter.com/immunomind">
    <span class="fa fa-lg fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/immunomind/immunarch">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="v21_singlecell_files/header-attrs-2.11/header-attrs.js"></script><script src="v21_singlecell_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Single-cell and paired chain data: scTCRseq and scBCRseq exploration</h1>
                        <h4 class="author">
<b>ImmunoMind</b> – improving design of T-cell therapies using multi-omics and AI. Research and biopharma partnerships, more details: <a href="https://immunomind.io">immunomind.io</a>
</h4>
            
            <h4 class="date"><a href="mailto:support@immunomind.io" class="email">support@immunomind.io</a></h4>
      
      <small class="dont-index">Source: <a href="https://github.com/immunomind/immunarch/blob/master/vignettes/web_only/v21_singlecell.Rmd"><code>vignettes/web_only/v21_singlecell.Rmd</code></a></small>
      <div class="hidden name"><code>v21_singlecell.Rmd</code></div>

    </div>

    
    
<!--
%\VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{Working with data in immunarch}
%\VignettePackage{immunarch}
-->
<div id="executive-summary" class="section level1">
<h1 class="hasAnchor">
<a href="#executive-summary" class="anchor"></a>Executive Summary</h1>
<blockquote>
<p>This is a vignette dedicated to provide an overview on how to work with single-cell paired chain data in <code>immunarch</code></p>
</blockquote>
<blockquote>
<p>Single-cell support is currently in the development version. In order to access it, you need to install the latest development version of the package by executing the following command:</p>
</blockquote>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span>; <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span><span class="op">(</span><span class="st">"immunomind/immunarch"</span>, ref<span class="op">=</span><span class="st">"dev"</span><span class="op">)</span></code></pre></div>
<blockquote>
<p>To read paired chain data into <code>immunarch</code> use the <code>repLoad</code> function with <code>.mode = "paired"</code>. Currently we support 10X Genomics only.</p>
</blockquote>
<blockquote>
<p>To subset immune repertoires by specific barcodes use the <code>select_barcodes</code> function. Output of <code>Seurat::Idents()</code> as a barcode vector works.</p>
</blockquote>
<blockquote>
<p>To create cluster-specific and patient-specific datasets using barcodes from the output of <code>Seurat::Idents()</code> use the <code>select_clusters</code> function.</p>
</blockquote>
</div>
<div id="use-the-data-packaged-with-immunarch" class="section level1">
<h1 class="hasAnchor">
<a href="#use-the-data-packaged-with-immunarch" class="anchor"></a>Use the data packaged with <code>immunarch</code>
</h1>
<p>Load the package into the R enviroment:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://immunarch.com/">immunarch</a></span><span class="op">)</span></code></pre></div>
<p>For testing purposes we attached a new paired chain dataset to <code>immunarch</code>. Load it by executing the following command:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">scdata</span><span class="op">)</span></code></pre></div>
</div>
<div id="load-the-paired-chain-data" class="section level1">
<h1 class="hasAnchor">
<a href="#load-the-paired-chain-data" class="anchor"></a>Load the paired chain data</h1>
<p>To load your own datasets, use the <code>repLoad</code> function. Currently we implemented paired chain data support for 10X Genomics data only. A working example of loading datasets into R:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">file_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span>package <span class="op">=</span> <span class="st">"immunarch"</span><span class="op">)</span>, <span class="st">"/extdata/sc/flu.csv.gz"</span><span class="op">)</span>
<span class="va">igdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/repLoad.html">repLoad</a></span><span class="op">(</span><span class="va">file_path</span>, .mode <span class="op">=</span> <span class="st">"paired"</span><span class="op">)</span></code></pre></div>
<pre><code>## 
## == Step 1/3: loading repertoire files... ==</code></pre>
<pre><code>## Processing "&lt;initial&gt;" ...</code></pre>
<pre><code>##   -- [1/1] Parsing "/tmp/RtmpEom39V/temp_libpath19746c19325ef2/immunarch/extdata/sc/flu.csv.gz" -- 10x (filt.contigs)</code></pre>
<pre><code>## Warning: The following named parsers don't match the column names:
## sample,barcode,is_cell,contig_id,high_confidence,length,chain,v_gene,d_gene,j_gene,c_gene,full_length,productive,cdr3,cdr3_nt,reads,umis,raw_clonotype_id,raw_consensus_id</code></pre>
<pre><code>## 
## == Step 2/3: checking metadata files and merging files... ==
## 
## Processing "&lt;initial&gt;" ...
##   -- Metadata file not found; creating a dummy metadata...
## 
## == Step 3/3: processing paired chain data... ==
## 
## Done!</code></pre>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">igdata</span><span class="op">$</span><span class="va">meta</span></code></pre></div>
<pre><code>## # A tibble: 1 × 1
##   Sample
##   &lt;chr&gt; 
## 1 flu</code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">igdata</span><span class="op">$</span><span class="va">data</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">7</span>, <span class="fl">16</span>, <span class="fl">17</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code>##   Clones Proportion
## 1      3      3e-04
## 2      3      3e-04
## 3      2      2e-04
## 4      2      2e-04
## 5      2      2e-04
## 6      2      2e-04
##                                                                                                         CDR3.nt
## 1          TGTGCACACACCACCGAACTCTATTGTACTAATGGTGTATGCTATGGGGGCTACTTTGACTACTGG;TGCCAACAGTATAATAGTTATTCGTGGACGTTC
## 2                                     TGTGCGAGGCTATGGGGTTGGGGATTACTCTACTGG;TGCACCTCATATGCAGGCAGCAACAATTTGGTATTC
## 3                                  TGTACCGCTCATGGTAGGGAGGGTACTTTTGATCTCTGG;TGTCAACAATATGATAATCTCCCTCGGTACACTTTT
## 4                            TGTGCAAGAGACGACTTTGCTTCGGGGGGTCGACACTTTGGCTGCTGG;TGTCAGCAGTCTGGTAACGCACCTCGAACTTTT
## 5                                           TGTGCAAGAGATCTGGACTACATGGACGTCTGG;TGTCAACAGAGTTACAGTACCCCTCGAACTTTT
## 6 TGTGCAAGAGGTTTAATACCCTCAGTTAGTGGCTACGACTACTACTACTACTACGGTATGGACGTCTGG;TGTCAATCAGCAGACAGCAGTGGTACTTATGAGGTATTC
##                                 CDR3.aa             V.name        D.name
## 1    CAHTTELYCTNGVCYGGYFDYW;CQQYNSYSWTF    IGHV2-5;IGKV1-5  IGHD2-8;None
## 2             CARLWGWGLLYW;CTSYAGSNNLVF  IGHV4-59;IGLV2-11 IGHD3-10;None
## 3            CTAHGREGTFDLW;CQQYDNLPRYTF  IGHV3-15;IGKV1-33 IGHD2-15;None
## 4          CARDDFASGGRHFGCW;CQQSGNAPRTF   IGHV1-2;IGKV3-20 IGHD3-10;None
## 5               CARDLDYMDVW;CQQSYSTPRTF IGHV3-13;IGKV1D-39     None;None
## 6 CARGLIPSVSGYDYYYYYGMDVW;CQSADSSGTYEVF  IGHV3-74;IGLV3-25 IGHD5-12;None
##        J.name   chain                                                  Barcode
## 1 IGHJ4;IGKJ1 IGH;IGK AGTAGTCAGTGTACTC-1;GGCGACTGTACCGAGA-1;TTGAACGGTCACCTAA-1
## 2 IGHJ4;IGLJ2 IGH;IGL AGAGCGACACCTTGTC-1;ATTGGTGAGACCTAGG-1;TCTTCGGAGGTGATTA-1
## 3 IGHJ3;IGKJ2 IGH;IGK                    CGCGGTATCCTCTAGC-1;TGAGCATCAGGAACGT-1
## 4 IGHJ4;IGKJ2 IGH;IGK                    ACTGTCCAGACGCAAC-1;CACACTCTCCGTTGTC-1
## 5 IGHJ6;IGKJ2 IGH;IGK                    AGGTCCGAGTCAAGCG-1;GGCGTGTTCTCTAGGA-1
## 6 IGHJ6;IGLJ2 IGH;IGL                    ACTGAGTCAAGCGATG-1;CTGCTGTCACGCATCG-1</code></pre>
</div>
<div id="subset-by-barcodes" class="section level1">
<h1 class="hasAnchor">
<a href="#subset-by-barcodes" class="anchor"></a>Subset by barcodes</h1>
<p>To subset the data by barcodes, use the <code>select_barcodes</code> function.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">barcodes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"AGTAGTCAGTGTACTC-1"</span>, <span class="st">"GGCGACTGTACCGAGA-1"</span>, <span class="st">"TTGAACGGTCACCTAA-1"</span><span class="op">)</span>

<span class="va">new_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/select_barcodes.html">select_barcodes</a></span><span class="op">(</span><span class="va">scdata</span><span class="op">$</span><span class="va">data</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="va">barcodes</span><span class="op">)</span>

<span class="va">new_df</span></code></pre></div>
<pre><code>##                                                                                                CDR3.nt
## 1 TGTGCACACACCACCGAACTCTATTGTACTAATGGTGTATGCTATGGGGGCTACTTTGACTACTGG;TGCCAACAGTATAATAGTTATTCGTGGACGTTC
##                              CDR3.aa          V.name D.name      J.name V.end
## 1 CAHTTELYCTNGVCYGGYFDYW;CQQYNSYSWTF IGHV2-5;IGKV1-5     NA IGHJ4;IGKJ1    NA
##   D.start D.end J.start VJ.ins VD.ins DJ.ins Sequence   chain raw_clonotype_id
## 1      NA    NA      NA     NA     NA     NA       NA IGH;IGK               14
##   ContigID Clones                                                  Barcode
## 1       NA      3 AGTAGTCAGTGTACTC-1;GGCGACTGTACCGAGA-1;TTGAACGGTCACCTAA-1
##   Proportion
## 1          1</code></pre>
<div id="patient-specific-datasets" class="section level2">
<h2 class="hasAnchor">
<a href="#patient-specific-datasets" class="anchor"></a>Patient-specific datasets</h2>
<p>To create a new dataset with cluster-specific immune repertoires, use the <code>select_clusters</code> function:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scdata_pat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/select_clusters.html">select_clusters</a></span><span class="op">(</span><span class="va">scdata</span>, <span class="va">scdata</span><span class="op">$</span><span class="va">bc_patient</span>, <span class="st">"Patient"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">scdata_pat</span><span class="op">$</span><span class="va">data</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "flu_PatientA" "flu_PatientB" "flu_PatientC"</code></pre>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scdata_pat</span><span class="op">$</span><span class="va">meta</span></code></pre></div>
<pre><code>## # A tibble: 3 × 3
##   Sample       Patient.source Patient 
##   &lt;chr&gt;        &lt;chr&gt;          &lt;chr&gt;   
## 1 flu_PatientA flu            PatientA
## 2 flu_PatientB flu            PatientB
## 3 flu_PatientC flu            PatientC</code></pre>
</div>
<div id="cluster-specific-datasets" class="section level2">
<h2 class="hasAnchor">
<a href="#cluster-specific-datasets" class="anchor"></a>Cluster-specific datasets</h2>
<p>To create a new dataset with cluster-specific immune repertoires, use the <code>select_clusters</code> function. You can apply this function after you created patient-specific datasets to get patient-specific cell cluster-specific immune repertoires, e.g., a Memory B Cell repertoire for a specific patient:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scdata_cl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/select_clusters.html">select_clusters</a></span><span class="op">(</span><span class="va">scdata_pat</span>, <span class="va">scdata</span><span class="op">$</span><span class="va">bc_cluster</span>, <span class="st">"Cluster"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">scdata_cl</span><span class="op">$</span><span class="va">data</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "flu_PatientA_Activ"  "flu_PatientA_Memory" "flu_PatientA_Naive" 
## [4] "flu_PatientB_Activ"  "flu_PatientB_Memory" "flu_PatientB_Naive" 
## [7] "flu_PatientC_Activ"  "flu_PatientC_Memory" "flu_PatientC_Naive"</code></pre>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scdata_cl</span><span class="op">$</span><span class="va">meta</span></code></pre></div>
<pre><code>## # A tibble: 9 × 5
##   Sample              Patient.source Patient  Cluster.source Cluster
##   &lt;chr&gt;               &lt;chr&gt;          &lt;chr&gt;    &lt;chr&gt;          &lt;chr&gt;  
## 1 flu_PatientA_Activ  flu            PatientA flu_PatientA   Activ  
## 2 flu_PatientA_Memory flu            PatientA flu_PatientA   Memory 
## 3 flu_PatientA_Naive  flu            PatientA flu_PatientA   Naive  
## 4 flu_PatientB_Activ  flu            PatientB flu_PatientB   Activ  
## 5 flu_PatientB_Memory flu            PatientB flu_PatientB   Memory 
## 6 flu_PatientB_Naive  flu            PatientB flu_PatientB   Naive  
## 7 flu_PatientC_Activ  flu            PatientC flu_PatientC   Activ  
## 8 flu_PatientC_Memory flu            PatientC flu_PatientC   Memory 
## 9 flu_PatientC_Naive  flu            PatientC flu_PatientC   Naive</code></pre>
</div>
</div>
<div id="explore-and-compute-statistics" class="section level1">
<h1 class="hasAnchor">
<a href="#explore-and-compute-statistics" class="anchor"></a>Explore and compute statistics</h1>
<p>Most functions will work out-of-the-box with paired chain data.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/repOverlap.html">repOverlap</a></span><span class="op">(</span><span class="va">scdata_cl</span><span class="op">$</span><span class="va">data</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../../reference/vis.html">vis</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/repDiversity.html">repDiversity</a></span><span class="op">(</span><span class="va">scdata_cl</span><span class="op">$</span><span class="va">data</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../../reference/vis.html">vis</a></span><span class="op">(</span><span class="op">)</span>

<span class="va">target</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CARAGYLRGFDYW;CQQYGSSPLTF"</span>, <span class="st">"CARATSFYYFHHW;CTSYTTRTTLIF"</span>, <span class="st">"CARDLSRGDYFPYFSYHMNVW;CQSDDTANHVIF"</span>, <span class="st">"CARGFDTNAFDIW;CTAWDDSLSGVVF"</span>, <span class="st">"CTREDYW;CMQTIQLRTF"</span><span class="op">)</span>
<span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/trackClonotypes.html">trackClonotypes</a></span><span class="op">(</span><span class="va">scdata_cl</span><span class="op">$</span><span class="va">data</span>, <span class="va">target</span>, .col <span class="op">=</span> <span class="st">"aa"</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../../reference/vis.html">vis</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning in melt.data.table(.data): id.vars and measure.vars are internally
## guessed when both are 'NULL'. All non-numeric/integer/logical type columns are
## considered id.vars, which in this case are columns [CDR3.aa, ...]. Consider
## providing at least one of 'id' or 'measure' vars in future.</code></pre>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="op">(</span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span><span class="op">)</span> <span class="op">/</span> <span class="va">p3</span></code></pre></div>
<pre><code>## Warning: The `.dots` argument of `group_by()` is deprecated as of dplyr 1.0.0.
## This warning is displayed once every 8 hours.
## Call `lifecycle::last_lifecycle_warnings()` to see where this warning was generated.</code></pre>
<p><img src="v21_singlecell_files/figure-html/unnamed-chunk-7-1.png" width="1728" style="display: block; margin: auto;"></p>
<p>Several functions may work incorrectly with paired chain data in this release of <code>immunarch</code>. Let us know via <a href="http://github.com/immunomind/immunarch/issues">GitHub Issues</a>!</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://www.linkedin.com/in/vdnaz">Vadim I. Nazarov</a>, <a href="https://www.linkedin.com/in/vasily-tsvetkov-227218ab">Vasily O. Tsvetkov</a>, <a href="https://www.linkedin.com/in/erumynskiy/">Eugene Rumynskiy</a>, <a href="https://www.linkedin.com/in/aleksandr-popov-634878105/">Aleksandr A. Popov</a>, <a href="https://www.linkedin.com/in/ivan-balashov/">Ivan Balashov</a>, <a href="https://www.linkedin.com/in/maria-volobueva-b0856a223/">Maria Volobueva</a>, <a href="https://immunomind.io">ImmunoMind</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '94207562f34b455c0796790d104e3549',
    indexName: 'immunarch',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
