<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Repertoire overlap and public clonotypes • immunarch</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><!-- docsearch --><script src="../../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><link href="../../extra.css" rel="stylesheet">
<meta property="og:title" content="Repertoire overlap and public clonotypes">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-128418614-2"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-128418614-2');
</script>
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">immunarch</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.6.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../articles/v1_introduction.html">Get started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">First steps</li>
    <li>
      <a href="../../articles/v1_introduction.html">Introduction to immunarch</a>
    </li>
    <li>
      <a href="../../articles/v2_data.html">Working with data in immunarch</a>
    </li>
    <li>
      <a href="../../articles/v3_basic_analysis.html">Basic analysis and clonality</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Explore and compare repertoires</li>
    <li>
      <a href="../../articles/web_only/v4_overlap.html">Repertoire overlap and public clonotypes</a>
    </li>
    <li>
      <a href="../../articles/web_only/v5_gene_usage.html">Gene usage analysis</a>
    </li>
    <li>
      <a href="../../articles/web_only/v6_diversity.html">Diversity estimation</a>
    </li>
    <li>
      <a href="../../articles/web_only/v11_db.html">Annotate clonotypes using immune receptor databases</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Preparing to publication</li>
    <li>
      <a href="../../articles/web_only/v7_fixvis.html">Make your plots publication-ready with fixVis</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Advanced methods</li>
    <li>
      <a href="../../articles/web_only/v8_tracking.html">Tracking clonotypes across time points</a>
    </li>
    <li>
      <a href="../../articles/web_only/v9_kmers.html">Kmer and sequence motif analysis and visualisation</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
<li>
  <a href="https://github.com/immunomind/immunarch/releases">Changelog</a>
</li>
<li>
  <a href="../../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://immunomind.io">ImmunoMind</a>
</li>
<li>
  <a href="http://twitter.com/immunomind">
    <span class="fa fa-lg fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/immunomind/immunarch">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Repertoire overlap and public clonotypes</h1>
                        <h4 class="author"><a href="https://immunomind.io">ImmunoMind</a></h4>
            
            <h4 class="date"><a href="mailto:support@immunomind.io" class="email">support@immunomind.io</a></h4>
      
      <small class="dont-index">Source: <a href="https://github.com/immunomind/immunarch/blob/master/vignettes/web_only/v4_overlap.Rmd"><code>vignettes/web_only/v4_overlap.Rmd</code></a></small>
      <div class="hidden name"><code>v4_overlap.Rmd</code></div>

    </div>

    
    
<!--
%\VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{Repertoire overlap and public clonotypes}
%\VignettePackage{immunarch}
-->
<div id="repertoire-overlap" class="section level1">
<h1 class="hasAnchor">
<a href="#repertoire-overlap" class="anchor"></a>Repertoire overlap</h1>
<p>Repertoire overlap is the most common approach to measure repertoire similarity. It is achieved by computation of specific statistics on clonotypes shared between given repertoires, also called “public” clonotypes. <code>immunarch</code> provides several indices: - number of public clonotypes (<code>.method = "public"</code>) - a classic measure of overlap similarity.</p>
<ul>
<li><p>overlap coefficient (<code>.method = "overlap"</code>) - a normalised measure of overlap similarity. It is defined as the size of the intersection divided by the smaller of the size of the two sets.</p></li>
<li><p>Jaccard index (<code>.method = "jaccard"</code>) - it measures similarity between finite sample sets, and is defined as the size of the intersection divided by the size of the union of the sample sets.</p></li>
<li><p>Tversky index (<code>.method = "tversky"</code>) - an asymmetric similarity measure on sets that compares a variant to a prototype. If using default arguments, it’s similar to Dice’s coefficient.</p></li>
<li><p>cosine similarity (<code>.method = "cosine"</code>) - a measure of similarity between two non-zero vectors</p></li>
<li><p>Morisita’s overlap index (<code>.method = "morisita"</code>) - a statistical measure of dispersion of individuals in a population. It is used to compare overlap among samples.</p></li>
<li><p>incremental overlap - overlaps of the N most abundant clonotypes with incrementally growing N (<code>.method = "inc+METHOD"</code>, e.g., <code>"inc+public"</code> or <code>"inc+morisita"</code>).</p></li>
</ul>
<p>The function that includes described methods is <code>repOverlap</code>. Again the output is easily visualised when passed to <code><a href="../../reference/vis.html">vis()</a></code> function that does all the work:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1">imm_ov1 =<span class="st"> </span><span class="kw"><a href="../../reference/repOverlap.html">repOverlap</a></span>(immdata<span class="op">$</span>data, <span class="dt">.method =</span> <span class="st">"public"</span>, <span class="dt">.verbose =</span> F)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2">imm_ov2 =<span class="st"> </span><span class="kw"><a href="../../reference/repOverlap.html">repOverlap</a></span>(immdata<span class="op">$</span>data, <span class="dt">.method =</span> <span class="st">"morisita"</span>, <span class="dt">.verbose =</span> F)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3"></a>
<a class="sourceLine" id="cb1-4" data-line-number="4"><span class="kw">grid.arrange</span>(<span class="kw"><a href="../../reference/vis.html">vis</a></span>(imm_ov1), <span class="kw"><a href="../../reference/vis.html">vis</a></span>(imm_ov2, <span class="dt">.text.size=</span><span class="dv">2</span>), <span class="dt">ncol =</span> <span class="dv">2</span>)</a></code></pre></div>
<p><img src="v4_overlap_files/figure-html/overlap-1.png" width="1152" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw"><a href="../../reference/vis.html">vis</a></span>(imm_ov1, <span class="st">"heatmap2"</span>)</a></code></pre></div>
<p><img src="v4_overlap_files/figure-html/overlap-2.png" width="1152" style="display: block; margin: auto;"></p>
<p>You can easily change the number of significant digits:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="kw">grid.arrange</span>(<span class="kw"><a href="../../reference/vis.html">vis</a></span>(imm_ov2, <span class="dt">.text.size=</span><span class="fl">2.5</span>, <span class="dt">.signif.digits=</span><span class="dv">1</span>), <span class="kw"><a href="../../reference/vis.html">vis</a></span>(imm_ov2, <span class="dt">.text.size=</span><span class="dv">2</span>, <span class="dt">.signif.digits=</span><span class="dv">2</span>), <span class="dt">ncol =</span> <span class="dv">2</span>)</a></code></pre></div>
<p><img src="v4_overlap_files/figure-html/overlap-signif-digits-1.png" width="1152" style="display: block; margin: auto;"></p>
<!---
Top-overlap [Work in Progress]

```r
warning("TODO")
```

```
## Warning: TODO
```
-->
<p>To analyse the computed overlap measures function apply <code>repOverlapAnalysis</code>.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="co"># Apply different analysis algorithms to the matrix of public clonotypes:</span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2"><span class="co"># "mds" - Multi-dimensional Scaling</span></a>
<a class="sourceLine" id="cb4-3" data-line-number="3"><span class="kw"><a href="../../reference/repOverlapAnalysis.html">repOverlapAnalysis</a></span>(imm_ov1, <span class="st">"mds"</span>)</a></code></pre></div>
<pre><code>## Standard deviations (1, .., p=4):
## [1] 0 0 0 0
## 
## Rotation (n x k) = (12 x 2):
##                [,1]       [,2]
## A2-i129 -20.2308709 -22.431389
## A2-i131   8.3055445  26.779321
## A2-i133  45.9341813   6.893304
## A2-i132 -55.0903957  18.572513
## A4-i191  23.7461189   1.118162
## A4-i192  -4.4041243 -38.028858
## MS1     -19.5494165  12.836320
## MS2      -1.9063188   6.075283
## MS3      -9.8321059 -11.217724
## MS4       0.9127103  -1.154627
## MS5       5.6552254  27.415676
## MS6      26.4594518 -26.857981</code></pre>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="co"># "tsne" - t-Stochastic Neighbor Embedding</span></a>
<a class="sourceLine" id="cb6-2" data-line-number="2"><span class="kw"><a href="../../reference/repOverlapAnalysis.html">repOverlapAnalysis</a></span>(imm_ov1, <span class="st">"tsne"</span>)</a></code></pre></div>
<pre><code>##                DimI      DimII
## A2-i129  12.2697768  -72.98788
## A2-i131 -40.4181005 -349.95546
## A2-i133  15.9122415  175.00404
## A2-i132   6.2753698  156.02270
## A4-i191  18.6706749  161.66663
## A4-i192   5.6982493  -84.26917
## MS1      -0.5272897  -70.40293
## MS2       4.3843790  168.91833
## MS3       7.2157768  -75.25710
## MS4       9.6924766  166.22368
## MS5     -35.6424261 -347.90316
## MS6      -3.5311284  172.94031
## attr(,"class")
## [1] "matrix"      "immunr_tsne"</code></pre>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="co"># Visualise the results</span></a>
<a class="sourceLine" id="cb8-2" data-line-number="2"><span class="kw"><a href="../../reference/vis.html">vis</a></span>(<span class="kw"><a href="../../reference/repOverlapAnalysis.html">repOverlapAnalysis</a></span>(imm_ov1, <span class="st">"mds"</span>))</a></code></pre></div>
<p><img src="v4_overlap_files/figure-html/overlap-1-1.png" width="768" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1"><span class="co"># Apply different analysis algorithms to the matrix of public clonotypes:</span></a>
<a class="sourceLine" id="cb9-2" data-line-number="2"><span class="co"># "mds" - Multi-dimensional Scaling</span></a>
<a class="sourceLine" id="cb9-3" data-line-number="3"><span class="kw"><a href="../../reference/repOverlapAnalysis.html">repOverlapAnalysis</a></span>(imm_ov1, <span class="st">"mds"</span>)</a></code></pre></div>
<pre><code>## Standard deviations (1, .., p=4):
## [1] 0 0 0 0
## 
## Rotation (n x k) = (12 x 2):
##                [,1]       [,2]
## A2-i129 -20.2308709 -22.431389
## A2-i131   8.3055445  26.779321
## A2-i133  45.9341813   6.893304
## A2-i132 -55.0903957  18.572513
## A4-i191  23.7461189   1.118162
## A4-i192  -4.4041243 -38.028858
## MS1     -19.5494165  12.836320
## MS2      -1.9063188   6.075283
## MS3      -9.8321059 -11.217724
## MS4       0.9127103  -1.154627
## MS5       5.6552254  27.415676
## MS6      26.4594518 -26.857981</code></pre>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="co"># "tsne" - t-Stochastic Neighbor Embedding</span></a>
<a class="sourceLine" id="cb11-2" data-line-number="2"><span class="kw"><a href="../../reference/repOverlapAnalysis.html">repOverlapAnalysis</a></span>(imm_ov1, <span class="st">"tsne"</span>)</a></code></pre></div>
<pre><code>##               DimI       DimII
## A2-i129  151.07971   6.4235410
## A2-i131 -212.85491  33.7630904
## A2-i133  -40.78513 -11.0785740
## A2-i132  -24.57136 -18.2749108
## A4-i191  -35.54737 -21.1417265
## A4-i192  160.23235   0.5933717
## MS1      148.48450  -4.1120462
## MS2      -30.14214  -8.9633306
## MS3      152.78470   2.1650050
## MS4      -32.14518 -13.4768715
## MS5     -209.39075  36.3082597
## MS6      -27.14442  -2.2058083
## attr(,"class")
## [1] "matrix"      "immunr_tsne"</code></pre>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1"><span class="co"># Visualise the results</span></a>
<a class="sourceLine" id="cb13-2" data-line-number="2"><span class="kw"><a href="../../reference/vis.html">vis</a></span>(<span class="kw"><a href="../../reference/repOverlapAnalysis.html">repOverlapAnalysis</a></span>(imm_ov1, <span class="st">"mds"</span>))</a></code></pre></div>
<p><img src="v4_overlap_files/figure-html/overlap-2-1.png" width="960" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1"><span class="co"># Clusterise the MDS resulting components using K-means</span></a>
<a class="sourceLine" id="cb14-2" data-line-number="2"><span class="kw"><a href="../../reference/vis.html">vis</a></span>(<span class="kw"><a href="../../reference/repOverlapAnalysis.html">repOverlapAnalysis</a></span>(imm_ov1, <span class="st">"mds+kmeans"</span>))</a></code></pre></div>
<p><img src="v4_overlap_files/figure-html/overlap-2-2.png" width="960" style="display: block; margin: auto;"></p>
</div>
<div id="public-repertoire" class="section level1">
<h1 class="hasAnchor">
<a href="#public-repertoire" class="anchor"></a>Public repertoire</h1>
<p>In order to build a massive table with all clonotypes from the list of repertoires use the <code>pubRep</code> function.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1"><span class="co"># Pass "nt" as the second parameter to build the public repertoire table using CDR3 nucleotide sequences</span></a>
<a class="sourceLine" id="cb15-2" data-line-number="2">pr.nt =<span class="st"> </span><span class="kw"><a href="../../reference/pubRep.html">pubRep</a></span>(immdata<span class="op">$</span>data, <span class="st">"nt"</span>, <span class="dt">.verbose =</span> F)</a>
<a class="sourceLine" id="cb15-3" data-line-number="3">pr.nt</a></code></pre></div>
<pre><code>##                                                    CDR3.nt Samples A2-i129
##     1:                   TGCGCCAGCAGCTTGGAAGAGACCCAGTACTTC       8       2
##     2:                   TGTGCCAGCAGCTTCCAAGAGACCCAGTACTTC       7      NA
##     3:                   TGTGCCAGCAGTTACCAAGAGACCCAGTACTTC       7       1
##     4:                   TGCGCCAGCAGCTTCCAAGAGACCCAGTACTTC       6       2
##     5:                      TGTGCCAGCAGCCAAGAGACCCAGTACTTC       6       5
##    ---                                                                    
## 86979:             TGTGCTTCACAACTCTTATTGGACGAGACCCAGTACTTC       1      NA
## 86980: TGTGCTTCACAAGCCCTACAGGGCACTTTCCATAATTCACCCCTCCACTTT       1      NA
## 86981:                   TGTGCTTCAGGGCGGGCCTACGAGCAGTACTTC       1      NA
## 86982:             TGTGCTTCCGCCGGACCGGACCGGGAGACCCAGTACTTC       1      NA
## 86983:                TGTGCTTGCGGGACAGATAACTATGGCTACACCTTC       1      NA
##        A2-i131 A2-i133 A2-i132 A4-i191 A4-i192 MS1 MS2 MS3 MS4 MS5 MS6
##     1:      NA       2       1      NA       1  NA  NA   1   1   1   1
##     2:       1       1       2       1      NA   2  NA  NA   2  NA   1
##     3:       1       1      NA       1       1   1  NA   2  NA  NA  NA
##     4:      NA       1       1      NA      NA  NA   1  NA   1  NA   1
##     5:       3      NA       2       3       2  NA  NA  NA  NA   5  NA
##    ---                                                                
## 86979:       1      NA      NA      NA      NA  NA  NA  NA  NA  NA  NA
## 86980:      NA      NA      NA      NA      NA  NA  NA  NA  NA   1  NA
## 86981:      NA      NA      NA      NA      NA   1  NA  NA  NA  NA  NA
## 86982:      NA       1      NA      NA      NA  NA  NA  NA  NA  NA  NA
## 86983:      NA      NA      NA      NA       1  NA  NA  NA  NA  NA  NA</code></pre>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1"><span class="co"># Pass "aa+v" as the second parameter to build the public repertoire table using CDR3 aminoacid sequences and V alleles</span></a>
<a class="sourceLine" id="cb17-2" data-line-number="2"><span class="co"># In order to use only CDR3 aminoacid sequences, just pass "aa"</span></a>
<a class="sourceLine" id="cb17-3" data-line-number="3">pr.aav =<span class="st"> </span><span class="kw"><a href="../../reference/pubRep.html">pubRep</a></span>(immdata<span class="op">$</span>data, <span class="st">"aa+v"</span>, <span class="dt">.verbose =</span> F)</a>
<a class="sourceLine" id="cb17-4" data-line-number="4">pr.aav</a></code></pre></div>
<pre><code>##                    CDR3.aa   V.name Samples A2-i129 A2-i131 A2-i133
##     1:         CASSLEETQYF  TRBV5-1       8       2      NA       3
##     2:     CASSDSSGGANEQFF  TRBV6-4       6       1       1       2
##     3:     CASSDSSGSTDTQYF  TRBV6-4       6      NA      NA      NA
##     4:         CASSFQETQYF  TRBV5-1       6       3      NA       1
##     5:         CASSLGETQYF TRBV12-4       6       2      NA      NA
##    ---                                                             
## 86181:     CTSSRPTQGAYEQYF  TRBV7-2       1      NA      NA      NA
## 86182:    CTSSSRAGAGTDTQYF  TRBV7-2       1      NA      NA      NA
## 86183: CTSSYPGLAGLKRKETQYF  TRBV7-2       1      NA      NA      NA
## 86184:    CTSSYRQRPYQETQYF  TRBV7-2       1      NA      NA      NA
## 86185:      CTSSYSTSGVGQFF  TRBV7-2       1      NA      NA      NA
##        A2-i132 A4-i191 A4-i192 MS1 MS2 MS3 MS4 MS5 MS6
##     1:       1      NA       2  NA  NA   1   1   1   1
##     2:      NA       5      NA  NA  NA   2  NA  NA  15
##     3:       4       1       1  NA  NA   1   1  NA   2
##     4:       1      NA      NA  NA   1  NA   1  NA   1
##     5:       4       3      NA   1  NA  NA  NA   2   3
##    ---                                                
## 86181:      NA      NA      NA  NA  NA  NA  NA  NA   1
## 86182:      NA      NA      NA  NA  NA   1  NA  NA  NA
## 86183:       1      NA      NA  NA  NA  NA  NA  NA  NA
## 86184:      NA      NA      NA  NA  NA   1  NA  NA  NA
## 86185:      NA      NA      NA  NA  NA  NA   1  NA  NA</code></pre>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1"><span class="co"># You can also pass the ".coding" parameter to filter out all noncoding sequences first:</span></a>
<a class="sourceLine" id="cb19-2" data-line-number="2">pr.aav.cod =<span class="st"> </span><span class="kw"><a href="../../reference/pubRep.html">pubRep</a></span>(immdata<span class="op">$</span>data, <span class="st">"aa+v"</span>, <span class="dt">.coding=</span>T)</a></code></pre></div>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1"><span class="co"># Create a public repertoire with coding-only sequences using both CDR3 amino acid sequences and V genes</span></a>
<a class="sourceLine" id="cb20-2" data-line-number="2">pr =<span class="st"> </span><span class="kw"><a href="../../reference/pubRep.html">pubRep</a></span>(immdata<span class="op">$</span>data, <span class="st">"aa+v"</span>, <span class="dt">.coding =</span> T, <span class="dt">.verbose =</span> F)</a>
<a class="sourceLine" id="cb20-3" data-line-number="3"></a>
<a class="sourceLine" id="cb20-4" data-line-number="4"><span class="co"># Apply the filter subroutine to leave clonotypes presented only in healthy individuals</span></a>
<a class="sourceLine" id="cb20-5" data-line-number="5">pr1 =<span class="st"> </span><span class="kw"><a href="../../reference/pubRepFilter.html">pubRepFilter</a></span>(pr, immdata<span class="op">$</span>meta, <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dt">Status =</span> <span class="st">"C"</span>))</a>
<a class="sourceLine" id="cb20-6" data-line-number="6"></a>
<a class="sourceLine" id="cb20-7" data-line-number="7"><span class="co"># Apply the filter subroutine to leave clonotypes presented only in diseased individuals</span></a>
<a class="sourceLine" id="cb20-8" data-line-number="8">pr2 =<span class="st"> </span><span class="kw"><a href="../../reference/pubRepFilter.html">pubRepFilter</a></span>(pr, immdata<span class="op">$</span>meta, <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dt">Status =</span> <span class="st">"MS"</span>))</a>
<a class="sourceLine" id="cb20-9" data-line-number="9"></a>
<a class="sourceLine" id="cb20-10" data-line-number="10"><span class="co"># Divide one by another</span></a>
<a class="sourceLine" id="cb20-11" data-line-number="11">pr3 =<span class="st"> </span><span class="kw"><a href="../../reference/pubRepApply.html">pubRepApply</a></span>(pr1, pr2)</a>
<a class="sourceLine" id="cb20-12" data-line-number="12"></a>
<a class="sourceLine" id="cb20-13" data-line-number="13"><span class="co"># Plot it</span></a>
<a class="sourceLine" id="cb20-14" data-line-number="14">p =<span class="st"> </span><span class="kw">ggplot</span>() <span class="op">+</span><span class="st"> </span><span class="kw">geom_jitter</span>(<span class="kw">aes</span>(<span class="dt">x =</span> <span class="st">"Treatment"</span>, <span class="dt">y =</span> Result), <span class="dt">data=</span>pr3)</a>
<a class="sourceLine" id="cb20-15" data-line-number="15">p</a></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#repertoire-overlap">Repertoire overlap</a></li>
      <li><a href="#public-repertoire">Public repertoire</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Vadim I. Nazarov, Vasily O. Tsvetkov, Eugene Rumynskiy, Anna Lorenc, Daniel J. Moore, Victor Greiff, <a href="https://immunomind.io">ImmunoMind</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '94207562f34b455c0796790d104e3549',
    indexName: 'immunarch',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
