<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Repertoire overlap and public clonotypes • immunarch</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><!-- docsearch --><script src="../../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><link href="../../extra.css" rel="stylesheet">
<meta property="og:title" content="Repertoire overlap and public clonotypes">
<meta property="og:description" content="immunarch">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-128418614-2"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-128418614-2');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">immunarch</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.7.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Start Here
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/v1_introduction.html">Installation &amp; troubleshooting</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">First steps</li>
    <li>
      <a href="../../articles/v2_data.html">Data loading</a>
    </li>
    <li>
      <a href="../../articles/web_only/load_mixcr.html">How-to: Loading MiXCR Data</a>
    </li>
    <li>
      <a href="../../articles/web_only/load_10x.html">How-to: Loading 10x Genomics Data</a>
    </li>
    <li>
      <a href="../../articles/web_only/v21_singlecell.html">How-to: Single-cell and paired chain data</a>
    </li>
    <li>
      <a href="../../articles/web_only/v3_basic_analysis.html">Basic statistics and clonality</a>
    </li>
    <li>
      <a href="../../articles/web_only/repFilter_v3.html">Data filtering</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Repertoire-level exploration and comparison</li>
    <li>
      <a href="../../articles/web_only/v4_overlap.html">Repertoire overlap and public clonotypes</a>
    </li>
    <li>
      <a href="../../articles/web_only/v5_gene_usage.html">Gene usage</a>
    </li>
    <li>
      <a href="../../articles/web_only/v6_diversity.html">Diversity estimation</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Clonotype-level exploration and comparison</li>
    <li>
      <a href="../../articles/web_only/v8_tracking.html">Track clonotypes across samples and time</a>
    </li>
    <li>
      <a href="../../articles/web_only/v11_db.html">Annotate clonotypes using immune receptor databases</a>
    </li>
    <li>
      <a href="../../articles/web_only/v9_kmers.html">Kmer and sequence motif analysis and visualisation</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Preparing to publication</li>
    <li>
      <a href="../../articles/web_only/v7_fixvis.html">Make your plots publication-ready with fixVis</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">BCR analysis</li>
    <li>
      <a href="../../articles/web_only/BCRpipeline.html">BCR pipeline</a>
    </li>
    <li>
      <a href="../../articles/web_only/clustering.html">Clustering</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
<li>
  <a href="https://github.com/immunomind/covid19" class="external-link">COVID-19</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://immunomind.io" class="external-link">ImmunoMind</a>
</li>
<li>
  <a href="http://twitter.com/immunomind" class="external-link">
    <span class="fa fa-lg fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/immunomind/immunarch" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Repertoire overlap and public clonotypes</h1>
                        <h4 data-toc-skip class="author">
<b>ImmunoMind</b> – improving design of T-cell therapies using multi-omics and AI. Research and biopharma partnerships, more details: <a href="https://immunomind.io" class="external-link">immunomind.io</a>
</h4>
            
            <h4 data-toc-skip class="date"><a href="mailto:support@immunomind.io" class="email">support@immunomind.io</a></h4>
      
      <small class="dont-index">Source: <a href="https://github.com/immunomind/immunarch/blob/HEAD/vignettes/web_only/v4_overlap.Rmd" class="external-link"><code>vignettes/web_only/v4_overlap.Rmd</code></a></small>
      <div class="hidden name"><code>v4_overlap.Rmd</code></div>

    </div>

    
    
<!--
%\VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{Repertoire overlap and public clonotypes}
%\VignettePackage{immunarch}
-->
<div class="section level2">
<h2 id="repertoire-overlap">Repertoire overlap<a class="anchor" aria-label="anchor" href="#repertoire-overlap"></a>
</h2>
<p>Repertoire overlap is the most common approach to measure repertoire similarity. It is achieved by computation of specific statistics on clonotypes shared between given repertoires, also called “public” clonotypes. <code>immunarch</code> provides several indices: - number of public clonotypes (<code>.method = "public"</code>) - a classic measure of overlap similarity.</p>
<ul>
<li><p>overlap coefficient (<code>.method = "overlap"</code>) - a normalised measure of overlap similarity. It is defined as the size of the intersection divided by the smaller of the size of the two sets.</p></li>
<li><p>Jaccard index (<code>.method = "jaccard"</code>) - measures the similarity between finite sample sets, and is defined as the size of the intersection divided by the size of the union of the sample sets.</p></li>
<li><p>Tversky index (<code>.method = "tversky"</code>) - an asymmetric similarity measure on sets that compares a variant to a prototype. If using default arguments, it’s similar to Dice’s coefficient.</p></li>
<li><p>cosine similarity (<code>.method = "cosine"</code>) - a measure of similarity between two non-zero vectors</p></li>
<li><p>Morisita’s overlap index (<code>.method = "morisita"</code>) - a statistical measure of dispersion of individuals in a population. It is used to compare overlap among samples.</p></li>
<li><p>incremental overlap - overlaps of the N most abundant clonotypes with incrementally growing N (<code>.method = "inc+METHOD"</code>, e.g., <code>"inc+public"</code> or <code>"inc+morisita"</code>).</p></li>
</ul>
<p>The function that includes described methods is <code>repOverlap</code>. Again, the output is easily visualised when passed to <code><a href="../../reference/vis.html">vis()</a></code> function that does all the work:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">imm_ov1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/repOverlap.html">repOverlap</a></span><span class="op">(</span><span class="va">immdata</span><span class="op">$</span><span class="va">data</span>, .method <span class="op">=</span> <span class="st">"public"</span>, .verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="va">imm_ov2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/repOverlap.html">repOverlap</a></span><span class="op">(</span><span class="va">immdata</span><span class="op">$</span><span class="va">data</span>, .method <span class="op">=</span> <span class="st">"morisita"</span>, .verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/vis.html">vis</a></span><span class="op">(</span><span class="va">imm_ov1</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/vis.html">vis</a></span><span class="op">(</span><span class="va">imm_ov2</span>, .text.size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></span></code></pre></div>
<p><img src="v4_overlap_files/figure-html/overlap-1.png" width="1152" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/vis.html">vis</a></span><span class="op">(</span><span class="va">imm_ov1</span>, <span class="st">"heatmap2"</span><span class="op">)</span></span></code></pre></div>
<p><img src="v4_overlap_files/figure-html/overlap-2.png" width="1152" style="display: block; margin: auto;"></p>
<p>You can easily change the number of significant digits:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/vis.html">vis</a></span><span class="op">(</span><span class="va">imm_ov2</span>, .text.size <span class="op">=</span> <span class="fl">2.5</span>, .signif.digits <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/vis.html">vis</a></span><span class="op">(</span><span class="va">imm_ov2</span>, .text.size <span class="op">=</span> <span class="fl">2</span>, .signif.digits <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></span></code></pre></div>
<p><img src="v4_overlap_files/figure-html/overlap-signif-digits-1.png" width="1152" style="display: block; margin: auto;"></p>
<!---
Top-overlap [Work in Progress]

```r
warning("TODO")
```

```
## Warning: TODO
```
-->
<p>To analyse the computed overlap measures function apply <code>repOverlapAnalysis</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Apply different analysis algorithms to the matrix of public clonotypes:</span></span>
<span><span class="co"># "mds" - Multi-dimensional Scaling</span></span>
<span><span class="fu"><a href="../../reference/repOverlapAnalysis.html">repOverlapAnalysis</a></span><span class="op">(</span><span class="va">imm_ov1</span>, <span class="st">"mds"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Standard deviations (1, .., p=4):</span></span>
<span><span class="co">## [1] 0 0 0 0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Rotation (n x k) = (12 x 2):</span></span>
<span><span class="co">##                [,1]       [,2]</span></span>
<span><span class="co">## A2-i129 -18.7767715 -18.360817</span></span>
<span><span class="co">## A2-i131  29.9586985  -7.870441</span></span>
<span><span class="co">## A2-i133  28.1148594  22.629093</span></span>
<span><span class="co">## A2-i132 -44.3435640   6.221812</span></span>
<span><span class="co">## A4-i191  13.8586515   7.452149</span></span>
<span><span class="co">## A4-i192 -14.0065477  27.068830</span></span>
<span><span class="co">## MS1      -8.8469009  -8.151574</span></span>
<span><span class="co">## MS2      -0.9712073  -1.297017</span></span>
<span><span class="co">## MS3     -10.4398629   4.894354</span></span>
<span><span class="co">## MS4       0.5131505  10.471309</span></span>
<span><span class="co">## MS5      18.5153823 -12.628029</span></span>
<span><span class="co">## MS6       6.4241122 -30.429669</span></span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># "tsne" - t-Stochastic Neighbor Embedding</span></span>
<span><span class="fu"><a href="../../reference/repOverlapAnalysis.html">repOverlapAnalysis</a></span><span class="op">(</span><span class="va">imm_ov1</span>, <span class="st">"tsne"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##               DimI      DimII</span></span>
<span><span class="co">## A2-i129   70.38459   58.97613</span></span>
<span><span class="co">## A2-i131  110.54482 -290.06614</span></span>
<span><span class="co">## A2-i133  -43.82199   48.33251</span></span>
<span><span class="co">## A2-i132 -136.26217   46.97332</span></span>
<span><span class="co">## A4-i191  -93.62640  112.94984</span></span>
<span><span class="co">## A4-i192   41.05285   16.44851</span></span>
<span><span class="co">## MS1       92.71964   18.66990</span></span>
<span><span class="co">## MS2     -111.06691   66.03547</span></span>
<span><span class="co">## MS3       63.65165   39.73049</span></span>
<span><span class="co">## MS4      -97.60649   81.58552</span></span>
<span><span class="co">## MS5       91.28411 -290.93060</span></span>
<span><span class="co">## MS6       12.74629   91.29504</span></span>
<span><span class="co">## attr(,"class")</span></span>
<span><span class="co">## [1] "immunr_tsne" "matrix"      "array"</span></span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualise the results</span></span>
<span><span class="fu"><a href="../../reference/repOverlapAnalysis.html">repOverlapAnalysis</a></span><span class="op">(</span><span class="va">imm_ov1</span>, <span class="st">"mds"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="../../reference/vis.html">vis</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="v4_overlap_files/figure-html/overlap-1-1.png" width="768" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Apply different analysis algorithms to the matrix of public clonotypes:</span></span>
<span><span class="co"># "mds" - Multi-dimensional Scaling</span></span>
<span><span class="fu"><a href="../../reference/repOverlapAnalysis.html">repOverlapAnalysis</a></span><span class="op">(</span><span class="va">imm_ov1</span>, <span class="st">"mds"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Standard deviations (1, .., p=4):</span></span>
<span><span class="co">## [1] 0 0 0 0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Rotation (n x k) = (12 x 2):</span></span>
<span><span class="co">##                [,1]       [,2]</span></span>
<span><span class="co">## A2-i129 -18.7767715 -18.360817</span></span>
<span><span class="co">## A2-i131  29.9586985  -7.870441</span></span>
<span><span class="co">## A2-i133  28.1148594  22.629093</span></span>
<span><span class="co">## A2-i132 -44.3435640   6.221812</span></span>
<span><span class="co">## A4-i191  13.8586515   7.452149</span></span>
<span><span class="co">## A4-i192 -14.0065477  27.068830</span></span>
<span><span class="co">## MS1      -8.8469009  -8.151574</span></span>
<span><span class="co">## MS2      -0.9712073  -1.297017</span></span>
<span><span class="co">## MS3     -10.4398629   4.894354</span></span>
<span><span class="co">## MS4       0.5131505  10.471309</span></span>
<span><span class="co">## MS5      18.5153823 -12.628029</span></span>
<span><span class="co">## MS6       6.4241122 -30.429669</span></span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># "tsne" - t-Stochastic Neighbor Embedding</span></span>
<span><span class="fu"><a href="../../reference/repOverlapAnalysis.html">repOverlapAnalysis</a></span><span class="op">(</span><span class="va">imm_ov1</span>, <span class="st">"tsne"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                DimI      DimII</span></span>
<span><span class="co">## A2-i129   -5.305477  -37.43846</span></span>
<span><span class="co">## A2-i131  274.779200 -292.65370</span></span>
<span><span class="co">## A2-i133  -14.075846   38.29146</span></span>
<span><span class="co">## A2-i132  -96.198030  118.65035</span></span>
<span><span class="co">## A4-i191  -37.279125  182.28853</span></span>
<span><span class="co">## A4-i192  -61.883983  -45.17148</span></span>
<span><span class="co">## MS1      -28.570690  -88.57819</span></span>
<span><span class="co">## MS2      -65.103850  133.62756</span></span>
<span><span class="co">## MS3      -26.840462  -47.50085</span></span>
<span><span class="co">## MS4      -44.543856  146.92722</span></span>
<span><span class="co">## MS5      268.942408 -270.57046</span></span>
<span><span class="co">## MS6     -163.920290  162.12801</span></span>
<span><span class="co">## attr(,"class")</span></span>
<span><span class="co">## [1] "immunr_tsne" "matrix"      "array"</span></span></code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualise the results</span></span>
<span><span class="fu"><a href="../../reference/repOverlapAnalysis.html">repOverlapAnalysis</a></span><span class="op">(</span><span class="va">imm_ov1</span>, <span class="st">"mds"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="../../reference/vis.html">vis</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="v4_overlap_files/figure-html/overlap-2-1.png" width="960" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Clusterise the MDS resulting components using K-means</span></span>
<span><span class="fu"><a href="../../reference/repOverlapAnalysis.html">repOverlapAnalysis</a></span><span class="op">(</span><span class="va">imm_ov1</span>, <span class="st">"mds+kmeans"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="../../reference/vis.html">vis</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="v4_overlap_files/figure-html/overlap-2-2.png" width="960" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="public-repertoire">Public repertoire<a class="anchor" aria-label="anchor" href="#public-repertoire"></a>
</h2>
<p>In order to build a massive table with all clonotypes from the list of repertoires use the <code>pubRep</code> function.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Pass "nt" as the second parameter to build the public repertoire table using CDR3 nucleotide sequences</span></span>
<span><span class="va">pr.nt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/pubRep.html">pubRep</a></span><span class="op">(</span><span class="va">immdata</span><span class="op">$</span><span class="va">data</span>, <span class="st">"nt"</span>, .verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="va">pr.nt</span></span></code></pre></div>
<pre><code><span><span class="co">##                                                    CDR3.nt Samples A2-i129</span></span>
<span><span class="co">##     1:                   TGCGCCAGCAGCTTGGAAGAGACCCAGTACTTC       8       1</span></span>
<span><span class="co">##     2:                   TGTGCCAGCAGCTTCCAAGAGACCCAGTACTTC       7      NA</span></span>
<span><span class="co">##     3:                   TGTGCCAGCAGTTACCAAGAGACCCAGTACTTC       7       1</span></span>
<span><span class="co">##     4:                   TGCGCCAGCAGCTTCCAAGAGACCCAGTACTTC       6       2</span></span>
<span><span class="co">##     5:                      TGTGCCAGCAGCCAAGAGACCCAGTACTTC       6       4</span></span>
<span><span class="co">##    ---                                                                    </span></span>
<span><span class="co">## 75101:             TGTGCTTCACAACTCTTATTGGACGAGACCCAGTACTTC       1      NA</span></span>
<span><span class="co">## 75102: TGTGCTTCACAAGCCCTACAGGGCACTTTCCATAATTCACCCCTCCACTTT       1      NA</span></span>
<span><span class="co">## 75103:                   TGTGCTTCAGGGCGGGCCTACGAGCAGTACTTC       1      NA</span></span>
<span><span class="co">## 75104:             TGTGCTTCCGCCGGACCGGACCGGGAGACCCAGTACTTC       1      NA</span></span>
<span><span class="co">## 75105:                TGTGCTTGCGGGACAGATAACTATGGCTACACCTTC       1      NA</span></span>
<span><span class="co">##        A2-i131 A2-i133 A2-i132 A4-i191 A4-i192 MS1 MS2 MS3 MS4 MS5 MS6</span></span>
<span><span class="co">##     1:      NA       1       1      NA       1  NA  NA   1   1   1   1</span></span>
<span><span class="co">##     2:       1       1       2       1      NA   1  NA  NA   2  NA   1</span></span>
<span><span class="co">##     3:       1       1      NA       1       1   1  NA   2  NA  NA  NA</span></span>
<span><span class="co">##     4:      NA       1       1      NA      NA  NA   1  NA   1  NA   1</span></span>
<span><span class="co">##     5:       2      NA       2       3       1  NA  NA  NA  NA   4  NA</span></span>
<span><span class="co">##    ---                                                                </span></span>
<span><span class="co">## 75101:       1      NA      NA      NA      NA  NA  NA  NA  NA  NA  NA</span></span>
<span><span class="co">## 75102:      NA      NA      NA      NA      NA  NA  NA  NA  NA   1  NA</span></span>
<span><span class="co">## 75103:      NA      NA      NA      NA      NA   1  NA  NA  NA  NA  NA</span></span>
<span><span class="co">## 75104:      NA       1      NA      NA      NA  NA  NA  NA  NA  NA  NA</span></span>
<span><span class="co">## 75105:      NA      NA      NA      NA       1  NA  NA  NA  NA  NA  NA</span></span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Pass "aa+v" as the second parameter to build the public repertoire table using CDR3 aminoacid sequences and V alleles</span></span>
<span><span class="co"># In order to use only CDR3 aminoacid sequences, just pass "aa"</span></span>
<span><span class="va">pr.aav</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/pubRep.html">pubRep</a></span><span class="op">(</span><span class="va">immdata</span><span class="op">$</span><span class="va">data</span>, <span class="st">"aa+v"</span>, .verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="va">pr.aav</span></span></code></pre></div>
<pre><code><span><span class="co">##                    CDR3.aa   V.name Samples A2-i129 A2-i131 A2-i133 A2-i132</span></span>
<span><span class="co">##     1:         CASSLEETQYF  TRBV5-1       8       1      NA       2       1</span></span>
<span><span class="co">##     2:     CASSDSSGGANEQFF  TRBV6-4       6       1       1       2      NA</span></span>
<span><span class="co">##     3:         CASSFQETQYF  TRBV5-1       6       3      NA       1       1</span></span>
<span><span class="co">##     4:         CASSLGETQYF TRBV12-4       6       2      NA      NA       4</span></span>
<span><span class="co">##     5:     CASSDSGGSYNEQFF  TRBV6-4       5      NA      NA      NA       3</span></span>
<span><span class="co">##    ---                                                                     </span></span>
<span><span class="co">## 74440:     CTSSRPTQGAYEQYF  TRBV7-2       1      NA      NA      NA      NA</span></span>
<span><span class="co">## 74441:    CTSSSRAGAGTDTQYF  TRBV7-2       1      NA      NA      NA      NA</span></span>
<span><span class="co">## 74442: CTSSYPGLAGLKRKETQYF  TRBV7-2       1      NA      NA      NA       1</span></span>
<span><span class="co">## 74443:    CTSSYRQRPYQETQYF  TRBV7-2       1      NA      NA      NA      NA</span></span>
<span><span class="co">## 74444:      CTSSYSTSGVGQFF  TRBV7-2       1      NA      NA      NA      NA</span></span>
<span><span class="co">##        A4-i191 A4-i192 MS1 MS2 MS3 MS4 MS5 MS6</span></span>
<span><span class="co">##     1:      NA       2  NA  NA   1   1   1   1</span></span>
<span><span class="co">##     2:       3      NA  NA  NA   2  NA  NA  12</span></span>
<span><span class="co">##     3:      NA      NA  NA   1  NA   1  NA   1</span></span>
<span><span class="co">##     4:       3      NA   1  NA  NA  NA   2   1</span></span>
<span><span class="co">##     5:      NA       1   1  NA   1  NA  NA   1</span></span>
<span><span class="co">##    ---                                        </span></span>
<span><span class="co">## 74440:      NA      NA  NA  NA  NA  NA  NA   1</span></span>
<span><span class="co">## 74441:      NA      NA  NA  NA   1  NA  NA  NA</span></span>
<span><span class="co">## 74442:      NA      NA  NA  NA  NA  NA  NA  NA</span></span>
<span><span class="co">## 74443:      NA      NA  NA  NA   1  NA  NA  NA</span></span>
<span><span class="co">## 74444:      NA      NA  NA  NA  NA   1  NA  NA</span></span></code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># You can also pass the ".coding" parameter to filter out all noncoding sequences first:</span></span>
<span><span class="va">pr.aav.cod</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/pubRep.html">pubRep</a></span><span class="op">(</span><span class="va">immdata</span><span class="op">$</span><span class="va">data</span>, <span class="st">"aa+v"</span>, .coding <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create a public repertoire with coding-only sequences using both CDR3 amino acid sequences and V genes</span></span>
<span><span class="va">pr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/pubRep.html">pubRep</a></span><span class="op">(</span><span class="va">immdata</span><span class="op">$</span><span class="va">data</span>, <span class="st">"aa+v"</span>, .coding <span class="op">=</span> <span class="cn">T</span>, .verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Apply the filter subroutine to leave clonotypes presented only in healthy individuals</span></span>
<span><span class="va">pr1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/pubRepFilter.html">pubRepFilter</a></span><span class="op">(</span><span class="va">pr</span>, <span class="va">immdata</span><span class="op">$</span><span class="va">meta</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>Status <span class="op">=</span> <span class="st">"C"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Apply the filter subroutine to leave clonotypes presented only in diseased individuals</span></span>
<span><span class="va">pr2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/pubRepFilter.html">pubRepFilter</a></span><span class="op">(</span><span class="va">pr</span>, <span class="va">immdata</span><span class="op">$</span><span class="va">meta</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>Status <span class="op">=</span> <span class="st">"MS"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Divide one by another</span></span>
<span><span class="va">pr3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/pubRepApply.html">pubRepApply</a></span><span class="op">(</span><span class="va">pr1</span>, <span class="va">pr2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot it</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html" class="external-link">geom_jitter</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Treatment"</span>, y <span class="op">=</span> <span class="va">Result</span><span class="op">)</span>, data <span class="op">=</span> <span class="va">pr3</span><span class="op">)</span></span>
<span><span class="va">p</span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://www.linkedin.com/in/vdnaz" class="external-link">Vadim I. Nazarov</a>, <a href="https://www.linkedin.com/in/vasily-tsvetkov-227218ab" class="external-link">Vasily O. Tsvetkov</a>, <a href="https://www.linkedin.com/in/erumynskiy/" class="external-link">Eugene Rumynskiy</a>, <a href="https://www.linkedin.com/in/aleksandr-popov-634878105/" class="external-link">Aleksandr A. Popov</a>, <a href="https://www.linkedin.com/in/ivan-balashov/" class="external-link">Ivan Balashov</a>, <a href="https://www.linkedin.com/in/maria-volobueva-b0856a223/" class="external-link">Maria Samokhina</a>, <a href="https://immunomind.io" class="external-link">ImmunoMind</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '94207562f34b455c0796790d104e3549',
    indexName: 'immunarch',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
