<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Tracking clonotypes across time points in `immunarch` • immunarch</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><!-- docsearch --><script src="../../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><link href="../../extra.css" rel="stylesheet">
<meta property="og:title" content="Tracking clonotypes across time points in `immunarch`">
<meta property="og:description" content="immunarch">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-128418614-2"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-128418614-2');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">immunarch</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.6.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="https://github.com/immunomind/covid19">COVID-19</a>
</li>
<li>
  <a href="../../articles/v1_introduction.html">Get started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">First steps</li>
    <li>
      <a href="../../articles/v1_introduction.html">Introduction to immunarch</a>
    </li>
    <li>
      <a href="../../articles/v2_data.html">Working with data in immunarch</a>
    </li>
    <li>
      <a href="../../articles/v3_basic_analysis.html">Basic analysis and clonality</a>
    </li>
    <li>
      <a href="../../articles/web_only/v21_singlecell.html">Single-cell and paired chain data</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Explore and compare repertoires</li>
    <li>
      <a href="../../articles/web_only/v4_overlap.html">Repertoire overlap and public clonotypes</a>
    </li>
    <li>
      <a href="../../articles/web_only/v5_gene_usage.html">Gene usage analysis</a>
    </li>
    <li>
      <a href="../../articles/web_only/v6_diversity.html">Diversity estimation</a>
    </li>
    <li>
      <a href="../../articles/web_only/v11_db.html">Annotate clonotypes using immune receptor databases</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Preparing to publication</li>
    <li>
      <a href="../../articles/web_only/v7_fixvis.html">Make your plots publication-ready with fixVis</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Advanced methods</li>
    <li>
      <a href="../../articles/web_only/v8_tracking.html">Tracking clonotypes across time points</a>
    </li>
    <li>
      <a href="../../articles/web_only/v9_kmers.html">Kmer and sequence motif analysis and visualisation</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://immunomind.io">ImmunoMind</a>
</li>
<li>
  <a href="http://twitter.com/immunomind">
    <span class="fa fa-lg fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/immunomind/immunarch">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Tracking clonotypes across time points in <code>immunarch</code>
</h1>
                        <h4 class="author"><a href="https://immunomind.io">ImmunoMind</a></h4>
            
            <h4 class="date"><a href="mailto:support@immunomind.io" class="email">support@immunomind.io</a></h4>
      
      <small class="dont-index">Source: <a href="https://github.com/immunomind/immunarch/blob/master/vignettes/web_only/v8_tracking.Rmd"><code>vignettes/web_only/v8_tracking.Rmd</code></a></small>
      <div class="hidden name"><code>v8_tracking.Rmd</code></div>

    </div>

    
    
<!--
%\VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{Tracking clonotypes across time points}
%\VignettePackage{immunarch}
-->
<div id="tracking-of-clonotypes" class="section level1">
<h1 class="hasAnchor">
<a href="#tracking-of-clonotypes" class="anchor"></a>Tracking of clonotypes</h1>
<p>Clonotype tracking is popular approach to monitor changes in frequency of clonotypes of interest in vaccination and cancer immunology. For example, a researcher can track a clonotype of across different time points in pre- and post-vaccination repertoires. Or analyse growth of malignant clonotypes in tumor sample.</p>
<p>Various methods of clonotype tracking are integrated into the one <code>trackClonotypes</code> function. Currently, there are three methods to choose from. The output of <code>trackClonotypes</code> can be immediately visualised with the <code>vis</code> function.</p>
<div id="tracking-the-most-abundant-clonotypes" class="section level2">
<h2 class="hasAnchor">
<a href="#tracking-the-most-abundant-clonotypes" class="anchor"></a>Tracking the most abundant clonotypes</h2>
<p>The simplest approach is to choose the most abundant clonotypes from one of the input immune repertoires and track across all immune repertoires in a batch. Arguments <code>.which</code> and <code>.col</code> are used to choose the immune repertoire, the number of clonotypes to take from it and which columns to use.</p>
<p>To choose the 10 most abundant clonotypes from the first repertoire and track them using their CDR3 nucleotide sequence:</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="no">tc1</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../../reference/trackClonotypes.html">trackClonotypes</a></span>(<span class="no">immdata</span>$<span class="no">data</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="fl">1</span>, <span class="fl">5</span>), <span class="kw">.col</span> <span class="kw">=</span> <span class="st">"nt"</span>)</pre></body></html></div>
<p>Value <code><a href="https://rdrr.io/r/base/list.html">list(1, 5)</a></code> of the <code>.which</code> argument (second argument) means to choose 10 clonotypes from the 1st repertoire in the input list of repertoires <code>immdata$data</code>. Value <code>"nt"</code> of the <code>.col</code> argument means that the function should take CDR3 nucleotide sequences only.</p>
<p>To choose the 10 most abundant amino acid clonotype sequences and their V genes from the “MS1” repertoire to track:</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="no">tc2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../../reference/trackClonotypes.html">trackClonotypes</a></span>(<span class="no">immdata</span>$<span class="no">data</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="st">"MS1"</span>, <span class="fl">10</span>), <span class="kw">.col</span> <span class="kw">=</span> <span class="st">"aa+v"</span>)</pre></body></html></div>
<p>Value <code><a href="https://rdrr.io/r/base/list.html">list("MS1", "10")</a></code> of the <code>.which</code> argument means to choose 10 clonotypes from the repertoire named “MS1” in the input list of repertoires <code>immdata$data</code>. Value <code>"aa+v"</code> of the <code>.col</code> argument means that the function should take both CDR3 amino acid sequences and V gene segments of the most abundant clonotypes.</p>
<p>Visualisation of both approaches:</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="no">p1</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../../reference/vis.html">vis</a></span>(<span class="no">tc1</span>)
<span class="no">p2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../../reference/vis.html">vis</a></span>(<span class="no">tc2</span>)

<span class="no">p1</span> / <span class="no">p2</span></pre></body></html></div>
<p><img src="v8_tracking_files/figure-html/unnamed-chunk-3-1.png" width="1152" style="display: block; margin: auto;"></p>
</div>
<div id="tracking-clonotypes-with-specific-nucleotide-or-amino-acid-sequences" class="section level2">
<h2 class="hasAnchor">
<a href="#tracking-clonotypes-with-specific-nucleotide-or-amino-acid-sequences" class="anchor"></a>Tracking clonotypes with specific nucleotide or amino acid sequences</h2>
<p>In order to track specific clonotype sequences, you can provide nucleotide or amino acid sequences as the <code>.which</code> argument, along with the column <code>.col</code> specifying in which columns to search for sequences. For example, to track seven CDR3 amino acid sequences specifyed below you need to execute the following code:</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="no">target</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"CASSLEETQYF"</span>, <span class="st">"CASSDSSGGANEQFF"</span>, <span class="st">"CASSDSSGSTDTQYF"</span>, <span class="st">"CASSLAGGYNEQFF"</span>, <span class="st">"CASSDSAGGTDTQYF"</span>, <span class="st">"CASSLDSYEQYF"</span>, <span class="st">"CASSSAGGYNEQFF"</span>)
<span class="no">tc</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../../reference/trackClonotypes.html">trackClonotypes</a></span>(<span class="no">immdata</span>$<span class="no">data</span>, <span class="no">target</span>, <span class="kw">.col</span> <span class="kw">=</span> <span class="st">"aa"</span>)
<span class="fu"><a href="../../reference/vis.html">vis</a></span>(<span class="no">tc</span>)</pre></body></html></div>
<p><img src="v8_tracking_files/figure-html/unnamed-chunk-4-1.png" width="1152" style="display: block; margin: auto;"></p>
</div>
<div id="tracking-clonotypes-with-specific-sequences-and-gene-segments" class="section level2">
<h2 class="hasAnchor">
<a href="#tracking-clonotypes-with-specific-sequences-and-gene-segments" class="anchor"></a>Tracking clonotypes with specific sequences and gene segments</h2>
<p>An improvement over the previous approach, it is possible to track clonotypes using information about both sequences and gene segments. Support you have a data frame of sequences with specific CDR3 sequences and gene segments. We will simulate this by choosing the 10 most abundant clonotypes from the first repertoire in the batch:</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="no">target</span> <span class="kw">&lt;-</span> <span class="no">immdata</span>$<span class="no">data</span><span class="kw">[[</span><span class="fl">1</span>]] <span class="kw">%&gt;%</span>
  <span class="fu">select</span>(<span class="no">CDR3.aa</span>, <span class="no">V.name</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fl">10</span>)

<span class="no">target</span></pre></body></html></div>
<pre><code>## # A tibble: 10 x 2
##    CDR3.aa           V.name 
##    &lt;chr&gt;             &lt;chr&gt;  
##  1 CASSQEGTGYSGELFF  TRBV4-1
##  2 CASSYRVGTDTQYF    TRBV4-1
##  3 CATSTNRGGTPADTQYF TRBV15 
##  4 CATSIGGGSYEQYF    TRBV15 
##  5 CASQGDSFNSPLHF    TRBV4-1
##  6 CASSPWTGSMALHF    TRBV27 
##  7 CASSQDMGGRNTGELFF TRBV4-1
##  8 CASSEEPRLFGYTF    TRBV2  
##  9 CASSQPGQGGGDEQFF  TRBV4-1
## 10 CASSWVARGPYEQYF   TRBV6-6</code></pre>
<p>Supply this data frame as an argument value to the <code>.which</code> argument to track target clonotypes:</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="no">tc</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../../reference/trackClonotypes.html">trackClonotypes</a></span>(<span class="no">immdata</span>$<span class="no">data</span>, <span class="no">target</span>)
<span class="fu"><a href="../../reference/vis.html">vis</a></span>(<span class="no">tc</span>)</pre></body></html></div>
<p><img src="v8_tracking_files/figure-html/unnamed-chunk-6-1.png" width="1152" style="display: block; margin: auto;"></p>
<p>Note that you can use any columns in the <code>target</code> data frame, such as both CDR3 nucleotide and amino acid sequences and any gene segments.</p>
</div>
</div>
<div id="visualisation-of-tracking" class="section level1">
<h1 class="hasAnchor">
<a href="#visualisation-of-tracking" class="anchor"></a>Visualisation of tracking</h1>
<p>There are three ways to visualise clonotype tracking, depending on your research and aesthetic needs. To choose the type of plot, you need to provide the <code>".plot"</code> parameter to the <code><a href="../../reference/vis.html">vis()</a></code> function, specifying one of three plot types: - <code>.plot = "smooth"</code> - used by default, a visualisation using smooth lines and stacked bar plots; - <code>.plot = "area"</code> - visualise abundances using areas under the abundance lines; - <code>.plot = "line"</code> - visualise only lines, connecting levels of abundances of a same clonotype between time points.</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="no">target</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"CASSLEETQYF"</span>, <span class="st">"CASSDSSGGANEQFF"</span>, <span class="st">"CASSDSSGSTDTQYF"</span>, <span class="st">"CASSLAGGYNEQFF"</span>, <span class="st">"CASSDSAGGTDTQYF"</span>, <span class="st">"CASSLDSYEQYF"</span>, <span class="st">"CASSSAGGYNEQFF"</span>)
<span class="no">tc</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../../reference/trackClonotypes.html">trackClonotypes</a></span>(<span class="no">immdata</span>$<span class="no">data</span>, <span class="no">target</span>, <span class="kw">.col</span> <span class="kw">=</span> <span class="st">"aa"</span>)
<span class="fu"><a href="../../reference/vis.html">vis</a></span>(<span class="no">tc</span>, <span class="kw">.plot</span> <span class="kw">=</span> <span class="st">"smooth"</span>)</pre></body></html></div>
<p><img src="v8_tracking_files/figure-html/unnamed-chunk-7-1.png" width="1152" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="fu"><a href="../../reference/vis.html">vis</a></span>(<span class="no">tc</span>, <span class="kw">.plot</span> <span class="kw">=</span> <span class="st">"area"</span>)</pre></body></html></div>
<p><img src="v8_tracking_files/figure-html/unnamed-chunk-7-2.png" width="1152" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="fu"><a href="../../reference/vis.html">vis</a></span>(<span class="no">tc</span>, <span class="kw">.plot</span> <span class="kw">=</span> <span class="st">"line"</span>)</pre></body></html></div>
<p><img src="v8_tracking_files/figure-html/unnamed-chunk-7-3.png" width="1152" style="display: block; margin: auto;"></p>
<div id="changing-the-order-of-samples" class="section level2">
<h2 class="hasAnchor">
<a href="#changing-the-order-of-samples" class="anchor"></a>Changing the order of samples</h2>
<p>The <code>.order</code> argument of the <code>vis</code> function controls the order of samples in the visualisation. You can pass either indices of samples you plan to visualise or sample names.</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="co"># Passing indices</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">immdata</span>$<span class="no">data</span>)[<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>, <span class="fl">3</span>, <span class="fl">5</span>)] <span class="co"># check sample names</span></pre></body></html></div>
<pre><code>## [1] "A2-i129" "A2-i133" "A4-i191"</code></pre>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="fu"><a href="../../reference/vis.html">vis</a></span>(<span class="no">tc</span>, <span class="kw">.order</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>, <span class="fl">3</span>, <span class="fl">5</span>))</pre></body></html></div>
<p><img src="v8_tracking_files/figure-html/unnamed-chunk-8-1.png" width="1152" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="co"># You can change the order</span>
<span class="fu"><a href="../../reference/vis.html">vis</a></span>(<span class="no">tc</span>, <span class="kw">.order</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">5</span>, <span class="fl">1</span>, <span class="fl">3</span>))</pre></body></html></div>
<p><img src="v8_tracking_files/figure-html/unnamed-chunk-8-2.png" width="1152" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="co"># Passing sample names</span>
<span class="fu"><a href="../../reference/vis.html">vis</a></span>(<span class="no">tc</span>, <span class="kw">.order</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"A2-i129"</span>, <span class="st">"A2-i133"</span>, <span class="st">"A4-i191"</span>))</pre></body></html></div>
<p><img src="v8_tracking_files/figure-html/unnamed-chunk-8-3.png" width="1152" style="display: block; margin: auto;"></p>
<p>If your metadata contains information about time such as timepoints for vaccination or tumor samples, you can use it to re-order samples accordingly. In our examples <code>immdata$meta</code> does not contain information about timepoints, so we will simulate this case.</p>
<p>First, we create an additional column in the metadata with randomly choosen timepoints:</p>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="no">immdata</span>$<span class="no">meta</span>$<span class="no">Timepoint</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="fl">1</span>:<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">immdata</span>$<span class="no">data</span>))
<span class="no">immdata</span>$<span class="no">meta</span></pre></body></html></div>
<pre><code>## # A tibble: 12 x 7
##    Sample  ID    Sex     Age Status Lane  Timepoint
##    &lt;chr&gt;   &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;     &lt;int&gt;
##  1 A2-i129 C1    M        11 C      A             7
##  2 A2-i131 C2    M         9 C      A            10
##  3 A2-i133 C4    M        16 C      A             8
##  4 A2-i132 C3    F         6 C      A            12
##  5 A4-i191 C8    F        22 C      B             6
##  6 A4-i192 C9    F        24 C      B             4
##  7 MS1     MS1   M        12 MS     C             1
##  8 MS2     MS2   M        30 MS     C             5
##  9 MS3     MS3   M         8 MS     C            11
## 10 MS4     MS4   F        14 MS     C             3
## 11 MS5     MS5   F        15 MS     C             2
## 12 MS6     MS6   F        15 MS     C             9</code></pre>
<p>Next, we create a vector with samples in the right order, according to the “Timepoint” column (from smallest to greatest):</p>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="no">sample_order</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span>(<span class="no">immdata</span>$<span class="no">meta</span>$<span class="no">Timepoint</span>)</pre></body></html></div>
<p>Sanity check: timepoints are following the right order:</p>
<div class="sourceCode" id="cb19"><html><body><pre class="r"><span class="no">immdata</span>$<span class="no">meta</span>$<span class="no">Timepoint</span>[<span class="no">sample_order</span>]</pre></body></html></div>
<pre><code>##  [1]  1  2  3  4  5  6  7  8  9 10 11 12</code></pre>
<p>Samples, sorted by the timepoints:</p>
<div class="sourceCode" id="cb21"><html><body><pre class="r"><span class="no">immdata</span>$<span class="no">meta</span>$<span class="no">Sample</span>[<span class="no">sample_order</span>]</pre></body></html></div>
<pre><code>##  [1] "MS1"     "MS5"     "MS4"     "A4-i192" "MS2"     "A4-i191" "A2-i129"
##  [8] "A2-i133" "MS6"     "A2-i131" "MS3"     "A2-i132"</code></pre>
<p>And finally, we visualise the data:</p>
<div class="sourceCode" id="cb23"><html><body><pre class="r"><span class="fu"><a href="../../reference/vis.html">vis</a></span>(<span class="no">tc</span>, <span class="kw">.order</span> <span class="kw">=</span> <span class="no">sample_order</span>)</pre></body></html></div>
<p><img src="v8_tracking_files/figure-html/unnamed-chunk-13-1.png" width="1152" style="display: block; margin: auto;"></p>
<p>It is possible to create a one-liner with the full pipeline from ordering to plotting:</p>
<div class="sourceCode" id="cb24"><html><body><pre class="r"><span class="fu"><a href="../../reference/vis.html">vis</a></span>(<span class="no">tc</span>, <span class="kw">.order</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span>(<span class="no">immdata</span>$<span class="no">meta</span>$<span class="no">Timepoint</span>))</pre></body></html></div>
<p><img src="v8_tracking_files/figure-html/unnamed-chunk-14-1.png" width="1152" style="display: block; margin: auto;"></p>
</div>
<div id="changing-the-colour-palette" class="section level2">
<h2 class="hasAnchor">
<a href="#changing-the-colour-palette" class="anchor"></a>Changing the colour palette</h2>
<p>If you want to change the colour palette, add a ggplot2 <code>scale_fill_*</code> function to the plot. We recommend using <code>scale_fill_brewer</code>:</p>
<div class="sourceCode" id="cb25"><html><body><pre class="r"><span class="fu"><a href="../../reference/vis.html">vis</a></span>(<span class="no">tc</span>) + <span class="fu">scale_fill_brewer</span>(<span class="kw">palette</span> <span class="kw">=</span> <span class="st">"Spectral"</span>)</pre></body></html></div>
<p><img src="v8_tracking_files/figure-html/unnamed-chunk-15-1.png" width="1152" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb26"><html><body><pre class="r"><span class="fu"><a href="../../reference/vis.html">vis</a></span>(<span class="no">tc</span>) + <span class="fu">scale_fill_brewer</span>(<span class="kw">palette</span> <span class="kw">=</span> <span class="st">"RdBu"</span>)</pre></body></html></div>
<p><img src="v8_tracking_files/figure-html/unnamed-chunk-15-2.png" width="1152" style="display: block; margin: auto;"></p>
<p>Run <code>?scale_fill_brewer</code> in the R console to learn more about ColorBrewer and it’s colour schemes.</p>
</div>
</div>
<div id="get-in-contact-with-us" class="section level1">
<h1 class="hasAnchor">
<a href="#get-in-contact-with-us" class="anchor"></a>Get in contact with us</h1>
<p>Can not find an important feature? Have a question or found a bug? Contact us at <a href="mailto:support@immunomind.io" class="email">support@immunomind.io</a></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Vadim I. Nazarov, Vasily O. Tsvetkov, Eugene Rumynskiy, <a href="https://immunomind.io">ImmunoMind</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '94207562f34b455c0796790d104e3549',
    indexName: 'immunarch',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
