---
title: "COVID-19 Immune Repertoire Analysis"
author: '<a href="https://immunomind.io">ImmunoMind</a>'
date: "support@immunomind.io"
output:
  html_document:
    fig_height: 8
    fig_width: 10
    theme: spacelab
    toc: yes
  pdf_document:
    toc: yes
  word_document:
    toc: yes
---


<!--
%\VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{COVID-19 Immune Repertoire Analysis}
%\VignettePackage{immunarch}
-->


  ```{r setup, include=FALSE, echo=FALSE}
# knitr::knit_hooks$set(optipng = knitr::hook_optipng)
# knitr::opts_chunk$set(optipng = '-o7')

knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.align = "center")
knitr::opts_chunk$set(fig.width = 12)
knitr::opts_chunk$set(fig.height = 5)

library(immunarch)

embed_png <- function(path, dpi = NULL) {
  meta <- attr(png::readPNG(path, native = TRUE, info = TRUE), "info")
  if (!is.null(dpi)) meta$dpi <- rep(dpi, 2)
  knitr::asis_output(paste0(
    "<img src='", path, "'",
    " width=", round(meta$dim[1] / (meta$dpi[1] / 96)),
    " height=", round(meta$dim[2] / (meta$dpi[2] / 96)),
    " />"
  ))
}
knitr::opts_chunk$set(comment = "#>", collapse = TRUE)
```
# Introduction
This is a tutorial on analyzing COVID-19 immune-repertoire data in `immunarch`. 

We will be analyzing data from <a href= "https://doi.org/10.1101/2020.05.18.100545">this</a> longitudinal study of T-cell dynamics in COVID-19 data from Minervina, et al. 

## Experiment

This figure from the paper shows the experimental setup. The authors analyzed two patients (male and female) whom they had samples from 2018 and 2019 before infection, and they collected samples from a few intervals (15, 30, 37, and 45 days) after the onset of symptoms.

```{r, echo = FALSE}
embed_png("../images/experimentpipeline.png")
```

## Prerequisites
**Storage requirement**: The full dataset includes 120 samples that are around ~1GB each. In this tutorial we used a subset of the data that is around 250MB. 

Follow the instructions to install the `immunarch` package [here](https://immunarch.com/articles/v1_introduction.html)


# Understand the Data

```{r, echo = FALSE}
embed_png("../images/pipeline-wide.png")
```

## Load the Data

There are several versions of the data that we can start with. You can access the raw data from Sequence Read Archive. However, it is recommended that you use the pre-processed data from Zenodo as linked below. It will save you a lot of time and pain! Below are links to the different versions of the data:

1. The [raw read data](https://immunarch.com/articles/load_sra.html) from Sequence Read Archive (120 GB)
2.  <a href="https://zenodo.org/record/3835956">Pre-processed</a> files with MiXCRÂ 
3. <a href="https://github.com/pogorely/Minervina_COVID">Clonotype tables</a> from final analysis

We will be using the pre-processed MiXCR data from the second step. 

Download the data from <a href="https://zenodo.org/record/3835956">Zenodo</a> and load the MiXCR data into immunarch. This step takes while, so use repSave to save the data into immunarch format so next time will be faster. While you're waiting, check out our new single cell tutorial <a href="https://immunarch.com/articles/web_only/v21_singlecell.html">here</a>!

```{r, echo = FALSE}
embed_png("../images/dataformatgraphic.png")
```

First, load the immunarch library. Replace the path with your path to the downloaded data from Zenodo. The `repSave` method below will help save some time so next time you can load the data quicker.

Set up your metadata file like below:

| **Sample**   |**Sex**|**Status**|**Day**|
|:-------------:|:-----:|:-----:|:--------:|
|M_15_F1_beta  |M      |I      |15         |
|M_30_F2_beta  |M      |I      |30         |
|M_F1_2019_beta|M      |I      |-1         |
|W_30_F1_beta  |F      |I      |30         |
|W_F1_2018_beta|F      |N      |-2         |
|W_F1_2019_beta|F      |N      |-1         |


```{r, eval=F}
# load the immunarch library
library(immunarch)
pbmc_beta_path <- "/path/to/covidtutorial/beta/pbmc/"
pbmc_beta <- repLoad(pbmc_beta_path)
repSave(pbmc_beta, "/path/to/covidtutorial/processeddata/beta/PBMC/")

```


# Explore the Data

Next, let's explore the data! Here are some methods that we will be using.

- [`repExplore`](https://immunarch.com/reference/repExplore.html): calculates the basic statistics of repertoire: the number of unique immune receptor clonotypes, their relative abundances, and sequence length distribution across the input dataset.

- [`vis`](https://immunarch.com/reference/vis.html): Output from every function in immunarch can be visualised with a single function - vis. The vis automatically detects the type of the data and draws a proper visualisation. 

For the next figure, we grouped by "I" -infected and "N" -pre-infection. On the right, it is further broken down by male and female. Notice how, overall, the number of unique clonotypes is much higher for the female patient than the male patient. In addition, the infected samples had much higher number of unique clonotypes than the non-infected samples.

```{r, eval=F}
exp_vol <- repExplore(pbmc_beta$data, .method = "volume")

# separate with metadata
p1 <- vis(exp_vol, .by = c("Status"), .meta=pbmc_beta$meta)
p2 <- vis(exp_vol, .by = c("Status", "Sex"), .meta = pbmc_beta$meta)

# graph both next to each other
p1 + p2

```
```{r, echo = FALSE}
embed_png("../images/num_unique_clonotypes.png")
```

```
p3
```
```{r, echo = FALSE}
embed_png("../images/clonotype_abundances.png")
```

```
p4
```
```{r, echo = FALSE}
embed_png("../images/cdr3lengths.png")
```

# Clonotypes

```{r, echo = FALSE}
embed_png("../images/pipe-2.png")
```

## Diversity 

One of the ways to estimate the diversity of samples is to evaluate clonality. repClonality measures the amount of the most or the least frequent clonotypes. In the graphs below, we used the rare method to find the least prolific clonotypes.

- [`repClonality`](https://immunarch.com/reference/repClonality.html): measures the amount of the most or the least frequent clonotypes. 

```{r, eval=F}
pbmcbeta_pr <- repClonality(pbmc_beta$data, .method = "clonal.prop")
pbmcbeta_pr

```
The top method considers the most abundant cell clonotypes:

```
               Clones Percentage Clonal.count.prop
M_15_F1_beta        3       11.4      5.718632e-06
M_15_F2_beta        3       11.0      8.327389e-06
M_30_F1_beta        3       10.8      5.368340e-06
M_30_F2_beta        3       10.8      5.688066e-06
M_37_F1_beta        3       12.2      6.503684e-06
M_37_F2_beta        3       11.9      6.261754e-06
M_45_F1_beta        3       10.9      4.384414e-06
M_45_F2_beta        3       11.0      3.960997e-06
M_F1_2018_beta      4       11.2      1.040742e-05
M_F1_2019_beta      3       11.7      9.901317e-06
W_15_F1_beta        8       10.2      9.739102e-06
W_15_F2_beta        8       10.1      9.975386e-06
W_30_F1_beta        7       10.1      7.704686e-06
W_30_F2_beta        9       10.1      9.405450e-06
W_37_F1_beta        7       10.2      7.253676e-06
W_37_F2_beta        7       10.2      7.313906e-06
W_45_F1_beta       50       10.0      5.355154e-05
W_45_F2_beta       53       10.0      4.930723e-05
W_F1_2018_beta      4       10.7      1.921340e-05
W_F1_2019_beta     15       10.1      3.921630e-05
attr(,"class")
[1] "immunr_clonal_prop" "matrix"      

```
```{r, eval=F}
vis(pbmcbeta_top) + vis(pbmcbeta_top, .by = "Status", .meta = pbmc_beta$meta)

```
```{r, echo = FALSE}
embed_png("../images/topclonalprop.png")
```

While the rare method deals with the least prolific clonotypes:

```{r, eval=F}
pbmcbeta_rare <- repClonality(pbmc_beta$data, .method = "rare")

vis(pbmcbeta_rare) + vis(pbmcbeta_rare, .by = "Status", .meta = pbmc_beta$meta)

```
```{r, echo = FALSE}
embed_png("../images/rare_clonal_prop.png")
```

The post-infection samples (Status = "I") have significantly lower proportion of "rare" clonotypes which have a count of 1. On the other hand, they have a higher proportion of clonotypes with more than 101 counts. The samples with the highest proportion of clonotypes with higher frequencies are on days 15 and 37, while the samples with the highest proportion of rare clonotypes (counts of 3 or less) are in the pre-infection timepoints.

```{r, eval=F}
pbmcbeta_hom <- repClonality(pbmc_beta$data,
                        .method = "homeo",
                        .clone.types = c(Small = .0001, Medium = .001, Large = .01, Hyperexpanded = 1)
)

vis(pbmcbeta_hom) + vis(pbmcbeta_hom, .by = c("Status", "Sex"), .meta = pbmc_beta$meta)

```
```{r, echo = FALSE}
embed_png("../images/relative_abundance.png")
```

## PCA
We can use PCA to analyze the differences in gene usage. 

- [`immunr_pca`](https://immunarch.com/reference/immunr_pca.html): perform Principal Component Analysis and analyze the differences in gene usage between the samples.

```{r, eval=F}
# Separate the data by sex, let's use the female samples here
females <- pbmc_beta$meta %>% filter(Sex=="F") %>% select(Sample)
pbmc_beta_f <- pbmc_beta$data[females$Sample]
data(pbmc_beta_f)

# calculate gene usage
gu <- geneUsage(pbmc_beta_f$data)

# replace NA values with zeros
gu[is.na(gu)] <- 0

# transform into matrix
gu <- t(as.matrix(gu[, -1]))

# perform pca and visualize
immunr_pca(gu) %>% vis()

```
```{r, echo = FALSE}
embed_png("../images/public_clonotype_pca_female.png")
```

There is a clear difference between the post-infection samples and the pre-infection samples on the first principal component! The pink dots on the positive side of the x-axis are the 2018 and 2019 samples, pre-infection.

## Tracking clonotypes over time

Clonotype tracking is popular approach for detecting changes in frequency of clonotypes of interest in vaccination and cancer immunology. For example, a researcher can track a clonotype of across different time points in pre- and post-vaccination repertoires, or analyze growth of malignant clonotypes in a tumor sample.

- [`trackClonotypes`](https://immunarch.com/reference/trackClonotypes.html): pick clonotypes of interest by amino acid sequences, nucleotide sequences, V gene segments, J gene segments, or a combination of the above.

Let's look at the top 30 clonotypes using their nucleotide sequences, and plot the proportions in order by date. 

```{r, eval=F}
females <- pbmc_beta$meta %>% filter(Sex=="F") %>% select(Sample)
pbmc_beta_f <- pbmc_beta$data[females$Sample]
pbmc_beta_f$data <- pbmc_beta_f
pbmc_beta_f$meta <- pbmc_beta$meta
tc1 <- trackClonotypes(pbmc_beta$data, list("W_15_F1_beta", 30), .col = "nt")
sample_order <- order(pbmc_beta_f$meta$Day)
p1 <- vis(tc1, .order = sample_order)
```

```{r, echo = FALSE}
embed_png("../images/clonotype_tracking.png")
```
It is interesting to see the dip in proportion from 2019 (still pre-infection) and another drop in proportion by the 45th day after onset of symptoms.

# Public Repertoire Analysis

```{r, echo = FALSE}
embed_png("../images/pipe-3.png")
```

Repertoire overlap is the most common approach to measure repertoire similarity. It is achieved by computation of specific statistics on clonotypes shared between given repertoires, also called "public" clonotypes.

- [`pubRep`](https://immunarch.com/reference/pubRep.html): plot the distribution of public clonotype frequencies
- [`other overlap methods`](https://immunarch.com/articles/web_only/v9_kmers.html): Jaccard index, cosine similarity, and Morisita's overlap index.

```{r, eval=F}
pbmc_beta_top_500 <- pbmc_beta_f
pbmc_beta_top_500$data <- lapply(pbmc_beta_f$data, head, 500)
pr <- pubRep(pbmc_beta_top_500$data, .verbose = FALSE)
vis(pr, "freq")
```
```{r, echo = FALSE}
embed_png("../images/public_clonotype_frequencies.png")
```

```{r, eval=F}
vis(pr, "freq", .type = "none")
```
```{r, echo = FALSE}
embed_png("../images/distrib_public_clonotypes.png")
```

Graph the biological replicates for day 15 against each other. They are closely correlated, as seen below.

```{r, eval=F}
vis(pr, "clonotypes", 1, 2)
```
```{r, echo = FALSE}
embed_png("../images/biologicalreplicates_pub.png")
```

Plot the clonotype frequencies of day 37 and day 15. Less similar than the biological replicates.

```{r, eval=F}
vis(pr, "clonotypes", 1, 5)
```
```{r, echo = FALSE}
embed_png("../images/37x15_pub.png")
```

Plot the pre-infection data against day 15. Difference is much more noticeable here.

```{r, eval=F}
vis(pr, "clonotypes", 1, 9)
```
```{r, echo = FALSE}
embed_png("../images/2018x15_pub.png")
```

# Amino Acid Sequence Motif Analysis

```{r, echo = FALSE}
embed_png("../images/pipe-4.png")
```

`immunarch` has several options for sequence motif analysis, we demonstrate some of the options including the position self-information matrices below.

- [`getKmers`](https://immunarch.com/reference/getKmers.html): Calculate the kmer statistics of immune repertoires
- [`kmer_profile`](https://immunarch.com/reference/split_to_kmers.html): Analysis of immune repertoire kmer statistics: sequence profiles, etc.

```{r, eval=F}
kmers <- getKmers(pbmc_beta_f$data, 5)
vis(kmers)
```
```{r, echo = FALSE}
embed_png("../images/kmersdistrib_full.png")
```

```{r, eval=F}
kmers <- getKmers(pbmc_beta_f$data[[1]], 5)
kmer_profile(kmers)
```
```{r, echo = FALSE}
embed_png("../images/kmers_distrib_3.png")
```

```{r, eval=F}
kp <- kmer_profile(kmers, "self")
p1 <- vis(kp)
p2 <- vis(kp, .plot = "seq")

p1 + p2
```
```{r, echo = FALSE}
embed_png("../images/kmers_vis.png")
```