<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Changelog • immunarch</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>


<!-- docsearch -->
<script src="../docsearch.js"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous" />
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script>



<meta property="og:title" content="Changelog" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-128418614-2"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-128418614-2');
</script>


  </head>

  <body>
    <div class="container template-news">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">immunarch</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.5.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/v1_introduction.html">Get started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">First steps</li>
    <li>
      <a href="../articles/v1_introduction.html">Introduction to immunarch</a>
    </li>
    <li>
      <a href="../articles/v2_data.html">Working with data in immunarch</a>
    </li>
    <li>
      <a href="../articles/v3_basic_analysis.html">Basic analysis and clonality</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Explore and compare repertoires</li>
    <li>
      <a href="../articles/web_only/v4_overlap.html">Repertoire overlap and public clonotypes</a>
    </li>
    <li>
      <a href="../articles/web_only/v5_gene_usage.html">Gene usage analysis</a>
    </li>
    <li>
      <a href="../articles/web_only/v6_diversity.html">Diversity estimation</a>
    </li>
    <li>
      <a href="../articles/web_only/v11_db.html">Annotate clonotypes using immune receptor databases</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Preparing to publication</li>
    <li>
      <a href="../articles/web_only/v7_fixvis.html">Make your plots publication-ready with fixVis</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Advanced methods</li>
    <li>
      <a href="../articles/web_only/v8_tracking.html">Tracking clonotypes across time points</a>
    </li>
    <li>
      <a href="../articles/web_only/v9_kmers.html">Kmer and sequence motif analysis and visualisation</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="https://github.com/immunomind/immunarch/releases">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://immunomind.io">ImmunoMind</a>
</li>
<li>
  <a href="http://twitter.com/immunomind">
    <span class="fa fa-lg fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/immunomind/immunarch">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
      
      <form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
        </div>
      </form>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1>Changelog <small></small></h1>
      <small>Source: <a href='https://github.com/immunomind/immunarch/blob/master/NEWS.md'><code>NEWS.md</code></a></small>
    </div>

    <div id="immunarch-0-4-3" class="section level1">
<h1 class="page-header">
<a href="#immunarch-0-4-3" class="anchor"></a>immunarch 0.4.3</h1>
<div id="documentation" class="section level2">
<h2 class="hasAnchor">
<a href="#documentation" class="anchor"></a>Documentation</h2>
<p>Make README more clear with the focus on installation.</p>
<p>Add a vignette for clonotype tracking.</p>
<p>Add a vignette for kmer analysis.</p>
</div>
<div id="work-in-progress" class="section level2">
<h2 class="hasAnchor">
<a href="#work-in-progress" class="anchor"></a>Work in progress</h2>
<p>Kmers - analysis + plots, seqlogo / textlogo</p>
<p><code>pubRepStatistics</code> function for computation of public clonotypes across groups and repertoires.</p>
<p><code>vis_public_frequencies</code>, <code>vis_public_clonotypes</code>, <code>vis_upset</code> for public repertoire visualisations.</p>
</div>
</div>
    <div id="immunarch-0-4-1---0-4-2" class="section level1">
<h1 class="page-header">
<a href="#immunarch-0-4-1---0-4-2" class="anchor"></a>immunarch 0.4.1 - 0.4.2</h1>
<p>Minor releases with bug fixes and minor improvements.</p>
</div>
    <div id="immunarch-0-4-0" class="section level1">
<h1 class="page-header">
<a href="#immunarch-0-4-0" class="anchor"></a>immunarch 0.4.0</h1>
<div id="features" class="section level2">
<h2 class="hasAnchor">
<a href="#features" class="anchor"></a>Features</h2>
<p>The <code>trackClonotype</code> function for tracking clonotypes.</p>
<p>Visualisations for <code>trackClonotype</code>.</p>
<p>Speed up Morisita-Horn index computation for up to 3 times.</p>
<p>Add the incremental overlap function <code>inc_overlap</code>, available from <code>repOverlap</code>.</p>
<p>Add the downsampling option to incremental overlaps.</p>
<p>Add visualisations for incvremental overlaps.</p>
<p>Remove the <code>prop_sample</code> function completely.</p>
<p>Move all downsampling, resampling and sampling procedures to the <code>repSample</code> function.</p>
<p>Add normalisation to rarafection.</p>
<p>Remove the .dup argument from <code>repOverlap</code>, all equal clonotypes are now always merged and their counts summed up.</p>
<p>Remove .quant from <code>repOverlap</code>, the column is now automatically detected.</p>
<p>Add an argument .add.layer to <code>vis_hist</code> to add any additional ggplot2 layers to the plots in the output grid.</p>
<p>Add grouping in <code>vis_hist</code> if data is grouped and <code>.grid</code> is T.</p>
<p>Update MiXCR parser (yet again) to read MiXCR files from the May 2019 release.</p>
<p>Add <code>.target</code> argument to visualisation of incremental overlaps.</p>
<p>Add <code>.clones</code> to <code>repExplore</code>.</p>
<p>Add more MiXCR file variants parsing.</p>
<p><code>repLoad</code> returns sorted by “Clones” data frames now.</p>
</div>
<div id="bug-fixes" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-fixes" class="anchor"></a>Bug fixes</h2>
<p>Fix <code>geneUsage</code> when the first two columns with gene usages were swapped.</p>
<p>Fix <code>repExplore</code> doesn’t work with a single repertoire.</p>
<p>Fix <code>vis_heatmap</code> doesn’t work with <code>geneUsage</code> output.</p>
<p>Fix a bug in computation of Morisita-Horn index.</p>
<p>Fix <code>vis_bar</code> and <code>vis_box</code> doesn’t work with numeric grouping variables.</p>
<p>Fix <code>repOverlap</code> failing when working on data tables with <code>morisita</code>.</p>
<p>Fix <code>repOverlap</code> failing to work with data tables on <code>public</code>, <code>overlap</code>, <code>jaccard</code> and <code>tversky</code>.</p>
<p>Fix <code>repLoad</code> failing when parsing MiXCR files with zero clonotypes.</p>
<p>Fix a bug in incorrect grouping in visualisations.</p>
<p>Fix a bug when clonal homeostasis and clonotype tracking don’t work properly with filtered coding.</p>
<p>Fix D50, “top” and “clonal.prop” from repClonality returns wrong values when the input data frame is not sorted.</p>
</div>
<div id="minor-updates" class="section level2">
<h2 class="hasAnchor">
<a href="#minor-updates" class="anchor"></a>Minor updates</h2>
<p>Remove the “wei” option from <code>geneUsage</code> becase it’s useless.</p>
<p>New dependency <code>ggseqlogo</code> for visualising of seq-logo plots.</p>
<p>Add <code>.transpose</code> to <code>vis_heatmap</code>.</p>
<p><code>repOverlap</code> default .col value is “aa” for comfortable usage.</p>
<p>Remove “fill” aesthetics warnings from <code>vis_heatmap</code>.</p>
<p>Remove warnings “In parse_fun(.path[i]) : NAs introduced by coercion” when parsing MiXCR files.</p>
<p>A lot or minor fixing and documentation improving to prepare for the CRAN release.</p>
<p>Remove short function names.</p>
</div>
</div>
    <div id="immunarch-0-3-3" class="section level1">
<h1 class="page-header">
<a href="#immunarch-0-3-3" class="anchor"></a>immunarch 0.3.3</h1>
<div id="features-1" class="section level2">
<h2 class="hasAnchor">
<a href="#features-1" class="anchor"></a>Features</h2>
<p>Parser for ArcherDX.</p>
<p>Update a parser for MiXCR to make it work with the “targetSequences” column format.</p>
<p>Update the coding function family to make it work with CDR3 amino acid sequences only.</p>
</div>
<div id="minor-updates-1" class="section level2">
<h2 class="hasAnchor">
<a href="#minor-updates-1" class="anchor"></a>Minor updates</h2>
<p>Replace all “unresolved” genes in ImmunoSEQ parsed files with NAs.</p>
<p>Add an argument for the color palette to <code>vis_heatmap2</code>.</p>
<p>Make <code>.grid=F</code> by default in <code>vis_hist</code> for gene usage analysis.</p>
<p>Fix parsing functions to not remove strings after dots in filenames.</p>
<p>Update the post-parsing processing subroutine to remove all characters except for amino acid alphabet, <code><a href="https://rdrr.io/r/base/Arithmetic.html">*</a></code> and <code>~</code> for compatability with all <code>immunarch</code>’s functions.</p>
</div>
<div id="documentation-1" class="section level2">
<h2 class="hasAnchor">
<a href="#documentation-1" class="anchor"></a>Documentation</h2>
<p>Add a documentation to the coding function family.</p>
<p>Update README.</p>
</div>
</div>
    <div id="immunarch-0-3-2" class="section level1">
<h1 class="page-header">
<a href="#immunarch-0-3-2" class="anchor"></a>immunarch 0.3.2</h1>
<p>Remove MonetDBLite from dependencies because it got removed from CRAN.</p>
<div id="bug-fixes-1" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-fixes-1" class="anchor"></a>Bug fixes</h2>
<ul>
<li>Fix a bug in MiXCR parser.</li>
</ul>
</div>
</div>
    <div id="immunarch-0-3-1" class="section level1">
<h1 class="page-header">
<a href="#immunarch-0-3-1" class="anchor"></a>immunarch 0.3.1</h1>
<div id="features-2" class="section level2">
<h2 class="hasAnchor">
<a href="#features-2" class="anchor"></a>Features</h2>
<ul>
<li><p>Boxplots and barplots are now support statistical tests via the <code>.test</code> argument.</p></li>
<li><p>Add parsers for old VDJtools formats.</p></li>
</ul>
</div>
<div id="documentation-2" class="section level2">
<h2 class="hasAnchor">
<a href="#documentation-2" class="anchor"></a>Documentation</h2>
<ul>
<li><p>Update docs and vignettes with statistical tests information.</p></li>
<li><p>Add a note for list names to vignettes.</p></li>
<li><p>Documentation for clustering.</p></li>
<li><p>Documentation for dimension reduction.</p></li>
<li><p>Minor fix for <code>repOverlap</code> documentation.</p></li>
</ul>
</div>
<div id="bug-fixes-2" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-fixes-2" class="anchor"></a>Bug fixes</h2>
<ul>
<li><p>Fix a grouping bug in visualisations.</p></li>
<li><p>Fix statistical tests from <code>ggpubr</code>.</p></li>
<li><p>Fix for <code>geneUsage</code> with <code>.type="family"</code>.</p></li>
</ul>
</div>
</div>
    <div id="immunarch-0-3-0" class="section level1">
<h1 class="page-header">
<a href="#immunarch-0-3-0" class="anchor"></a>immunarch 0.3.0</h1>
<div id="features-3" class="section level2">
<h2 class="hasAnchor">
<a href="#features-3" class="anchor"></a>Features</h2>
<ul>
<li><p><code>fixVis</code> now supports the following legends: size, shape, color, fill, linetype.</p></li>
<li><p><code>fixVis</code> can plot figures to R console / RStudio “Plots” tab.</p></li>
<li><p><code>fixVis</code> now supports the number of columns in legends.</p></li>
<li><p>Support for the AIRR file format.</p></li>
<li><p>Experimental support for the 10xGenomics format.</p></li>
<li><p>Save and load <code>immunarch</code> format via <code>repSave</code> and <code>repLoad</code>.</p></li>
<li><p>Save and load VDJtools format via <code>repSave</code> and <code>repLoad</code>.</p></li>
</ul>
</div>
<div id="bug-fixes-3" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-fixes-3" class="anchor"></a>Bug fixes</h2>
<ul>
<li><p><code>.a</code> and <code>.b</code> didn’t passed to Tversky index.</p></li>
<li><p><code>fixVis</code> - fix a bug when users apply X/Y settings to the other axis.</p></li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
        <li><a href="#immunarch-0-4-3">0.4.3</a></li>
        <li><a href="#immunarch-0-4-1---0-4-2">0.4.1</a></li>
        <li><a href="#immunarch-0-4-0">0.4.0</a></li>
        <li><a href="#immunarch-0-3-3">0.3.3</a></li>
        <li><a href="#immunarch-0-3-2">0.3.2</a></li>
        <li><a href="#immunarch-0-3-1">0.3.1</a></li>
        <li><a href="#immunarch-0-3-0">0.3.0</a></li>
      </ul>
    </div>
  </div>

</div>


      <footer>
      <div class="copyright">
  <p>Developed by Vadim I. Nazarov, Vasily O. Tsvetkov, Eugene Rumynskiy, Anna Lorenc, Daniel J. Moore, Victor Greiff, <a href='https://immunomind.io'>ImmunoMind</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script>
<script>
  docsearch({
    
    
    apiKey: '94207562f34b455c0796790d104e3549',
    indexName: 'immunarch',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>



  </body>
</html>


