<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Load clonotype databases such as VDJDB and McPAS into the R workspace — dbLoad • immunarch</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>


<!-- docsearch -->
<script src="../docsearch.js"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous" />
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script>


  <link href="../extra.css" rel="stylesheet">
  

<meta property="og:title" content="Load clonotype databases such as VDJDB and McPAS into the R workspace — dbLoad" />
<meta property="og:description" content="The function automatically detects the database format and loads it into R.
Additionally, the function provides a general query interface to databases that allows
filtering by species, chain types (i.e., locus) and pathology (i.e., antigen species).
Currently we support three popular databases:
VDJDB - https://github.com/antigenomics/vdjdb-db
McPAS-TCR - http://friedmanlab.weizmann.ac.il/McPAS-TCR/
TBAdb from PIRD - https://db.cngb.org/pird/tbadb/" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-128418614-2"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-128418614-2');
</script>


  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">immunarch</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.6.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Start Here
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/v1_introduction.html">Installation &amp; troubleshooting</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">First steps</li>
    <li>
      <a href="../articles/v2_data.html">Data loading</a>
    </li>
    <li>
      <a href="../articles/web_only/load_mixcr.html">How-to: Loading MiXCR Data</a>
    </li>
    <li>
      <a href="../articles/web_only/load_10x.html">How-to: Loading 10x Genomics Data</a>
    </li>
    <li>
      <a href="../articles/web_only/v21_singlecell.html">How-to: Single-cell and paired chain data</a>
    </li>
    <li>
      <a href="../articles/web_only/v3_basic_analysis.html">Basic statistics and clonality</a>
    </li>
    <li>
      <a href="../articles/web_only/repFilter_v3.html">Data filtering</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Repertoire-level exploration and comparison</li>
    <li>
      <a href="../articles/web_only/v4_overlap.html">Repertoire overlap and public clonotypes</a>
    </li>
    <li>
      <a href="../articles/web_only/v5_gene_usage.html">Gene usage</a>
    </li>
    <li>
      <a href="../articles/web_only/v6_diversity.html">Diversity estimation</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Clonotype-level exploration and comparison</li>
    <li>
      <a href="../articles/web_only/v8_tracking.html">Track clonotypes across samples and time</a>
    </li>
    <li>
      <a href="../articles/web_only/v11_db.html">Annotate clonotypes using immune receptor databases</a>
    </li>
    <li>
      <a href="../articles/web_only/v9_kmers.html">Kmer and sequence motif analysis and visualisation</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Preparing to publication</li>
    <li>
      <a href="../articles/web_only/v7_fixvis.html">Make your plots publication-ready with fixVis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="https://github.com/immunomind/covid19">COVID-19</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://immunomind.io">ImmunoMind</a>
</li>
<li>
  <a href="http://twitter.com/immunomind">
    <span class="fa fa-lg fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/immunomind/immunarch">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
      
      <form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
        </div>
      </form>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Load clonotype databases such as VDJDB and McPAS into the R workspace</h1>
    <small class="dont-index">Source: <a href='https://github.com/immunomind/immunarch/blob/master/R/annotation.R'><code>R/annotation.R</code></a></small>
    <div class="hidden name"><code>dbLoad.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The function automatically detects the database format and loads it into R.
Additionally, the function provides a general query interface to databases that allows
filtering by species, chain types (i.e., locus) and pathology (i.e., antigen species).</p>
<p>Currently we support three popular databases:</p>
<p>VDJDB - <a href='https://github.com/antigenomics/vdjdb-db'>https://github.com/antigenomics/vdjdb-db</a></p>
<p>McPAS-TCR - <a href='http://friedmanlab.weizmann.ac.il/McPAS-TCR/'>http://friedmanlab.weizmann.ac.il/McPAS-TCR/</a></p>
<p>TBAdb from PIRD - <a href='https://db.cngb.org/pird/tbadb/'>https://db.cngb.org/pird/tbadb/</a></p>
    </div>

    <pre class="usage"><span class='fu'>dbLoad</span><span class='op'>(</span><span class='va'>.path</span>, <span class='va'>.db</span>, .species <span class='op'>=</span> <span class='cn'>NA</span>, .chain <span class='op'>=</span> <span class='cn'>NA</span>, .pathology <span class='op'>=</span> <span class='cn'>NA</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>.path</th>
      <td><p>Character. A path to the database file, e.g., "/Users/researcher/Downloads/McPAS-TCR.csv".</p></td>
    </tr>
    <tr>
      <th>.db</th>
      <td><p>Character. A database type: either "vdjdb", "vdjdb-search", "mcpas" or "tbadb".</p>
<p>"vdjdb" for VDJDB; "vdjdb-search" for search table obtained from the web interface of VDJDB;
"mcpas" for McPAS-TCR; "tbadb" for PIRD TBAdb.</p></td>
    </tr>
    <tr>
      <th>.species</th>
      <td><p>Character. A string or a vector of strings specifying which species need to be in the database, e.g., "HomoSapiens".
Pass NA (by default) to load all available species.</p></td>
    </tr>
    <tr>
      <th>.chain</th>
      <td><p>Character. A string or a vector of strings specifying which chains need to be in the database, e.g., "TRB".
Pass NA (by default) to load all available chains.</p></td>
    </tr>
    <tr>
      <th>.pathology</th>
      <td><p>Character. A string or a vector of strings specifying which disease, virus, bacteria or any condition
needs to be in the database, e.g., "CMV".
Pass NA (by default) to load all available conditions.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Data frame with the input database records.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># Example file path</span>
<span class='va'>file_path</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span>package <span class='op'>=</span> <span class='st'>"immunarch"</span><span class='op'>)</span>, <span class='st'>"/extdata/db/vdjdb.example.txt"</span><span class='op'>)</span>

<span class='co'># Load the database with human-only TRB-only receptors for all known antigens</span>
<span class='va'>db</span> <span class='op'>&lt;-</span> <span class='fu'>dbLoad</span><span class='op'>(</span><span class='va'>file_path</span>, <span class='st'>"vdjdb"</span>, <span class='st'>"HomoSapiens"</span>, <span class='st'>"TRB"</span><span class='op'>)</span>
<span class='va'>db</span>
</div><div class='output co'>#&gt; <span style='color: #949494;'># A tibble: 10 × 19</span>
#&gt;    gene  cdr3    species antigen.epitope antigen.gene antigen.species complex.id
#&gt;    <span style='color: #949494; font-style: italic;'>&lt;chr&gt;</span> <span style='color: #949494; font-style: italic;'>&lt;chr&gt;</span>   <span style='color: #949494; font-style: italic;'>&lt;chr&gt;</span>   <span style='color: #949494; font-style: italic;'>&lt;chr&gt;</span>           <span style='color: #949494; font-style: italic;'>&lt;chr&gt;</span>        <span style='color: #949494; font-style: italic;'>&lt;chr&gt;</span>                <span style='color: #949494; font-style: italic;'>&lt;dbl&gt;</span>
#&gt; <span style='color: #BCBCBC;'> 1</span> TRB   CASSQD… HomoSa… RLRAEAQVK       EBNA3A       EBV                  <span style='text-decoration: underline;'>19</span>268
#&gt; <span style='color: #BCBCBC;'> 2</span> TRB   CSASIL… HomoSa… KLGGALQAK       IE1          CMV                   <span style='text-decoration: underline;'>8</span>584
#&gt; <span style='color: #BCBCBC;'> 3</span> TRB   CASSYF… HomoSa… KLGGALQAK       IE1          CMV                   <span style='text-decoration: underline;'>3</span>445
#&gt; <span style='color: #BCBCBC;'> 4</span> TRB   CASSAF… HomoSa… NLVPMVATV       pp65         CMV                      0
#&gt; <span style='color: #BCBCBC;'> 5</span> TRB   CASSLW… HomoSa… KLGGALQAK       IE1          CMV                  <span style='text-decoration: underline;'>19</span>396
#&gt; <span style='color: #BCBCBC;'> 6</span> TRB   CASSLT… HomoSa… NLVPMVATV       pp65         CMV                      0
#&gt; <span style='color: #BCBCBC;'> 7</span> TRB   CASTAK… HomoSa… KLGGALQAK       IE1          CMV                  <span style='text-decoration: underline;'>10</span>972
#&gt; <span style='color: #BCBCBC;'> 8</span> TRB   CASSGA… HomoSa… KLGGALQAK       IE1          CMV                   <span style='text-decoration: underline;'>6</span>231
#&gt; <span style='color: #BCBCBC;'> 9</span> TRB   CASSLI… HomoSa… KLGGALQAK       IE1          CMV                  <span style='text-decoration: underline;'>12</span>587
#&gt; <span style='color: #BCBCBC;'>10</span> TRB   CATSSS… HomoSa… KLGGALQAK       IE1          CMV                  <span style='text-decoration: underline;'>13</span>267
#&gt; <span style='color: #949494;'># … with 12 more variables: v.segm &lt;chr&gt;, j.segm &lt;chr&gt;, v.end &lt;dbl&gt;,</span>
#&gt; <span style='color: #949494;'>#   j.start &lt;dbl&gt;, mhc.a &lt;chr&gt;, mhc.b &lt;chr&gt;, mhc.class &lt;chr&gt;,</span>
#&gt; <span style='color: #949494;'>#   reference.id &lt;chr&gt;, vdjdb.score &lt;dbl&gt;, Species &lt;chr&gt;, Chain &lt;chr&gt;,</span>
#&gt; <span style='color: #949494;'>#   Pathology &lt;chr&gt;</span></div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://www.linkedin.com/in/vdnaz'>Vadim I. Nazarov</a>, <a href='https://www.linkedin.com/in/vasily-tsvetkov-227218ab'>Vasily O. Tsvetkov</a>, <a href='https://www.linkedin.com/in/erumynskiy/'>Eugene Rumynskiy</a>, <a href='https://www.linkedin.com/in/aleksandr-popov-634878105/'>Aleksandr A. Popov</a>, <a href='https://www.linkedin.com/in/ivan-balashov/'>Ivan Balashov</a>, <a href='https://www.linkedin.com/in/maria-volobueva-b0856a223/'>Maria Volobueva</a>, <a href='https://immunomind.io'>ImmunoMind</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script>
<script>
  docsearch({
    
    
    apiKey: '94207562f34b455c0796790d104e3549',
    indexName: 'immunarch',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>



  </body>
</html>


