<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ immunarch</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous"><link href="../docsearch.css" rel="stylesheet"><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-128418614-2"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-128418614-2');
</script></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">immunarch</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.7.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Start Here
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/v1_introduction.html">Installation &amp; troubleshooting</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">First steps</li>
    <li>
      <a href="../articles/v2_data.html">Data loading</a>
    </li>
    <li>
      <a href="../articles/web_only/load_mixcr.html">How-to: Loading MiXCR Data</a>
    </li>
    <li>
      <a href="../articles/web_only/load_10x.html">How-to: Loading 10x Genomics Data</a>
    </li>
    <li>
      <a href="../articles/web_only/v21_singlecell.html">How-to: Single-cell and paired chain data</a>
    </li>
    <li>
      <a href="../articles/web_only/v3_basic_analysis.html">Basic statistics and clonality</a>
    </li>
    <li>
      <a href="../articles/web_only/repFilter_v3.html">Data filtering</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Repertoire-level exploration and comparison</li>
    <li>
      <a href="../articles/web_only/v4_overlap.html">Repertoire overlap and public clonotypes</a>
    </li>
    <li>
      <a href="../articles/web_only/v5_gene_usage.html">Gene usage</a>
    </li>
    <li>
      <a href="../articles/web_only/v6_diversity.html">Diversity estimation</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Clonotype-level exploration and comparison</li>
    <li>
      <a href="../articles/web_only/v8_tracking.html">Track clonotypes across samples and time</a>
    </li>
    <li>
      <a href="../articles/web_only/v11_db.html">Annotate clonotypes using immune receptor databases</a>
    </li>
    <li>
      <a href="../articles/web_only/v9_kmers.html">Kmer and sequence motif analysis and visualisation</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Preparing to publication</li>
    <li>
      <a href="../articles/web_only/v7_fixvis.html">Make your plots publication-ready with fixVis</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">BCR analysis</li>
    <li>
      <a href="../articles/web_only/BCRpipeline.html">BCR pipeline</a>
    </li>
    <li>
      <a href="../articles/web_only/clustering.html">Clustering</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="https://github.com/immunomind/covid19" class="external-link">COVID-19</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://immunomind.io" class="external-link">ImmunoMind</a>
</li>
<li>
  <a href="http://twitter.com/immunomind" class="external-link">
    <span class="fa fa-lg fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/immunomind/immunarch" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul><form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off"></div>
      </form>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="bulk-and-single-cell-data">Bulk and single-cell data <a href="#bulk-and-single-cell-data" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="loading-and-saving-any-data">Loading and saving any data <a href="#loading-and-saving-any-data" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="repLoad.html">repLoad()</a></code> </p>
        </td>
        <td><p>Load immune repertoire files into the R workspace</p></td>
      </tr><tr><td>
          <p><code><a href="repSave.html">repSave()</a></code> </p>
        </td>
        <td><p>Save immune repertoires to the disk</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="filtering-data">Filtering data <a href="#filtering-data" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="repFilter.html">repFilter()</a></code> </p>
        </td>
        <td><p>Main function for data filtering</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="single-cell">Single-cell <a href="#single-cell" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p><p>More coming soon!</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="select_barcodes.html">select_barcodes()</a></code> </p>
        </td>
        <td><p>Select specific clonotypes using barcodes from single-cell metadata</p></td>
      </tr><tr><td>
          <p><code><a href="select_clusters.html">select_clusters()</a></code> </p>
        </td>
        <td><p>Split the immune repertoire data to clusters from single-cell barcodes</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="preprocessing">Preprocessing <a href="#preprocessing" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="bunch_translate.html">bunch_translate()</a></code> </p>
        </td>
        <td><p>Nucleotide to amino acid sequence translation</p></td>
      </tr><tr><td>
          <p><code><a href="coding.html">coding()</a></code> <code><a href="coding.html">noncoding()</a></code> <code><a href="coding.html">inframes()</a></code> <code><a href="coding.html">outofframes()</a></code> </p>
        </td>
        <td><p>Filter out coding and non-coding clonotype sequences</p></td>
      </tr><tr><td>
          <p><code><a href="repSample.html">repSample()</a></code> </p>
        </td>
        <td><p>Downsampling and resampling of immune repertoires</p></td>
      </tr><tr><td>
          <p><code><a href="top.html">top()</a></code> </p>
        </td>
        <td><p>Get the N most abundant clonotypes</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="bcr-data">BCR data <a href="#bcr-data" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="clustering">Clustering <a href="#clustering" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="seqDist.html">seqDist()</a></code> </p>
        </td>
        <td><p>Function for computing distance for sequences</p></td>
      </tr><tr><td>
          <p><code><a href="seqCluster.html">seqCluster()</a></code> </p>
        </td>
        <td><p>Function for assigning clusters based on sequences similarity</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="bcr-pipeline">BCR pipeline <a href="#bcr-pipeline" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="repGermline.html">repGermline()</a></code> </p>
        </td>
        <td><p>This function creates germlines for clonal lineages. B cell clonal lineage represents a set of B cells
that presumably have a common origin (arising from the same VDJ rearrangement event) and a common ancestor.
Each clonal lineage has its own germline sequence that represents the ancestral sequence
for each BCR in clonal lineage. In other words, germline sequence is a sequence of B-cells immediately
after VDJ recombination, before B-cell maturation and hypermutation process. Germline sequence is useful
for assessing the degree of mutation and maturity of the repertoire.</p></td>
      </tr><tr><td>
          <p><code><a href="repAlignLineage.html">repAlignLineage()</a></code> </p>
        </td>
        <td><p>This function aligns all sequences (incliding germline) that belong to one clonal lineage and one cluster.
After clustering and building the clonal lineage and germline, the next step is to analyze the degree of mutation
and maturity of each clonal lineage. This allows for finding high mature cells and cells with a large
number of offspring. The phylogenetic analysis will find mutations that increase the affinity of BCR.
Making alignment of the sequence is the first step towards sequence analysis including BCR.</p></td>
      </tr><tr><td>
          <p><code><a href="repClonalFamily.html">repClonalFamily()</a></code> </p>
        </td>
        <td><p>This function uses the PHYLIP package to make phylogenetic analysis. For making trees it uses
maximum parsimony methods.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="basic-immune-repertoire-statistics">Basic immune repertoire statistics <a href="#basic-immune-repertoire-statistics" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="exploratory-data-analysis">Exploratory data analysis <a href="#exploratory-data-analysis" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="repExplore.html">repExplore()</a></code> </p>
        </td>
        <td><p>Main function for exploratory data analysis: compute the distribution of lengths, clones, etc.</p></td>
      </tr><tr><td>
          <p><code><a href="vis.immunr_exp_vol.html">vis(<i>&lt;immunr_exp_vol&gt;</i>)</a></code> </p>
        </td>
        <td><p>Visualise results of the exploratory analysis</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="clonality-analysis">Clonality analysis <a href="#clonality-analysis" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="repClonality.html">repClonality()</a></code> </p>
        </td>
        <td><p>Clonality analysis of immune repertoires</p></td>
      </tr><tr><td>
          <p><code><a href="vis.immunr_clonal_prop.html">vis(<i>&lt;immunr_clonal_prop&gt;</i>)</a></code> </p>
        </td>
        <td><p>Visualise results of the clonality analysis</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="clonotype-annotation-and-dynamics">Clonotype annotation and dynamics <a href="#clonotype-annotation-and-dynamics" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="clonotype-annotation">Clonotype annotation <a href="#clonotype-annotation" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p><p>Annotate clonotypes in immune repertoires using external immune receptor databases (VDJDB, McPAS and PIRD).</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="dbAnnotate.html">dbAnnotate()</a></code> </p>
        </td>
        <td><p>Annotate clonotypes in immune repertoires using clonotype databases such as VDJDB and MCPAS</p></td>
      </tr><tr><td>
          <p><code><a href="dbLoad.html">dbLoad()</a></code> </p>
        </td>
        <td><p>Load clonotype databases such as VDJDB and McPAS into the R workspace</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="immune-repertoire-dynamics">Immune repertoire dynamics <a href="#immune-repertoire-dynamics" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p><p>Track the differences in clonotype abundances over time.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="trackClonotypes.html">trackClonotypes()</a></code> </p>
        </td>
        <td><p>Track clonotypes across time and data points</p></td>
      </tr><tr><td>
          <p><code><a href="vis.immunr_dynamics.html">vis(<i>&lt;immunr_dynamics&gt;</i>)</a></code> </p>
        </td>
        <td><p>Visualise clonotype dynamics</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="compare-repertoires">Compare repertoires <a href="#compare-repertoires" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="repertoire-diversity-analysis">Repertoire diversity analysis <a href="#repertoire-diversity-analysis" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="repDiversity.html">repDiversity()</a></code> </p>
        </td>
        <td><p>The main function for immune repertoire diversity estimation</p></td>
      </tr><tr><td>
          <p><code><a href="vis.immunr_chao1.html">vis(<i>&lt;immunr_chao1&gt;</i>)</a></code> </p>
        </td>
        <td><p>Visualise diversity.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="gene-usage">Gene usage <a href="#gene-usage" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p><p>V-gene and J-gene usage statistics, analysis and visualisations</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="geneUsage.html">geneUsage()</a></code> </p>
        </td>
        <td><p>Main function for estimation of V-gene and J-gene statistics</p></td>
      </tr><tr><td>
          <p><code><a href="geneUsageAnalysis.html">geneUsageAnalysis()</a></code> </p>
        </td>
        <td><p>Post-analysis of V-gene and J-gene statistics: PCA, clustering, etc.</p></td>
      </tr><tr><td>
          <p><code><a href="gene_stats.html">gene_stats()</a></code> </p>
        </td>
        <td><p>WIP</p></td>
      </tr><tr><td>
          <p><code><a href="vis.immunr_gene_usage.html">vis(<i>&lt;immunr_gene_usage&gt;</i>)</a></code> </p>
        </td>
        <td><p>Histograms and boxplots (general case / gene usage)</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="overlap">Overlap <a href="#overlap" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="repOverlap.html">repOverlap()</a></code> </p>
        </td>
        <td><p>Main function for public clonotype statistics calculations</p></td>
      </tr><tr><td>
          <p><code><a href="repOverlapAnalysis.html">repOverlapAnalysis()</a></code> </p>
        </td>
        <td><p>Post-analysis of public clonotype statistics: PCA, clustering, etc.</p></td>
      </tr><tr><td>
          <p><code><a href="inc_overlap.html">inc_overlap()</a></code> </p>
        </td>
        <td><p>Incremental counting of repertoire similarity</p></td>
      </tr><tr><td>
          <p><code><a href="vis.immunr_inc_overlap.html">vis(<i>&lt;immunr_inc_overlap&gt;</i>)</a></code> </p>
        </td>
        <td><p>Visualise incremental overlaps</p></td>
      </tr><tr><td>
          <p><code><a href="vis.immunr_ov_matrix.html">vis(<i>&lt;immunr_ov_matrix&gt;</i>)</a></code> </p>
        </td>
        <td><p>Repertoire overlap and gene usage visualisations</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="public-repertoire">Public repertoire <a href="#public-repertoire" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="pubRep.html">pubRep()</a></code> </p>
        </td>
        <td><p>Create a repertoire of public clonotypes</p></td>
      </tr><tr><td>
          <p><code><a href="pubRepApply.html">pubRepApply()</a></code> </p>
        </td>
        <td><p>Apply transformations to public repertoires</p></td>
      </tr><tr><td>
          <p><code><a href="pubRepFilter.html">pubRepFilter()</a></code> </p>
        </td>
        <td><p>Filter out clonotypes from public repertoires</p></td>
      </tr><tr><td>
          <p><code><a href="pubRepStatistics.html">pubRepStatistics()</a></code> </p>
        </td>
        <td><p>Statistics of number of public clonotypes for each possible combinations of repertoires</p></td>
      </tr><tr><td>
          <p><code><a href="public_matrix.html">public_matrix()</a></code> </p>
        </td>
        <td><p>Get a matrix with public clonotype frequencies</p></td>
      </tr><tr><td>
          <p><code><a href="vis.immunr_public_repertoire.html">vis(<i>&lt;immunr_public_repertoire&gt;</i>)</a></code> </p>
        </td>
        <td><p>Public repertoire visualisation</p></td>
      </tr><tr><td>
          <p><code><a href="vis.immunr_public_statistics.html">vis(<i>&lt;immunr_public_statistics&gt;</i>)</a></code> </p>
        </td>
        <td><p>Visualise sharing of clonotypes among samples</p></td>
      </tr><tr><td>
          <p><code><a href="vis_public_clonotypes.html">vis_public_clonotypes()</a></code> </p>
        </td>
        <td><p>Visualisation of public clonotypes</p></td>
      </tr><tr><td>
          <p><code><a href="vis_public_frequencies.html">vis_public_frequencies()</a></code> </p>
        </td>
        <td><p>Public repertoire visualisation</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="advanced-immune-repertoire-analysis">Advanced immune repertoire analysis <a href="#advanced-immune-repertoire-analysis" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="kmers-analysis">Kmers analysis <a href="#kmers-analysis" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="getKmers.html">getKmers()</a></code> </p>
        </td>
        <td><p>Calculate the k-mer statistics of immune repertoires</p></td>
      </tr><tr><td>
          <p><code><a href="split_to_kmers.html">split_to_kmers()</a></code> <code><a href="split_to_kmers.html">kmer_profile()</a></code> </p>
        </td>
        <td><p>Analysis immune repertoire kmer statistics: sequence profiles, etc.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="post-analysis">Post-analysis <a href="#post-analysis" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p><p>Advanced methods for post-analysis of gene usage, overlap and other statistics.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="immunr_hclust.html">immunr_hclust()</a></code> <code><a href="immunr_hclust.html">immunr_kmeans()</a></code> <code><a href="immunr_hclust.html">immunr_dbscan()</a></code> </p>
        </td>
        <td><p>Clustering of objects or distance matrices</p></td>
      </tr><tr><td>
          <p><code><a href="immunr_pca.html">immunr_pca()</a></code> <code><a href="immunr_pca.html">immunr_mds()</a></code> <code><a href="immunr_pca.html">immunr_tsne()</a></code> </p>
        </td>
        <td><p>Dimensionality reduction</p></td>
      </tr><tr><td>
          <p><code><a href="vis.immunr_hclust.html">vis(<i>&lt;immunr_hclust&gt;</i>)</a></code> </p>
        </td>
        <td><p>Visualisation of hierarchical clustering</p></td>
      </tr><tr><td>
          <p><code><a href="vis.immunr_kmeans.html">vis(<i>&lt;immunr_kmeans&gt;</i>)</a></code> </p>
        </td>
        <td><p>Visualisation of K-means and DBSCAN clustering</p></td>
      </tr><tr><td>
          <p><code><a href="vis.immunr_mds.html">vis(<i>&lt;immunr_mds&gt;</i>)</a></code> </p>
        </td>
        <td><p>PCA / MDS / tSNE visualisation (mainly overlap / gene usage)</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="visualisations">Visualisations <a href="#visualisations" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="general-visualisation-functions">General visualisation functions <a href="#general-visualisation-functions" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p><p>Functions for visualisations of different data types. For analysis-specific visualisation see related sections.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="vis.html">vis()</a></code> </p>
        </td>
        <td><p>One function to visualise them all</p></td>
      </tr><tr><td>
          <p><code><a href="vis.immunr_chao1.html">vis(<i>&lt;immunr_chao1&gt;</i>)</a></code> </p>
        </td>
        <td><p>Visualise diversity.</p></td>
      </tr><tr><td>
          <p><code><a href="vis.immunr_clonal_prop.html">vis(<i>&lt;immunr_clonal_prop&gt;</i>)</a></code> </p>
        </td>
        <td><p>Visualise results of the clonality analysis</p></td>
      </tr><tr><td>
          <p><code><a href="vis.immunr_dynamics.html">vis(<i>&lt;immunr_dynamics&gt;</i>)</a></code> </p>
        </td>
        <td><p>Visualise clonotype dynamics</p></td>
      </tr><tr><td>
          <p><code><a href="vis.immunr_exp_vol.html">vis(<i>&lt;immunr_exp_vol&gt;</i>)</a></code> </p>
        </td>
        <td><p>Visualise results of the exploratory analysis</p></td>
      </tr><tr><td>
          <p><code><a href="vis.immunr_gene_usage.html">vis(<i>&lt;immunr_gene_usage&gt;</i>)</a></code> </p>
        </td>
        <td><p>Histograms and boxplots (general case / gene usage)</p></td>
      </tr><tr><td>
          <p><code><a href="vis.immunr_hclust.html">vis(<i>&lt;immunr_hclust&gt;</i>)</a></code> </p>
        </td>
        <td><p>Visualisation of hierarchical clustering</p></td>
      </tr><tr><td>
          <p><code><a href="vis.immunr_inc_overlap.html">vis(<i>&lt;immunr_inc_overlap&gt;</i>)</a></code> </p>
        </td>
        <td><p>Visualise incremental overlaps</p></td>
      </tr><tr><td>
          <p><code><a href="vis.immunr_kmeans.html">vis(<i>&lt;immunr_kmeans&gt;</i>)</a></code> </p>
        </td>
        <td><p>Visualisation of K-means and DBSCAN clustering</p></td>
      </tr><tr><td>
          <p><code><a href="vis.immunr_kmer_table.html">vis(<i>&lt;immunr_kmer_table&gt;</i>)</a></code> </p>
        </td>
        <td><p>Most frequent kmers visualisation.</p></td>
      </tr><tr><td>
          <p><code><a href="vis.immunr_mds.html">vis(<i>&lt;immunr_mds&gt;</i>)</a></code> </p>
        </td>
        <td><p>PCA / MDS / tSNE visualisation (mainly overlap / gene usage)</p></td>
      </tr><tr><td>
          <p><code><a href="vis.immunr_ov_matrix.html">vis(<i>&lt;immunr_ov_matrix&gt;</i>)</a></code> </p>
        </td>
        <td><p>Repertoire overlap and gene usage visualisations</p></td>
      </tr><tr><td>
          <p><code><a href="vis.immunr_public_repertoire.html">vis(<i>&lt;immunr_public_repertoire&gt;</i>)</a></code> </p>
        </td>
        <td><p>Public repertoire visualisation</p></td>
      </tr><tr><td>
          <p><code><a href="vis.immunr_public_statistics.html">vis(<i>&lt;immunr_public_statistics&gt;</i>)</a></code> </p>
        </td>
        <td><p>Visualise sharing of clonotypes among samples</p></td>
      </tr><tr><td>
          <p><code><a href="vis_bar.html">vis_bar()</a></code> </p>
        </td>
        <td><p>Bar plots</p></td>
      </tr><tr><td>
          <p><code><a href="vis_box.html">vis_box()</a></code> </p>
        </td>
        <td><p>Flexible box-plots for visualisation of distributions</p></td>
      </tr><tr><td>
          <p><code><a href="vis_circos.html">vis_circos()</a></code> </p>
        </td>
        <td><p>Visualisation of matrices using circos plots</p></td>
      </tr><tr><td>
          <p><code><a href="vis_heatmap.html">vis_heatmap()</a></code> </p>
        </td>
        <td><p>Visualisation of matrices and data frames using ggplo2-based heatmaps</p></td>
      </tr><tr><td>
          <p><code><a href="vis_heatmap2.html">vis_heatmap2()</a></code> </p>
        </td>
        <td><p>Visualisation of matrices using pheatmap-based heatmaps</p></td>
      </tr><tr><td>
          <p><code><a href="vis_hist.html">vis_hist()</a></code> </p>
        </td>
        <td><p>Visualisation of distributions using histograms</p></td>
      </tr><tr><td>
          <p><code><a href="vis_immunr_kmer_profile_main.html">vis_immunr_kmer_profile_main()</a></code> </p>
        </td>
        <td><p>Visualise kmer profiles</p></td>
      </tr><tr><td>
          <p><code><a href="vis_public_clonotypes.html">vis_public_clonotypes()</a></code> </p>
        </td>
        <td><p>Visualisation of public clonotypes</p></td>
      </tr><tr><td>
          <p><code><a href="vis_public_frequencies.html">vis_public_frequencies()</a></code> </p>
        </td>
        <td><p>Public repertoire visualisation</p></td>
      </tr><tr><td>
          <p><code><a href="vis_textlogo.html">vis_textlogo()</a></code> <code><a href="vis_textlogo.html">vis_seqlogo()</a></code> </p>
        </td>
        <td><p>Sequence logo plots for amino acid profiles.</p></td>
      </tr><tr><td>
          <p><code><a href="spectratype.html">spectratype()</a></code> </p>
        </td>
        <td><p>Immune repertoire spectratyping</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="publication-ready-plots">Publication-ready plots <a href="#publication-ready-plots" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="fixVis.html">fixVis()</a></code> </p>
        </td>
        <td><p>Manipulate ggplot plots and create publication-ready plots</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="utilities">Utilities <a href="#utilities" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="data">Data <a href="#data" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="aa_properties.html">aa_properties</a></code> </p>
        </td>
        <td><p>Tables with amino acid properties</p></td>
      </tr><tr><td>
          <p><code><a href="aa_table.html">AA_TABLE</a></code> </p>
        </td>
        <td><p>Amino acid / codon table</p></td>
      </tr><tr><td>
          <p><code><a href="bcrdata.html">bcrdata</a></code> </p>
        </td>
        <td><p>BCR dataset</p></td>
      </tr><tr><td>
          <p><code><a href="gene_segments.html">gene_segments</a></code> </p>
        </td>
        <td><p>Gene segments table</p></td>
      </tr><tr><td>
          <p><code><a href="immdata.html">immdata</a></code> </p>
        </td>
        <td><p>Single chain immune repertoire dataset</p></td>
      </tr><tr><td>
          <p><code><a href="immunr_data_format.html">immunr_data_format</a></code> </p>
        </td>
        <td><p>Specification of the data format used by immunarch dataframes</p></td>
      </tr><tr><td>
          <p><code><a href="scdata.html">scdata</a></code> </p>
        </td>
        <td><p>Paired chain immune repertoire dataset</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="public-utility-utilities">Public utility utilities <a href="#public-utility-utilities" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="apply_symm.html">apply_symm()</a></code> <code><a href="apply_symm.html">apply_asymm()</a></code> </p>
        </td>
        <td><p>Apply function to each pair of data frames from a list.</p></td>
      </tr><tr><td>
          <p><code><a href="entropy.html">entropy()</a></code> <code><a href="entropy.html">kl_div()</a></code> <code><a href="entropy.html">js_div()</a></code> <code><a href="entropy.html">cross_entropy()</a></code> </p>
        </td>
        <td><p>Information measures</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="internal-utility-functions">Internal utility functions <a href="#internal-utility-functions" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="add_class.html">add_class()</a></code> </p>
        </td>
        <td><p>Add a new class attribute</p></td>
      </tr><tr><td>
          <p><code><a href="check_distribution.html">check_distribution()</a></code> </p>
        </td>
        <td><p>Check and normalise distributions</p></td>
      </tr><tr><td>
          <p><code><a href="dot-quant_column_choice.html">.quant_column_choice()</a></code> </p>
        </td>
        <td><p>Get a column's name using the input alias</p></td>
      </tr><tr><td>
          <p><code><a href="group_from_metadata.html">group_from_metadata()</a></code> </p>
        </td>
        <td><p>Get a character vector of samples' groups from the input metadata file</p></td>
      </tr><tr><td>
          <p><code><a href="has_class.html">has_class()</a></code> </p>
        </td>
        <td><p>Check for the specific class</p></td>
      </tr><tr><td>
          <p><code><a href="matrixdiagcopy.html">matrixdiagcopy()</a></code> </p>
        </td>
        <td><p>Copy the upper matrix triangle to the lower one</p></td>
      </tr><tr><td>
          <p><code><a href="set_pb.html">set_pb()</a></code> <code><a href="set_pb.html">add_pb()</a></code> </p>
        </td>
        <td><p>Set and update progress bars</p></td>
      </tr><tr><td>
          <p><code><a href="switch_type.html">switch_type()</a></code> <code><a href="switch_type.html">process_col_argument()</a></code> </p>
        </td>
        <td><p>Return a column's name</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="https://www.linkedin.com/in/vdnaz" class="external-link">Vadim I. Nazarov</a>, <a href="https://www.linkedin.com/in/vasily-tsvetkov-227218ab" class="external-link">Vasily O. Tsvetkov</a>, <a href="https://www.linkedin.com/in/erumynskiy/" class="external-link">Eugene Rumynskiy</a>, <a href="https://www.linkedin.com/in/aleksandr-popov-634878105/" class="external-link">Aleksandr A. Popov</a>, <a href="https://www.linkedin.com/in/ivan-balashov/" class="external-link">Ivan Balashov</a>, <a href="https://www.linkedin.com/in/maria-volobueva-b0856a223/" class="external-link">Maria Samokhina</a>, <a href="https://immunomind.io" class="external-link">ImmunoMind</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

      </footer></div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '94207562f34b455c0796790d104e3549',
    indexName: 'immunarch',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script></body></html>

