<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>This function creates germlines for clonal lineages. B cell clonal lineage represents a set of B cells
that presumably have a common origin (arising from the same VDJ rearrangement event) and a common ancestor.
Each clonal lineage has its own germline sequence that represents the ancestral sequence
for each BCR in clonal lineage. In other words, germline sequence is a sequence of B-cells immediately
after VDJ recombination, before B-cell maturation and hypermutation process. Germline sequence is useful
for assessing the degree of mutation and maturity of the repertoire. — repGermline • immunarch</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous"><link href="../docsearch.css" rel="stylesheet"><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="This function creates germlines for clonal lineages. B cell clonal lineage represents a set of B cells
that presumably have a common origin (arising from the same VDJ rearrangement event) and a common ancestor.
Each clonal lineage has its own germline sequence that represents the ancestral sequence
for each BCR in clonal lineage. In other words, germline sequence is a sequence of B-cells immediately
after VDJ recombination, before B-cell maturation and hypermutation process. Germline sequence is useful
for assessing the degree of mutation and maturity of the repertoire. — repGermline"><meta property="og:description" content="Creates germlines for clonal lineages"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-128418614-2"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-128418614-2');
</script></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">immunarch</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.6.8</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Start Here
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/v1_introduction.html">Installation &amp; troubleshooting</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">First steps</li>
    <li>
      <a href="../articles/v2_data.html">Data loading</a>
    </li>
    <li>
      <a href="../articles/web_only/load_mixcr.html">How-to: Loading MiXCR Data</a>
    </li>
    <li>
      <a href="../articles/web_only/load_10x.html">How-to: Loading 10x Genomics Data</a>
    </li>
    <li>
      <a href="../articles/web_only/v21_singlecell.html">How-to: Single-cell and paired chain data</a>
    </li>
    <li>
      <a href="../articles/web_only/v3_basic_analysis.html">Basic statistics and clonality</a>
    </li>
    <li>
      <a href="../articles/web_only/repFilter_v3.html">Data filtering</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Repertoire-level exploration and comparison</li>
    <li>
      <a href="../articles/web_only/v4_overlap.html">Repertoire overlap and public clonotypes</a>
    </li>
    <li>
      <a href="../articles/web_only/v5_gene_usage.html">Gene usage</a>
    </li>
    <li>
      <a href="../articles/web_only/v6_diversity.html">Diversity estimation</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Clonotype-level exploration and comparison</li>
    <li>
      <a href="../articles/web_only/v8_tracking.html">Track clonotypes across samples and time</a>
    </li>
    <li>
      <a href="../articles/web_only/v11_db.html">Annotate clonotypes using immune receptor databases</a>
    </li>
    <li>
      <a href="../articles/web_only/v9_kmers.html">Kmer and sequence motif analysis and visualisation</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Preparing to publication</li>
    <li>
      <a href="../articles/web_only/v7_fixvis.html">Make your plots publication-ready with fixVis</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">BCR analysis</li>
    <li>
      <a href="../articles/web_only/BCRpipeline.html">BCR pipeline</a>
    </li>
    <li>
      <a href="../articles/web_only/clustering.html">Clustering</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="https://github.com/immunomind/covid19">COVID-19</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://immunomind.io">ImmunoMind</a>
</li>
<li>
  <a href="http://twitter.com/immunomind">
    <span class="fa fa-lg fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/immunomind/immunarch">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul><form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off"></div>
      </form>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>This function creates germlines for clonal lineages. B cell clonal lineage represents a set of B cells
that presumably have a common origin (arising from the same VDJ rearrangement event) and a common ancestor.
Each clonal lineage has its own germline sequence that represents the ancestral sequence
for each BCR in clonal lineage. In other words, germline sequence is a sequence of B-cells immediately
after VDJ recombination, before B-cell maturation and hypermutation process. Germline sequence is useful
for assessing the degree of mutation and maturity of the repertoire.</h1>
    <small class="dont-index">Source: <a href="https://github.com/immunomind/immunarch/blob/master/R/germline.R"><code>R/germline.R</code></a></small>
    <div class="hidden name"><code>repGermline.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Creates germlines for clonal lineages</p>
    </div>

    <div class="ref-usage sourceCode"><pre class="sourceCode r"><code><span class="fu">repGermline</span><span class="op">(</span><span class="va">.data</span>, <span class="va">species</span>, <span class="va">min_nuc_outside_cdr3</span>, <span class="va">ref_only_first</span><span class="op">)</span></code></pre></div>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments"><colgroup><col class="name"><col class="desc"></colgroup><tr><th>.data</th>
      <td><p>The data to be processed. Can be <a href="https://rdrr.io/r/base/data.frame.html">data.frame</a>, <a href="https://Rdatatable.gitlab.io/data.table/reference/data.table.html">data.table</a>
or a list of these objects.</p>
<p>It must have columns in the immunarch compatible format <a href="immunr_data_format.html">immunarch_data_format</a>.</p></td>
    </tr><tr><th>species</th>
      <td><p>Species from which the data was acquired. Available options:
"HomoSapiens" (default), "MusMusculus", "BosTaurus", "CamelusDromedarius",
"CanisLupusFamiliaris", "DanioRerio", "MacacaMulatta", "MusMusculusDomesticus",
"MusMusculusCastaneus", "MusMusculusMolossinus", "MusMusculusMusculus", "MusSpretus",
"OncorhynchusMykiss", "OrnithorhynchusAnatinus", "OryctolagusCuniculus", "RattusNorvegicus",
"SusScrofa".</p></td>
    </tr><tr><th>min_nuc_outside_cdr3</th>
      <td><p>This parameter sets how many nucleotides should have V or J chain
outside of CDR3 to be considered good for further alignment.</p></td>
    </tr><tr><th>ref_only_first</th>
      <td><p>This parameter, if TRUE, means to take only first sequence from reference
for each allele name; if FALSE, all sequences will be taken, and output table will
increase in size as a result.</p></td>
    </tr></table><h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Data with added columns V.first.allele, J.first.allele (with first alleles of V and J genes),
V.sequence, J.sequence (with V and J reference sequences),
Germline.sequence (with combined germline sequence)</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <div class="ref-examples sourceCode"><pre class="sourceCode r"><code><span class="r-in"></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">bcrdata</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="va">bcrdata</span><span class="op">$</span><span class="va">data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span class="r-in">  <span class="fu">repGermline</span><span class="op">(</span><span class="op">)</span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Alleles IGHV4-55*01 from sample full_clones not found in the reference and will be dropped!</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> Probably, species argument is wrong (current value: HomoSapiens) or the data contains non-BCR genes.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $full_clones</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 999 × 39</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># Rowwise: </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    J.first.allele V.first.allele Clones Proportion CDR3.nt CDR3.aa V.name D.name</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> IGHJ1*01       IGHV2-5*01         24    2.21<span style="color: #949494;">e</span><span style="color: #BB0000;">-6</span> TGTGTC… CVEGQP… IGHV2… IGHD6…</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> IGHJ1*01       IGHV4-30-2*01      59    5.42<span style="color: #949494;">e</span><span style="color: #BB0000;">-6</span> TGTGTG… CVRDQR… IGHV4… IGHD2…</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> IGHJ1*01       IGHV1-3*01          8    7.35<span style="color: #949494;">e</span><span style="color: #BB0000;">-7</span> TGTGCG… CARGIT… IGHV1… IGHD3…</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> IGHJ1*01       IGHV1-69*01       824    7.57<span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span> TGTGCG… CATPAE… IGHV1… IGHD5…</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> IGHJ1*01       IGHV1-69*01        42    3.86<span style="color: #949494;">e</span><span style="color: #BB0000;">-6</span> TGTGTG… CVRDSA… IGHV1… IGHD6…</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> IGHJ2*01       IGHV1-46*01         7    6.43<span style="color: #949494;">e</span><span style="color: #BB0000;">-7</span> TGTGCG… CARASR… IGHV1… IGHD5…</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> IGHJ2*01       IGHV1-18*01        21    1.93<span style="color: #949494;">e</span><span style="color: #BB0000;">-6</span> TGTGCG… CARSYY… IGHV1… IGHD3…</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> IGHJ2*01       IGHV3-21*01         3    2.76<span style="color: #949494;">e</span><span style="color: #BB0000;">-7</span> TGTGCA… CARVPS… IGHV3… IGHD3…</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> IGHJ3*01       IGHV1-18*01        26    2.39<span style="color: #949494;">e</span><span style="color: #BB0000;">-6</span> TGTGCG… CARVLT… IGHV1… IGHD3…</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> IGHJ3*01       IGHV1-2*01         10    9.19<span style="color: #949494;">e</span><span style="color: #BB0000;">-7</span> TGTGCG… CARGGG… IGHV1… IGHD3…</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># … with 989 more rows, and 31 more variables: J.name &lt;chr&gt;, V.end &lt;int&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   D.start &lt;int&gt;, D.end &lt;int&gt;, J.start &lt;int&gt;, VJ.ins &lt;int&gt;, VD.ins &lt;int&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   DJ.ins &lt;int&gt;, Sequence &lt;chr&gt;, CDR3.start &lt;chr&gt;, CDR3.end &lt;chr&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   C.name &lt;chr&gt;, C.start &lt;list&gt;, C.end &lt;list&gt;, CDR1.nt &lt;chr&gt;, CDR1.aa &lt;chr&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   CDR2.nt &lt;chr&gt;, CDR2.aa &lt;chr&gt;, FR1.nt &lt;chr&gt;, FR1.aa &lt;chr&gt;, FR2.nt &lt;chr&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   FR2.aa &lt;chr&gt;, FR3.nt &lt;chr&gt;, FR3.aa &lt;chr&gt;, FR4.nt &lt;chr&gt;, FR4.aa &lt;chr&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   V3.Deletions &lt;chr&gt;, J3.Deletions &lt;chr&gt;, V.sequence &lt;chr&gt;, …</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
</code></pre></div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="https://www.linkedin.com/in/vdnaz" class="external-link">Vadim I. Nazarov</a>, <a href="https://www.linkedin.com/in/vasily-tsvetkov-227218ab" class="external-link">Vasily O. Tsvetkov</a>, <a href="https://www.linkedin.com/in/erumynskiy/" class="external-link">Eugene Rumynskiy</a>, <a href="https://www.linkedin.com/in/aleksandr-popov-634878105/" class="external-link">Aleksandr A. Popov</a>, <a href="https://www.linkedin.com/in/ivan-balashov/" class="external-link">Ivan Balashov</a>, <a href="https://www.linkedin.com/in/maria-volobueva-b0856a223/" class="external-link">Maria Samokhina</a>, <a href="https://immunomind.io" class="external-link">ImmunoMind</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 1.9000.9000.9000.</p>
</div>

      </footer></div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '94207562f34b455c0796790d104e3549',
    indexName: 'immunarch',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script></body></html>

