<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Load immune repertoire files into the R workspace — repLoad • immunarch</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous"><link href="../docsearch.css" rel="stylesheet"><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Load immune repertoire files into the R workspace — repLoad"><meta property="og:description" content='The repLoad function loads repertoire files
into R workspace in the immunarch format where you can immediately use them for
the analysis. repLoad automatically detects the right format for
your files, so all you need is simply provide the path to your files.
See "Details" for more information on supported formats. See "Examples" for
diving right into it.'><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-128418614-2"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-128418614-2');
</script></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">immunarch</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Start Here
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/v1_introduction.html">Installation &amp; troubleshooting</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">First steps</li>
    <li>
      <a href="../articles/v2_data.html">Data loading</a>
    </li>
    <li>
      <a href="../articles/web_only/load_mixcr.html">How-to: Loading MiXCR Data</a>
    </li>
    <li>
      <a href="../articles/web_only/load_10x.html">How-to: Loading 10x Genomics Data</a>
    </li>
    <li>
      <a href="../articles/web_only/v21_singlecell.html">How-to: Single-cell and paired chain data</a>
    </li>
    <li>
      <a href="../articles/web_only/v3_basic_analysis.html">Basic statistics and clonality</a>
    </li>
    <li>
      <a href="../articles/web_only/repFilter_v3.html">Data filtering</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Repertoire-level exploration and comparison</li>
    <li>
      <a href="../articles/web_only/v4_overlap.html">Repertoire overlap and public clonotypes</a>
    </li>
    <li>
      <a href="../articles/web_only/v5_gene_usage.html">Gene usage</a>
    </li>
    <li>
      <a href="../articles/web_only/v6_diversity.html">Diversity estimation</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Clonotype-level exploration and comparison</li>
    <li>
      <a href="../articles/web_only/v8_tracking.html">Track clonotypes across samples and time</a>
    </li>
    <li>
      <a href="../articles/web_only/v11_db.html">Annotate clonotypes using immune receptor databases</a>
    </li>
    <li>
      <a href="../articles/web_only/v9_kmers.html">Kmer and sequence motif analysis and visualisation</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Preparing to publication</li>
    <li>
      <a href="../articles/web_only/v7_fixvis.html">Make your plots publication-ready with fixVis</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">BCR analysis</li>
    <li>
      <a href="../articles/web_only/BCRpipeline.html">BCR pipeline</a>
    </li>
    <li>
      <a href="../articles/web_only/clustering.html">Clustering</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="https://github.com/immunomind/covid19" class="external-link">COVID-19</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://immunomind.io" class="external-link">ImmunoMind</a>
</li>
<li>
  <a href="http://twitter.com/immunomind" class="external-link">
    <span class="fa fa-lg fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/immunomind/immunarch" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul><form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off"></div>
      </form>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Load immune repertoire files into the R workspace</h1>
    <small class="dont-index">Source: <a href="https://github.com/immunomind/immunarch/blob/HEAD/R/io.R" class="external-link"><code>R/io.R</code></a></small>
    <div class="hidden name"><code>repLoad.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The <code>repLoad</code> function loads repertoire files
into R workspace in the immunarch format where you can immediately use them for
the analysis. <code>repLoad</code> automatically detects the right format for
your files, so all you need is simply provide the path to your files.</p>
<p>See "Details" for more information on supported formats. See "Examples" for
diving right into it.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">repLoad</span><span class="op">(</span><span class="va">.path</span>, .mode <span class="op">=</span> <span class="st">"paired"</span>, .coding <span class="op">=</span> <span class="cn">TRUE</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>.path</dt>
<dd><p>A character string specifying the path to the input data.
Input data can be one of the following:</p>
<p>- a single repertoire file.
In this case <code>repLoad</code> returns an R <a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a>;</p>
<p>- a vector of paths to repertoire files.
Same as in the case with no metadata file presented in the next section below;</p>
<p>- a path to the folder with repertoire files and, if available, metadata file "metadata.txt".
If the metadata file if presented, then the <code>repLoad</code> returns a list with two elements "data" and "meta".
"data" is an another list with repertoire R <a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a>s. "meta" is a data frame with the metadata.
If the metadata file "metadata.txt" is not presented, then the <code>repLoad</code> creates a dummy metadata file with
sample names and returns a list with two elements "data" and "meta".
If input data has multiple chains or cell types stored in the same file
(for example, like in 10xGenomics repertoire files), such repertoire files will be splitted to different
R data frames with only one type of chain and cell presented. The metadata file will have additional columns specifying
cell and chain types for different samples.</p></dd>


<dt>.mode</dt>
<dd><p>Either "single" for single chain data or "paired" for paired chain data.</p>
<p>Currently "single" works for every format, and "paired" works only for 10X Genomics data.</p>
<p>By default, 10X Genomics data will be loaded as paired chain data, and other files will be loaded as single chain data.</p></dd>


<dt>.coding</dt>
<dd><p>A logical value. Set TRUE to get coding-only clonotypes (by defaul). Set FALSE to get all clonotypes.</p></dd>


<dt>...</dt>
<dd><p>Extra arguments for parsing functions</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A list with two named elements:</p>


<p>- "data" is a list of input samples;</p>


<p>- "meta" is a data frame with sample metadata.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>The metadata has to be a tab delimited file with first column named "Sample".
It can have any number of additional columns with arbitrary names.
The first column should contain base names of files without extensions in
your folder. Example:</p><table class="table table"><tr><td>Sample</td><td>Sex</td><td>Age</td><td>Status</td></tr><tr><td>immunoseq_1</td><td>M</td><td>1</td><td>C</td></tr><tr><td>immunoseq_2</td><td>M</td><td>2</td><td>C</td></tr><tr><td>immunoseq_3</td><td>FALSE</td><td>3</td><td>A</td></tr></table><p>Currently, Immunarch support the following formats:</p>
<p>- "immunoseq" - ImmunoSEQ of any version. http://www.adaptivebiotech.com/immunoseq</p>
<p>- "mitcr" - MiTCR. https://github.com/milaboratory/mitcr</p>
<p>- "mixcr" - MiXCR (the "all" files) of any version. https://github.com/milaboratory/mixcr</p>
<p>- "migec" - MiGEC. http://migec.readthedocs.io/en/latest/</p>
<p>- "migmap" - For parsing IgBLAST results postprocessed with MigMap. https://github.com/mikessh/migmap</p>
<p>- "tcr" - tcR, our previous package. https://imminfo.github.io/tcr/</p>
<p>- "vdjtools" - VDJtools of any version. http://vdjtools-doc.readthedocs.io/en/latest/</p>
<p>- "imgt" - IMGT HighV-QUEST. http://www.imgt.org/HighV-QUEST/</p>
<p>- "airr" - adaptive immune receptor repertoire (AIRR) data format. http://docs.airr-community.org/en/latest/datarep/overview.html</p>
<p>- "10x" - 10XGenomics clonotype annotations tables. https://support.10xgenomics.com/single-cell-vdj/software/pipelines/latest/output/annotation</p>
<p>- "archer" - ArcherDX clonotype tables. https://archerdx.com/</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><a href="immunr_data_format.html">immunr_data_format</a> for immunarch data format; <a href="repSave.html">repSave</a> for file saving;
<a href="repOverlap.html">repOverlap</a>, <a href="geneUsage.html">geneUsage</a> and <a href="repDiversity.html">repDiversity</a> for starting with immune repertoires basic statistics.</p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># To load the data from a single file (note that you don't need to specify the data format):</span></span></span>
<span class="r-in"><span><span class="va">file_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package <span class="op">=</span> <span class="st">"immunarch"</span><span class="op">)</span>, <span class="st">"/extdata/io/Sample1.tsv.gz"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">immdata</span> <span class="op">&lt;-</span> <span class="fu">repLoad</span><span class="op">(</span><span class="va">file_path</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> == Step 1/3: loading repertoire files... ==</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Processing "&lt;initial&gt;" ...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   -- [1/1] Parsing "/tmp/RtmpYauDlO/temp_libpath125bc1594e6916/immunarch/extdata/io/Sample1.tsv.gz" -- </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> immunarch</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> == Step 2/3: checking metadata files and merging files... ==</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Processing "&lt;initial&gt;" ...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   -- Metadata file not found; creating a dummy metadata...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> == Step 3/3: processing paired chain data... ==</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Done!</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Suppose you have a following structure in your folder:</span></span></span>
<span class="r-in"><span><span class="co"># &gt;_ ls</span></span></span>
<span class="r-in"><span><span class="co"># immunoseq1.txt</span></span></span>
<span class="r-in"><span><span class="co"># immunoseq2.txt</span></span></span>
<span class="r-in"><span><span class="co"># immunoseq3.txt</span></span></span>
<span class="r-in"><span><span class="co"># metadata.txt</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># To load the whole folder with every file in it type:</span></span></span>
<span class="r-in"><span><span class="va">file_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package <span class="op">=</span> <span class="st">"immunarch"</span><span class="op">)</span>, <span class="st">"/extdata/io/"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">immdata</span> <span class="op">&lt;-</span> <span class="fu">repLoad</span><span class="op">(</span><span class="va">file_path</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> == Step 1/3: loading repertoire files... ==</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Processing "/tmp/RtmpYauDlO/temp_libpath125bc1594e6916/immunarch/extdata/io/" ...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   -- [1/5] Parsing "/tmp/RtmpYauDlO/temp_libpath125bc1594e6916/immunarch/extdata/io//Sample1.tsv.gz" -- </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> immunarch</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   -- [2/5] Parsing "/tmp/RtmpYauDlO/temp_libpath125bc1594e6916/immunarch/extdata/io//Sample2.tsv.gz" -- </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> immunarch</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   -- [3/5] Parsing "/tmp/RtmpYauDlO/temp_libpath125bc1594e6916/immunarch/extdata/io//Sample3.tsv.gz" -- </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> immunarch</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   -- [4/5] Parsing "/tmp/RtmpYauDlO/temp_libpath125bc1594e6916/immunarch/extdata/io//Sample4.tsv.gz" -- </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> immunarch</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   -- [5/5] Parsing "/tmp/RtmpYauDlO/temp_libpath125bc1594e6916/immunarch/extdata/io//metadata.txt" -- </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> metadata</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> == Step 2/3: checking metadata files and merging files... ==</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Processing "/tmp/RtmpYauDlO/temp_libpath125bc1594e6916/immunarch/extdata/io/" ...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   -- Everything is OK!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> == Step 3/3: processing paired chain data... ==</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Done!</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">immdata</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "data" "meta"</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># We recommend creating a metadata file named "metadata.txt" in the folder.</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># In that case, when you load your data you will see:</span></span></span>
<span class="r-in"><span><span class="co"># &gt; immdata &lt;- repLoad("path/to/your/folder/")</span></span></span>
<span class="r-in"><span><span class="co"># &gt; names(immdata)</span></span></span>
<span class="r-in"><span><span class="co"># [1] "data" "meta"</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># If you do not have "metadata.txt", you will see the same output,</span></span></span>
<span class="r-in"><span><span class="co"># but your metadata will be almost empty:</span></span></span>
<span class="r-in"><span><span class="co"># &gt; immdata &lt;- repLoad("path/to/your/folder/")</span></span></span>
<span class="r-in"><span><span class="co"># &gt; names(immdata)</span></span></span>
<span class="r-in"><span><span class="co"># [1] "data" "meta"</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="https://www.linkedin.com/in/vdnaz" class="external-link">Vadim I. Nazarov</a>, <a href="https://www.linkedin.com/in/vasily-tsvetkov-227218ab" class="external-link">Vasily O. Tsvetkov</a>, <a href="https://www.linkedin.com/in/sergey-f-709b0415b/" class="external-link">Siarhei Fiadziushchanka</a>, <a href="https://www.linkedin.com/in/erumynskiy/" class="external-link">Eugene Rumynskiy</a>, <a href="https://www.linkedin.com/in/aleksandr-popov-634878105/" class="external-link">Aleksandr A. Popov</a>, <a href="https://www.linkedin.com/in/ivan-balashov/" class="external-link">Ivan Balashov</a>, <a href="https://www.linkedin.com/in/maria-volobueva-b0856a223/" class="external-link">Maria Samokhina</a>, <a href="https://immunomind.io" class="external-link">ImmunoMind</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '94207562f34b455c0796790d104e3549',
    indexName: 'immunarch',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script></body></html>

