<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Track clonotypes across time and data points — trackClonotypes • immunarch</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>


<!-- docsearch -->
<script src="../docsearch.js"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous" />
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script>


  <link href="../extra.css" rel="stylesheet">
  

<meta property="og:title" content="Track clonotypes across time and data points — trackClonotypes" />
<meta property="og:description" content="Track the temporal dynamics of clonotypes in repertoires. For example, tracking across multiple
time points after vaccination.
Note: duplicated clonotypes are merged and their counts are summed up." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-128418614-2"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-128418614-2');
</script>


  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">immunarch</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.6.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Start Here
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/v1_introduction.html">Installation &amp; troubleshooting</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">First steps</li>
    <li>
      <a href="../articles/v2_data.html">Data loading</a>
    </li>
    <li>
      <a href="../articles/web_only/load_mixcr.html">How-to: Loading MiXCR Data</a>
    </li>
    <li>
      <a href="../articles/web_only/load_10x.html">How-to: Loading 10x Genomics Data</a>
    </li>
    <li>
      <a href="../articles/web_only/v21_singlecell.html">How-to: Single-cell and paired chain data</a>
    </li>
    <li>
      <a href="../articles/web_only/v3_basic_analysis.html">Basic statistics and clonality</a>
    </li>
    <li>
      <a href="../articles/web_only/repFilter_v3.html">Data filtering</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Repertoire-level exploration and comparison</li>
    <li>
      <a href="../articles/web_only/v4_overlap.html">Repertoire overlap and public clonotypes</a>
    </li>
    <li>
      <a href="../articles/web_only/v5_gene_usage.html">Gene usage</a>
    </li>
    <li>
      <a href="../articles/web_only/v6_diversity.html">Diversity estimation</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Clonotype-level exploration and comparison</li>
    <li>
      <a href="../articles/web_only/v8_tracking.html">Track clonotypes across samples and time</a>
    </li>
    <li>
      <a href="../articles/web_only/v11_db.html">Annotate clonotypes using immune receptor databases</a>
    </li>
    <li>
      <a href="../articles/web_only/v9_kmers.html">Kmer and sequence motif analysis and visualisation</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Preparing to publication</li>
    <li>
      <a href="../articles/web_only/v7_fixvis.html">Make your plots publication-ready with fixVis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="https://github.com/immunomind/covid19">COVID-19</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://immunomind.io">ImmunoMind</a>
</li>
<li>
  <a href="http://twitter.com/immunomind">
    <span class="fa fa-lg fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/immunomind/immunarch">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
      
      <form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
        </div>
      </form>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Track clonotypes across time and data points</h1>
    <small class="dont-index">Source: <a href='https://github.com/immunomind/immunarch/blob/master/R/dynamics.R'><code>R/dynamics.R</code></a></small>
    <div class="hidden name"><code>trackClonotypes.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Track the temporal dynamics of clonotypes in repertoires. For example, tracking across multiple
time points after vaccination.</p>
<p>Note: duplicated clonotypes are merged and their counts are summed up.</p>
    </div>

    <pre class="usage"><span class='fu'>trackClonotypes</span><span class='op'>(</span><span class='va'>.data</span>, .which <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>15</span><span class='op'>)</span>, .col <span class='op'>=</span> <span class='st'>"aa"</span>, .norm <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>.data</th>
      <td><p>The data to process. It can be a <a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a>, a
data.table, or a list of these objects.</p>
<p>Every object must have columns in the immunarch compatible format.
<a href='immunr_data_format.html'>immunarch_data_format</a></p>
<p>Competent users may provide advanced data representations:
DBI database connections, Apache Spark DataFrame from copy_to or a list
of these objects. They are supported with the same limitations as basic objects.</p>
<p>Note: each connection must represent a separate repertoire.</p></td>
    </tr>
    <tr>
      <th>.which</th>
      <td><p>An argument that regulates which clonotypes to choose for tracking. There are three options for this argument:</p>
<p>1) pass a list with two elements <code><a href='https://rdrr.io/r/base/list.html'>list(X, Y)</a></code>, where <code>X</code> is the name or the index of a target repertoire from ".data", and
<code>Y</code> is the number of the most abundant clonotypes to take from <code>X</code>.</p>
<p>2) pass a character vector of sequences to take from all data frames;</p>
<p>3) pass a data frame (data table, database) with one or more columns - first for sequences, and other for gene segments (if applicable).</p>
<p>See the "Examples" below with examples for each option.</p></td>
    </tr>
    <tr>
      <th>.col</th>
      <td><p>A character vector of length 1. Specifies an identifier for a column, from which the function
chooses clonotype sequences. Specify "nt" for nucleotide sequences, "aa" for amino acid sequences,
"aa+v" for amino acid sequences and Variable genes, "nt+j" for nucleotide
sequences with Joining genes, or any combination of the above.
Used only if ".which" has option 1) or option 2).</p></td>
    </tr>
    <tr>
      <th>.norm</th>
      <td><p>Logical. If TRUE then use Proportion instead of the number of Clones per clonotype to store
in the function output.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Data frame with input sequences and counts or proportions for each of the input repertoire.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># Load an example data that comes with immunarch</span>
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>immdata</span><span class='op'>)</span>

<span class='co'># Make the data smaller in order to speed up the examples</span>
<span class='va'>immdata</span><span class='op'>$</span><span class='va'>data</span> <span class='op'>&lt;-</span> <span class='va'>immdata</span><span class='op'>$</span><span class='va'>data</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>2</span>, <span class='fl'>3</span>, <span class='fl'>7</span>, <span class='fl'>8</span>, <span class='fl'>9</span><span class='op'>)</span><span class='op'>]</span>
<span class='va'>immdata</span><span class='op'>$</span><span class='va'>meta</span> <span class='op'>&lt;-</span> <span class='va'>immdata</span><span class='op'>$</span><span class='va'>meta</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>2</span>, <span class='fl'>3</span>, <span class='fl'>7</span>, <span class='fl'>8</span>, <span class='fl'>9</span><span class='op'>)</span>, <span class='op'>]</span>

<span class='co'># Option 1</span>
<span class='co'># Choose the first 10 amino acid clonotype sequences</span>
<span class='co'># from the first repertoire to track</span>
<span class='va'>tc</span> <span class='op'>&lt;-</span> <span class='fu'>trackClonotypes</span><span class='op'>(</span><span class='va'>immdata</span><span class='op'>$</span><span class='va'>data</span>, <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>10</span><span class='op'>)</span>, .col <span class='op'>=</span> <span class='st'>"aa"</span><span class='op'>)</span>
<span class='co'># Choose the first 20 nucleotide clonotype sequences</span>
<span class='co'># and their V genes from the "MS1" repertoire to track</span>
<span class='va'>tc</span> <span class='op'>&lt;-</span> <span class='fu'>trackClonotypes</span><span class='op'>(</span><span class='va'>immdata</span><span class='op'>$</span><span class='va'>data</span>, <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span><span class='st'>"MS1"</span>, <span class='fl'>20</span><span class='op'>)</span>, .col <span class='op'>=</span> <span class='st'>"nt+v"</span><span class='op'>)</span>

<span class='co'># Option 2</span>
<span class='co'># Choose clonotypes with amino acid sequences "CASRGLITDTQYF" or "CSASRGSPNEQYF"</span>
<span class='va'>tc</span> <span class='op'>&lt;-</span> <span class='fu'>trackClonotypes</span><span class='op'>(</span><span class='va'>immdata</span><span class='op'>$</span><span class='va'>data</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"CASRGLITDTQYF"</span>, <span class='st'>"CSASRGSPNEQYF"</span><span class='op'>)</span>, .col <span class='op'>=</span> <span class='st'>"aa"</span><span class='op'>)</span>

<span class='co'># Option 3</span>
<span class='co'># Choose the first 10 clonotypes from the first repertoire</span>
<span class='co'># with amino acid sequences and V segments</span>
<span class='va'>target</span> <span class='op'>&lt;-</span> <span class='va'>immdata</span><span class='op'>$</span><span class='va'>data</span><span class='op'>[[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>]</span> <span class='op'>%&gt;%</span>
  <span class='fu'>select</span><span class='op'>(</span><span class='va'>CDR3.aa</span>, <span class='va'>V.name</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='fl'>10</span><span class='op'>)</span>
<span class='va'>tc</span> <span class='op'>&lt;-</span> <span class='fu'>trackClonotypes</span><span class='op'>(</span><span class='va'>immdata</span><span class='op'>$</span><span class='va'>data</span>, <span class='va'>target</span><span class='op'>)</span>

<span class='co'># Visualise the output regardless of the chosen option</span>
<span class='co'># Therea are three way to visualise it, regulated by the .plot argument</span>
<span class='fu'><a href='vis.html'>vis</a></span><span class='op'>(</span><span class='va'>tc</span>, .plot <span class='op'>=</span> <span class='st'>"smooth"</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='warning'>Warning: id.vars and measure.vars are internally guessed when both are 'NULL'. All non-numeric/integer/logical type columns are considered id.vars, which in this case are columns [CDR3.aa, V.name, ...]. Consider providing at least one of 'id' or 'measure' vars in future.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: The `.dots` argument of `group_by()` is deprecated as of dplyr 1.0.0.</span></div><div class='img'><img src='trackClonotypes-1.png' alt='' width='700' height='433' /></div><div class='input'><span class='fu'><a href='vis.html'>vis</a></span><span class='op'>(</span><span class='va'>tc</span>, .plot <span class='op'>=</span> <span class='st'>"area"</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='warning'>Warning: id.vars and measure.vars are internally guessed when both are 'NULL'. All non-numeric/integer/logical type columns are considered id.vars, which in this case are columns [CDR3.aa, V.name, ...]. Consider providing at least one of 'id' or 'measure' vars in future.</span></div><div class='img'><img src='trackClonotypes-2.png' alt='' width='700' height='433' /></div><div class='input'><span class='fu'><a href='vis.html'>vis</a></span><span class='op'>(</span><span class='va'>tc</span>, .plot <span class='op'>=</span> <span class='st'>"line"</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='warning'>Warning: id.vars and measure.vars are internally guessed when both are 'NULL'. All non-numeric/integer/logical type columns are considered id.vars, which in this case are columns [CDR3.aa, V.name, ...]. Consider providing at least one of 'id' or 'measure' vars in future.</span></div><div class='img'><img src='trackClonotypes-3.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='co'># Visualising timepoints</span>
<span class='co'># First, we create an additional column in the metadata with randomly choosen timepoints:</span>
<span class='va'>immdata</span><span class='op'>$</span><span class='va'>meta</span><span class='op'>$</span><span class='va'>Timepoint</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/sample.html'>sample</a></span><span class='op'>(</span><span class='fl'>1</span><span class='op'>:</span><span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span><span class='op'>(</span><span class='va'>immdata</span><span class='op'>$</span><span class='va'>data</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>immdata</span><span class='op'>$</span><span class='va'>meta</span>
</div><div class='output co'>#&gt; <span style='color: #949494;'># A tibble: 6 × 7</span>
#&gt;   Sample  ID    Sex     Age Status Lane  Timepoint
#&gt;   <span style='color: #949494; font-style: italic;'>&lt;chr&gt;</span>   <span style='color: #949494; font-style: italic;'>&lt;chr&gt;</span> <span style='color: #949494; font-style: italic;'>&lt;chr&gt;</span> <span style='color: #949494; font-style: italic;'>&lt;dbl&gt;</span> <span style='color: #949494; font-style: italic;'>&lt;chr&gt;</span>  <span style='color: #949494; font-style: italic;'>&lt;chr&gt;</span>     <span style='color: #949494; font-style: italic;'>&lt;int&gt;</span>
#&gt; <span style='color: #BCBCBC;'>1</span> A2-i129 C1    M        11 C      A             1
#&gt; <span style='color: #BCBCBC;'>2</span> A2-i131 C2    M         9 C      A             5
#&gt; <span style='color: #BCBCBC;'>3</span> A2-i133 C4    M        16 C      A             4
#&gt; <span style='color: #BCBCBC;'>4</span> MS1     MS1   M        12 MS     C             6
#&gt; <span style='color: #BCBCBC;'>5</span> MS2     MS2   M        30 MS     C             3
#&gt; <span style='color: #BCBCBC;'>6</span> MS3     MS3   M         8 MS     C             2</div><div class='input'><span class='co'># Next, we create a vector with samples in the right order,</span>
<span class='co'># according to the "Timepoint" column (from smallest to greatest):</span>
<span class='va'>sample_order</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/order.html'>order</a></span><span class='op'>(</span><span class='va'>immdata</span><span class='op'>$</span><span class='va'>meta</span><span class='op'>$</span><span class='va'>Timepoint</span><span class='op'>)</span>
<span class='co'># Sanity check: timepoints are following the right order:</span>
<span class='va'>immdata</span><span class='op'>$</span><span class='va'>meta</span><span class='op'>$</span><span class='va'>Timepoint</span><span class='op'>[</span><span class='va'>sample_order</span><span class='op'>]</span>
</div><div class='output co'>#&gt; [1] 1 2 3 4 5 6</div><div class='input'><span class='co'># Samples, sorted by the timepoints:</span>
<span class='va'>immdata</span><span class='op'>$</span><span class='va'>meta</span><span class='op'>$</span><span class='va'>Sample</span><span class='op'>[</span><span class='va'>sample_order</span><span class='op'>]</span>
</div><div class='output co'>#&gt; [1] "A2-i129" "MS3"     "MS2"     "A2-i133" "A2-i131" "MS1"    </div><div class='input'><span class='co'># And finally, we visualise the data:</span>
<span class='fu'><a href='vis.html'>vis</a></span><span class='op'>(</span><span class='va'>tc</span>, .order <span class='op'>=</span> <span class='va'>sample_order</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='warning'>Warning: id.vars and measure.vars are internally guessed when both are 'NULL'. All non-numeric/integer/logical type columns are considered id.vars, which in this case are columns [CDR3.aa, V.name, ...]. Consider providing at least one of 'id' or 'measure' vars in future.</span></div><div class='img'><img src='trackClonotypes-4.png' alt='' width='700' height='433' /></div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://www.linkedin.com/in/vdnaz'>Vadim I. Nazarov</a>, <a href='https://www.linkedin.com/in/vasily-tsvetkov-227218ab'>Vasily O. Tsvetkov</a>, <a href='https://www.linkedin.com/in/erumynskiy/'>Eugene Rumynskiy</a>, <a href='https://www.linkedin.com/in/aleksandr-popov-634878105/'>Aleksandr A. Popov</a>, <a href='https://www.linkedin.com/in/ivan-balashov/'>Ivan Balashov</a>, <a href='https://www.linkedin.com/in/maria-volobueva-b0856a223/'>Maria Volobueva</a>, <a href='https://immunomind.io'>ImmunoMind</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script>
<script>
  docsearch({
    
    
    apiKey: '94207562f34b455c0796790d104e3549',
    indexName: 'immunarch',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>



  </body>
</html>


