<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Visualisation of distributions using histograms — vis_hist • immunarch</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>


<!-- docsearch -->
<script src="../docsearch.js"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous" />
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script>


  <link href="../extra.css" rel="stylesheet">
  

<meta property="og:title" content="Visualisation of distributions using histograms — vis_hist" />
<meta property="og:description" content="Visualisation of distributions using ggplot2-based histograms." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-128418614-2"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-128418614-2');
</script>


  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">immunarch</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.6.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="https://github.com/immunomind/covid19">COVID-19</a>
</li>
<li>
  <a href="../articles/v1_introduction.html">Get started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">First steps</li>
    <li>
      <a href="../articles/v1_introduction.html">Introduction to immunarch</a>
    </li>
    <li>
      <a href="../articles/v2_data.html">Working with data in immunarch</a>
    </li>
    <li>
      <a href="../articles/v3_basic_analysis.html">Basic analysis and clonality</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Explore and compare repertoires</li>
    <li>
      <a href="../articles/web_only/v4_overlap.html">Repertoire overlap and public clonotypes</a>
    </li>
    <li>
      <a href="../articles/web_only/v5_gene_usage.html">Gene usage analysis</a>
    </li>
    <li>
      <a href="../articles/web_only/v6_diversity.html">Diversity estimation</a>
    </li>
    <li>
      <a href="../articles/web_only/v11_db.html">Annotate clonotypes using immune receptor databases</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Preparing to publication</li>
    <li>
      <a href="../articles/web_only/v7_fixvis.html">Make your plots publication-ready with fixVis</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Advanced methods</li>
    <li>
      <a href="../articles/web_only/v8_tracking.html">Tracking clonotypes across time points</a>
    </li>
    <li>
      <a href="../articles/web_only/v9_kmers.html">Kmer and sequence motif analysis and visualisation</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://immunomind.io">ImmunoMind</a>
</li>
<li>
  <a href="http://twitter.com/immunomind">
    <span class="fa fa-lg fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/immunomind/immunarch">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
      
      <form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
        </div>
      </form>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Visualisation of distributions using histograms</h1>
    <small class="dont-index">Source: <a href='https://github.com/immunomind/immunarch/blob/master/R/vis.R'><code>R/vis.R</code></a></small>
    <div class="hidden name"><code>vis_hist.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Visualisation of distributions using ggplot2-based histograms.</p>
    </div>

    <pre class="usage"><span class='fu'>vis_hist</span>(
  <span class='no'>.data</span>,
  <span class='kw'>.by</span> <span class='kw'>=</span> <span class='fl'>NA</span>,
  <span class='kw'>.meta</span> <span class='kw'>=</span> <span class='fl'>NA</span>,
  <span class='kw'>.title</span> <span class='kw'>=</span> <span class='st'>"Gene usage"</span>,
  <span class='kw'>.ncol</span> <span class='kw'>=</span> <span class='fl'>NA</span>,
  <span class='kw'>.points</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>.test</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>.coord.flip</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>.grid</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>.labs</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"Gene"</span>, <span class='fl'>NA</span>),
  <span class='kw'>.return.grob</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>.melt</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>.legend</span> <span class='kw'>=</span> <span class='fl'>NA</span>,
  <span class='kw'>.add.layer</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='no'>...</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>.data</th>
      <td><p>Input matrix or data frame.</p></td>
    </tr>
    <tr>
      <th>.by</th>
      <td><p>Pass NA if you want to plot samples without grouping.</p>
<p>You can pass a character vector with one or several column names from ".meta"
to group your data before plotting. In this case you should provide ".meta".</p>
<p>You can pass a character vector that exactly matches the number of samples in
your data, each value should correspond to a sample's property. It will be used
to group data based on the values provided. Note that in this case you should
pass NA to ".meta".</p></td>
    </tr>
    <tr>
      <th>.meta</th>
      <td><p>A metadata object. An R dataframe with sample names and their properties,
such as age, serostatus or hla.</p></td>
    </tr>
    <tr>
      <th>.title</th>
      <td><p>The text for the title of the plot.</p></td>
    </tr>
    <tr>
      <th>.ncol</th>
      <td><p>A number of columns to display. Provide NA (by default) if you want the function
to automatically detect the optimal number of columns.</p></td>
    </tr>
    <tr>
      <th>.points</th>
      <td><p>A logical value defining whether points will be visualised or not.</p></td>
    </tr>
    <tr>
      <th>.test</th>
      <td><p>A logical vector whether statistical tests should be applied. See "Details" for more information.</p></td>
    </tr>
    <tr>
      <th>.coord.flip</th>
      <td><p>If TRUE then swap x- and y-axes.</p></td>
    </tr>
    <tr>
      <th>.grid</th>
      <td><p>If TRUE then plot separate visualisations for each sample.</p></td>
    </tr>
    <tr>
      <th>.labs</th>
      <td><p>A character vector of length two with names for x-axis and y-axis, respectively.</p></td>
    </tr>
    <tr>
      <th>.return.grob</th>
      <td><p>If TRUE then returh the gridArrange grob instead of the plot.</p></td>
    </tr>
    <tr>
      <th>.melt</th>
      <td><p>If TRUE then apply melt to the ".data" before plotting.
In this case ".data" is supposed to be a data frame with the first character column reserved
for names of genes and other numeric columns reserved to counts or frequencies of genes.
Each numeric column should be associated with a specific repertoire sample.</p></td>
    </tr>
    <tr>
      <th>.legend</th>
      <td><p>If TRUE then plots the legend. If FALSE removes the legend from the plot.
If NA automatically detects the best way to display legend.</p></td>
    </tr>
    <tr>
      <th>.add.layer</th>
      <td><p>Addditional ggplot2 layers, that added to each plot in the output plot or grid of plots.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Is not used here.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A ggplot2 object.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>If data is grouped, then statistical tests for comparing means of groups will be performed, unless <code>.test = FALSE</code> is supplied.
In case there are only two groups, the Wilcoxon rank sum test (https://en.wikipedia.org/wiki/Wilcoxon_signed-rank_test) is performed
(R function <code><a href='https://rdrr.io/r/stats/wilcox.test.html'>wilcox.test</a></code> with an argument <code>exact = FALSE</code>) for testing if there is a difference in mean rank values between two groups.
In case there more than two groups, the Kruskal-Wallis test (https://en.wikipedia.org/wiki/Kruskal<!-- %E2%80%93Wallis_one-way_analysis_of_variance) is performed (R function \code{\link{kruskal.test}}), that is equivalent to ANOVA for ranks and it tests whether samples from different groups originated from the same distribution. -->
A significant Kruskal-Wallis test indicates that at least one sample stochastically dominates one other sample.
Adjusted for multiple comparisons P-values are plotted on the top of groups.
P-value adjusting is done using the Holm method (https://en.wikipedia.org/wiki/Holm<!-- %E2%80%93Bonferroni_method) (also known as Holm-Bonferroni correction). -->
You can execute the command <code><a href='https://rdrr.io/r/stats/p.adjust.html'>?p.adjust</a></code> in the R console to see more.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><a href='vis.immunr_gene_usage.html'>vis.immunr_gene_usage</a>, <a href='geneUsage.html'>geneUsage</a></p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='no'>immdata</span>)
<span class='no'>imm_gu</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='geneUsage.html'>geneUsage</a></span>(<span class='no'>immdata</span>$<span class='no'>data</span><span class='kw'>[[</span><span class='fl'>1</span>]])
<span class='fu'><a href='vis.html'>vis</a></span>(<span class='no'>imm_gu</span>,
  <span class='kw'>.plot</span> <span class='kw'>=</span> <span class='st'>"hist"</span>, <span class='kw'>.add.layer</span> <span class='kw'>=</span>
    <span class='fu'>theme</span>(<span class='kw'>axis.text.x</span> <span class='kw'>=</span> <span class='fu'>element_text</span>(<span class='kw'>angle</span> <span class='kw'>=</span> <span class='fl'>75</span>, <span class='kw'>vjust</span> <span class='kw'>=</span> <span class='fl'>1</span>))
)</div><div class='output co'>#&gt; <span class='message'>Using Names as id variables</span></div><div class='img'><img src='vis_hist-1.png' alt='' width='700' height='433' /></div><div class='input'><span class='no'>imm_gu</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='geneUsage.html'>geneUsage</a></span>(<span class='no'>immdata</span>$<span class='no'>data</span>[<span class='fl'>1</span>:<span class='fl'>4</span>])
<span class='fu'><a href='vis.html'>vis</a></span>(<span class='no'>imm_gu</span>,
  <span class='kw'>.plot</span> <span class='kw'>=</span> <span class='st'>"hist"</span>, <span class='kw'>.grid</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>.add.layer</span> <span class='kw'>=</span>
    <span class='fu'>theme</span>(<span class='kw'>axis.text.x</span> <span class='kw'>=</span> <span class='fu'>element_text</span>(<span class='kw'>angle</span> <span class='kw'>=</span> <span class='fl'>75</span>, <span class='kw'>vjust</span> <span class='kw'>=</span> <span class='fl'>1</span>))
)</div><div class='output co'>#&gt; <span class='message'>Using Names as id variables</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Removed 2 rows containing missing values (position_stack).</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Removed 2 rows containing missing values (position_stack).</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Removed 1 rows containing missing values (position_stack).</span></div><div class='img'><img src='vis_hist-2.png' alt='' width='700' height='433' /></div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Vadim I. Nazarov, Vasily O. Tsvetkov, Eugene Rumynskiy, Anna Lorenc, Daniel J. Moore, Victor Greiff, <a href='https://immunomind.io'>ImmunoMind</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script>
<script>
  docsearch({
    
    
    apiKey: '94207562f34b455c0796790d104e3549',
    indexName: 'immunarch',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>



  </body>
</html>


