<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Annotate clonotypes using immune receptor databases • immunarch</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><!-- docsearch --><script src="../../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Annotate clonotypes using immune receptor databases">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-128418614-2"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-128418614-2');
</script>
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">immunarch</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.5.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../../articles/v1_introduction.html">Get started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">First steps</li>
    <li>
      <a href="../../articles/v1_introduction.html">Introduction to immunarch</a>
    </li>
    <li>
      <a href="../../articles/v2_data.html">Working with data in immunarch</a>
    </li>
    <li>
      <a href="../../articles/v3_basic_analysis.html">Basic analysis and clonality</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Explore and compare repertoires</li>
    <li>
      <a href="../../articles/web_only/v4_overlap.html">Repertoire overlap and public clonotypes</a>
    </li>
    <li>
      <a href="../../articles/web_only/v5_gene_usage.html">Gene usage analysis</a>
    </li>
    <li>
      <a href="../../articles/web_only/v6_diversity.html">Diversity estimation</a>
    </li>
    <li>
      <a href="../../articles/web_only/v11_db.html">Annotate clonotypes using immune receptor databases</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Preparing to publication</li>
    <li>
      <a href="../../articles/web_only/v7_fixvis.html">Make your plots publication-ready with fixVis</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Advanced methods</li>
    <li>
      <a href="../../articles/web_only/v8_tracking.html">Tracking clonotypes across time points</a>
    </li>
    <li>
      <a href="../../articles/web_only/v9_kmers.html">Kmer and sequence motif analysis and visualisation</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
<li>
  <a href="https://github.com/immunomind/immunarch/releases">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://immunomind.io">ImmunoMind</a>
</li>
<li>
  <a href="http://twitter.com/immunomind">
    <span class="fa fa-lg fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/immunomind/immunarch">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Annotate clonotypes using immune receptor databases</h1>
                        <h4 class="author"><a href="https://immunomind.io">ImmunoMind</a></h4>
            
            <h4 class="date"><a href="mailto:support@immunomind.io" class="email">support@immunomind.io</a></h4>
      
      <small class="dont-index">Source: <a href="https://github.com/immunomind/immunarch/blob/master/vignettes/web_only/v11_db.Rmd"><code>vignettes/web_only/v11_db.Rmd</code></a></small>
      <div class="hidden name"><code>v11_db.Rmd</code></div>

    </div>

    
    
<!--
%\VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{Annotate clonotypes using immune receptor databases}
%\VignettePackage{immunarch}
-->
<div id="introduction-to-immune-receptor-databases" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction-to-immune-receptor-databases" class="anchor"></a>Introduction to immune receptor databases</h1>
<p>Databases with aggregated information about immune receptor specificity provide a straightforward way to annotate your data and find condition-associated receptors. <code>immunarch</code> supports tools to annotate your data using the most popular AIRR databases - VDJDB, McPAS-TCR and PIRD TBAdb.</p>
<p>Database annotation is a two-step process. First, you need to download database files - either full database files or filtered data obtained from the web interface of databases. After that, you can use <code>immunarch</code> functions to annotate your data and visualise the results. Below you can find a guide to annotation covering both steps.</p>
</div>
<div id="downloading-databases" class="section level1">
<h1 class="hasAnchor">
<a href="#downloading-databases" class="anchor"></a>Downloading databases</h1>
<div id="vdjdb" class="section level2">
<h2 class="hasAnchor">
<a href="#vdjdb" class="anchor"></a>VDJDB</h2>
<p>VDJDB is a curated database of T-cell receptor sequences of known antigen specificity. The database is GitHub-based and available here: <a href="https://github.com/antigenomics/vdjdb-db" class="uri">https://github.com/antigenomics/vdjdb-db</a></p>
<p>Citation: <em>Shugay M et al. VDJdb: a curated database of T-cell receptor sequences with known antigen specificity. Nucleic Acids Research 2017</em></p>
<div id="how-to-filter-and-download-data" class="section level3">
<h3 class="hasAnchor">
<a href="#how-to-filter-and-download-data" class="anchor"></a>How to filter and download data</h3>
<p>It can be useful to filter out immune receptors that are not relevant from the database before working with it. For instance, if you analyse human T-cell beta repertoires, it is not necessary to keep immune receptors from other species, as well as non-TRB data. Use the web interface to VDJDB located at <a href="https://vdjdb.cdr3.net/search" class="uri">https://vdjdb.cdr3.net/search</a> to filter out data. Having filtered the data and pressed the “Refresh table” button, locate the “Export” button and select the “TSV” label inside. You will download the filtered database file with a name like “SearchTable-2019-10-17 12_36_11.989.tsv”, which can be used for annotation with <code>immunarch</code>.</p>
</div>
<div id="how-to-download-full-vdjdb" class="section level3">
<h3 class="hasAnchor">
<a href="#how-to-download-full-vdjdb" class="anchor"></a>How to download full VDJDB</h3>
<p>You can use the previous method to download the full database if you set all checkmarks in the “General” section of the “CDR3” tab. However, if you want to download the raw database files, here is the step-by-step guide to the sophisticated process of VDJDB downloading and unpacking.</p>
<ol style="list-style-type: decimal">
<li><p>First, you need to install JDK 8 - Java Development Kit. If you already have it, skip this step. If you don’t, just search for the proper installation instructions for your system.</p></li>
<li><p>Second, you need to install Groovy - a language that is used for processing VDJDB. Go to <a href="https://groovy.apache.org/download.html#distro">this link</a> and download the distribution or windows installer depending on your system. For Windows users the best way is to download Windows installer. For Linux users the easiest way is to use OS package manager such as apt, dpkg, pacman, etc. For Mac users the most seamless way is to use <a href="https://brew.sh">Homebrew</a>.</p></li>
<li><p>Download the VDJDB repository from GitHub via this link: <a href="https://github.com/antigenomics/vdjdb-db/archive/master.zip" class="uri">https://github.com/antigenomics/vdjdb-db/archive/master.zip</a></p></li>
<li><p>Unzip the archive and go to the unpacked “vdjdb-db-master” folder.</p></li>
<li><p>Go to the “src” folder.</p></li>
<li><p>Open your Terminal or Console and execute the following command: <code>groovy -cp . BuildDatabase.groovy --no2fix</code>.</p></li>
<li><p>After some processing, the database files will be available at the “database” folder inside the “vdjdb-db-master” folder. You will need to provide paths to this files for the <code>immunarch</code> annotation functions.</p></li>
</ol>
</div>
</div>
<div id="mcpas-tcr" class="section level2">
<h2 class="hasAnchor">
<a href="#mcpas-tcr" class="anchor"></a>McPAS-TCR</h2>
<p>McPAS-TCR is a manually curated catalogue of pathology associated T-cell receptor sequences. The database is available at <a href="http://friedmanlab.weizmann.ac.il/McPAS-TCR/" class="uri">http://friedmanlab.weizmann.ac.il/McPAS-TCR/</a></p>
<p>Citation: <em>Tickotsky N, Sagiv T, Prilusky J, Shifrut E, Friedman N (2017). McPAS-TCR: A manually-curated catalogue of pathology-associated T cell receptor sequences. Bioinformatics 33:2924-2929</em></p>
<div id="how-to-filter-and-download-data-1" class="section level3">
<h3 class="hasAnchor">
<a href="#how-to-filter-and-download-data-1" class="anchor"></a>How to filter and download data</h3>
<p>The filtering feature of the database’s web interface is located in the “Search Database” tab. After processing the data, press the “Download csv” button. The downloaded file named “McPAS-TCR_search.csv” can be used for annotation with <code>immunarch</code>.</p>
</div>
<div id="how-to-download-full-mcpas-tcr" class="section level3">
<h3 class="hasAnchor">
<a href="#how-to-download-full-mcpas-tcr" class="anchor"></a>How to download full McPAS-TCR</h3>
<p>To download McPAS-TCR you just need to go to <a href="http://friedmanlab.weizmann.ac.il/McPAS-TCR/" class="uri">http://friedmanlab.weizmann.ac.il/McPAS-TCR/</a> and press the “Download the complete database” button there. Note that sometimes you need to press it twice or press it in a new browser tab to start the downloading process.</p>
</div>
</div>
<div id="tbadb-from-pird" class="section level2">
<h2 class="hasAnchor">
<a href="#tbadb-from-pird" class="anchor"></a>TBAdb from PIRD</h2>
<p>TBAdb is a manually curated database of T-cell receptor (TCR) and B-cell receptor (BCR) targeting specific antigen or diseases. The database contains three parts, namely TCR-AB, TCR-GD and BCR. These three parts are aimed at collecting sequences and specificity information of TCRA and TCRB, TCR- gamma and TCR-delta and BCR separately. The database is available at <a href="https://db.cngb.org/pird/tbadb/" class="uri">https://db.cngb.org/pird/tbadb/</a></p>
<p>Citation: <em>ZHANG W, Wang L, Liu K, Wei X, Yang K, Du W, Wang S, Guo N, Ma C, Luo L, et al. PIRD: Pan immune repertoire database. Bioinformatics(2019)</em></p>
<div id="how-to-filter-and-download-data-2" class="section level3">
<h3 class="hasAnchor">
<a href="#how-to-filter-and-download-data-2" class="anchor"></a>How to filter and download data</h3>
<p>Currently there is no way to download the filtered data from TBAdb. The query functionality is available at <a href="https://db.cngb.org/pird/query/" class="uri">https://db.cngb.org/pird/query/</a> in the “TBAdb” tab.</p>
</div>
<div id="how-to-download-full-tbadb" class="section level3">
<h3 class="hasAnchor">
<a href="#how-to-download-full-tbadb" class="anchor"></a>How to download full TBAdb</h3>
<p>To download TBAdb you need to go to <a href="https://db.cngb.org/pird/tbadb/" class="uri">https://db.cngb.org/pird/tbadb/</a> and press the “Download TBAdb” button. Note that you should agree with the licensing agreement in order to download the database file.</p>
</div>
</div>
</div>
<div id="annotation-of-clonotypes" class="section level1">
<h1 class="hasAnchor">
<a href="#annotation-of-clonotypes" class="anchor"></a>Annotation of clonotypes</h1>
<p>After downloading the database, we can proceed to the annotation part with R. To demonstrate the applicability of R and <code>immunarch</code>, we will use a common task of annotation of repertoires with Cytomegalovirus (CMV) infection.</p>
<div id="preprocessing-databases-with-r" class="section level2">
<h2 class="hasAnchor">
<a href="#preprocessing-databases-with-r" class="anchor"></a>Preprocessing databases with R</h2>
<p>For the start, we need to load databases into R and filter out non-human, non-TRB and non-CMV data from the input database. With databases, we follow the same philosophy as with <code>repLoad</code> and <code>vis</code> functions: the function <code>dbLoad</code> provides a singular interface to loading and basic quering for all supported databases.</p>
<p>For demonstration purposes, we will process each of the supported databases below.</p>
<div id="vdjdb-1" class="section level3">
<h3 class="hasAnchor">
<a href="#vdjdb-1" class="anchor"></a>VDJDB</h3>
<p>Download the VDJDB database following the instructions above. In the examples below, we use URLs to snippets of databases as file paths. In your own code you need to provide paths to your local database files, e.g., “/Users/yourname/Downloads/vdjdb-db-master/vdjdb.slim.txt”. Do not use the links below since they are only for testing purposes and not the actual databases!</p>
<p>Note that VDJDB data obtained from the web interface differs from VDJDB obtained from raw files. Check the next section for working with VDJDB search tables.</p>
<p>The most basic way to load VDJDB to R:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1">vdjdb =<span class="st"> </span><span class="kw"><a href="../../reference/dbLoad.html">dbLoad</a></span>(<span class="st">"https://gitlab.com/immunomind/immunarch/raw/dev-0.5.0/private/vdjdb.slim.txt.gz"</span>, <span class="st">"vdjdb"</span>)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2">vdjdb</a></code></pre></div>
<pre><code>## # A tibble: 61,049 x 19
##    gene  cdr3  species antigen.epitope antigen.gene antigen.species complex.id
##    &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt;           &lt;chr&gt;        &lt;chr&gt;                &lt;dbl&gt;
##  1 TRB   CASS… Macaca… STPESANL        Tat          SIV                0.     
##  2 TRB   CASS… HomoSa… RLRAEAQVK       EBNA3A       EBV                1.93e 4
##  3 TRB   CASS… Macaca… TTPESANL        Tat          SIV                0.     
##  4 TRA   CASN… HomoSa… GILGFVFTL       M            InfluenzaA         0.     
##  5 TRB   CASS… MusMus… HGIRNASFI       M45          MCMV               2.24e24
##  6 TRB   CSAS… HomoSa… KLGGALQAK       IE1          CMV                8.58e 3
##  7 TRA   CAVL… HomoSa… GILGFVFTL       M            InfluenzaA         0.     
##  8 TRB   CASS… HomoSa… KLGGALQAK       IE1          CMV                3.44e 3
##  9 TRB   CAST… MusMus… SSYRRPVGI       PB1          InfluenzaA         2.28e 4
## 10 TRB   CASS… HomoSa… NLVPMVATV       pp65         CMV                0.     
## # … with 61,039 more rows, and 12 more variables: v.segm &lt;chr&gt;, j.segm &lt;chr&gt;,
## #   v.end &lt;dbl&gt;, j.start &lt;dbl&gt;, mhc.a &lt;chr&gt;, mhc.b &lt;chr&gt;, mhc.class &lt;chr&gt;,
## #   reference.id &lt;chr&gt;, vdjdb.score &lt;dbl&gt;, Species &lt;chr&gt;, Chain &lt;chr&gt;,
## #   Pathology &lt;chr&gt;</code></pre>
<p>To load VDJDB and filter out information you need to provide <code>.species</code>, <code>.chain</code> and <code>.pathology</code> arguments:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1">vdjdb =<span class="st"> </span><span class="kw"><a href="../../reference/dbLoad.html">dbLoad</a></span>(<span class="st">"https://gitlab.com/immunomind/immunarch/raw/dev-0.5.0/private/vdjdb.slim.txt.gz"</span>, <span class="st">"vdjdb"</span>, <span class="dt">.species =</span> <span class="st">"HomoSapiens"</span>, <span class="dt">.chain =</span> <span class="st">"TRB"</span>, <span class="dt">.pathology =</span> <span class="st">"CMV"</span>)</a>
<a class="sourceLine" id="cb3-2" data-line-number="2">vdjdb</a></code></pre></div>
<pre><code>## # A tibble: 18,039 x 19
##    gene  cdr3  species antigen.epitope antigen.gene antigen.species complex.id
##    &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt;           &lt;chr&gt;        &lt;chr&gt;                &lt;dbl&gt;
##  1 TRB   CSAS… HomoSa… KLGGALQAK       IE1          CMV                   8584
##  2 TRB   CASS… HomoSa… KLGGALQAK       IE1          CMV                   3445
##  3 TRB   CASS… HomoSa… NLVPMVATV       pp65         CMV                      0
##  4 TRB   CASS… HomoSa… KLGGALQAK       IE1          CMV                  19396
##  5 TRB   CASS… HomoSa… NLVPMVATV       pp65         CMV                      0
##  6 TRB   CAST… HomoSa… KLGGALQAK       IE1          CMV                  10972
##  7 TRB   CASS… HomoSa… KLGGALQAK       IE1          CMV                   6231
##  8 TRB   CASS… HomoSa… KLGGALQAK       IE1          CMV                  12587
##  9 TRB   CATS… HomoSa… KLGGALQAK       IE1          CMV                  13267
## 10 TRB   CASS… HomoSa… NLVPMVATV       pp65         CMV                      0
## # … with 18,029 more rows, and 12 more variables: v.segm &lt;chr&gt;, j.segm &lt;chr&gt;,
## #   v.end &lt;dbl&gt;, j.start &lt;dbl&gt;, mhc.a &lt;chr&gt;, mhc.b &lt;chr&gt;, mhc.class &lt;chr&gt;,
## #   reference.id &lt;chr&gt;, vdjdb.score &lt;dbl&gt;, Species &lt;chr&gt;, Chain &lt;chr&gt;,
## #   Pathology &lt;chr&gt;</code></pre>
</div>
<div id="vdjdb-search-tables" class="section level3">
<h3 class="hasAnchor">
<a href="#vdjdb-search-tables" class="anchor"></a>VDJDB search tables</h3>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1">vdjdb_st =<span class="st"> </span><span class="kw"><a href="../../reference/dbLoad.html">dbLoad</a></span>(<span class="st">"https://gitlab.com/immunomind/immunarch/raw/dev-0.5.0/private/SearchTable-2019-10-17%2012_36_11.989.tsv.gz"</span>, <span class="st">"vdjdb-search"</span>, <span class="dt">.species =</span> <span class="st">"HomoSapiens"</span>, <span class="dt">.chain =</span> <span class="st">"TRB"</span>, <span class="dt">.pathology =</span> <span class="st">"CMV"</span>)</a>
<a class="sourceLine" id="cb5-2" data-line-number="2">vdjdb_st</a></code></pre></div>
<pre><code>## # A tibble: 4,999 x 19
##    complex.id Gene  CDR3  V     J     Species `MHC A` `MHC B` `MHC class`
##         &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;      
##  1          0 TRB   CASS… TRBV… TRBJ… HomoSa… HLA-A*… B2M     MHCI       
##  2          0 TRB   CAWS… TRBV… TRBJ… HomoSa… HLA-A*… B2M     MHCI       
##  3          0 TRB   CASS… TRBV… TRBJ… HomoSa… HLA-A*… B2M     MHCI       
##  4          0 TRB   CASS… TRBV… TRBJ… HomoSa… HLA-B*… B2M     MHCI       
##  5          0 TRB   CASS… TRBV… TRBJ… HomoSa… HLA-B*… B2M     MHCI       
##  6          0 TRB   CASS… TRBV… TRBJ… HomoSa… HLA-B*… B2M     MHCI       
##  7          0 TRB   CASV… TRBV… TRBJ… HomoSa… HLA-B*… B2M     MHCI       
##  8          0 TRB   CASS… TRBV… TRBJ… HomoSa… HLA-B*… B2M     MHCI       
##  9          0 TRB   CASS… TRBV… TRBJ… HomoSa… HLA-B*… B2M     MHCI       
## 10          0 TRB   CASG… TRBV… TRBJ… HomoSa… HLA-B*… B2M     MHCI       
## # … with 4,989 more rows, and 10 more variables: Epitope &lt;chr&gt;, `Epitope
## #   gene` &lt;chr&gt;, `Epitope species` &lt;chr&gt;, Reference &lt;chr&gt;, Method &lt;chr&gt;,
## #   Meta &lt;chr&gt;, CDR3fix &lt;chr&gt;, Score &lt;dbl&gt;, Chain &lt;chr&gt;, Pathology &lt;chr&gt;</code></pre>
</div>
<div id="mcpas-tcr-1" class="section level3">
<h3 class="hasAnchor">
<a href="#mcpas-tcr-1" class="anchor"></a>McPAS-TCR</h3>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">mcpas =<span class="st"> </span><span class="kw"><a href="../../reference/dbLoad.html">dbLoad</a></span>(<span class="st">"https://gitlab.com/immunomind/immunarch/raw/dev-0.5.0/private/McPAS-TCR.csv.gz"</span>, <span class="st">"mcpas"</span>, <span class="dt">.species =</span> <span class="st">"Human"</span>, <span class="dt">.chain =</span> <span class="st">"TRB"</span>, <span class="dt">.pathology =</span> <span class="st">"Cytomegalovirus (CMV)"</span>)</a>
<a class="sourceLine" id="cb7-2" data-line-number="2">mcpas</a></code></pre></div>
<pre><code>## # A tibble: 2,723 x 29
##    CDR3.alpha.aa CDR3.beta.aa Species Category Pathology Pathology.Mesh.…
##    &lt;chr&gt;         &lt;chr&gt;        &lt;chr&gt;   &lt;chr&gt;    &lt;chr&gt;     &lt;chr&gt;           
##  1 &lt;NA&gt;          CASLAPGTTNE… Human   Pathoge… Cytomega… D003586         
##  2 &lt;NA&gt;          CASLQAGANEQF Human   Pathoge… Cytomega… D003586         
##  3 &lt;NA&gt;          CASLSGGGEQF  Human   Pathoge… Cytomega… D003586         
##  4 &lt;NA&gt;          CASLVASGQET… Human   Pathoge… Cytomega… D003586         
##  5 &lt;NA&gt;          CASSHRDSGNT… Human   Pathoge… Cytomega… D003586         
##  6 &lt;NA&gt;          CASSSANYGYTF Human   Pathoge… Cytomega… D003586         
##  7 &lt;NA&gt;          CATSDPLTASY… Human   Pathoge… Cytomega… D003586         
##  8 CARNTGNQF     CACSLRSQGTD… Human   Pathoge… Cytomega… D003586         
##  9 CAGNTGNQFYFG  CASSAWDRSSG… Human   Pathoge… Cytomega… D003586         
## 10 CAYPYNNNDMRF  CASSELGGAGT… Human   Pathoge… Cytomega… D003586         
## # … with 2,713 more rows, and 23 more variables:
## #   Additional.study.details &lt;chr&gt;, Antigen.identification.method &lt;dbl&gt;,
## #   NGS &lt;chr&gt;, Antigen.protein &lt;chr&gt;, Protein.ID &lt;chr&gt;, Epitope.peptide &lt;chr&gt;,
## #   Epitope.ID &lt;dbl&gt;, MHC &lt;chr&gt;, Tissue &lt;chr&gt;, T.Cell.Type &lt;chr&gt;,
## #   T.cell.characteristics &lt;chr&gt;, CDR3.alpha.nt &lt;chr&gt;, TRAV &lt;chr&gt;, TRAJ &lt;chr&gt;,
## #   TRBV &lt;chr&gt;, TRBD &lt;chr&gt;, TRBJ &lt;chr&gt;, Reconstructed.J.annotation &lt;chr&gt;,
## #   CDR3.beta.nt &lt;chr&gt;, Mouse.strain &lt;chr&gt;, PubMed.ID &lt;dbl&gt;, Remarks &lt;chr&gt;,
## #   Chain &lt;chr&gt;</code></pre>
</div>
<div id="tbadb" class="section level3">
<h3 class="hasAnchor">
<a href="#tbadb" class="anchor"></a>TBAdb</h3>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1">tbadb =<span class="st"> </span><span class="kw"><a href="../../reference/dbLoad.html">dbLoad</a></span>(<span class="st">"https://gitlab.com/immunomind/immunarch/raw/dev-0.5.0/private/TBAdb.xlsx"</span>, <span class="st">"tbadb"</span>, <span class="dt">.species =</span> <span class="st">"Homo Sapiens"</span>, <span class="dt">.chain =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"TRB"</span>, <span class="st">"TRA-TRB"</span>), <span class="dt">.pathology =</span> <span class="st">"CMV"</span>)</a>
<a class="sourceLine" id="cb9-2" data-line-number="2">tbadb</a></code></pre></div>
</div>
</div>
<div id="repertoire-annotation" class="section level2">
<h2 class="hasAnchor">
<a href="#repertoire-annotation" class="anchor"></a>Repertoire annotation</h2>
<p>The key <code>immunarch</code> function for annotation is <code>dbAnnotate</code>. As an input it requires repertoires to search in, a database to lookup from, and columns of interest such as CDR3 amino acid sequence or V gene segment names columns. If you want to try it on the test data packaged with <code>immunarch</code>, execute the following line of code before proceeding further:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(immdata)</a></code></pre></div>
<p>Just in a single line of code you are able to find all clonotypes with matching CDR3 amino acid sequences in the input data and VDJDB database:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="kw"><a href="../../reference/dbAnnotate.html">dbAnnotate</a></span>(immdata<span class="op">$</span>data, vdjdb, <span class="st">"CDR3.aa"</span>, <span class="st">"cdr3"</span>)</a></code></pre></div>
<pre><code>##            CDR3.aa Samples A2-i129 A2-i131 A2-i133 A2-i132 A4-i191 A4-i192 MS1
##   1:   CASSLGETQYF      11       6       4       2       9       6       0   1
##   2:    CASSQETQYF      10       7       3       1       2       3       3   1
##   3:   CASSFQETQYF       9       3       2       2       4       2       0   2
##   4:  CASSLEGYEQYF       9       0       1       2       1       3       0   1
##   5:   CASSSSYEQYF       9       1       0       0       1       2       2   1
##  ---                                                                          
## 635:  CSVGTGTYEQYF       1       0       0       0       0       1       0   0
## 636: CSVQGGAYNEQFF       1       0       1       0       0       0       0   0
## 637: CSVQGGSYNEQFF       1       0       1       0       0       0       0   0
## 638:  CSVVATNEKLFF       1       0       0       1       0       0       0   0
## 639: CSVVGTGNTEAFF       1       0       0       0       0       0       0   0
##      MS2 MS3 MS4 MS5 MS6
##   1:   3   1   2   5   3
##   2:   1   0   0   5   1
##   3:   1   0   4   0   2
##   4:   0   1   1   1   1
##   5:   1   0   1   2   3
##  ---                    
## 635:   0   0   0   0   0
## 636:   0   0   0   0   0
## 637:   0   0   0   0   0
## 638:   0   0   0   0   0
## 639:   0   0   0   1   0</code></pre>
<p>The “Samples” column specifies the number of samples in which the clonotype found. Other numbers in columns correspond to the clonal count of the clonotype in a specific input sample.</p>
<p>In the next example we will search the McPAS-TCR database for condition-associated sequences using both CDR3 amino acid sequences and V gene segments:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1"><span class="kw"><a href="../../reference/dbAnnotate.html">dbAnnotate</a></span>(immdata<span class="op">$</span>data, mcpas, <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"CDR3.aa"</span>, <span class="st">"V.name"</span>), <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"CDR3.beta.aa"</span>, <span class="st">"TRBV"</span>))</a></code></pre></div>
<pre><code>##                CDR3.aa   V.name Samples A2-i129 A2-i131 A2-i133 A2-i132 A4-i191
##    1: CASSLAPGATNEKLFF  TRBV7-6       4       0       0       0       0       0
##    2:     CASSLGENIQYF   TRBV13       4       0       0       0       0       1
##    3:      CAISESYEQYF TRBV10-3       3       0       1       0       0       0
##    4:     CASSLGRETQYF   TRBV28       3       0       0       0       0       0
##    5:   CSVGTGGTNEKLFF TRBV29-1       3       0       0       0       0       0
##   ---                                                                          
## 2123:           KNPTAF   TRBV19       1       0       0       0       0       0
## 2124:       LLGGQETQYF  TRBV7-4       1       0       0       0       0       0
## 2125:     WASSFQGFTEAF   TRBV28       1       0       0       0       0       0
## 2126:    WASSQALPYEQYF TRBV12-4       1       0       0       0       0       0
## 2127:  WASSQQTGTIGGYTF  TRBV6-5       1       0       0       0       0       0
##       A4-i192 MS1 MS2 MS3 MS4 MS5 MS6
##    1:       1   0   1   0   0   1   0
##    2:       1   0   0   0   0   1   0
##    3:       1   0   0   0   0   0   0
##    4:       0   0   0   0   0  89   1
##    5:       0   0   1   0   0   0   1
##   ---                                
## 2123:       0   0   0   0   0   0   0
## 2124:       0   0   0   0   0   0   0
## 2125:       0   0   0   0   0   0   0
## 2126:       0   0   0   0   0   0   0
## 2127:       0   0   0   0   0   0   0</code></pre>
<p>If you seek to search a database for a specific set of sequences, create a data frame with them and use it as a database file:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1">local_db =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="dt">Seq =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"CASSDSSGGANEQFF"</span>, <span class="st">"CSARLAGGQETQYF"</span>), <span class="dt">V =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"TRBV6-4"</span>, <span class="st">"TRBV20-1"</span>), <span class="dt">stringsAsFactors =</span> F)</a>
<a class="sourceLine" id="cb15-2" data-line-number="2"></a>
<a class="sourceLine" id="cb15-3" data-line-number="3"><span class="kw"><a href="../../reference/dbAnnotate.html">dbAnnotate</a></span>(immdata<span class="op">$</span>data, local_db, <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"CDR3.aa"</span>, <span class="st">"V.name"</span>), <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Seq"</span>, <span class="st">"V"</span>))</a></code></pre></div>
<pre><code>##            CDR3.aa   V.name Samples A2-i129 A2-i131 A2-i133 A2-i132 A4-i191
## 1: CASSDSSGGANEQFF  TRBV6-4       7       1       1       2       0       5
## 2:  CSARLAGGQETQYF TRBV20-1       7       1       3       0       2       1
##    A4-i192 MS1 MS2 MS3 MS4 MS5 MS6
## 1:       0   0   0   2   0   0  15
## 2:       0   0   0   2   0   0   1</code></pre>
<div id="visualisation" class="section level3">
<h3 class="hasAnchor">
<a href="#visualisation" class="anchor"></a>Visualisation</h3>
<p>Visualisation with the <code><a href="../../reference/vis.html">vis()</a></code> function will be supported in the next major release of <code>immunarch</code>. You can use <code>ggplot2</code> to visualise distributions of found clonotypes.</p>
</div>
</div>
<div id="advanced-filtering" class="section level2">
<h2 class="hasAnchor">
<a href="#advanced-filtering" class="anchor"></a>Advanced filtering</h2>
<p><code>immunarch</code> provides a very basic query interface that permits filtering by species types, chain types and pathology types only. To perform advanced filtering such as filtering by antigen epitope, you need to use R. In the most cases, filtering with the <code>dplyr</code> package is the most seamless way. Here is an example on how to use <code>dplyr</code> to filter out specific antigen epitopes from VDJDB:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1"><span class="co"># Load the dplyr library</span></a>
<a class="sourceLine" id="cb17-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(dplyr)</a>
<a class="sourceLine" id="cb17-3" data-line-number="3"></a>
<a class="sourceLine" id="cb17-4" data-line-number="4"><span class="co"># Load the database with immunarch</span></a>
<a class="sourceLine" id="cb17-5" data-line-number="5">vdjdb =<span class="st"> </span><span class="kw"><a href="../../reference/dbLoad.html">dbLoad</a></span>(<span class="st">"https://gitlab.com/immunomind/immunarch/raw/dev-0.5.0/private/vdjdb.slim.txt.gz"</span>, <span class="st">"vdjdb"</span>, <span class="dt">.species =</span> <span class="st">"HomoSapiens"</span>, <span class="dt">.chain =</span> <span class="st">"TRB"</span>, <span class="dt">.pathology =</span> <span class="st">"CMV"</span>)</a>
<a class="sourceLine" id="cb17-6" data-line-number="6"></a>
<a class="sourceLine" id="cb17-7" data-line-number="7"><span class="co"># Check which antigen epitopes are presented in the database</span></a>
<a class="sourceLine" id="cb17-8" data-line-number="8"><span class="kw"><a href="https://rdrr.io/r/base/table.html">table</a></span>(vdjdb<span class="op">$</span>antigen.epitope)</a></code></pre></div>
<pre><code>## 
## ARNLVPMVATVQGQN       AYAQKIFKI    CPSQEPMSIYVY       CVETMCNEY       DEEDAIAAY 
##               3              39               2               2               2 
## EDVPSGKLFMHVTLG      EFFWDANDIY       ELKRKMIYM       ELRRKMMYM        FPTKDVAL 
##               1               1               5              10              10 
##       IPSINVHHY       KLGGALQAK LSEFCRVLCCYVLEE       MLNIPSINV        NEGVKAAW 
##              93           12667               2              73              49 
##       NLVPMVATV       QIKVRVDMV       QIKVRVKMV       QYDPVAALF      RPHERNGFTV 
##            4496              15              24              39               4 
##     RPHERNGFTVL      TPRVTGGGAM       VLEETSVML       VMAPRTLIL       VTEHDTLLY 
##              22             207              14               1             202 
##       YILEETSVM     YSEHPTFTSQY 
##               3              53</code></pre>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1"><span class="co"># Filter out all non NLVPMVATV epitopes</span></a>
<a class="sourceLine" id="cb19-2" data-line-number="2">vdjdb =<span class="st"> </span>vdjdb <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(antigen.epitope <span class="op">==</span><span class="st"> "NLVPMVATV"</span>)</a>
<a class="sourceLine" id="cb19-3" data-line-number="3">vdjdb</a></code></pre></div>
<pre><code>## # A tibble: 4,496 x 19
##    gene  cdr3  species antigen.epitope antigen.gene antigen.species complex.id
##    &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt;           &lt;chr&gt;        &lt;chr&gt;                &lt;dbl&gt;
##  1 TRB   CASS… HomoSa… NLVPMVATV       pp65         CMV                      0
##  2 TRB   CASS… HomoSa… NLVPMVATV       pp65         CMV                      0
##  3 TRB   CASS… HomoSa… NLVPMVATV       pp65         CMV                      0
##  4 TRB   CASS… HomoSa… NLVPMVATV       pp65         CMV                      0
##  5 TRB   CSAD… HomoSa… NLVPMVATV       pp65         CMV                      0
##  6 TRB   CASS… HomoSa… NLVPMVATV       pp65         CMV                      0
##  7 TRB   CASS… HomoSa… NLVPMVATV       pp65         CMV                      0
##  8 TRB   CASS… HomoSa… NLVPMVATV       pp65         CMV                      0
##  9 TRB   CSVE… HomoSa… NLVPMVATV       pp65         CMV                      0
## 10 TRB   CASS… HomoSa… NLVPMVATV       pp65         CMV                      0
## # … with 4,486 more rows, and 12 more variables: v.segm &lt;chr&gt;, j.segm &lt;chr&gt;,
## #   v.end &lt;dbl&gt;, j.start &lt;dbl&gt;, mhc.a &lt;chr&gt;, mhc.b &lt;chr&gt;, mhc.class &lt;chr&gt;,
## #   reference.id &lt;chr&gt;, vdjdb.score &lt;dbl&gt;, Species &lt;chr&gt;, Chain &lt;chr&gt;,
## #   Pathology &lt;chr&gt;</code></pre>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1"><span class="co"># Check if everything is OK and there is no other epitopes</span></a>
<a class="sourceLine" id="cb21-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/base/table.html">table</a></span>(vdjdb<span class="op">$</span>antigen.epitope)</a></code></pre></div>
<pre><code>## 
## NLVPMVATV 
##      4496</code></pre>
</div>
</div>
<div id="get-in-contact-with-us" class="section level1">
<h1 class="hasAnchor">
<a href="#get-in-contact-with-us" class="anchor"></a>Get in contact with us</h1>
<p>Can not find an important feature? Have a question or found a bug? Contact us at <a href="mailto:support@immunomind.io" class="email">support@immunomind.io</a></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction-to-immune-receptor-databases">Introduction to immune receptor databases</a></li>
      <li>
<a href="#downloading-databases">Downloading databases</a><ul class="nav nav-pills nav-stacked">
<li><a href="#vdjdb">VDJDB</a></li>
      <li><a href="#mcpas-tcr">McPAS-TCR</a></li>
      <li><a href="#tbadb-from-pird">TBAdb from PIRD</a></li>
      </ul>
</li>
      <li>
<a href="#annotation-of-clonotypes">Annotation of clonotypes</a><ul class="nav nav-pills nav-stacked">
<li><a href="#preprocessing-databases-with-r">Preprocessing databases with R</a></li>
      <li><a href="#repertoire-annotation">Repertoire annotation</a></li>
      <li><a href="#advanced-filtering">Advanced filtering</a></li>
      </ul>
</li>
      <li><a href="#get-in-contact-with-us">Get in contact with us</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Vadim I. Nazarov, Vasily O. Tsvetkov, Eugene Rumynskiy, Anna Lorenc, Daniel J. Moore, Victor Greiff, <a href="https://immunomind.io">ImmunoMind</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '94207562f34b455c0796790d104e3549',
    indexName: 'immunarch',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
