<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Kmer and sequence motif analysis and visualisation • immunarch</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><meta property="og:title" content="Kmer and sequence motif analysis and visualisation">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-128418614-2"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-128418614-2');
</script>
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">immunarch</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.5.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../../articles/v1_introduction.html">Get started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">First steps</li>
    <li>
      <a href="../../articles/v1_introduction.html">Introduction to immunarch</a>
    </li>
    <li>
      <a href="../../articles/v2_data.html">Working with data in immunarch</a>
    </li>
    <li>
      <a href="../../articles/v3_basic_analysis.html">Basic analysis and clonality</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Explore and compare repertoires</li>
    <li>
      <a href="../../articles/web_only/v4_overlap.html">Repertoire overlap and public clonotypes</a>
    </li>
    <li>
      <a href="../../articles/web_only/v5_gene_usage.html">Gene usage analysis</a>
    </li>
    <li>
      <a href="../../articles/web_only/v6_diversity.html">Diversity estimation</a>
    </li>
    <li>
      <a href="../../articles/web_only/v11_db.html">Annotate clonotypes using immune receptor databases</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Preparing to publication</li>
    <li>
      <a href="../../articles/web_only/v7_fixvis.html">Make your plots publication-ready with fixVis</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Advanced methods</li>
    <li>
      <a href="../../articles/web_only/v8_tracking.html">Tracking clonotypes across time points</a>
    </li>
    <li>
      <a href="../../articles/web_only/v9_kmers.html">Kmer and sequence motif analysis and visualisation</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
<li>
  <a href="https://github.com/immunomind/immunarch/releases">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://immunomind.io">ImmunoMind</a>
</li>
<li>
  <a href="http://twitter.com/immunomind">
    <span class="fa fa-lg fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/immunomind/immunarch">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Kmer and sequence motif analysis and visualisation</h1>
                        <h4 class="author"><a href="https://immunomind.io">ImmunoMind</a></h4>
            
            <h4 class="date"><a href="mailto:support@immunomind.io" class="email">support@immunomind.io</a></h4>
      
      <small class="dont-index">Source: <a href="https://github.com/immunomind/immunarch/blob/master/vignettes/web_only/v9_kmers.Rmd"><code>vignettes/web_only/v9_kmers.Rmd</code></a></small>
      <div class="hidden name"><code>v9_kmers.Rmd</code></div>

    </div>

    
    
<!--
%\VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{Kmer and sequence motif analysis and visualisation}
%\VignettePackage{immunarch}
-->
<div id="kmer-statistics-computation" class="section level1">
<h1 class="hasAnchor">
<a href="#kmer-statistics-computation" class="anchor"></a>Kmer statistics computation</h1>
<p>Counting kmer occurrences in <code>immunarch</code> is very easy. All you need to do is to run the <code><a href="../../reference/getKmers.html">getKmers()</a></code> function on your data:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1">kmers =<span class="st"> </span><span class="kw"><a href="../../reference/getKmers.html">getKmers</a></span>(immdata<span class="op">$</span>data[[<span class="dv">1</span>]], <span class="dv">3</span>)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2">kmers</a></code></pre></div>
<pre><code>## # A tibble: 4,572 x 2
##    Kmer  Count
##    &lt;chr&gt; &lt;int&gt;
##  1 AAA       7
##  2 AAD       6
##  3 AAE       9
##  4 AAF       2
##  5 AAG      45
##  6 AAI       2
##  7 AAK       5
##  8 AAL       4
##  9 AAM       1
## 10 AAN      15
## # … with 4,562 more rows</code></pre>
<p>If is possible to compute occurrences of kmers in a batch of immune repertoires. In order to do that, you just need to provide a list of immune repertoires to the function. NA means that there is no such kmer found in a sample, specifyed by the column name.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1">kmers =<span class="st"> </span><span class="kw"><a href="../../reference/getKmers.html">getKmers</a></span>(immdata<span class="op">$</span>data, <span class="dv">5</span>)</a>
<a class="sourceLine" id="cb3-2" data-line-number="2">kmers</a></code></pre></div>
<pre><code>## # A tibble: 189,757 x 13
##    Kmer  `A2-i129` `A2-i131` `A2-i133` `A2-i132` `A4-i191` `A4-i192`   MS1   MS2
##    &lt;chr&gt;     &lt;int&gt;     &lt;int&gt;     &lt;int&gt;     &lt;int&gt;     &lt;int&gt;     &lt;int&gt; &lt;int&gt; &lt;int&gt;
##  1 AAAAG        NA        NA        NA        NA        NA        NA    NA     1
##  2 AAAAL        NA        NA        NA        NA        NA        NA    NA    NA
##  3 AAAAW        NA        NA        NA        NA        NA         1    NA    NA
##  4 AAADE        NA        NA        NA        NA        NA        NA    NA    NA
##  5 AAADN        NA         1        NA        NA        NA        NA    NA    NA
##  6 AAADT        NA        NA        NA        NA        NA        NA    NA    NA
##  7 AAAEA        NA        NA        NA        NA        NA        NA    NA    NA
##  8 AAAEN        NA        NA        NA         1        NA        NA    NA    NA
##  9 AAAET        NA        NA        NA        NA        NA        NA    NA    NA
## 10 AAAFE        NA        NA        NA         1        NA        NA    NA    NA
## # … with 189,747 more rows, and 4 more variables: MS3 &lt;int&gt;, MS4 &lt;int&gt;,
## #   MS5 &lt;int&gt;, MS6 &lt;int&gt;</code></pre>
<p>Note that by default <code><a href="../../reference/getKmers.html">getKmers()</a></code> filter out all non-coding sequences before counting the kmer statistics. You can use both coding and non-coding sequences by setting the <code>.coding</code> argument to FALSE:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1">kmers =<span class="st"> </span><span class="kw"><a href="../../reference/getKmers.html">getKmers</a></span>(immdata<span class="op">$</span>data[[<span class="dv">1</span>]], <span class="dv">3</span>, <span class="dt">.coding=</span>F)</a>
<a class="sourceLine" id="cb5-2" data-line-number="2">kmers</a></code></pre></div>
<pre><code>## # A tibble: 4,848 x 2
##    Kmer  Count
##    &lt;chr&gt; &lt;int&gt;
##  1 **G       1
##  2 *~G       4
##  3 *~L       1
##  4 *AA       1
##  5 *D~       1
##  6 *ED       1
##  7 *EE       1
##  8 *G~       1
##  9 *GG       1
## 10 *GP       1
## # … with 4,838 more rows</code></pre>
<div id="kmer-statistics-visualisation" class="section level2">
<h2 class="hasAnchor">
<a href="#kmer-statistics-visualisation" class="anchor"></a>Kmer statistics visualisation</h2>
<p>To visualise your kmer statistics, the <code><a href="../../reference/vis.html">vis()</a></code> function comes to help:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">kmers =<span class="st"> </span><span class="kw"><a href="../../reference/getKmers.html">getKmers</a></span>(immdata<span class="op">$</span>data, <span class="dv">5</span>)</a>
<a class="sourceLine" id="cb7-2" data-line-number="2"><span class="kw"><a href="../../reference/vis.html">vis</a></span>(kmers)</a></code></pre></div>
<p><img src="v9_kmers_files/figure-html/unnamed-chunk-4-1.png" width="1152" style="display: block; margin: auto;"></p>
<p>The <code><a href="../../reference/vis.html">vis()</a></code> function for kmers has a number of arguments to manipulate the plot. First, the <code>.head</code> argument specifies the number of the most abundant kmers to visualise.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1">p1 =<span class="st"> </span><span class="kw"><a href="../../reference/vis.html">vis</a></span>(kmers, <span class="dt">.head =</span> <span class="dv">5</span>)</a>
<a class="sourceLine" id="cb8-2" data-line-number="2">p2 =<span class="st"> </span><span class="kw"><a href="../../reference/vis.html">vis</a></span>(kmers, <span class="dt">.head =</span> <span class="dv">10</span>)</a>
<a class="sourceLine" id="cb8-3" data-line-number="3">p3 =<span class="st"> </span><span class="kw"><a href="../../reference/vis.html">vis</a></span>(kmers, <span class="dt">.head =</span> <span class="dv">30</span>)</a>
<a class="sourceLine" id="cb8-4" data-line-number="4"><span class="kw">grid.arrange</span>(p1, p2, <span class="dt">ncol=</span><span class="dv">2</span>)</a></code></pre></div>
<p><img src="v9_kmers_files/figure-html/unnamed-chunk-5-1.png" width="1152" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1">p3</a></code></pre></div>
<p><img src="v9_kmers_files/figure-html/unnamed-chunk-5-2.png" width="1152" style="display: block; margin: auto;"></p>
<p>Second, there are three options to choose from for positions of bars: “stack”, “dodge” and “fill”, adjusted by providing the correposnding option to the <code>.position</code> argument:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1">p1 =<span class="st"> </span><span class="kw"><a href="../../reference/vis.html">vis</a></span>(kmers, <span class="dt">.head =</span> <span class="dv">10</span>, <span class="dt">.position =</span> <span class="st">"stack"</span>)</a>
<a class="sourceLine" id="cb10-2" data-line-number="2">p2 =<span class="st"> </span><span class="kw"><a href="../../reference/vis.html">vis</a></span>(kmers, <span class="dt">.head =</span> <span class="dv">10</span>, <span class="dt">.position =</span> <span class="st">"fill"</span>)</a>
<a class="sourceLine" id="cb10-3" data-line-number="3">p3 =<span class="st"> </span><span class="kw"><a href="../../reference/vis.html">vis</a></span>(kmers, <span class="dt">.head =</span> <span class="dv">10</span>, <span class="dt">.position =</span> <span class="st">"dodge"</span>)</a>
<a class="sourceLine" id="cb10-4" data-line-number="4"><span class="kw">grid.arrange</span>(p1, p2, <span class="dt">ncol=</span><span class="dv">2</span>)</a></code></pre></div>
<p><img src="v9_kmers_files/figure-html/unnamed-chunk-6-1.png" width="1152" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1">p3</a></code></pre></div>
<p><img src="v9_kmers_files/figure-html/unnamed-chunk-6-2.png" width="1152" style="display: block; margin: auto;"></p>
<p>Option “stack” stacks all bars on top of each other so you can see the full distribution of kmers. Option “fill” stack all bars on top of each other as well, but normalises it in a such way so you see distribution of counts per-kmer, i.e., you can clearly see which repertoire has more kmer counts than others for a specific kmer. Option “dodge” groups kmer bars of different samples so you can clearly see, which samples has more kmer occurrences overall.</p>
<p>Additional argument is <code>.log</code> needed if your distribution of kmer counts is vastly imbalanced for some of repertoires. It permits to use the log-transformation of y-axis so you can see differences in orders of magnitude in kmer counts rather.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1">p1 =<span class="st"> </span><span class="kw"><a href="../../reference/vis.html">vis</a></span>(kmers, <span class="dt">.head =</span> <span class="dv">10</span>, <span class="dt">.position =</span> <span class="st">"stack"</span>)</a>
<a class="sourceLine" id="cb12-2" data-line-number="2">p2 =<span class="st"> </span><span class="kw"><a href="../../reference/vis.html">vis</a></span>(kmers, <span class="dt">.head =</span> <span class="dv">10</span>, <span class="dt">.position =</span> <span class="st">"stack"</span>, <span class="dt">.log =</span> T)</a>
<a class="sourceLine" id="cb12-3" data-line-number="3"></a>
<a class="sourceLine" id="cb12-4" data-line-number="4"><span class="kw">grid.arrange</span>(p1, p2, <span class="dt">ncol=</span><span class="dv">2</span>)</a></code></pre></div>
<p><img src="v9_kmers_files/figure-html/unnamed-chunk-7-1.png" width="1152" style="display: block; margin: auto;"></p>
</div>
</div>
<div id="sequence-motifs-analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#sequence-motifs-analysis" class="anchor"></a>Sequence motifs analysis</h1>
<p><code>immunarch</code> utilises common approaches to sequence motif analysis and uses different types of matrices to represent sequence motifs:</p>
<ul>
<li>position frequency matrix (PFM) - a matrix with occurences of each amino acid in each position;</li>
<li>position probability matrix (PPM) - a matrix with probabilities of occurences of each amino acid in each position;</li>
<li>position weight matrix (PWM) - a matrix with log likelihoods of PPM elements;</li>
<li>a matrix with self-information of elements in PWM.</li>
</ul>
<p>To compute and visualise sequence motifs, first you need to compute kmer statistics for one of the input immune repertoires, and then apply the <code><a href="../../reference/split_to_kmers.html">kmer_profile()</a></code> function to compute sequence motif matrices:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1">kmers =<span class="st"> </span><span class="kw"><a href="../../reference/getKmers.html">getKmers</a></span>(immdata<span class="op">$</span>data[[<span class="dv">1</span>]], <span class="dv">5</span>)</a>
<a class="sourceLine" id="cb13-2" data-line-number="2"><span class="kw"><a href="../../reference/split_to_kmers.html">kmer_profile</a></span>(kmers)</a></code></pre></div>
<pre><code>##    [,1]  [,2]  [,3]  [,4]  [,5]
## A 10336 10366  4355  3743  3639
## C  7476    31    32    21    12
## D  2428  2431  2431  2429  2376
## E  3876  6814  6820  6818  6644
## F   795   799   799  3005 10436
## G 10672 11153 11153 11111 10827
## H   452   452   452   756   732
## I   854  1173  1276  1034   981
## K   834  1274  1275  1275  1206
## L  3515  3522  4716  4704  4633
## M   273   277   279   279   256
## N  2809  2811  2812  2807  2733
## P  2646  2957  2957  2952  2896
## Q  2839  2851  8177  8176  8136
## R  4001  4016  4020  4011  3367
## S 16260 16272 15192  9452  4041
## T  3946  6556  6558  6968  6710
## V  1993  2247  2248  1823  1763
## W   553   555   556   510   497
## Y  2405  2406  2855  7089  7078
## attr(,"class")
## [1] "matrix"                  "immunr_kmer_profile_pfm"</code></pre>
<p>Currently we are not supporting sequence motifs analysis for more than one sample, but we are working on it. In order to compute and visualise sequence motif matrices for all of your samples you need to process them one-by-one, which can be easily done in for-loops or via the <code><a href="https://rdrr.io/r/base/lapply.html">lapply()</a></code> function.</p>
<p>Argument <code>.method</code> specifies which matrix to compute:</p>
<ul>
<li>
<code>.method = "freq"</code> - position frequency matrix (PFM);</li>
<li>
<code>.method = "prob"</code> - position probability matrix (PPM);</li>
<li>
<code>.method = "wei"</code> - position weight matrix (PWM);</li>
<li>
<code>.method = "self"</code> - self-information matrix.</li>
</ul>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1"><span class="kw"><a href="../../reference/split_to_kmers.html">kmer_profile</a></span>(kmers, <span class="st">"freq"</span>)</a></code></pre></div>
<pre><code>##    [,1]  [,2]  [,3]  [,4]  [,5]
## A 10336 10366  4355  3743  3639
## C  7476    31    32    21    12
## D  2428  2431  2431  2429  2376
## E  3876  6814  6820  6818  6644
## F   795   799   799  3005 10436
## G 10672 11153 11153 11111 10827
## H   452   452   452   756   732
## I   854  1173  1276  1034   981
## K   834  1274  1275  1275  1206
## L  3515  3522  4716  4704  4633
## M   273   277   279   279   256
## N  2809  2811  2812  2807  2733
## P  2646  2957  2957  2952  2896
## Q  2839  2851  8177  8176  8136
## R  4001  4016  4020  4011  3367
## S 16260 16272 15192  9452  4041
## T  3946  6556  6558  6968  6710
## V  1993  2247  2248  1823  1763
## W   553   555   556   510   497
## Y  2405  2406  2855  7089  7078
## attr(,"class")
## [1] "matrix"                  "immunr_kmer_profile_pfm"</code></pre>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1"><span class="kw"><a href="../../reference/split_to_kmers.html">kmer_profile</a></span>(kmers, <span class="st">"prob"</span>)</a></code></pre></div>
<pre><code>##          [,1]         [,2]         [,3]         [,4]         [,5]
## A 0.130896749 0.1312766739 0.0551524132 0.0474019477 0.0460848752
## C 0.094677254 0.0003925889 0.0004052531 0.0002659473 0.0001519699
## D 0.030748578 0.0307865709 0.0307865709 0.0307612426 0.0300900422
## E 0.049086281 0.0862935805 0.0863695655 0.0863442372 0.0841406735
## F 0.010068007 0.0101186632 0.0101186632 0.0380557983 0.1321631650
## G 0.135151907 0.1412433671 0.1412433671 0.1407114725 0.1371148513
## H 0.005724200 0.0057241999 0.0057241999 0.0095741043 0.0092701645
## I 0.010815192 0.0148550587 0.0161594671 0.0130947406 0.0124235401
## K 0.010561909 0.0161341388 0.0161468029 0.0161468029 0.0152729760
## L 0.044514519 0.0446031686 0.0597241746 0.0595722047 0.0586730494
## M 0.003457315 0.0035079721 0.0035333004 0.0035333004 0.0032420247
## N 0.035573623 0.0355989514 0.0356116156 0.0355482948 0.0346111470
## P 0.033509365 0.0374479186 0.0374479186 0.0373845978 0.0366754049
## Q 0.035953548 0.0361055178 0.1035548295 0.1035421653 0.1030355990
## R 0.050669301 0.0508592632 0.0509099198 0.0507959424 0.0426402239
## S 0.205919228 0.2060711979 0.1923939060 0.1197016324 0.0511758672
## T 0.049972772 0.0830262275 0.0830515558 0.0882438610 0.0849765080
## V 0.025239669 0.0284563656 0.0284690298 0.0230867622 0.0223269126
## W 0.007003280 0.0070286083 0.0070412725 0.0064587212 0.0062940871
## Y 0.030457303 0.0304699669 0.0361561744 0.0897762243 0.0896369186
## attr(,"class")
## [1] "matrix"                  "immunr_kmer_profile_ppm"</code></pre>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1"><span class="kw"><a href="../../reference/split_to_kmers.html">kmer_profile</a></span>(kmers, <span class="st">"wei"</span>)</a></code></pre></div>
<pre><code>##       [,1]      [,2]      [,3]       [,4]       [,5]
## A 9.013462 9.0176436 7.7665289 7.54805124  7.5073982
## C 8.546123 0.6322682 0.6780719 0.07038933 -0.7369656
## D 6.923625 6.9254061 6.9254061 6.92421868  6.8923910
## E 7.598425 8.4123581 8.4136279 8.41320479  8.3759083
## F 5.312883 5.3201236 5.3201236 7.23122118  9.0273531
## G 9.059615 9.1232161 9.1232161 9.11777295  9.0804178
## H 4.498251 4.4982509 4.4982509 5.24031433  5.1937717
## I 5.416164 5.8740592 5.9954845 5.69209238  5.6161812
## K 5.381975 5.9932215 5.9943534 5.99435344  5.9140861
## L 7.457381 7.4602511 7.8814199 7.87774425  7.8558029
## M 3.770829 3.7918141 3.8021932 3.80219322  3.6780719
## N 7.133913 7.1349396 7.1354528 7.13288525  7.0943416
## P 7.047669 7.2079904 7.2079904 7.20554891  7.1779178
## Q 7.149239 7.1553242 8.6754278 8.67525139  8.6681759
## R 7.644217 7.6496155 7.6510517 7.64781816  7.3953199
## S 9.667112 9.6681759 9.5690961 8.88447582  7.6585685
## T 7.624247 8.3566720 8.3571121 8.44460081  8.3901690
## V 6.638798 6.8118563 6.8124982 6.51017075  6.4618887
## W 4.789208 4.7944159 4.7970130 4.67242534  4.6351739
## Y 6.909893 6.9104928 7.1573469 8.46943832  8.4671980
## attr(,"class")
## [1] "matrix"                  "immunr_kmer_profile_pwm"</code></pre>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1"><span class="kw"><a href="../../reference/split_to_kmers.html">kmer_profile</a></span>(kmers, <span class="st">"self"</span>)</a></code></pre></div>
<pre><code>##         [,1]        [,2]        [,3]        [,4]        [,5]
## A 0.38398546 0.384551059 0.230560923 0.208516895 0.204596700
## C 0.32198203 0.004442023 0.004566752 0.003158543 0.001927575
## D 0.15446046 0.154596459 0.154596459 0.154505798 0.152092226
## E 0.21345348 0.305013544 0.305172450 0.305119492 0.300470790
## F 0.06679194 0.067054736 0.067054736 0.179461247 0.385864628
## G 0.39023035 0.398835248 0.398835248 0.398099228 0.393045652
## H 0.04263791 0.042637907 0.042637907 0.064210136 0.062603176
## I 0.07063182 0.090213314 0.096172690 0.081905883 0.078650698
## K 0.06933878 0.096058461 0.096115583 0.096115583 0.092139955
## L 0.19985151 0.200121482 0.242811091 0.242412218 0.240040728
## M 0.02826747 0.028608028 0.028777912 0.028777912 0.026807943
## N 0.17121757 0.171302918 0.171345584 0.171132187 0.167954725
## P 0.16417216 0.177464588 0.177464588 0.177255787 0.174906594
## Q 0.17249513 0.173004530 0.338783069 0.338759907 0.337831599
## R 0.21801704 0.218559832 0.218704403 0.218379017 0.194082999
## S 0.46946486 0.469592000 0.457486734 0.366584486 0.219462207
## T 0.21601800 0.298088155 0.298142544 0.309061793 0.302243775
## V 0.13397628 0.146126419 0.146173176 0.125517885 0.122464733
## W 0.05012775 0.050272439 0.050344733 0.046984198 0.046021025
## Y 0.15341551 0.153461021 0.173174125 0.312198864 0.311915245
## attr(,"class")
## [1] "matrix"                   "immunr_kmer_profile_self"</code></pre>
<div id="sequence-motif-visualisation" class="section level2">
<h2 class="hasAnchor">
<a href="#sequence-motif-visualisation" class="anchor"></a>Sequence motif visualisation</h2>
<p>Visualisation of sequence motif matrices is done by <code><a href="../../reference/vis.html">vis()</a></code>. There are two types of plots to choose from - sequence logo and “text logo”. The argument <code>.plot</code> regulates the type of plot: <code>.plot = "seq"</code> for sequence logo plots and <code>.plot = "text"</code> (by default) for “text logo” plots.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" data-line-number="1">kp =<span class="st"> </span><span class="kw"><a href="../../reference/split_to_kmers.html">kmer_profile</a></span>(kmers, <span class="st">"self"</span>)</a>
<a class="sourceLine" id="cb23-2" data-line-number="2">p1 =<span class="st"> </span><span class="kw"><a href="../../reference/vis.html">vis</a></span>(kp)</a>
<a class="sourceLine" id="cb23-3" data-line-number="3">p2 =<span class="st"> </span><span class="kw"><a href="../../reference/vis.html">vis</a></span>(kp, <span class="dt">.plot =</span> <span class="st">"seq"</span>)</a>
<a class="sourceLine" id="cb23-4" data-line-number="4"><span class="kw">grid.arrange</span>(p1, p2, <span class="dt">ncol=</span><span class="dv">2</span>)</a></code></pre></div>
<p><img src="v9_kmers_files/figure-html/unnamed-chunk-10-1.png" width="1152" style="display: block; margin: auto;"></p>
</div>
</div>
<div id="get-in-contact-with-us" class="section level1">
<h1 class="hasAnchor">
<a href="#get-in-contact-with-us" class="anchor"></a>Get in contact with us</h1>
<p>Can not find an important feature? Have a question or found a bug? Contact us at <a href="mailto:support@immunomind.io" class="email">support@immunomind.io</a></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#kmer-statistics-computation">Kmer statistics computation</a><ul class="nav nav-pills nav-stacked">
<li><a href="#kmer-statistics-visualisation">Kmer statistics visualisation</a></li>
      </ul>
</li>
      <li>
<a href="#sequence-motifs-analysis">Sequence motifs analysis</a><ul class="nav nav-pills nav-stacked">
<li><a href="#sequence-motif-visualisation">Sequence motif visualisation</a></li>
      </ul>
</li>
      <li><a href="#get-in-contact-with-us">Get in contact with us</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Vadim I. Nazarov, Vasily O. Tsvetkov, Eugene Rumynskiy, <a href="https://immunomind.io">ImmunoMind</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
