<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Fast and Effortless Analysis of Large-Scale Immune Repertoire Data • immunarch</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Fast and Effortless Analysis of Large-Scale Immune Repertoire Data">
<meta property="og:description" content="Fast and Effortless Analysis of Large-Scale Immune Repertoire Data.">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-128418614-2"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-128418614-2');
</script>
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">immunarch</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="articles/1_introduction.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/1_introduction.html">Introduction to `immunarch`</a>
    </li>
    <li>
      <a href="articles/2_data.html">Working with data in `immunarch`</a>
    </li>
    <li>
      <a href="articles/3_basic_analysis.html">Basic analysis and clonality</a>
    </li>
    <li>
      <a href="articles/4_overlap.html">Repertoire overlap and public clonotypes</a>
    </li>
    <li>
      <a href="articles/5_gene_usage.html">Gene usage analysis</a>
    </li>
    <li>
      <a href="articles/6_diversity.html">Diversity estimation</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://immunomind.io">ImmunoMind</a>
</li>
<li>
  <a href="https://github.com/immunomind/immunarch">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="contents col-md-9">

<div id="immunarch---fast-and-effortless-analysis-of-large-scale-immune-repertoire-data" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#immunarch---fast-and-effortless-analysis-of-large-scale-immune-repertoire-data" class="anchor"></a><code>immunarch</code> - Fast and Effortless Analysis of Large-Scale Immune Repertoire Data</h1></div>
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p><code>immunarch</code> is an R package designed to analyse TCR and BCR (immunoglobulin) repertoires, which constitute a large amount of data. The mission of <code>immunarch</code> is to make immune sequencing data analysis as effortless as possible—and help you focus on research instead of coding.</p>
</div>
<div id="features" class="section level2">
<h2 class="hasAnchor">
<a href="#features" class="anchor"></a>Features</h2>
<ol>
<li>
<p>Fast and easy manipulation of immune repertoire data:</p>
<ul>
<li><p>The package automatically detects the format of your files—no more guessing what format is <em>that</em> file, just pass them to the package;</p></li>
<li><p>Supports all popular TCR and BCR analysis and post-analysis formats: ImmunoSEQ, IMGT, MiTCR, MiXCR, MiGEC, MigMap, VDJtools, tcR. More coming in the future;</p></li>
<li><p>Works on any data source you are comfortable with: R data frames, data tables from <a href="http://r-datatable.com">data.table</a>, databases like <a href="https://github.com/MonetDB">MonetDB</a>, Apache Spark data frames via <a href="https://spark.rstudio.com/">sparklyr</a>.</p></li>
</ul>
</li>
<li>
<p>Immune repertoire analysis made simple:</p>
<ul>
<li><p>Most methods are incorporated in a couple of main functions with clear naming—no more remembering tens and tens of functions with obscure names;</p></li>
<li><p>Repertoire overlap analysis <em>(common indices including overlap coefficient, Jaccard index and Morisita’s overlap index)</em></p></li>
<li><p>Gene usage estimation <em>(correlation, Jensen-Shannon Divergence, clustering)</em></p></li>
<li><p>Diversity evaluation <em>(ecological diversity index, Gini index, inverse Simpson index, rarefaction analysis)</em></p></li>
<li><p>Coming in the next releases: CDR3 amino acid physical and chemical properties assessment, Kmer distribution measures and statistics, mutation networks, tracking clonotypes across time points</p></li>
</ul>
</li>
</ol>
<ul>
<li>
<p>Publication-ready plots with a built-in tool for tweaking visualisations:</p>
<ul>
<li><p>Rich visualisation procedures with <a href="https://ggplot2.tidyverse.org/">ggplot2</a>;</p></li>
<li><p>Built-in tool <code>fixVis</code> makes your plots publication-ready: easily change font sizes, text angles, titles, legends and many more with clear-cut GUI;</p></li>
</ul>
</li>
</ul>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>You can find the list of releases of <code>immunarch</code> here: <a href="https://github.com/immunomind/immunarch/releases" class="uri">https://github.com/immunomind/immunarch/releases</a></p>
<p>In order to install <code>immunarch</code>, you need to download it first. If you want to download the latest version, you need to download the package file, available here <a href="https://github.com/immunomind/immunarch/releases/download/latest/immunarch.tar.gz" class="uri">https://github.com/immunomind/immunarch/releases/download/latest/immunarch.tar.gz</a></p>
<p>Note that You <em>should not</em> un-archive it!</p>
<p>After downloading the file, you need to install a number of packages with R commands listed below, and run the newly installed <code>devtools</code> package to install <code>immunarch</code> locally. Upon completion the dependencies will have been already downloaded and installed.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages</a></span>(<span class="st">"devtools"</span>, <span class="dt">dependencies =</span> T)
devtools<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/devtools/topics/reexports">install_local</a></span>(<span class="st">"path/to/your/folder/with/immunarch.tar.gz"</span>, <span class="dt">dependencies=</span>T)</code></pre></div>
<p>That’s it, you can start using <code>immunarch</code> now!</p>
</div>
<div id="installation-troubleshooting" class="section level2">
<h2 class="hasAnchor">
<a href="#installation-troubleshooting" class="anchor"></a>Installation troubleshooting</h2>
<p>If you run in any trouble, try the following steps:</p>
<ol>
<li><p>Check your R version. Run <code>version</code> command in the console to get your R versions. If the R version is below 3.4.0 (for example, <code>R version 3.1.0</code>), try updating your R version to the latest one.</p></li>
<li><p>Check if your packages are outdated and update them. In RStudio you can run the “Update” button on top of the package list. In R console you can run the <code><a href="https://www.rdocumentation.org/packages/utils/topics/update.packages">old.packages()</a></code> command to view a list of outdated packages.</p></li>
<li><p>If you are on Mac and have issues like old packages can’t be updated, or error messages such as <code>ld: warning: directory not found for option</code> or <code>ld: library not found for -lgfortran</code>, <a href="https://thecoatlessprofessor.com/programming/rcpp-rcpparmadillo-and-os-x-mavericks--lgfortran-and--lquadmath-error/">this link</a> will help you to fix the issue.</p></li>
<li><p>If you are working under Linux and have issues with igraph library or have Fortran errors, see <a href="https://ashokragavendran.wordpress.com/2017/10/24/error-installing-rigraph-unable-to-load-shared-object-igraph-so-libgfortran-so-4-cannot-open-shared-object-file-no-such-file-or-directory/">this link</a></p></li>
<li><p>If troubles still persist, message us on <a href="mailto:support@immunomind.io">support@immunomind.io</a> or create an issue in <a href="https://github.com/immunomind/immunarch/issues" class="uri">https://github.com/immunomind/immunarch/issues</a> with the code that represents the issue and the output you get in the console.</p></li>
</ol>
</div>
<div id="quick-start" class="section level2">
<h2 class="hasAnchor">
<a href="#quick-start" class="anchor"></a>Quick start</h2>
<p>Importing data into R is fairly simple. The gist of the typical TCR or BCR explorational data analysis workflow can be reduced to the next few lines of code:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Load the data to the package</span>
immdata =<span class="st"> </span><span class="kw"><a href="reference/repLoad.html">repLoad</a></span>(<span class="st">"path/to/your/folder/with/repertoires"</span>)
<span class="co"># If you folder contains metadata.txt file, immdata will have two elements:</span>
<span class="co"># - immdata$data with a list of parsed repertoires</span>
<span class="co"># - immdata$meta with the metadata file</span>

<span class="co"># Compute and visualise overlap statistics</span>
ov =<span class="st"> </span><span class="kw"><a href="reference/repOverlap.html">repOverlap</a></span>(immdata<span class="op">$</span>data)
<span class="kw"><a href="reference/vis.html">vis</a></span>(ov)

<span class="co"># Cluster samples using K-means algorithm applied to the number of overlapped clonotypes</span>
<span class="co"># and visualise the results</span>
ov.kmeans =<span class="st"> </span><span class="kw"><a href="reference/repOverlapAnalysis.html">repOverlapAnalysis</a></span>(ov, <span class="dt">.method =</span> <span class="st">"kmeans"</span>)
<span class="kw"><a href="reference/vis.html">vis</a></span>(ov.kmeans)

<span class="co"># Compute and visualise gene usage with samples, grouped by their disease status</span>
gu =<span class="st"> </span><span class="kw"><a href="reference/geneUsage.html">geneUsage</a></span>(immdata<span class="op">$</span>data)
<span class="kw"><a href="reference/vis.html">vis</a></span>(gu, <span class="dt">.by=</span><span class="st">"Status"</span>, <span class="dt">.meta=</span>immdata<span class="op">$</span>meta)

<span class="co"># Compute Jensen-Shannon divergence among gene distributions of samples, </span>
<span class="co"># cluster samples using the hierarchical clustering and visualise the results</span>
gu.clust =<span class="st"> </span><span class="kw"><a href="reference/geneUsageAnalysis.html">geneUsageAnalysis</a></span>(gu, <span class="dt">.method =</span> <span class="st">"js+hclust"</span>)
<span class="kw"><a href="reference/vis.html">vis</a></span>(gu.clust)

<span class="co"># Compare diversity of repertoires and visualise samples, grouped by two parameters</span>
div =<span class="st"> </span><span class="kw"><a href="reference/repDiversity.html">repDiversity</a></span>(immdata<span class="op">$</span>data, <span class="dt">.method =</span> <span class="st">"chao1"</span>)
<span class="kw"><a href="reference/vis.html">vis</a></span>(div, <span class="dt">.by=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"Status"</span>, <span class="st">"Treatment"</span>), <span class="dt">.meta=</span>immdata<span class="op">$</span>meta)

<span class="co"># Tweak and fix the visalisation of diversity estimates to make the plot publication-ready</span>
div.plot =<span class="st"> </span><span class="kw"><a href="reference/vis.html">vis</a></span>(div, <span class="dt">.by=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"Status"</span>, <span class="st">"Treatment"</span>), <span class="dt">.meta=</span>immdata<span class="op">$</span>meta)
<span class="kw"><a href="reference/fixVis.html">fixVis</a></span>(div.plot)</code></pre></div>
<p>If you want to test the package without parsing any data, you can load a small test dataset provided along with the package. Load the data with the following command:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/data">data</a></span>(immdata)</code></pre></div>
</div>
</div>
<div id="license" class="section level1">
<h1 class="hasAnchor">
<a href="#license" class="anchor"></a>License</h1>
<p>The package is freely distributed under the AGPL v3 license. You can read more about it <a href="https://tldrlegal.com/license/gnu-affero-general-public-license-v3-(agpl-3.0)">here</a>.</p>
<p>Additionally, we provide an annual subscription that includes next services:</p>
<ul>
<li>100 hours of consultations on the TCR &amp; BCR repertoire analysis (contact us to purchase more);</li>
<li>Priority email and call support;</li>
<li>Package modifications and feature implementations are issued promptly;</li>
<li>Setup a cloud or cluster installation of <em>immunarch</em>, including the development of cloud <em>immunarch</em>-based software;</li>
<li>Use <em>immunarch</em> team expertise in your projects;</li>
<li>If you need a GPL-compatible license other than AGPL v3, contact us.</li>
</ul>
<p>Contact us at <a href="mailto:support@immunomind.io">support@immunomind.io</a> for more information.</p>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Report a bug at <br><a href="https://github.com/immunomind/immunarch/issues">https://​github.com/​immunomind/​immunarch/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/AGPL-3">AGPL-3</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li> ImmunoMind <br><small class="roles"> Maintainer </small>  </li>
</ul>
</div>

  </div>
</div>

      <footer><div class="copyright">
  <p>Developed by  ImmunoMind.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
