<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Working with data in <code>immunarch</code> • immunarch</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Working with data in &lt;code&gt;immunarch&lt;/code&gt;">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-128418614-2"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-128418614-2');
</script>
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">immunarch</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.3.9005</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/1_introduction.html">Get started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/1_introduction.html">Introduction to `immunarch`</a>
    </li>
    <li>
      <a href="../articles/2_data.html">Working with data in `immunarch`</a>
    </li>
    <li>
      <a href="../articles/3_basic_analysis.html">Basic analysis and clonality</a>
    </li>
    <li>
      <a href="../articles/4_overlap.html">Repertoire overlap and public clonotypes</a>
    </li>
    <li>
      <a href="../articles/5_gene_usage.html">Gene usage analysis</a>
    </li>
    <li>
      <a href="../articles/6_diversity.html">Diversity estimation</a>
    </li>
    <li>
      <a href="../articles/7_fixvis.html">Make your plots publication-ready with fixVis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://immunomind.io">ImmunoMind</a>
</li>
<li>
  <a href="http://twitter.com/immunomind">
    <span class="fa fa-lg fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/immunomind/immunarch">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Working with data in <code>immunarch</code>
</h1>
                        <h4 class="author"><a href="https://immunomind.io">ImmunoMind</a></h4>
            
            <h4 class="date"><a href="mailto:support@immunomind.io" class="email">support@immunomind.io</a></h4>
      
      <small class="dont-index">Source: <a href="https://github.com/immunomind/immunarch/blob/master/vignettes/2_data.Rmd"><code>vignettes/2_data.Rmd</code></a></small>
      <div class="hidden name"><code>2_data.Rmd</code></div>

    </div>

    
    
<!--
%\VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{Working with data in immunarch}
%\VignettePackage{immunarch}
-->
<div id="immunarch-data-format" class="section level1">
<h1 class="hasAnchor">
<a href="#immunarch-data-format" class="anchor"></a>Immunarch data format</h1>
<p><code>immunarch</code> comes with its own data format, including tab-delimited columns that can be specified as follows:</p>
<ul>
<li><p>“Clones” - count or number of barcodes (events, UMIs) or reads;</p></li>
<li><p>“Proportion” - proportion of barcodes (events, UMIs) or reads;</p></li>
<li><p>“CDR3.nt” - CDR3 nucleotide sequence;</p></li>
<li><p>“CDR3.aa” - CDR3 amino acid sequence;</p></li>
<li><p>“V.name” - names of aligned Variable gene segments;</p></li>
<li><p>“D.name” - names of aligned Diversity gene segments or NA;</p></li>
<li><p>“J.name” - names of aligned Joining gene segments;</p></li>
<li><p>“V.end” - last positions of aligned V gene segments (1-based);</p></li>
<li><p>“D.start” - positions of D’5 end of aligned D gene segments (1-based);</p></li>
<li><p>“D.end” - positions of D’3 end of aligned D gene segments (1-based);</p></li>
<li><p>“J.start” - first positions of aligned J gene segments (1-based);</p></li>
<li><p>“VJ.ins” - number of inserted nucleotides (N-nucleotides) at V-J junction (-1 for receptors with VDJ recombination);</p></li>
<li><p>“VD.ins” - number of inserted nucleotides (N-nucleotides) at V-D junction (-1 for receptors with VJ recombination);</p></li>
<li><p>“DJ.ins” - number of inserted nucleotides (N-nucleotides) at D-J junction (-1 for receptors with VJ recombination);</p></li>
<li><p>“Sequence” - full nucleotide sequence.</p></li>
</ul>
</div>
<div id="input-output" class="section level1">
<h1 class="hasAnchor">
<a href="#input-output" class="anchor"></a>Input / output</h1>
<p>The package provides several IO functions:</p>
<ul>
<li><p><code>repLoad</code> - to load the repertoires, having compatible format.</p></li>
<li><p><code>repSave</code> - to save changes and to write the repertoire data to a file in a specific format (<code>immunarch</code>, VDJtools or MonetDB).</p></li>
</ul>
<p><code>repLoad</code> has the <code>.format</code> argument that sets the format for input repertoire files. Do not provide it if you want <code>immunarch</code> to detect the formats and parse files automatically! In case you want to force the package to parse the data in a specific format, you can choose one of the several options for the argument:</p>
<ul>
<li><p><code>"immunarch"</code> - current software tool, in case you forgot it :)</p></li>
<li><p><code>"immunoseq"</code> - <a href="http://www.adaptivebiotech.com/immunoseq" class="uri">http://www.adaptivebiotech.com/immunoseq</a></p></li>
<li><p><code>"mitcr"</code> - <a href="https://github.com/milaboratory/mitcr" class="uri">https://github.com/milaboratory/mitcr</a></p></li>
<li><p><code>"mixcr"</code> - <a href="https://github.com/milaboratory/mixcr" class="uri">https://github.com/milaboratory/mixcr</a></p></li>
<li><p><code>"migec"</code> - <a href="http://migec.readthedocs.io/en/latest/" class="uri">http://migec.readthedocs.io/en/latest/</a></p></li>
<li><p><code>"migmap"</code> - <a href="https://github.com/mikessh/migmap" class="uri">https://github.com/mikessh/migmap</a></p></li>
<li><p><code>"tcr"</code> - <a href="https://imminfo.github.io/tcr/" class="uri">https://imminfo.github.io/tcr/</a></p></li>
<li><p><code>"vdjtools"</code> - <a href="http://vdjtools-doc.readthedocs.io/en/latest/" class="uri">http://vdjtools-doc.readthedocs.io/en/latest/</a></p></li>
<li><p><code>"imgt"</code> - <a href="http://www.imgt.org/HighV-QUEST/" class="uri">http://www.imgt.org/HighV-QUEST/</a></p></li>
<li><p><code>"airr"</code> - <a href="http://docs.airr-community.org/en/latest/datarep/overview.html" class="uri">http://docs.airr-community.org/en/latest/datarep/overview.html</a></p></li>
<li><p><code>"10x"</code> - <a href="https://support.10xgenomics.com/single-cell-vdj/software/pipelines/latest/output/annotation" class="uri">https://support.10xgenomics.com/single-cell-vdj/software/pipelines/latest/output/annotation</a></p></li>
<li><p><code>"archer"</code> - ArcherDX clonotype tables. <a href="https://archerdx.com/" class="uri">https://archerdx.com/</a></p></li>
</ul>
<p>These parsers will be available soon.</p>
<ul>
<li><p><code>"imseq"</code> - <a href="http://www.imtools.org/" class="uri">http://www.imtools.org/</a></p></li>
<li><p><code>"rtcr"</code> - <a href="https://github.com/uubram/RTCR" class="uri">https://github.com/uubram/RTCR</a></p></li>
<li><p><code>"vidjil"</code> - <a href="http://www.vidjil.org/" class="uri">http://www.vidjil.org/</a></p></li>
</ul>
<p>Please contact us if there are more file formats you want to be supported.</p>
<p>For parsing IgBLAST results process the data with MigMap first.</p>
<p>You can load the data either from a single file or from a folder with repertoire files. A single file can be loaded as follows:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="co"># To load the data from a single file without forcing the data format:</span></a>
<a class="sourceLine" id="cb1-2" data-line-number="2">immdata &lt;-<span class="st"> </span><span class="kw"><a href="../reference/repLoad.html">repLoad</a></span>(<span class="st">"path/to/your/folder/immunoseq_1.txt"</span>)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3"></a>
<a class="sourceLine" id="cb1-4" data-line-number="4"><span class="co"># To load the data from a single ImmunoSEQ file go with:</span></a>
<a class="sourceLine" id="cb1-5" data-line-number="5">immdata &lt;-<span class="st"> </span><span class="kw"><a href="../reference/repLoad.html">repLoad</a></span>(<span class="st">"path/to/your/folder/immunoseq_1.txt"</span>, <span class="dt">.format =</span> <span class="st">"immunoseq"</span>)</a></code></pre></div>
<p>In the second case you may want to provide a metadata file and locate it in the folder. It is necessary to name it exactly “metadata.txt”.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="co"># For instance you have a following structure in your folder:</span></a>
<a class="sourceLine" id="cb2-2" data-line-number="2"><span class="co"># &gt;_ ls</span></a>
<a class="sourceLine" id="cb2-3" data-line-number="3"><span class="co"># immunoseq1.txt</span></a>
<a class="sourceLine" id="cb2-4" data-line-number="4"><span class="co"># immunoseq2.txt</span></a>
<a class="sourceLine" id="cb2-5" data-line-number="5"><span class="co"># immunoseq3.txt</span></a>
<a class="sourceLine" id="cb2-6" data-line-number="6"><span class="co"># metadata.txt</span></a></code></pre></div>
<p>With the metadata <code>repLoad</code> will create a list in the environment with 2 elements, namely <code>data</code> and <code>meta</code>. All the data will be accessible simply from <code>immdata$data</code>.</p>
<p>Otherwise <code>repLoad</code> will create a list with the number of elements matching the number of your files. They will be accessible directly from <code>immdata</code>.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="co"># To load the whole folder with every file in it type:</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2"></a>
<a class="sourceLine" id="cb3-3" data-line-number="3">immdata &lt;-<span class="st"> </span><span class="kw"><a href="../reference/repLoad.html">repLoad</a></span>(<span class="st">"path/to/your/folder/"</span>)</a>
<a class="sourceLine" id="cb3-4" data-line-number="4"></a>
<a class="sourceLine" id="cb3-5" data-line-number="5"><span class="co"># In order to do that your folder must contain metadata file named</span></a>
<a class="sourceLine" id="cb3-6" data-line-number="6"><span class="co"># exactly "metadata.txt".</span></a>
<a class="sourceLine" id="cb3-7" data-line-number="7"></a>
<a class="sourceLine" id="cb3-8" data-line-number="8"><span class="co"># In R, when you load your data:</span></a>
<a class="sourceLine" id="cb3-9" data-line-number="9"><span class="co"># &gt; immdata &lt;- repLoad("path/to/your/folder/")</span></a>
<a class="sourceLine" id="cb3-10" data-line-number="10"><span class="co"># &gt; names(immdata)</span></a>
<a class="sourceLine" id="cb3-11" data-line-number="11"><span class="co"># [1] "data" "meta"</span></a>
<a class="sourceLine" id="cb3-12" data-line-number="12"></a>
<a class="sourceLine" id="cb3-13" data-line-number="13"><span class="co"># Suppose you do not have "metadata.txt":</span></a>
<a class="sourceLine" id="cb3-14" data-line-number="14"><span class="co"># &gt; immdata &lt;- repLoad("path/to/your/folder/")</span></a>
<a class="sourceLine" id="cb3-15" data-line-number="15"><span class="co"># &gt; names(immdata)</span></a>
<a class="sourceLine" id="cb3-16" data-line-number="16"><span class="co"># [1] "immunoseq_1" "immunoseq_2" "immunoseq_3"</span></a></code></pre></div>
<p>The metadata has to be tab delimited file with first column named “Sample” and any number of additional columns with arbitrary names. The first column should contain base names of files without extensions in your folder.</p>
<table class="table">
<thead><tr class="header">
<th align="center"><strong>Sample</strong></th>
<th align="center"><strong>Sex</strong></th>
<th align="center"><strong>Age</strong></th>
<th align="center"><strong>Status</strong></th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center">immunoseq_1</td>
<td align="center">M</td>
<td align="center">1</td>
<td align="center">C</td>
</tr>
<tr class="even">
<td align="center">immunoseq_2</td>
<td align="center">M</td>
<td align="center">2</td>
<td align="center">C</td>
</tr>
<tr class="odd">
<td align="center">immunoseq_3</td>
<td align="center">F</td>
<td align="center">3</td>
<td align="center">A</td>
</tr>
</tbody>
</table>
<p>In order to import data from the external databases you have to create a connection to this database and then load the data. Make sure that the table format in your database matches the <code>immunarch</code>’s format.</p>
<p>To illustrate the use of external database, here is an example demonstrating data loading to the local MonetDB database:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="co"># Your list of repertoires in immunarch's format</span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2">DATA</a>
<a class="sourceLine" id="cb4-3" data-line-number="3"><span class="co"># Metadata data frame</span></a>
<a class="sourceLine" id="cb4-4" data-line-number="4">META</a>
<a class="sourceLine" id="cb4-5" data-line-number="5"></a>
<a class="sourceLine" id="cb4-6" data-line-number="6"><span class="co"># Create a temporary directory</span></a>
<a class="sourceLine" id="cb4-7" data-line-number="7">dbdir =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/tempfile">tempdir</a></span>()</a>
<a class="sourceLine" id="cb4-8" data-line-number="8"></a>
<a class="sourceLine" id="cb4-9" data-line-number="9"><span class="co"># Create a DBI connection to MonetDB in the temporary directory.</span></a>
<a class="sourceLine" id="cb4-10" data-line-number="10">con =<span class="st"> </span>DBI<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/DBI/topics/dbConnect">dbConnect</a></span>(MonetDBLite<span class="op">::</span><span class="kw">MonetDBLite</span>(), <span class="dt">embedded =</span> dbdir)</a>
<a class="sourceLine" id="cb4-11" data-line-number="11"></a>
<a class="sourceLine" id="cb4-12" data-line-number="12"><span class="co"># Write each repertoire to MonetDB. Each table has corresponding name from the DATA</span></a>
<a class="sourceLine" id="cb4-13" data-line-number="13"><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(DATA)) {</a>
<a class="sourceLine" id="cb4-14" data-line-number="14">  DBI<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/DBI/topics/dbWriteTable">dbWriteTable</a></span>(con, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(DATA)[i], DATA[[i]], <span class="dt">overwrite=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb4-15" data-line-number="15">}</a>
<a class="sourceLine" id="cb4-16" data-line-number="16"></a>
<a class="sourceLine" id="cb4-17" data-line-number="17"><span class="co"># Create a source in the temporary directory with MonetDB</span></a>
<a class="sourceLine" id="cb4-18" data-line-number="18">ms =<span class="st"> </span>MonetDBLite<span class="op">::</span><span class="kw">src_monetdblite</span>(<span class="dt">dbdir =</span> dbdir)</a>
<a class="sourceLine" id="cb4-19" data-line-number="19">res_db =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>()</a>
<a class="sourceLine" id="cb4-20" data-line-number="20"></a>
<a class="sourceLine" id="cb4-21" data-line-number="21"><span class="co"># Load the data from MonetDB to dplyr tables</span></a>
<a class="sourceLine" id="cb4-22" data-line-number="22"><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(DATA)) {</a>
<a class="sourceLine" id="cb4-23" data-line-number="23">  res_db[[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(DATA)[i]]] =<span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/tbl.html">tbl</a></span>(ms, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(DATA)[i])</a>
<a class="sourceLine" id="cb4-24" data-line-number="24">}</a>
<a class="sourceLine" id="cb4-25" data-line-number="25"></a>
<a class="sourceLine" id="cb4-26" data-line-number="26"><span class="co"># Your data is ready to use</span></a>
<a class="sourceLine" id="cb4-27" data-line-number="27"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">data =</span> res_db, <span class="dt">meta =</span> META)</a></code></pre></div>
<p>You might want to make a list with <code>data</code>, which is a list of tables of repertoires, and the <code>meta</code>, containing metadata:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="co"># Load the data to the immdata variables. Metadata file "metadata.txt" will be found automatically.</span></a>
<a class="sourceLine" id="cb5-2" data-line-number="2">immdata =<span class="st"> </span><span class="kw"><a href="../reference/repLoad.html">repLoad</a></span>(<span class="st">"your_folder"</span>, <span class="st">"optionally_your_format"</span>)</a>
<a class="sourceLine" id="cb5-3" data-line-number="3"><span class="co"># Repertoires</span></a>
<a class="sourceLine" id="cb5-4" data-line-number="4">immdata<span class="op">$</span>data</a>
<a class="sourceLine" id="cb5-5" data-line-number="5"><span class="co"># Metadata</span></a>
<a class="sourceLine" id="cb5-6" data-line-number="6">immdata<span class="op">$</span>meta</a></code></pre></div>
<p><code>immunarch</code> is compatible with following sources:</p>
<ul>
<li><p>R data frames (for common applications)</p></li>
<li><p>R data tables (for speed, although requires lots of RAM)</p></li>
<li><p>MonetDB-like databases that support both DBI and dplyr (best choice, although you have to be a bit familiar with dplyr)</p></li>
<li><p>Apache Spark (if you are expierienced and comfortable with it)</p></li>
</ul>
</div>
<div id="basic-data-manipulations-with-dplyr-and-immunarch" class="section level1">
<h1 class="hasAnchor">
<a href="#basic-data-manipulations-with-dplyr-and-immunarch" class="anchor"></a>Basic data manipulations with dplyr and immunarch</h1>
<p>You can find the introduction to <code>dplyr</code> here: <a href="https://cran.rstudio.com/web/packages/dplyr/vignettes/introduction.html" class="uri">https://cran.rstudio.com/web/packages/dplyr/vignettes/introduction.html</a></p>
<div id="get-the-most-abundant-clonotypes" class="section level2">
<h2 class="hasAnchor">
<a href="#get-the-most-abundant-clonotypes" class="anchor"></a>Get the most abundant clonotypes</h2>
<p>The function returns the most abundant clonotypes for the given repertoire:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="kw"><a href="../reference/top.html">top</a></span>(immdata<span class="op">$</span>data[[<span class="dv">1</span>]])</a></code></pre></div>
<pre><code>## # A tibble: 11 x 15
##    Clones Proportion CDR3.nt CDR3.aa V.name D.name J.name V.end D.start
##     &lt;dbl&gt;      &lt;dbl&gt; &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;  &lt;int&gt;   &lt;int&gt;
##  1    206    0.0206  TGCGCC… CASSQE… TRBV4… TRBD1  TRBJ2…    16      18
##  2    193    0.0193  TGCGCC… CASSYR… TRBV4… TRBD1  TRBJ2…    11      13
##  3     79    0.0079  TGTGCC… CATSTN… TRBV15 TRBD1  TRBJ2…    11      16
##  4     64    0.0064  TGTGCC… CATSIG… TRBV15 TRBD2  TRBJ2…    11      19
##  5     62    0.0062  TGCGCC… CASQGD… TRBV4… TRBD1  TRBJ1…     8      13
##  6     56    0.0056  TGTGCC… CASSPW… TRBV27 TRBD1  TRBJ1…    11      16
##  7     54    0.0054  TGCGCC… CASSQD… TRBV4… TRBD1  TRBJ2…    16      21
##  8     35    0.0035  TGTGCC… CASSEE… TRBV2  TRBD1  TRBJ1…    15      17
##  9     34    0.0034  TGCGCC… CASSQP… TRBV4… TRBD1  TRBJ2…    14      23
## 10     29    0.00290 TGTGCC… CASSWV… TRBV6… TRBD1  TRBJ2…    12      20
## 11     29    0.00290 TGTGCC… CASSFR… TRBV27 TRBD1  TRBJ2…    13      16
## # … with 6 more variables: D.end &lt;int&gt;, J.start &lt;int&gt;, VJ.ins &lt;dbl&gt;,
## #   VD.ins &lt;dbl&gt;, DJ.ins &lt;dbl&gt;, Sequence &lt;lgl&gt;</code></pre>
</div>
<div id="filter-functional-non-functional-in-frame-out-of-frame-clonotypes" class="section level2">
<h2 class="hasAnchor">
<a href="#filter-functional-non-functional-in-frame-out-of-frame-clonotypes" class="anchor"></a>Filter functional / non-functional / in-frame / out-of-frame clonotypes</h2>
<p>Conveniently, functions are vectorised over the list of data frames; and <code><a href="../reference/coding.html">coding(immdata$data)</a></code> in the example below returns a list of data frames with coding sequences:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="kw"><a href="../reference/coding.html">coding</a></span>(immdata<span class="op">$</span>data[[<span class="dv">1</span>]])</a></code></pre></div>
<p>The next one operates in a similar fashion:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1"><span class="kw"><a href="../reference/coding.html">noncoding</a></span>(immdata<span class="op">$</span>data[[<span class="dv">1</span>]])</a></code></pre></div>
<p>Now, the computation of the number of filtered sequences is straightforward:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">nrow</a></span>(<span class="kw"><a href="../reference/coding.html">inframes</a></span>(immdata<span class="op">$</span>data[[<span class="dv">1</span>]]))</a></code></pre></div>
<p>And for the out-of-frame clonotypes:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">nrow</a></span>(<span class="kw"><a href="../reference/coding.html">outofframes</a></span>(immdata<span class="op">$</span>data[[<span class="dv">1</span>]]))</a></code></pre></div>
</div>
<div id="get-subset-of-clonotypes-with-a-specific-v-gene" class="section level2">
<h2 class="hasAnchor">
<a href="#get-subset-of-clonotypes-with-a-specific-v-gene" class="anchor"></a>Get subset of clonotypes with a specific V gene</h2>
<p>It is simple to subset data frame according to labels in the specified index. In the example the resulting data frame contains only records with ‘TRBV10-1’ V gene:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/filter">filter</a></span>(immdata<span class="op">$</span>data[[<span class="dv">1</span>]], V.name <span class="op">==</span><span class="st"> 'TRBV10-1'</span>)</a></code></pre></div>
<pre><code>## # A tibble: 26 x 15
##    Clones Proportion CDR3.nt CDR3.aa V.name D.name J.name V.end D.start
##     &lt;dbl&gt;      &lt;dbl&gt; &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;  &lt;int&gt;   &lt;int&gt;
##  1      2     0.0002 TGCGCC… CASSES… TRBV1… TRBD2  TRBJ2…    16      20
##  2      2     0.0002 TGCGCC… CASSDG… TRBV1… TRBD1  TRBJ2…    13      15
##  3      2     0.0002 TGCGCC… CASSGD… TRBV1… TRBD2  TRBJ2…     8      10
##  4      1     0.0001 TGCGCC… CASSEA… TRBV1… TRBD2  TRBJ2…    14      21
##  5      1     0.0001 TGCGCC… CATLRS… TRBV1… TRBD1  TRBJ2…     6       7
##  6      1     0.0001 TGCGCC… CASSES… TRBV1… TRBD2  TRBJ2…    16      20
##  7      1     0.0001 TGCGCC… CASSES… TRBV1… TRBD2  TRBJ2…    16      17
##  8      1     0.0001 TGCGCC… CASRAS… TRBV1… TRBD2  TRBJ2…    10      13
##  9      1     0.0001 TGCGCC… CASRGS… TRBV1… TRBD1  TRBJ2…    10      11
## 10      1     0.0001 TGCGCC… CASRRD… TRBV1… TRBD1  TRBJ2…     8      13
## # … with 16 more rows, and 6 more variables: D.end &lt;int&gt;, J.start &lt;int&gt;,
## #   VJ.ins &lt;dbl&gt;, VD.ins &lt;dbl&gt;, DJ.ins &lt;dbl&gt;, Sequence &lt;lgl&gt;</code></pre>
<!---
## Additional functions to help with dplyr

- `todf` - convert the input to R data frame.

- `todt` - convert the input to R data table.

- `db_add` - add a new column to the existing database.

- `db_rem` - remove a specific column from the existing database.

- `db_write` - write / rewrite a database table.

With these functions you can write the subset of your repertoires to new tables with `db_write`, after you have filtered out the noncoding sequences, for instance. 

-->
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#immunarch-data-format">Immunarch data format</a></li>
      <li><a href="#input-output">Input / output</a></li>
      <li>
<a href="#basic-data-manipulations-with-dplyr-and-immunarch">Basic data manipulations with dplyr and immunarch</a><ul class="nav nav-pills nav-stacked">
<li><a href="#get-the-most-abundant-clonotypes">Get the most abundant clonotypes</a></li>
      <li><a href="#filter-functional-non-functional-in-frame-out-of-frame-clonotypes">Filter functional / non-functional / in-frame / out-of-frame clonotypes</a></li>
      <li><a href="#get-subset-of-clonotypes-with-a-specific-v-gene">Get subset of clonotypes with a specific V gene</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by  ImmunoMind.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
