<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Repertoire overlap and public clonotypes • immunarch</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Repertoire overlap and public clonotypes">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-128418614-2"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-128418614-2');
</script>
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">immunarch</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/1_introduction.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/1_introduction.html">Introduction to `immunarch`</a>
    </li>
    <li>
      <a href="../articles/2_data.html">Working with data in `immunarch`</a>
    </li>
    <li>
      <a href="../articles/3_basic_analysis.html">Basic analysis and clonality</a>
    </li>
    <li>
      <a href="../articles/4_overlap.html">Repertoire overlap and public clonotypes</a>
    </li>
    <li>
      <a href="../articles/5_gene_usage.html">Gene usage analysis</a>
    </li>
    <li>
      <a href="../articles/6_diversity.html">Diversity estimation</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://immunomind.io">ImmunoMind</a>
</li>
<li>
  <a href="https://github.com/immunomind/immunarch">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Repertoire overlap and public clonotypes</h1>
                        <h4 class="author"><a href="https://immunomind.io">ImmunoMind</a></h4>
            
            <h4 class="date"><a href="mailto:support@immunomind.io">support@immunomind.io</a></h4>
      
      
      <div class="hidden name"><code>4_overlap.Rmd</code></div>

    </div>

    
    
<!--
%\VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{Introduction to immunarch}
%\VignettePackage{immunarch}
-->
<div id="repertoire-overlap" class="section level1">
<h1 class="hasAnchor">
<a href="#repertoire-overlap" class="anchor"></a>Repertoire overlap</h1>
<p>Repertoire overlap is the most common approach to measure repertoire similarity. It is achieved by computation of specific statistics on clonotypes shared between given repertoires, also called “public” clonotypes. <code>immunarch</code> provides several indices: - number of public clonotypes (<code>.method = "public"</code>) - a classic measure of overlap similarity.</p>
<ul>
<li><p>overlap coefficient (<code>.method = "overlap"</code>) - a normalised measure of overlap similarity. It is defined as the size of the intersection divided by the smaller of the size of the two sets.</p></li>
<li><p>Jaccard index (<code>.method = "jaccard"</code>) - it measures similarity between finite sample sets, and is defined as the size of the intersection divided by the size of the union of the sample sets.</p></li>
<li><p>Tversky index (<code>.method = "tversky"</code>) - an asymmetric similarity measure on sets that compares a variant to a prototype. If using default arguments, it’s similar to Dice’s coefficient.</p></li>
<li><p>cosine similarity (<code>.method = "cosine"</code>) - a measure of similarity between two non-zero vectors</p></li>
<li><p>Morisita’s overlap index (<code>.method = "morisita"</code>) - a statistical measure of dispersion of individuals in a population. It is used to compare overlap among samples.</p></li>
<li><p>top-overlap - overlaps of the N most abundant clonotypes with sequentially growing N (<code>.method = "top+METHOD"</code>)</p></li>
</ul>
<p>The function that includes described methods is <code>repOverlap</code>. Again the output is easily visualised when passed to <code><a href="../reference/vis.html">vis()</a></code> function that does all the work:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">imm_ov1 =<span class="st"> </span><span class="kw"><a href="../reference/repOverlap.html">repOverlap</a></span>(immdata<span class="op">$</span>data, <span class="dt">.method =</span> <span class="st">"public"</span>, <span class="dt">.verbose =</span> F)
imm_ov2 =<span class="st"> </span><span class="kw"><a href="../reference/repOverlap.html">repOverlap</a></span>(immdata<span class="op">$</span>data, <span class="dt">.method =</span> <span class="st">"morisita"</span>, <span class="dt">.verbose =</span> F)

<span class="kw">grid.arrange</span>(<span class="kw"><a href="../reference/vis.html">vis</a></span>(imm_ov1), <span class="kw"><a href="../reference/vis.html">vis</a></span>(imm_ov2, <span class="dt">.text.size=</span><span class="fl">1.5</span>), <span class="dt">ncol =</span> <span class="dv">2</span>)</code></pre></div>
<p><img src="4_overlap_files/figure-html/overlap-1.png" width="1152" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/vis.html">vis</a></span>(imm_ov1, <span class="st">"heatmap2"</span>)</code></pre></div>
<p><img src="4_overlap_files/figure-html/overlap-2.png" width="1152" style="display: block; margin: auto;"></p>
<!---
Top-overlap [Work in Progress]

```r
warning("TODO")
```

```
## Warning: TODO
```
-->
<p>To analyse the computed overlap measures function apply <code>repOverlapAnalysis</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Apply different analysis algorithms to the matrix of public clonotypes:</span>
<span class="co"># "mds" - Multi-dimensional Scaling</span>
<span class="kw"><a href="../reference/repOverlapAnalysis.html">repOverlapAnalysis</a></span>(imm_ov1, <span class="st">"mds"</span>)</code></pre></div>
<pre><code>## Standard deviations:
## [1] 0 0 0 0
## 
## Rotation:
##                [,1]       [,2]
## A2-i129   1.8651791   7.771695
## A2-i131  -2.0246183  -8.435687
## A2-i133  -4.6152434  12.017445
## A2-i132  15.6362099   2.186638
## A4-i191  -9.5429713   4.745392
## A4-i192   5.8953987 -10.163369
## MS1      -5.5601454  -1.810935
## MS2       0.3411929  -1.954662
## MS3       5.1691898   1.792923
## MS4      -2.4522675   3.000137
## MS5       5.6609794  -2.721500
## MS6     -10.3729039  -6.428076</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># "tsne" - t-Stochastic Neighbor Embedding</span>
<span class="kw"><a href="../reference/repOverlapAnalysis.html">repOverlapAnalysis</a></span>(imm_ov1, <span class="st">"tsne"</span>)</code></pre></div>
<pre><code>##                 DimI     DimII
## A2-i129  44.52741936 107.89380
## A2-i131  -0.01203784 -43.56757
## A2-i133  16.11936882 -15.31748
## A2-i132  20.56026375 -69.81114
## A4-i191 -70.40314643 -36.32286
## A4-i192  21.85831350  56.46504
## MS1     -58.83288112 -53.90767
## MS2      14.29439695 -41.33225
## MS3      23.78253956 101.73544
## MS4     -60.82217424 -71.23519
## MS5      10.53004891 106.01326
## MS6      38.39788876 -40.61336
## attr(,"class")
## [1] "matrix"      "immunr_tsne"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Visualise the results</span>
<span class="kw"><a href="../reference/vis.html">vis</a></span>(<span class="kw"><a href="../reference/repOverlapAnalysis.html">repOverlapAnalysis</a></span>(imm_ov1, <span class="st">"mds"</span>))</code></pre></div>
<p><img src="4_overlap_files/figure-html/overlap-1-1.png" width="768" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Apply different analysis algorithms to the matrix of public clonotypes:</span>
<span class="co"># "mds" - Multi-dimensional Scaling</span>
<span class="kw"><a href="../reference/repOverlapAnalysis.html">repOverlapAnalysis</a></span>(imm_ov1, <span class="st">"mds"</span>)</code></pre></div>
<pre><code>## Standard deviations:
## [1] 0 0 0 0
## 
## Rotation:
##                [,1]       [,2]
## A2-i129   1.8651791   7.771695
## A2-i131  -2.0246183  -8.435687
## A2-i133  -4.6152434  12.017445
## A2-i132  15.6362099   2.186638
## A4-i191  -9.5429713   4.745392
## A4-i192   5.8953987 -10.163369
## MS1      -5.5601454  -1.810935
## MS2       0.3411929  -1.954662
## MS3       5.1691898   1.792923
## MS4      -2.4522675   3.000137
## MS5       5.6609794  -2.721500
## MS6     -10.3729039  -6.428076</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># "tsne" - t-Stochastic Neighbor Embedding</span>
<span class="kw"><a href="../reference/repOverlapAnalysis.html">repOverlapAnalysis</a></span>(imm_ov1, <span class="st">"tsne"</span>)</code></pre></div>
<pre><code>##               DimI     DimII
## A2-i129 -220.91065 -90.53168
## A2-i131   80.06455 -80.74273
## A2-i133   42.87002  30.27545
## A2-i132   35.42945 -33.78564
## A4-i191  115.05325 133.46523
## A4-i192 -112.36657 -39.91310
## MS1      135.68849  93.47148
## MS2       65.99152 -16.41437
## MS3     -207.97902 -48.18723
## MS4      169.53829  87.66579
## MS5     -216.78872 -20.46095
## MS6      113.40939 -14.84226
## attr(,"class")
## [1] "matrix"      "immunr_tsne"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Visualise the results</span>
<span class="kw"><a href="../reference/vis.html">vis</a></span>(<span class="kw"><a href="../reference/repOverlapAnalysis.html">repOverlapAnalysis</a></span>(imm_ov1, <span class="st">"mds"</span>))</code></pre></div>
<p><img src="4_overlap_files/figure-html/overlap-2-1.png" width="960" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Clusterise the MDS resulting components using K-means</span>
<span class="kw"><a href="../reference/vis.html">vis</a></span>(<span class="kw"><a href="../reference/repOverlapAnalysis.html">repOverlapAnalysis</a></span>(imm_ov1, <span class="st">"mds+kmeans"</span>))</code></pre></div>
<p><img src="4_overlap_files/figure-html/overlap-2-2.png" width="960" style="display: block; margin: auto;"></p>
</div>
<div id="public-repertoire" class="section level1">
<h1 class="hasAnchor">
<a href="#public-repertoire" class="anchor"></a>Public repertoire</h1>
<p>In order to build a massive table with all clonotypes from the list of repertoires use the <code>pubRep</code> function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Pass "nt" as the second parameter to build the public repertoire table using CDR3 nucleotide sequences</span>
pr.nt =<span class="st"> </span><span class="kw">pubRep</span>(immdata<span class="op">$</span>data, <span class="st">"nt"</span>, <span class="dt">.verbose =</span> F)
pr.nt</code></pre></div>
<pre><code>## Source: local data table [88,528 x 14]
## 
## # A tibble: 88,528 x 14
##    CDR3.nt Samples `A2-i129` `A2-i131` `A2-i133` `A2-i132` `A4-i191`
##    &lt;chr&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;
##  1 TGCGCC…       8         2        NA         2         1        NA
##  2 TGTGCC…       7        NA         1         1         2         1
##  3 TGTGCC…       7         1         1         1        NA         1
##  4 TGCGCC…       6         2        NA         1         1        NA
##  5 TGTGCC…       6         5         3        NA         2         3
##  6 TGTGCC…       6        NA         1         1         5         1
##  7 TGTGCC…       6        NA        NA        NA         1         1
##  8 TGTGCC…       6         1         1         2        NA         5
##  9 TGCAGT…       5         1         3        NA         2        NA
## 10 TGCGCC…       5         1        NA        NA        NA        NA
## # … with 88,518 more rows, and 7 more variables: `A4-i192` &lt;dbl&gt;,
## #   MS1 &lt;dbl&gt;, MS2 &lt;dbl&gt;, MS3 &lt;dbl&gt;, MS4 &lt;dbl&gt;, MS5 &lt;dbl&gt;, MS6 &lt;dbl&gt;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Pass "aa+v" as the second parameter to build the public repertoire table using CDR3 aminoacid sequences and V alleles</span>
<span class="co"># In order to use only CDR3 aminoacid sequences, just pass "aa"</span>
pr.aav =<span class="st"> </span><span class="kw">pubRep</span>(immdata<span class="op">$</span>data, <span class="st">"aa+v"</span>, <span class="dt">.verbose =</span> F)
pr.aav</code></pre></div>
<pre><code>## Source: local data table [87,730 x 15]
## 
## # A tibble: 87,730 x 15
##    CDR3.aa V.name Samples `A2-i129` `A2-i131` `A2-i133` `A2-i132` `A4-i191`
##    &lt;chr&gt;   &lt;chr&gt;    &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;
##  1 CASSLE… TRBV5…       8         2        NA         3         1        NA
##  2 CASSDS… TRBV6…       6         1         1         2        NA         5
##  3 CASSDS… TRBV6…       6        NA        NA        NA         4         1
##  4 CASSFQ… TRBV5…       6         3        NA         1         1        NA
##  5 CASSLG… TRBV1…       6         2        NA        NA         4         3
##  6 CASSLQ… TRBV1…       6        NA         1         1         4        NA
##  7 CAS~FF  TRBV1…       6        NA         2         1         2        NA
##  8 CSARLA… TRBV2…       6         1         3        NA         2         1
##  9 CASSDS… TRBV6…       5        NA        NA        NA         3        NA
## 10 CASSFG… TRBV1…       5        NA         1        NA         5        NA
## # … with 87,720 more rows, and 7 more variables: `A4-i192` &lt;dbl&gt;,
## #   MS1 &lt;dbl&gt;, MS2 &lt;dbl&gt;, MS3 &lt;dbl&gt;, MS4 &lt;dbl&gt;, MS5 &lt;dbl&gt;, MS6 &lt;dbl&gt;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># You can also pass the ".coding" parameter to filter out all noncoding sequences first:</span>
pr.aav.cod =<span class="st"> </span><span class="kw">pubRep</span>(immdata<span class="op">$</span>data, <span class="st">"aa+v"</span>, <span class="dt">.coding=</span>T)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Create a public repertoire with coding-only sequences using both CDR3 amino acid sequences and V genes</span>
pr =<span class="st"> </span><span class="kw">pubRep</span>(immdata<span class="op">$</span>data, <span class="st">"aa+v"</span>, <span class="dt">.coding =</span> T, <span class="dt">.verbose =</span> F)

<span class="co"># Apply the filter subroutine to leave clonotypes presented only in healthy individuals</span>
pr1 =<span class="st"> </span><span class="kw">pubRepFilter</span>(pr, immdata<span class="op">$</span>meta, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dt">Status =</span> <span class="st">"C"</span>))

<span class="co"># Apply the filter subroutine to leave clonotypes presented only in diseased individuals</span>
pr2 =<span class="st"> </span><span class="kw">pubRepFilter</span>(pr, immdata<span class="op">$</span>meta, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dt">Status =</span> <span class="st">"MS"</span>))

<span class="co"># Divide one by another</span>
pr3 =<span class="st"> </span><span class="kw">pubRepApply</span>(pr1, pr2)

<span class="co"># Plot it</span>
p =<span class="st"> </span><span class="kw">ggplot</span>() <span class="op">+</span><span class="st"> </span><span class="kw">geom_jitter</span>(<span class="kw">aes</span>(<span class="dt">x =</span> <span class="st">"Treatment"</span>, <span class="dt">y =</span> Result), <span class="dt">data=</span>pr3)
p</code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#repertoire-overlap">Repertoire overlap</a></li>
      <li><a href="#public-repertoire">Public repertoire</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by  ImmunoMind.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
