<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Gene usage analysis • immunarch</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Gene usage analysis">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">immunarch</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/1_introduction.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/1_introduction.html">Introduction to `immunarch`</a>
    </li>
    <li>
      <a href="../articles/2_data.html">Working with data in `immunarch`</a>
    </li>
    <li>
      <a href="../articles/3_basic_analysis.html">Basic analysis and clonality</a>
    </li>
    <li>
      <a href="../articles/4_overlap.html">Repertoire overlap and public clonotypes</a>
    </li>
    <li>
      <a href="../articles/5_gene_usage.html">Gene usage analysis</a>
    </li>
    <li>
      <a href="../articles/6_diversity.html">Diversity estimation</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://immunomind.io">ImmunoMind</a>
</li>
<li>
  <a href="https://github.com/immunomind/immunarch">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Gene usage analysis</h1>
                        <h4 class="author"><a href="https://immunomind.io">ImmunoMind</a></h4>
            
            <h4 class="date"><a href="mailto:support@immunomind.io">support@immunomind.io</a></h4>
      
      
      <div class="hidden name"><code>5_gene_usage.Rmd</code></div>

    </div>

    
    
<!--
%\VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{Introduction to immunarch}
%\VignettePackage{immunarch}
-->
<div id="gene-usage-computation" class="section level1">
<h1 class="hasAnchor">
<a href="#gene-usage-computation" class="anchor"></a>Gene usage computation</h1>
<p><code>immunarch</code> comes with a gene segments data table containing known gene segments for several species. In order to get the current statistics of genes, call the <code><a href="../reference/gene_stats.html">gene_stats()</a></code> function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/gene_stats.html">gene_stats</a></span>()</code></pre></div>
<pre><code>##     alias                 species ighd ighj ighv igij igkj igkv iglj iglv
## 1      bt               BosTaurus    0    0    0    0    0    0    0    0
## 2      cd      CamelusDromedarius    0    0    0    0    0    0    0    0
## 3     clf    CanisLupusFamiliaris    0    0    0    0    0    0    0    0
## 4      dr              DanioRerio    7    7    0    3    0    0    0    0
## 5      hs             HomoSapiens   30   13  224    0    5   64    7   69
## 6  macmul           MacacaMulatta   24    7   19    0    4   83    5    0
## 7     mmc    MusMusculusCastaneus    0    0    0    0    0    4    0    0
## 8     mmd   MusMusculusDomesticus    0    0    0    0    0    2    0    0
## 9  musmus             MusMusculus   27    8  234    0    8  109    3    5
## 10     oa OrnithorhynchusAnatinus    3   10    0    0    0    0    0    0
## 11     oc    OryctolagusCuniculus   10   11   39    0    8   26    2   20
## 12     om      OncorhynchusMykiss    9    7    6    0    0    0    0    0
## 13     rn        RattusNorvegicus   30    4  113    0    6  132    2    8
## 14   smth   MusMusculusMolossinus    0    0    0    0    0    1    0    0
## 15   smth     MusMusculusMusculus    0    0    0    0    0    1    0    0
## 16   smth              MusSpretus    0    0    0    0    0    2    0    2
## 17     ss               SusScrofa    5    5   15    0    8   19    4   14
##    traj trav trbd trbj trbv trdd trdj trdv trgj trgv vprv
## 1    46    0    0    0    0    5    3    2    6   14    0
## 2     0    0    0    0    0    0    0    7    2    2    0
## 3     0    0    2    8   19    0    0    0    7    8    0
## 4     0    0    0    0    0    0    0    0    0    0    0
## 5    52   47    3   14   60    3    4    6    3    9    1
## 6     0    0    2   15   58    0    0    0    0    0    0
## 7     0    0    0    0    0    0    0    0    0    0    0
## 8     0    0    0    0    0    0    0    0    0    0    0
## 9    41  145    2   14   23    2    3    7    0   11    0
## 10    0    0    0    0    0    0    0    0    0    0    0
## 11    0    0    0    0    0    0    0    0    0    0    0
## 12    0    0    1    9    0    0    0    0    0    0    0
## 13    0    0    0    0    0    0    0    0    0    0    0
## 14    0    0    0    0    0    0    0    0    0    0    0
## 15    0    0    0    0    0    0    0    0    0    0    0
## 16    0    0    0    0    0    0    0    0    0    0    0
## 17    0    0    0    0    0    0    0    0    0    0    0</code></pre>
<p>To compute the distributions of genes, <code>immunarch</code> includes the <code>geneUsage</code> function. It receives a repertoire or a list of repertoires as input and genes and species for which you want to get the statistics. E.g., if you plan to use <code>TRBV</code> genes of <code>Homo Sapiens</code>, you need to use the <code>hs.trbv</code> string in the function, where <code>hs</code> comes from the <code>alias</code> column and <code>trbv</code> is the gene name. Of if you plan to use <code>IGHV</code> genes of <code>Mus Musculus</code>, you need to use <code>musmus.ighv</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Next four function calls are equal. "hs" is from the "alias" column.</span>
imm_gu =<span class="st"> </span><span class="kw"><a href="../reference/geneUsage.html">geneUsage</a></span>(immdata<span class="op">$</span>data, <span class="st">"hs.trbv"</span>)
<span class="co"># imm_gu = geneUsage(immdata$data, "HomoSapiens.trbv")</span>
<span class="co"># imm_gu = geneUsage(immdata$data, "hs.TRBV")</span>
<span class="co"># imm_gu = geneUsage(immdata$data, "HomoSapiens.TRBV")</span>

imm_gu</code></pre></div>
<pre><code>## # A tibble: 48 x 13
##    Names `A2-i129` `A2-i131` `A2-i133` `A2-i132` `A4-i191` `A4-i192`   MS1
##    &lt;chr&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt; &lt;dbl&gt;
##  1 TRBV…        54        69         8        34        35        20    29
##  2 TRBV…       264       320       154       128       429       256   217
##  3 TRBV…        29        22        33        20        22        18    16
##  4 TRBV…       208       193       153       181       110       130   107
##  5 TRBV…        10        15         8        30         2         9     5
##  6 TRBV…       682       541       376       495       396       640   470
##  7 TRBV…        46        63         9        46        20        19     8
##  8 TRBV…        50        59        52        87        33        47    45
##  9 TRBV…        79       109        55        80        44        37    25
## 10 TRBV…       105       166       115        98        88        81   125
## # … with 38 more rows, and 5 more variables: MS2 &lt;dbl&gt;, MS3 &lt;dbl&gt;,
## #   MS4 &lt;dbl&gt;, MS5 &lt;dbl&gt;, MS6 &lt;dbl&gt;</code></pre>
<p>Gene distributions could be computed either using counts of individual clonotypes (<code>.quant = "count"</code>) or not using them (<code>.quant = NA</code>).</p>
<p>In order to compute allele-level or family-level distributions, change the <code>.type</code> parameter.</p>
<p>Due to the ambiguity of gene alignments for some clonotypes, <code>geneUsage</code> has the following options to deal with ambiguous data:</p>
<ul>
<li><p><code>.ambig = "exc"</code> - filters out all clonotypes with ambiguous gene alignments.</p></li>
<li><p><code>.ambig = "inc"</code> - includes all possible combinations of ambiguous gene alignments from the data. NOTE: ImmunoSEQ formats use non-standart gene segment names, so it is preferable to use this argument value with ImmunoSEQ formats. This argument is ON by default to ease the gene manipulation. Feel free to change it to <code>"exc"</code> in case of other data formats.</p></li>
<li><p><code>.ambig = "wei"</code> - introduces weighted approach (divides by n (<code>1/n</code>) the frequency for each entry of the corresponding gene if there are <code>n</code> genes for a clonotype).</p></li>
<li><p><code>.ambig = "maj"</code> - chooses only the first gene segment.</p></li>
</ul>
<p>Parameter <code>.norm</code> controls whether <code>immunarch</code> will normalise the data to ensure the sum of all frequencies to be equal 1 or not.</p>
<p>You can visualise the histogram of gene usage in two different ways:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">imm_gu =<span class="st"> </span><span class="kw"><a href="../reference/geneUsage.html">geneUsage</a></span>(immdata<span class="op">$</span>data, <span class="st">"hs.trbv"</span>, <span class="dt">.norm =</span> T, <span class="dt">.ambig =</span> <span class="st">"exc"</span>)

<span class="kw"><a href="../reference/vis.html">vis</a></span>(imm_gu, <span class="dt">.plot =</span> <span class="st">"hist"</span>, <span class="dt">.grid =</span> T)</code></pre></div>
<pre><code>## Warning: Connection is garbage-collected, use dbDisconnect() to avoid this.</code></pre>
<pre><code>## Warning: Removed 1 rows containing missing values (position_stack).

## Warning: Removed 1 rows containing missing values (position_stack).</code></pre>
<p><img src="5_gene_usage_files/figure-html/unnamed-chunk-2-1.png" width="1440" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/vis.html">vis</a></span>(imm_gu, <span class="dt">.plot =</span> <span class="st">"hist"</span>, <span class="dt">.grid =</span> F, <span class="dt">.by =</span> <span class="st">"Status"</span>, <span class="dt">.meta =</span> immdata<span class="op">$</span>meta)</code></pre></div>
<p><img src="5_gene_usage_files/figure-html/unnamed-chunk-3-1.png" width="1152" style="display: block; margin: auto;"></p>
<p>Another practical approach to the visualisation of group distributions are box plots:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/vis.html">vis</a></span>(imm_gu, <span class="dt">.by =</span> <span class="st">"Status"</span>, <span class="dt">.meta =</span> immdata<span class="op">$</span>meta, <span class="dt">.plot =</span> <span class="st">"box"</span>)</code></pre></div>
<p><img src="5_gene_usage_files/figure-html/unnamed-chunk-4-1.png" width="1152" style="display: block; margin: auto;"></p>
<p>Sometimes tree maps could be used to reveal the differences in repertoires. They display the overall picture and the comparisons of related items, both at the same time, enabling intelligible exploration of the details:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">imm_gu =<span class="st"> </span><span class="kw"><a href="../reference/geneUsage.html">geneUsage</a></span>(immdata<span class="op">$</span>data, <span class="st">"hs.trbv"</span>, <span class="dt">.norm =</span> T, <span class="dt">.ambig =</span> <span class="st">"exc"</span>)

<span class="kw"><a href="../reference/vis.html">vis</a></span>(imm_gu, <span class="dt">.plot =</span> <span class="st">"tree"</span>)</code></pre></div>
<p><img src="5_gene_usage_files/figure-html/unnamed-chunk-5-1.png" width="1440" style="display: block; margin: auto;"></p>
</div>
<div id="gene-usage-analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#gene-usage-analysis" class="anchor"></a>Gene usage analysis</h1>
<p>To analyse the gene usage <code>immunarch</code> introduces the <code>geneUsageAnalysis</code> function. The <code>.method</code> parameter controls how the data is going to be preprocessed and analysed. <code>geneUsageAnalysis</code> includes following methods for preprocessing:</p>
<ul>
<li><p>“js” - Jensen-Shannon Divergence.</p></li>
<li><p>“cor” - correlation.</p></li>
<li><p>“cosine” - cosine similarity.</p></li>
<li><p>“pca” - principal component analysis.</p></li>
<li><p>“mds” - multi-dimensional scaling.</p></li>
<li><p>“tsne” - t-Distributed Stochastic Neighbor Embedding.</p></li>
</ul>
<p>And a few methods for the actual analysis:</p>
<ul>
<li><p>“hclust” - clusters the data using hierarchical clustering.</p></li>
<li><p>“kmeans” - clusters the data using K-means.</p></li>
<li><p>“dbscan” - clusters the data using DBSCAN.</p></li>
<li><p>“permut” - permutation testing of differences between groups.</p></li>
<li><p>“anova” - computes ANOVA for each gene separately on data splitted to groups (without preprocessing). Results could be used with Tukey post-hoc test in order to detect significant differences among groups per gene.</p></li>
<li><p>“kruskall” - compute Kruskall for each gene separately on data splitted to groups (without preprocessing). Results could be used with Dunn test in order to detect significant differences between groups.</p></li>
</ul>
<p>You can call several methods in a single line of code, which is probably the most powerful feature of the package. For instance, <code>"js+hclust"</code> first computes Jensen-Shannon divergence and then applies hierarchical clustering on the resulting distance matrix, whereas <code>"anova"</code> computes ANOVA on each gene separately after repertoires have been grouped:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">imm_gu =<span class="st"> </span><span class="kw"><a href="../reference/geneUsage.html">geneUsage</a></span>(immdata<span class="op">$</span>data, <span class="st">"hs.trbv"</span>, <span class="dt">.norm =</span> T, <span class="dt">.ambig =</span> <span class="st">"exc"</span>)

imm_gu_js =<span class="st"> </span><span class="kw"><a href="../reference/geneUsageAnalysis.html">geneUsageAnalysis</a></span>(imm_gu, <span class="dt">.method =</span> <span class="st">"js"</span>, <span class="dt">.verbose =</span> F)
imm_gu_cor =<span class="st"> </span><span class="kw"><a href="../reference/geneUsageAnalysis.html">geneUsageAnalysis</a></span>(imm_gu, <span class="dt">.method =</span> <span class="st">"cor"</span>, <span class="dt">.verbose =</span> F)

gridExtra<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/gridExtra/topics/arrangeGrob">grid.arrange</a></span>(<span class="kw"><a href="../reference/vis.html">vis</a></span>(imm_gu_js, <span class="dt">.title =</span> <span class="st">"Gene usage JS-divergence"</span>, <span class="dt">.leg.title =</span> <span class="st">"JS"</span>, <span class="dt">.text.size=</span><span class="fl">1.5</span>), <span class="kw"><a href="../reference/vis.html">vis</a></span>(imm_gu_cor, <span class="dt">.title =</span> <span class="st">"Gene usage correlation"</span>, <span class="dt">.leg.title =</span> <span class="st">"Cor"</span>, <span class="dt">.text.size=</span><span class="fl">1.5</span>), <span class="dt">ncol =</span> <span class="dv">2</span>)</code></pre></div>
<p><img src="5_gene_usage_files/figure-html/unnamed-chunk-6-1.png" width="1152" style="display: block; margin: auto;"></p>
<p>Now let us visualise the output after both preprocessing and analysis:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">imm_gu_js[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/NA">is.na</a></span>(imm_gu_js)] =<span class="st"> </span><span class="dv">0</span>

<span class="kw"><a href="../reference/vis.html">vis</a></span>(<span class="kw"><a href="../reference/geneUsageAnalysis.html">geneUsageAnalysis</a></span>(imm_gu, <span class="st">"cosine+hclust"</span>, <span class="dt">.verbose =</span> F))</code></pre></div>
<p><img src="5_gene_usage_files/figure-html/unnamed-chunk-7-1.png" width="960" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#vis(geneUsageAnalysis(imm_gu, "js+dbscan", .verbose = F))</span></code></pre></div>
<p>On top of that you can add clustering:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">imm_cl_pca =<span class="st"> </span><span class="kw"><a href="../reference/geneUsageAnalysis.html">geneUsageAnalysis</a></span>(imm_gu, <span class="st">"js+pca+kmeans"</span>, <span class="dt">.verbose =</span> F)
imm_cl_mds =<span class="st"> </span><span class="kw"><a href="../reference/geneUsageAnalysis.html">geneUsageAnalysis</a></span>(imm_gu, <span class="st">"js+mds+kmeans"</span>, <span class="dt">.verbose =</span> F)
imm_cl_tsne =<span class="st"> </span><span class="kw"><a href="../reference/geneUsageAnalysis.html">geneUsageAnalysis</a></span>(imm_gu, <span class="st">"js+tsne+kmeans"</span>, <span class="dt">.perp =</span> .<span class="dv">01</span>, <span class="dt">.verbose =</span> F)</code></pre></div>
<pre><code>## Perplexity should be lower than K!</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">grid.arrange</span>(<span class="kw"><a href="../reference/vis.html">vis</a></span>(imm_cl_pca, <span class="dt">.plot =</span> <span class="st">"clust"</span>), <span class="kw"><a href="../reference/vis.html">vis</a></span>(imm_cl_mds, <span class="dt">.plot =</span> <span class="st">"clust"</span>), <span class="kw"><a href="../reference/vis.html">vis</a></span>(imm_cl_tsne, <span class="dt">.plot =</span> <span class="st">"clust"</span>), <span class="dt">ncol =</span> <span class="dv">3</span>)</code></pre></div>
<p><img src="5_gene_usage_files/figure-html/unnamed-chunk-8-1.png" width="1152" style="display: block; margin: auto;"></p>
</div>
<div id="spectratyping" class="section level1">
<h1 class="hasAnchor">
<a href="#spectratyping" class="anchor"></a>Spectratyping</h1>
<p>Spectratype is a useful way to represent distributions of genes per length. Parameter <code>.quant</code> controls the quantity that used to compute proportions of genes - either by clonotype (<code>id</code>) or by number of clones per clonotype (<code>count</code>)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p1 =<span class="st"> </span><span class="kw"><a href="../reference/vis.html">vis</a></span>(<span class="kw"><a href="../reference/spectratype.html">spectratype</a></span>(immdata<span class="op">$</span>data[[<span class="dv">1</span>]], <span class="dt">.quant =</span> <span class="st">"id"</span>, <span class="dt">.col =</span> <span class="st">"aa"</span>, <span class="dt">.gene =</span> <span class="st">"v"</span>))
p2 =<span class="st"> </span><span class="kw"><a href="../reference/vis.html">vis</a></span>(<span class="kw"><a href="../reference/spectratype.html">spectratype</a></span>(immdata<span class="op">$</span>data[[<span class="dv">1</span>]], <span class="dt">.quant =</span> <span class="st">"count"</span>, <span class="dt">.col =</span> <span class="st">"aa"</span>, <span class="dt">.gene =</span> <span class="st">"v"</span>))

<span class="kw">grid.arrange</span>(p1, p2, <span class="dt">ncol =</span> <span class="dv">2</span>)</code></pre></div>
<p><img src="5_gene_usage_files/figure-html/spectr-1.png" width="1152" style="display: block; margin: auto;"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#gene-usage-computation">Gene usage computation</a></li>
      <li><a href="#gene-usage-analysis">Gene usage analysis</a></li>
      <li><a href="#spectratyping">Spectratyping</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by  ImmunoMind.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
