<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Analysis of gene usage — geneUsageAnalysis • immunarch</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>


<!-- docsearch -->
<script src="../docsearch.js"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous" />
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script>



<meta property="og:title" content="Analysis of gene usage — geneUsageAnalysis" />
<meta property="og:description" content="The geneUsageAnalysis function deploys several
data analysis methods, including PCA, multidimensional scaling,
Jensen-Shannon divergence, k-means, hierarchical clustering, DBscan, and different
correlation coefficients." />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-128418614-2"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-128418614-2');
</script>


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">immunarch</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.5.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/v1_introduction.html">Get started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">First steps</li>
    <li>
      <a href="../articles/v1_introduction.html">Introduction to immunarch</a>
    </li>
    <li>
      <a href="../articles/v2_data.html">Working with data in immunarch</a>
    </li>
    <li>
      <a href="../articles/v3_basic_analysis.html">Basic analysis and clonality</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Explore and compare repertoires</li>
    <li>
      <a href="../articles/web_only/v4_overlap.html">Repertoire overlap and public clonotypes</a>
    </li>
    <li>
      <a href="../articles/web_only/v5_gene_usage.html">Gene usage analysis</a>
    </li>
    <li>
      <a href="../articles/web_only/v6_diversity.html">Diversity estimation</a>
    </li>
    <li>
      <a href="../articles/web_only/v11_db.html">Annotate clonotypes using immune receptor databases</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Preparing to publication</li>
    <li>
      <a href="../articles/web_only/v7_fixvis.html">Make your plots publication-ready with fixVis</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Advanced methods</li>
    <li>
      <a href="../articles/web_only/v8_tracking.html">Tracking clonotypes across time points</a>
    </li>
    <li>
      <a href="../articles/web_only/v9_kmers.html">Kmer and sequence motif analysis and visualisation</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="https://github.com/immunomind/immunarch/releases">Changelog</a>
</li>
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://immunomind.io">ImmunoMind</a>
</li>
<li>
  <a href="http://twitter.com/immunomind">
    <span class="fa fa-lg fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/immunomind/immunarch">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
      
      <form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
        </div>
      </form>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Analysis of gene usage</h1>
    <small class="dont-index">Source: <a href='https://github.com/immunomind/immunarch/blob/master/R/gene_usage_analysis.R'><code>R/gene_usage_analysis.R</code></a></small>
    <div class="hidden name"><code>geneUsageAnalysis.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The <code>geneUsageAnalysis</code> function deploys several
data analysis methods, including PCA, multidimensional scaling,
Jensen-Shannon divergence, k-means, hierarchical clustering, DBscan, and different
correlation coefficients.</p>
    </div>

    <pre class="usage"><span class='fu'>geneUsageAnalysis</span>(
  <span class='no'>.data</span>,
  <span class='kw'>.method</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"js+hclust"</span>, <span class='st'>"pca+kmeans"</span>, <span class='st'>"anova"</span>, <span class='st'>"js+pca+kmeans"</span>),
  <span class='kw'>.base</span> <span class='kw'>=</span> <span class='fl'>2</span>,
  <span class='kw'>.norm.entropy</span> <span class='kw'>=</span> <span class='no'>F</span>,
  <span class='kw'>.cor</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"pearson"</span>, <span class='st'>"kendall"</span>, <span class='st'>"spearman"</span>),
  <span class='kw'>.do.norm</span> <span class='kw'>=</span> <span class='no'>T</span>,
  <span class='kw'>.laplace</span> <span class='kw'>=</span> <span class='fl'>1e-12</span>,
  <span class='kw'>.verbose</span> <span class='kw'>=</span> <span class='no'>T</span>,
  <span class='kw'>.k</span> <span class='kw'>=</span> <span class='fl'>2</span>,
  <span class='kw'>.eps</span> <span class='kw'>=</span> <span class='fl'>0.01</span>,
  <span class='kw'>.perp</span> <span class='kw'>=</span> <span class='fl'>1</span>,
  <span class='kw'>.theta</span> <span class='kw'>=</span> <span class='fl'>0.1</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>.data</th>
      <td><p>The <code>geneUsageAnalysis</code> function runs on the output from
<code><a href='geneUsage.html'>geneUsage</a></code>.</p></td>
    </tr>
    <tr>
      <th>.method</th>
      <td><p>A string that defines the type of analysis to perform. Can be "pca",
"mds", "js", "kmeans", "hclust", "dbscan" or "cor" if you want to calculate
correlation coefficient. In the latter case you have to provide <code>.cor</code> argument.</p></td>
    </tr>
    <tr>
      <th>.base</th>
      <td><p>A numerical value that defines the logarithm base for Jensen-Shannon
divergence.</p></td>
    </tr>
    <tr>
      <th>.norm.entropy</th>
      <td><p>A logical value. Set TRUE to normalise your data if you haven't
done it already.</p></td>
    </tr>
    <tr>
      <th>.cor</th>
      <td><p>A string that defines the correlation coefficient for analysis. Can be
"pearson", "kendall" or "spearman".</p></td>
    </tr>
    <tr>
      <th>.do.norm</th>
      <td><p>A logical value. If TRUE it forces Laplace smoothing, if NA it checks
if smoothing is necessary, if FALSE does nothing.</p></td>
    </tr>
    <tr>
      <th>.laplace</th>
      <td><p>The numeric value, which is used as a pseudocount for Laplace
smoothing.</p></td>
    </tr>
    <tr>
      <th>.verbose</th>
      <td><p>A logical value.</p></td>
    </tr>
    <tr>
      <th>.k</th>
      <td><p>The number of clusters to create, passed as <code>k</code> to <a href='https://rdrr.io/pkg/factoextra/man/hcut.html'>hcut</a> or as <code>centers</code> to <a href='https://rdrr.io/r/stats/kmeans.html'>kmeans</a>.</p></td>
    </tr>
    <tr>
      <th>.eps</th>
      <td><p>A numerical value, DBscan epsylon parameter, see
<code><a href='immunr_hclust.html'>immunr_dbscan</a></code>.</p></td>
    </tr>
    <tr>
      <th>.perp</th>
      <td><p>A numerical value, t-SNE perplexity, see <code><a href='immunr_pca.html'>immunr_tsne</a></code>.</p></td>
    </tr>
    <tr>
      <th>.theta</th>
      <td><p>A numerical value, t-SNE theta parameter, see <code><a href='immunr_pca.html'>immunr_tsne</a></code>.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Work in progress.</p>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#value">Value</a></li>
    </ul>

  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Vadim I. Nazarov, Vasily O. Tsvetkov, Eugene Rumynskiy, Anna Lorenc, Daniel J. Moore, Victor Greiff, <a href='https://immunomind.io'>ImmunoMind</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script>
<script>
  docsearch({
    
    
    apiKey: '94207562f34b455c0796790d104e3549',
    indexName: 'immunarch',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>



  </body>
</html>


