<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ immunarch</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Function reference" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-128418614-2"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-128418614-2');
</script>

  </head>

  <body>
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">immunarch</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/1_introduction.html">Get started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/1_introduction.html">Introduction to `immunarch`</a>
    </li>
    <li>
      <a href="../articles/2_data.html">Working with data in `immunarch`</a>
    </li>
    <li>
      <a href="../articles/3_basic_analysis.html">Basic analysis and clonality</a>
    </li>
    <li>
      <a href="../articles/4_overlap.html">Repertoire overlap and public clonotypes</a>
    </li>
    <li>
      <a href="../articles/5_gene_usage.html">Gene usage analysis</a>
    </li>
    <li>
      <a href="../articles/6_diversity.html">Diversity estimation</a>
    </li>
    <li>
      <a href="../articles/7_fixvis.html">Make your plots publication-ready with fixVis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://immunomind.io">ImmunoMind</a>
</li>
<li>
  <a href="http://twitter.com/immunomind">
    <span class="fa fa-lg fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/immunomind/immunarch">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="aa_properties.html">aa_properties</a></code> </p>
        </td>
        <td><p>Tables with amino acid properties</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="aa_table.html">AA_TABLE</a></code> </p>
        </td>
        <td><p>Amino acid / codon table</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="add_class.html">add_class()</a></code> </p>
        </td>
        <td><p>Add a new class attribute to the input object.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="apply_symm.html">apply.symm()</a></code> <code><a href="apply_symm.html">apply.asymm()</a></code> </p>
        </td>
        <td><p>Apply function to each pair of data frames from a list.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="bunch_translate.html">bunch_translate()</a></code> </p>
        </td>
        <td><p>Translate</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cdrProp.html">cdrProp()</a></code> </p>
        </td>
        <td><p>AA properties calculation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="check_distribution.html">check_distribution()</a></code> </p>
        </td>
        <td><p>Check that distribution is sums up to one.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="check_group_names.html">check_group_names()</a></code> </p>
        </td>
        <td><p>Check if everything is ok with group names</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="coding.html">coding()</a></code> </p>
        </td>
        <td><p>Work in progress</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="dot-make_names.html">.make_names()</a></code> </p>
        </td>
        <td><p>Additional functions for help in parsing</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="entropy.html">entropy()</a></code> </p>
        </td>
        <td><p>Information measures</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fixVis.html">fixVis()</a></code> </p>
        </td>
        <td><p>Make publication-ready plots</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="geneUsage.html">geneUsage()</a></code> </p>
        </td>
        <td><p>Gene usage</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="geneUsageAnalysis.html">geneUsageAnalysis()</a></code> </p>
        </td>
        <td><p>Analysis of gene usage</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gene_segments.html">gene_segments</a></code> </p>
        </td>
        <td><p>Gene segments table.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gene_stats.html">gene_stats()</a></code> </p>
        </td>
        <td><p>WIP</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getKmers.html">getKmers()</a></code> </p>
        </td>
        <td><p>Kmer analysis of repertoires.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_rank.html">get_rank()</a></code> </p>
        </td>
        <td><p>Get ranks / indices for counts of clonotypes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gibbs_sampling.html">gibbs_sampling()</a></code> </p>
        </td>
        <td><p>Simple Gibbs sampling on kmers.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="group_from_metadata.html">group_from_metadata()</a></code> </p>
        </td>
        <td><p>Get a character vector of samples' groups from the input metadata file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="has_class.html">has_class()</a></code> </p>
        </td>
        <td><p>Check class</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="immunr_data_format.html">immunr_data_format</a></code> </p>
        </td>
        <td><p>Specification of the data format used by immunarch dataframes.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="immunr_hclust.html">immunr_hclust()</a></code> </p>
        </td>
        <td><p>Clustering the distance matrix data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="immunr_pca.html">immunr_pca()</a></code> </p>
        </td>
        <td><p>Preprocessing the distance matrix data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="immunr_permut.html">immunr_permut()</a></code> </p>
        </td>
        <td><p>Statistical analysis of groups</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="kmerAnalysis.html">kmerAnalysis()</a></code> </p>
        </td>
        <td><p>WIP</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="load_test_data.html">load_test_data()</a></code> </p>
        </td>
        <td><p>Various functions for testing purposes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="lookup_clonotypes.html">lookup_clonotypes()</a></code> </p>
        </td>
        <td><p>Search for clonotypes in external databases</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="num_shared_clonotypes.html">num_shared_clonotypes()</a></code> </p>
        </td>
        <td><p>Number of shared clonotypes.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="overlap_coef.html">overlap_coef()</a></code> <code><a href="overlap_coef.html">jaccard_index()</a></code> <code><a href="overlap_coef.html">tversky_index()</a></code> <code><a href="overlap_coef.html">cosine_sim()</a></code> <code><a href="overlap_coef.html">morisita_index()</a></code> <code><a href="overlap_coef.html">horn_index()</a></code> </p>
        </td>
        <td><p>Various measures for overlapping sets of objects.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="repClonality.html">repClonality()</a></code> </p>
        </td>
        <td><p>Assess clonal proportions of repertoire</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="repDiversity.html">repDiversity()</a></code> </p>
        </td>
        <td><p>Estimate diversity of repertoire</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="repExplore.html">repExplore()</a></code> </p>
        </td>
        <td><p>Explore basics of repertoire</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="repLoad.html">repLoad()</a></code> </p>
        </td>
        <td><p>Load repertoire files to the workspace</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="repOverlap.html">repOverlap()</a></code> </p>
        </td>
        <td><p>Calculate the overlap between repertoires</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="repOverlapAnalysis.html">repOverlapAnalysis()</a></code> </p>
        </td>
        <td><p>Analysis of overlaps or any distance matrices</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="repSave.html">repSave()</a></code> </p>
        </td>
        <td><p>Save repertoire files to the disk.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="resample_col.html">resample()</a></code> <code><a href="resample_col.html">downsample()</a></code> <code><a href="resample_col.html">prop_sample()</a></code> </p>
        </td>
        <td><p>Resample data frame using values from the column with number of clonesets.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="set_pb.html">set_pb()</a></code> </p>
        </td>
        <td><p>Set and update progress bars</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="spectratype.html">spectratype()</a></code> </p>
        </td>
        <td><p>Analyse spectratype of immune repertoire</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="split_immunoseq_files.html">split_immunoseq_files()</a></code> </p>
        </td>
        <td><p>Split all-samples-in-one ImmunoSEQ files by sample</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="split_to_kmers.html">split_to_kmers()</a></code> </p>
        </td>
        <td><p>Kmers processing.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="switch_type.html">switch_type()</a></code> </p>
        </td>
        <td><p>WIP</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="top.html">top()</a></code> </p>
        </td>
        <td><p>Work in progress</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="top_overlap.html">top_overlap()</a></code> </p>
        </td>
        <td><p>Sequential overlap of top-by-count clonotypes.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="trackClonotypes.html">trackClonotypes()</a></code> <code><a href="trackClonotypes.html">tr.clo()</a></code> <code><a href="trackClonotypes.html">poisson_model()</a></code> <code><a href="trackClonotypes.html">norm_model()</a></code> </p>
        </td>
        <td><p>Clonotype tracking across different samples.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis.html">vis</a></code> </p>
        </td>
        <td><p>One function to visualise them all</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis.immunr_cdr_prop.html">vis.immunr_cdr_prop</a></code> </p>
        </td>
        <td><p>Work in progress</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis.immunr_chao1.html">vis.immunr_chao1</a></code> </p>
        </td>
        <td><p>Visualise diversity.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis.immunr_clonal_prop.html">vis.immunr_clonal_prop</a></code> </p>
        </td>
        <td><p>Visualise clonality</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis.immunr_dynamics.html">vis.immunr_dynamics</a></code> </p>
        </td>
        <td><p>Repertoire dynamics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis.immunr_exp_vol.html">vis.immunr_exp_vol</a></code> </p>
        </td>
        <td><p>Visualise exploratory analysis.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis.immunr_gene_usage.html">vis.immunr_gene_usage</a></code> </p>
        </td>
        <td><p>Histograms and boxplots (general case / gene usage)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis.immunr_hclust.html">vis.immunr_hclust()</a></code> <code><a href="vis.immunr_hclust.html">vis.immunr_kmeans()</a></code> <code><a href="vis.immunr_hclust.html">vis.immunr_dbscan()</a></code> </p>
        </td>
        <td><p>Visualisation of clustering results</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis.immunr_kmer_table.html">vis.immunr_kmer_table</a></code> </p>
        </td>
        <td><p>Most frequent kmers visualisation.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis.immunr_mds.html">vis.immunr_mds</a></code> </p>
        </td>
        <td><p>PCA / MDS / tSNE visualisation (mainly overlap / gene usage)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis.immunr_mutation_network.html">vis.immunr_mutation_network</a></code> </p>
        </td>
        <td><p>Work in progress</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis.immunr_ov_matrix.html">vis.immunr_ov_matrix</a></code> </p>
        </td>
        <td><p>Overlap / gene usage distance visualisation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis.immunr_public_repertoire.html">vis.immunr_public_repertoire</a></code> </p>
        </td>
        <td><p>Public repertoire visualisation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis.immunr_top_overlap.html">vis.immunr_top_overlap</a></code> </p>
        </td>
        <td><p>Visualise incremental overlaps</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_box.html">vis_box</a></code> </p>
        </td>
        <td><p>Visualisation of distributions using boxplots</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_circos.html">vis_circos</a></code> </p>
        </td>
        <td><p>Visualisation of matrices using circos plots</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_heatmap.html">vis_heatmap</a></code> </p>
        </td>
        <td><p>Visualisation of matrices and data frames using ggplot2-based heatmaps</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_heatmap2.html">vis_heatmap2</a></code> </p>
        </td>
        <td><p>Visualisation of matrices using heatmap3-based heatmaps</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_hist.html">vis_hist</a></code> </p>
        </td>
        <td><p>Visualisation of distributions using histograms</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_logo.html">vis_logo</a></code> </p>
        </td>
        <td><p>Logo - plots for amino acid and nucletide profiles.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_radar.html">vis_radar</a></code> </p>
        </td>
        <td><p>Visualisation of distance matrices with polar area plots</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_treemap.html">vis_treemap</a></code> </p>
        </td>
        <td><p>Visualisation of data frames and matrices using treemaps</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#section-all-functions">All functions</a></li>
    </ul>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by  ImmunoMind.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

