<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ immunarch</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>


<!-- docsearch -->
<script src="../docsearch.js"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous" />
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script>



<meta property="og:title" content="Function reference" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-128418614-2"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-128418614-2');
</script>


  </head>

  <body>
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">immunarch</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.5.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/v1_introduction.html">Get started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">First steps</li>
    <li>
      <a href="../articles/v1_introduction.html">Introduction to immunarch</a>
    </li>
    <li>
      <a href="../articles/v2_data.html">Working with data in immunarch</a>
    </li>
    <li>
      <a href="../articles/v3_basic_analysis.html">Basic analysis and clonality</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Explore and compare repertoires</li>
    <li>
      <a href="../articles/web_only/v4_overlap.html">Repertoire overlap and public clonotypes</a>
    </li>
    <li>
      <a href="../articles/web_only/v5_gene_usage.html">Gene usage analysis</a>
    </li>
    <li>
      <a href="../articles/web_only/v6_diversity.html">Diversity estimation</a>
    </li>
    <li>
      <a href="../articles/web_only/v11_db.html">Annotate clonotypes using immune receptor databases</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Preparing to publication</li>
    <li>
      <a href="../articles/web_only/v7_fixvis.html">Make your plots publication-ready with fixVis</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Advanced methods</li>
    <li>
      <a href="../articles/web_only/v8_tracking.html">Tracking clonotypes across time points</a>
    </li>
    <li>
      <a href="../articles/web_only/v9_kmers.html">Kmer and sequence motif analysis and visualisation</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="https://github.com/immunomind/immunarch/releases">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://immunomind.io">ImmunoMind</a>
</li>
<li>
  <a href="http://twitter.com/immunomind">
    <span class="fa fa-lg fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/immunomind/immunarch">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
      
      <form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
        </div>
      </form>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="aa_properties.html">aa_properties</a></code> </p>
        </td>
        <td><p>Tables with amino acid properties</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="aa_table.html">AA_TABLE</a></code> </p>
        </td>
        <td><p>Amino acid / codon table</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="add_class.html">add_class()</a></code> </p>
        </td>
        <td><p>Add a new class attribute to the input object.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="apply_symm.html">apply_symm()</a></code> <code><a href="apply_symm.html">apply_asymm()</a></code> </p>
        </td>
        <td><p>Apply function to each pair of data frames from a list.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="bunch_translate.html">bunch_translate()</a></code> </p>
        </td>
        <td><p>Nucleotide to amino acid sequence translation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="check_distribution.html">check_distribution()</a></code> </p>
        </td>
        <td><p>Check that distribution is sums up to one.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="coding.html">coding()</a></code> <code><a href="coding.html">noncoding()</a></code> <code><a href="coding.html">inframes()</a></code> <code><a href="coding.html">outofframes()</a></code> </p>
        </td>
        <td><p>Working with coding and non-coding clonotype sequences</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="dbAnnotate.html">dbAnnotate()</a></code> </p>
        </td>
        <td><p>Annotate clonotypes using databases</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="dbLoad.html">dbLoad()</a></code> </p>
        </td>
        <td><p>Load external immune receptor databases into R</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="dot-quant_column_choice.html">.quant_column_choice()</a></code> </p>
        </td>
        <td><p>Get a column's name using the input alias</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="entropy.html">entropy()</a></code> <code><a href="entropy.html">kl_div()</a></code> <code><a href="entropy.html">js_div()</a></code> <code><a href="entropy.html">cross_entropy()</a></code> </p>
        </td>
        <td><p>Information measures</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filter_barcodes.html">filter_barcodes()</a></code> </p>
        </td>
        <td><p>Filter data by barcodes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fixVis.html">fixVis()</a></code> </p>
        </td>
        <td><p>Make publication-ready plots</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="geneUsage.html">geneUsage()</a></code> </p>
        </td>
        <td><p>Gene usage</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="geneUsageAnalysis.html">geneUsageAnalysis()</a></code> </p>
        </td>
        <td><p>Analysis of gene usage</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gene_segments.html">gene_segments</a></code> </p>
        </td>
        <td><p>Gene segments table.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gene_stats.html">gene_stats()</a></code> </p>
        </td>
        <td><p>WIP</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getKmers.html">getKmers()</a></code> </p>
        </td>
        <td><p>Kmer analysis of repertoires.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="group_from_metadata.html">group_from_metadata()</a></code> </p>
        </td>
        <td><p>Get a character vector of samples' groups from the input metadata file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="has_class.html">has_class()</a></code> </p>
        </td>
        <td><p>Check class</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="immdata.html">immdata</a></code> </p>
        </td>
        <td><p>Immune repertoire dataset</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="immunr_data_format.html">immunr_data_format</a></code> </p>
        </td>
        <td><p>Specification of the data format used by immunarch dataframes.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="immunr_hclust.html">immunr_hclust()</a></code> <code><a href="immunr_hclust.html">immunr_kmeans()</a></code> <code><a href="immunr_hclust.html">immunr_dbscan()</a></code> </p>
        </td>
        <td><p>Clustering of objects or distance matrices</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="immunr_pca.html">immunr_pca()</a></code> <code><a href="immunr_pca.html">immunr_mds()</a></code> <code><a href="immunr_pca.html">immunr_tsne()</a></code> </p>
        </td>
        <td><p>Dimensionality reduction</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="inc_overlap.html">inc_overlap()</a></code> </p>
        </td>
        <td><p>Incremental counting of repertoire similarity</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="matrixdiagcopy.html">matrixdiagcopy()</a></code> </p>
        </td>
        <td><p>Copy the upper matrix triangle to the lower one</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="num_shared_clonotypes.html">num_shared_clonotypes()</a></code> </p>
        </td>
        <td><p>Number of shared clonotypes.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="overlap_coef.html">overlap_coef()</a></code> <code><a href="overlap_coef.html">jaccard_index()</a></code> <code><a href="overlap_coef.html">tversky_index()</a></code> <code><a href="overlap_coef.html">cosine_sim()</a></code> <code><a href="overlap_coef.html">morisita_index()</a></code> <code><a href="overlap_coef.html">horn_index()</a></code> </p>
        </td>
        <td><p>Various measures for overlapping sets of objects.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pubRep.html">pubRep()</a></code> </p>
        </td>
        <td><p>Public repertoire</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pubRepApply.html">pubRepApply()</a></code> </p>
        </td>
        <td><p>Apply transformations to public repertoires</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pubRepFilter.html">pubRepFilter()</a></code> </p>
        </td>
        <td><p>Filter out clonotypes from public repertoires</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pubRepStatistics.html">pubRepStatistics()</a></code> </p>
        </td>
        <td><p>Statistics of number of public clonotypes for each possible combinations of repertoires</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="public_matrix.html">public_matrix()</a></code> </p>
        </td>
        <td><p>Get a matrix with public clonotype frequencies</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="repClonality.html">repClonality()</a></code> </p>
        </td>
        <td><p>Assess clonal proportions of repertoire</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="repDiversity.html">repDiversity()</a></code> </p>
        </td>
        <td><p>Estimate diversity of repertoire</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="repExplore.html">repExplore()</a></code> </p>
        </td>
        <td><p>Explore basics of repertoire</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="repLoad.html">repLoad()</a></code> </p>
        </td>
        <td><p>Load repertoire files to the workspace</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="repOverlap.html">repOverlap()</a></code> </p>
        </td>
        <td><p>Calculate the overlap between repertoires</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="repOverlapAnalysis.html">repOverlapAnalysis()</a></code> </p>
        </td>
        <td><p>Analysis of overlaps or any distance matrices</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="repSample.html">repSample()</a></code> </p>
        </td>
        <td><p>Sampling of repertoires</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="repSave.html">repSave()</a></code> </p>
        </td>
        <td><p>Save repertoire files to the disk.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="set_pb.html">set_pb()</a></code> <code><a href="set_pb.html">add_pb()</a></code> </p>
        </td>
        <td><p>Set and update progress bars</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="spectratype.html">spectratype()</a></code> </p>
        </td>
        <td><p>Analyse spectratype of immune repertoires</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="split_immunoseq_files.html">split_immunoseq_files()</a></code> </p>
        </td>
        <td><p>Split all-samples-in-one ImmunoSEQ files by sample</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="split_to_kmers.html">split_to_kmers()</a></code> <code><a href="split_to_kmers.html">kmer_profile()</a></code> </p>
        </td>
        <td><p>Kmers processing.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="switch_type.html">switch_type()</a></code> <code><a href="switch_type.html">process_col_argument()</a></code> </p>
        </td>
        <td><p>Return a column's name</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="top.html">top()</a></code> </p>
        </td>
        <td><p>Get the N most abundant clonotypes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="trackClonotypes.html">trackClonotypes()</a></code> </p>
        </td>
        <td><p>Clonotype tracking</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis.html">vis()</a></code> </p>
        </td>
        <td><p>One function to visualise them all</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis.immunr_chao1.html">vis(<i>&lt;immunr_chao1&gt;</i>)</a></code> </p>
        </td>
        <td><p>Visualise diversity.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis.immunr_clonal_prop.html">vis(<i>&lt;immunr_clonal_prop&gt;</i>)</a></code> </p>
        </td>
        <td><p>Visualise clonality</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis.immunr_dynamics.html">vis(<i>&lt;immunr_dynamics&gt;</i>)</a></code> </p>
        </td>
        <td><p>Repertoire dynamics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis.immunr_exp_vol.html">vis(<i>&lt;immunr_exp_vol&gt;</i>)</a></code> </p>
        </td>
        <td><p>Visualise exploratory analysis.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis.immunr_gene_usage.html">vis(<i>&lt;immunr_gene_usage&gt;</i>)</a></code> </p>
        </td>
        <td><p>Histograms and boxplots (general case / gene usage)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis.immunr_hclust.html">vis(<i>&lt;immunr_hclust&gt;</i>)</a></code> </p>
        </td>
        <td><p>Visualisation of hierarchical clustering</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis.immunr_inc_overlap.html">vis(<i>&lt;immunr_inc_overlap&gt;</i>)</a></code> </p>
        </td>
        <td><p>Visualise incremental overlaps</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis.immunr_kmeans.html">vis(<i>&lt;immunr_kmeans&gt;</i>)</a></code> </p>
        </td>
        <td><p>Visualisation of K-means and DBSCAN clustering</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis.immunr_kmer_table.html">vis(<i>&lt;immunr_kmer_table&gt;</i>)</a></code> </p>
        </td>
        <td><p>Most frequent kmers visualisation.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis.immunr_mds.html">vis(<i>&lt;immunr_mds&gt;</i>)</a></code> </p>
        </td>
        <td><p>PCA / MDS / tSNE visualisation (mainly overlap / gene usage)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis.immunr_ov_matrix.html">vis(<i>&lt;immunr_ov_matrix&gt;</i>)</a></code> </p>
        </td>
        <td><p>Overlap / gene usage distance visualisation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis.immunr_public_repertoire.html">vis(<i>&lt;immunr_public_repertoire&gt;</i>)</a></code> </p>
        </td>
        <td><p>Public repertoire visualisation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis.immunr_public_statistics.html">vis(<i>&lt;immunr_public_statistics&gt;</i>)</a></code> </p>
        </td>
        <td><p>Visualise sharing of clonotypes among samples</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_bar.html">vis_bar()</a></code> </p>
        </td>
        <td><p>Bar plots</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_box.html">vis_box()</a></code> </p>
        </td>
        <td><p>Visualisation of distributions using boxplots</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_circos.html">vis_circos()</a></code> </p>
        </td>
        <td><p>Visualisation of matrices using circos plots</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_heatmap.html">vis_heatmap()</a></code> </p>
        </td>
        <td><p>Visualisation of matrices and data frames using ggplot2-based heatmaps</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_heatmap2.html">vis_heatmap2()</a></code> </p>
        </td>
        <td><p>Visualisation of matrices using heatmap3-based heatmaps</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_hist.html">vis_hist()</a></code> </p>
        </td>
        <td><p>Visualisation of distributions using histograms</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_immunr_kmer_profile_main.html">vis_immunr_kmer_profile_main()</a></code> </p>
        </td>
        <td><p>Visualise kmer profiles</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_public_clonotypes.html">vis_public_clonotypes()</a></code> </p>
        </td>
        <td><p>Visualisation of public clonotypes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_public_frequencies.html">vis_public_frequencies()</a></code> </p>
        </td>
        <td><p>Public repertoire visualisation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_radar.html">vis_radar()</a></code> </p>
        </td>
        <td><p>Visualisation of distance matrices with polar area plots</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_textlogo.html">vis_textlogo()</a></code> <code><a href="vis_textlogo.html">vis_seqlogo()</a></code> </p>
        </td>
        <td><p>Sequence logo plots for amino acid profiles.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_treemap.html">vis_treemap()</a></code> </p>
        </td>
        <td><p>Visualisation of data frames and matrices using treemaps</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#section-all-functions">All functions</a></li>
    </ul>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Vadim I. Nazarov, Vasily O. Tsvetkov, Eugene Rumynskiy, Anna Lorenc, Daniel J. Moore, Victor Greiff, <a href='https://immunomind.io'>ImmunoMind</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script>
<script>
  docsearch({
    
    
    apiKey: '94207562f34b455c0796790d104e3549',
    indexName: 'immunarch',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>



  </body>
</html>


