<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Load repertoire files to the workspace — repLoad • immunarch</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Load repertoire files to the workspace — repLoad" />
<meta property="og:description" content="The repLoad function loads repertoire files
into R workspace in the immunarch format where you can immediately use them for
the analysis. repLoad automatically detects the right format for
your files, so all you need is simply provide the path to your files.
See &quot;Details&quot; for more information on supported formats. See &quot;Examples&quot; for
diving right into it." />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-128418614-2"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-128418614-2');
</script>


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">immunarch</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.5.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/v1_introduction.html">Get started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">First steps</li>
    <li>
      <a href="../articles/v1_introduction.html">Introduction to immunarch</a>
    </li>
    <li>
      <a href="../articles/v2_data.html">Working with data in immunarch</a>
    </li>
    <li>
      <a href="../articles/v3_basic_analysis.html">Basic analysis and clonality</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Explore and compare repertoires</li>
    <li>
      <a href="../articles/web_only/v4_overlap.html">Repertoire overlap and public clonotypes</a>
    </li>
    <li>
      <a href="../articles/web_only/v5_gene_usage.html">Gene usage analysis</a>
    </li>
    <li>
      <a href="../articles/web_only/v6_diversity.html">Diversity estimation</a>
    </li>
    <li>
      <a href="../articles/web_only/v11_db.html">Annotate clonotypes using immune receptor databases</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Preparing to publication</li>
    <li>
      <a href="../articles/web_only/v7_fixvis.html">Make your plots publication-ready with fixVis</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Advanced methods</li>
    <li>
      <a href="../articles/web_only/v8_tracking.html">Tracking clonotypes across time points</a>
    </li>
    <li>
      <a href="../articles/web_only/v9_kmers.html">Kmer and sequence motif analysis and visualisation</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="https://github.com/immunomind/immunarch/releases">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://immunomind.io">ImmunoMind</a>
</li>
<li>
  <a href="http://twitter.com/immunomind">
    <span class="fa fa-lg fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/immunomind/immunarch">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Load repertoire files to the workspace</h1>
    <small class="dont-index">Source: <a href='https://github.com/immunomind/immunarch/blob/master/R/io.R'><code>R/io.R</code></a></small>
    <div class="hidden name"><code>repLoad.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The <code>repLoad</code> function loads repertoire files
into R workspace in the immunarch format where you can immediately use them for
the analysis. <code>repLoad</code> automatically detects the right format for
your files, so all you need is simply provide the path to your files.</p>
<p>See "Details" for more information on supported formats. See "Examples" for
diving right into it.</p>
    </div>

    <pre class="usage"><span class='fu'>repLoad</span>(<span class='no'>.path</span>, <span class='kw'>.format</span> <span class='kw'>=</span> <span class='fl'>NA</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>.path</th>
      <td><p>A character string specifying the path to the input data.
Input data can be one of the following:</p>
<p>- a single repertoire file.
In this case <code>repLoad</code> returns an R <a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a>;</p>
<p>- a vector of paths to repertoire files.
Same as in the case with no metadata file presented in the next section below;</p>
<p>- a path to the folder with repertoire files and, if available, metadata file "metadata.txt".
If the metadata file if presented, then the <code>repLoad</code> returns a list with two elements "data" and "meta".
"data" is an another list with repertoire R <a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a>s. "meta" is a data frame with the metadata.
If the metadata file "metadata.txt" is not presented, then the <code>repLoad</code> creates a dummy metadata file with
sample names and returns a list with two elements "data" and "meta".
If input data has multiple chains or cell types stored in the same file
(for example, like in 10xGenomics repertoire files), such repertoire files will be splitted to different
R data frames with only one type of chain and cell presented. The metadata file will have additional columns specifying
cell and chain types for different samples.</p></td>
    </tr>
    <tr>
      <th>.format</th>
      <td><p>A character string specifying what format to use. See "Details" for more information on supported formats.</p>
<p>Leave NA (which is default) if you want `immunarch` to detect formats automatically.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The metadata has to be a tab delimited file with first column named "Sample".
It can have any number of additional columns with arbitrary names.
The first column should contain base names of files without extensions in
your folder. Example:</p><table class='table'>
<tr><td>Sample</td><td>Sex</td><td>Age</td><td>Status</td></tr>
<tr><td>immunoseq_1</td><td>M</td><td>1</td><td>C</td></tr>
<tr><td>immunoseq_2</td><td>M</td><td>2</td><td>C</td></tr>
<tr><td>immunoseq_3</td><td>F</td><td>3</td><td>A</td></tr>
</table>


<p><code>repLoad</code> has the ".format" argument that sets the format for input repertoire files.
Do not pass it if you want immunarch to detect the formats and parse files automatically!
In case you want to force the package to parse the data in a specific format,
you can choose one of the several options for the argument:</p>
<p>- "immunoseq" - ImmunoSEQ of any version. http://www.adaptivebiotech.com/immunoseq</p>
<p>- "mitcr" - MiTCR. https://github.com/milaboratory/mitcr</p>
<p>- "mixcr" - MiXCR of any version. https://github.com/milaboratory/mixcr</p>
<p>- "migec" - MiGEC. http://migec.readthedocs.io/en/latest/</p>
<p>- "migmap" - For parsing IgBLAST results postprocessed with MigMap. https://github.com/mikessh/migmap</p>
<p>- "tcr" - tcR, our previous package. https://imminfo.github.io/tcr/</p>
<p>- "vdjtools" - VDJtools of any version. http://vdjtools-doc.readthedocs.io/en/latest/</p>
<p>- "imgt" - IMGT HighV-QUEST. http://www.imgt.org/HighV-QUEST/</p>
<p>- "airr" - adaptive immune receptor repertoire (AIRR) data format. http://docs.airr-community.org/en/latest/datarep/overview.html</p>
<p>- "10x" - 10XGenomics clonotype annotations tables. https://support.10xgenomics.com/single-cell-vdj/software/pipelines/latest/output/annotation</p>
<p>- "archer" - ArcherDX clonotype tables. https://archerdx.com/</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><a href='immunr_data_format.html'>immunr_data_format</a> for immunarch data format; <a href='repSave.html'>repSave</a> for file saving;
<a href='repOverlap.html'>repOverlap</a>, <a href='geneUsage.html'>geneUsage</a> and <a href='repDiversity.html'>repDiversity</a> for starting with immune repertoires basic statistics.</p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> (<span class='fl'>FALSE</span>) {
<span class='co'># To load the data from a single file without specifying the data format:</span>
<span class='no'>immdata</span> <span class='kw'>&lt;-</span> <span class='fu'>repLoad</span>(<span class='st'>"path/to/your/folder/immunoseq_1.txt"</span>)
<span class='co'># To load the data from a single ImmunoSEQ file, go with:</span>
<span class='no'>immdata</span> <span class='kw'>&lt;-</span> <span class='fu'>repLoad</span>(<span class='st'>"path/to/your/folder/immunoseq_1.txt"</span>, <span class='kw'>.format</span> <span class='kw'>=</span> <span class='st'>"immunoseq"</span>)

<span class='co'># Suppose you have a following structure in your folder:</span>
<span class='co'># &gt;_ ls</span>
<span class='co'># immunoseq1.txt</span>
<span class='co'># immunoseq2.txt</span>
<span class='co'># immunoseq3.txt</span>
<span class='co'># metadata.txt</span>

<span class='co'># To load the whole folder with every file in it type:</span>
<span class='no'>immdata</span> <span class='kw'>&lt;-</span> <span class='fu'>repLoad</span>(<span class='st'>"path/to/your/folder/"</span>)

<span class='co'># We recommend creating a metadata file named exactly "metadata.txt" in the folder.</span>

<span class='co'># In that case, when you load your data you will see:</span>
<span class='co'># &gt; immdata &lt;- repLoad("path/to/your/folder/")</span>
<span class='co'># &gt; names(immdata)</span>
<span class='co'># [1] "data" "meta"</span>

<span class='co'># If you do not have "metadata.txt", you will see the same output,</span>
<span class='co'># but your metadata will be almost empty:</span>
<span class='co'># &gt; immdata &lt;- repLoad("path/to/your/folder/")</span>
<span class='co'># &gt; names(immdata)</span>
<span class='co'># [1] "data" "meta"</span>
}</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#details">Details</a></li>
      <li><a href="#see-also">See also</a></li>
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Vadim I. Nazarov, Vasily O. Tsvetkov, Eugene Rumynskiy, <a href='https://immunomind.io'>ImmunoMind</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


