<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Data loading • immunarch</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Data loading">
<meta property="og:description" content="immunarch">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-128418614-2"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-128418614-2');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">immunarch</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.6.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Start Here
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/v1_introduction.html">Installation &amp; troubleshooting</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">First steps</li>
    <li>
      <a href="../articles/v2_data.html">Data loading</a>
    </li>
    <li>
      <a href="../articles/web_only/load_mixcr.html">How-to: Loading MiXCR Data</a>
    </li>
    <li>
      <a href="../articles/web_only/load_10x.html">How-to: Loading 10x Genomics Data</a>
    </li>
    <li>
      <a href="../articles/web_only/v21_singlecell.html">How-to: Single-cell and paired chain data</a>
    </li>
    <li>
      <a href="../articles/web_only/v3_basic_analysis.html">Basic statistics and clonality</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Repertoire-level exploration and comparison</li>
    <li>
      <a href="../articles/web_only/v4_overlap.html">Repertoire overlap and public clonotypes</a>
    </li>
    <li>
      <a href="../articles/web_only/v5_gene_usage.html">Gene usage</a>
    </li>
    <li>
      <a href="../articles/web_only/v6_diversity.html">Diversity estimation</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Clonotype-level exploration and comparison</li>
    <li>
      <a href="../articles/web_only/v8_tracking.html">Track clonotypes across samples and time</a>
    </li>
    <li>
      <a href="../articles/web_only/v11_db.html">Annotate clonotypes using immune receptor databases</a>
    </li>
    <li>
      <a href="../articles/web_only/v9_kmers.html">Kmer and sequence motif analysis and visualisation</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Preparing to publication</li>
    <li>
      <a href="../articles/web_only/v7_fixvis.html">Make your plots publication-ready with fixVis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="https://github.com/immunomind/covid19">COVID-19</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://immunomind.io">ImmunoMind</a>
</li>
<li>
  <a href="http://twitter.com/immunomind">
    <span class="fa fa-lg fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/immunomind/immunarch">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="v2_data_files/header-attrs-2.9/header-attrs.js"></script><script src="v2_data_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Data loading</h1>
                        <h4 class="author">
<b>ImmunoMind</b> – improving design of T-cell therapies using multi-omics and AI. Research and biopharma partnerships, more details: <a href="https://immunomind.io">immunomind.io</a>
</h4>
            
            <h4 class="date"><a href="mailto:support@immunomind.io" class="email">support@immunomind.io</a></h4>
      
      <small class="dont-index">Source: <a href="https://github.com/immunomind/immunarch/blob/master/vignettes/v2_data.Rmd"><code>vignettes/v2_data.Rmd</code></a></small>
      <div class="hidden name"><code>v2_data.Rmd</code></div>

    </div>

    
    
<!--
%\VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{Working with data in immunarch}
%\VignettePackage{immunarch}
-->
<div id="input-output" class="section level1">
<h1 class="hasAnchor">
<a href="#input-output" class="anchor"></a>Input / output</h1>
<p>The package provides several IO functions:</p>
<ul>
<li><p><code>repLoad</code> - to load the repertoires, having compatible format.</p></li>
<li><p><code>repSave</code> - to save changes and to write the repertoire data to a file in a specific format (<code>immunarch</code>, VDJtools).</p></li>
</ul>
<p><code>repLoad</code> detects the input file format automatically. <code>immunarch</code> currently support the following immune repertoire data formats:</p>
<ul>
<li><p><code>"immunarch"</code> - current software tool, in case you forgot it :)</p></li>
<li><p><code>"immunoseq"</code> - <a href="https://www.immunoseq.com" class="uri">https://www.immunoseq.com</a></p></li>
<li><p><code>"mitcr"</code> - <a href="https://github.com/milaboratory/mitcr" class="uri">https://github.com/milaboratory/mitcr</a></p></li>
<li><p><code>"mixcr"</code> - <a href="https://github.com/milaboratory/mixcr" class="uri">https://github.com/milaboratory/mixcr</a></p></li>
<li><p><code>"migec"</code> - <a href="http://migec.readthedocs.io/en/latest/" class="uri">http://migec.readthedocs.io/en/latest/</a></p></li>
<li><p><code>"migmap"</code> - <a href="https://github.com/mikessh/migmap" class="uri">https://github.com/mikessh/migmap</a></p></li>
<li><p><code>"tcr"</code> - <a href="https://imminfo.github.io/tcr/" class="uri">https://imminfo.github.io/tcr/</a></p></li>
<li><p><code>"vdjtools"</code> - <a href="https://vdjtools-doc.readthedocs.io/en/master/" class="uri">https://vdjtools-doc.readthedocs.io/en/master/</a></p></li>
<li><p><code>"imgt"</code> - <a href="http://www.imgt.org/HighV-QUEST/" class="uri">http://www.imgt.org/HighV-QUEST/</a></p></li>
<li><p><code>"airr"</code> - <a href="http://docs.airr-community.org/en/latest/datarep/overview.html" class="uri">http://docs.airr-community.org/en/latest/datarep/overview.html</a></p></li>
<li><p><code>"10x"</code> - <a href="https://support.10xgenomics.com/single-cell-vdj/software/pipelines/latest/output/annotation" class="uri">https://support.10xgenomics.com/single-cell-vdj/software/pipelines/latest/output/annotation</a></p></li>
<li><p><code>"archer"</code> - ArcherDX clonotype tables.</p></li>
<li><p><code>"imseq"</code> - <a href="http://www.imtools.org/" class="uri">http://www.imtools.org/</a></p></li>
<li><p><code>"rtcr"</code> - <a href="https://github.com/uubram/RTCR" class="uri">https://github.com/uubram/RTCR</a></p></li>
<li><p><code>"vidjil"</code> - <a href="http://www.vidjil.org/" class="uri">http://www.vidjil.org/</a></p></li>
</ul>
<p>Please contact us if there are more file formats you want to be supported.</p>
<p>For parsing IgBLAST results process the data with MigMap first.</p>
<p>You can load the data either from a single file, a list of repertoire file paths or from a folder with repertoire files. A single file can be loaded as follows:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># To load the data from a single file (note that you don't need to specify the data format):</span>
<span class="va">file_path</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">"immunarch"</span><span class="op">)</span>, <span class="st">"/extdata/io/Sample1.tsv.gz"</span><span class="op">)</span>
<span class="va">immdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/repLoad.html">repLoad</a></span><span class="op">(</span><span class="va">file_path</span><span class="op">)</span></code></pre></div>
<p>In other cases you may want to provide a metadata file and locate it in the folder. It is necessary to name it exactly “metadata.txt”.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># For instance you have a following structure in your folder:</span>
<span class="co"># &gt;_ ls</span>
<span class="co"># immunoseq1.txt</span>
<span class="co"># immunoseq2.txt</span>
<span class="co"># immunoseq3.txt</span>
<span class="co"># metadata.txt</span></code></pre></div>
<p>With the metadata <code>repLoad</code> will create a list in the environment with 2 elements, namely <code>data</code> and <code>meta</code>. All the data will be accessible simply from <code>immdata$data</code>.</p>
<p>Otherwise <code>repLoad</code> will create a dummy metadata file with only sample names.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># To load the whole folder with every file in it type:</span>
<span class="va">file_path</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">"immunarch"</span><span class="op">)</span>, <span class="st">"/extdata/io/"</span><span class="op">)</span>
<span class="va">immdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/repLoad.html">repLoad</a></span><span class="op">(</span><span class="va">file_path</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">immdata</span><span class="op">)</span><span class="op">)</span>

<span class="co"># In order to do that your folder must contain metadata file named</span>
<span class="co"># exactly "metadata.txt".</span>

<span class="co"># In R, when you load your data:</span>
<span class="co"># &gt; immdata &lt;- repLoad("path/to/your/folder/")</span>
<span class="co"># &gt; names(immdata)</span>
<span class="co"># [1] "data" "meta"</span>

<span class="co"># Suppose you do not have "metadata.txt":</span>
<span class="co"># &gt; immdata &lt;- repLoad("path/to/your/folder/")</span>
<span class="co"># &gt; names(immdata)</span>
<span class="co"># [1] "data" "meta"</span></code></pre></div>
<p>Dummy metadata data frame look like this:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">as_tibble</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>Sample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"immunoseq1"</span>, <span class="st">"immunoseq2"</span>, <span class="st">"immunoseq3"</span><span class="op">)</span>, stringsAsFactors <span class="op">=</span> <span class="cn">F</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 3 x 1
##   Sample    
##   &lt;chr&gt;     
## 1 immunoseq1
## 2 immunoseq2
## 3 immunoseq3</code></pre>
<p>The metadata file “metadata.txt” has to be tab delimited file with first column named “Sample” and any number of additional columns with arbitrary names. The first column should contain base names of files without extensions in your folder.</p>
<table class="table">
<thead><tr class="header">
<th align="center"><strong>Sample</strong></th>
<th align="center"><strong>Sex</strong></th>
<th align="center"><strong>Age</strong></th>
<th align="center"><strong>Status</strong></th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center">immunoseq_1</td>
<td align="center">M</td>
<td align="center">1</td>
<td align="center">C</td>
</tr>
<tr class="even">
<td align="center">immunoseq_2</td>
<td align="center">M</td>
<td align="center">2</td>
<td align="center">C</td>
</tr>
<tr class="odd">
<td align="center">immunoseq_3</td>
<td align="center">F</td>
<td align="center">3</td>
<td align="center">A</td>
</tr>
</tbody>
</table>
<p>In order to import data from the external databases you have to create a connection to this database and then load the data. Make sure that the table format in your database matches the <code>immunarch</code>’s format.</p>
<p>To illustrate the use of external database, here is an example demonstrating data loading to the local MonetDB database:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Your list of repertoires in immunarch's format</span>
<span class="va">DATA</span>
<span class="co"># Metadata data frame</span>
<span class="va">META</span>

<span class="co"># Create a temporary directory</span>
<span class="va">dbdir</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempdir</a></span><span class="op">(</span><span class="op">)</span>

<span class="co"># Create a DBI connection to MonetDB in the temporary directory.</span>
<span class="va">con</span> <span class="op">=</span> <span class="fu">DBI</span><span class="fu">::</span><span class="fu"><a href="https://dbi.r-dbi.org/reference/dbConnect.html">dbConnect</a></span><span class="op">(</span><span class="fu">MonetDBLite</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MonetDBLite/man/MonetDBLite.html">MonetDBLite</a></span><span class="op">(</span><span class="op">)</span>, embedded <span class="op">=</span> <span class="va">dbdir</span><span class="op">)</span>

<span class="co"># Write each repertoire to MonetDB. Each table has corresponding name from the DATA</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">DATA</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu">DBI</span><span class="fu">::</span><span class="fu"><a href="https://dbi.r-dbi.org/reference/dbWriteTable.html">dbWriteTable</a></span><span class="op">(</span><span class="va">con</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">DATA</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="va">DATA</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, overwrite<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
<span class="op">}</span>

<span class="co"># Create a source in the temporary directory with MonetDB</span>
<span class="va">ms</span> <span class="op">=</span> <span class="fu">MonetDBLite</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MonetDBLite/man/src_monetdb.html">src_monetdblite</a></span><span class="op">(</span>dbdir <span class="op">=</span> <span class="va">dbdir</span><span class="op">)</span>
<span class="va">res_db</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>

<span class="co"># Load the data from MonetDB to dplyr tables</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">DATA</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
  <span class="va">res_db</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">DATA</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/tbl.html">tbl</a></span><span class="op">(</span><span class="va">ms</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">DATA</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span>
<span class="op">}</span>

<span class="co"># Your data is ready to use</span>
<span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">res_db</span>, meta <span class="op">=</span> <span class="va">META</span><span class="op">)</span></code></pre></div>
<p><code>immunarch</code> is compatible with following sources:</p>
<ul>
<li><p>R data frames (for common applications)</p></li>
<li><p>R data tables (for speed, although requires lots of RAM)</p></li>
<li><p>MonetDB-like databases that support both DBI and dplyr (best choice, although you have to be a bit familiar with dplyr)</p></li>
<li><p>Apache Spark (if you are expierienced and comfortable with it)</p></li>
</ul>
</div>
<div id="basic-data-manipulations-with-dplyr-and-immunarch" class="section level1">
<h1 class="hasAnchor">
<a href="#basic-data-manipulations-with-dplyr-and-immunarch" class="anchor"></a>Basic data manipulations with dplyr and immunarch</h1>
<p>You can find the introduction to <code>dplyr</code> here: <a href="https://CRAN.R-project.org/package=dplyr/vignettes/dplyr.html" class="uri">https://CRAN.R-project.org/package=dplyr/vignettes/dplyr.html</a></p>
<div id="get-the-most-abundant-clonotypes" class="section level2">
<h2 class="hasAnchor">
<a href="#get-the-most-abundant-clonotypes" class="anchor"></a>Get the most abundant clonotypes</h2>
<p>The function returns the most abundant clonotypes for the given repertoire:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/top.html">top</a></span><span class="op">(</span><span class="va">immdata</span><span class="op">$</span><span class="va">data</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 10 x 15
##    Clones Proportion CDR3.nt    CDR3.aa V.name D.name J.name V.end D.start D.end
##     &lt;dbl&gt;      &lt;dbl&gt; &lt;chr&gt;      &lt;chr&gt;   &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;  &lt;int&gt;   &lt;int&gt; &lt;int&gt;
##  1    173    0.0204  TGCGCCAGC… CASSQE… TRBV4… TRBD1  TRBJ2…    16      18    26
##  2    163    0.0192  TGCGCCAGC… CASSYR… TRBV4… TRBD1  TRBJ2…    11      13    18
##  3     66    0.00776 TGTGCCACC… CATSTN… TRBV15 TRBD1  TRBJ2…    11      16    22
##  4     54    0.00635 TGTGCCACC… CATSIG… TRBV15 TRBD2  TRBJ2…    11      19    25
##  5     48    0.00565 TGTGCCAGC… CASSPW… TRBV27 TRBD1  TRBJ1…    11      16    23
##  6     48    0.00565 TGCGCCAGC… CASQGD… TRBV4… TRBD1  TRBJ1…     8      13    19
##  7     40    0.00471 TGCGCCAGC… CASSQD… TRBV4… TRBD1  TRBJ2…    16      21    26
##  8     31    0.00365 TGTGCCAGC… CASSEE… TRBV2  TRBD1  TRBJ1…    15      17    20
##  9     30    0.00353 TGCGCCAGC… CASSQP… TRBV4… TRBD1  TRBJ2…    14      23    28
## 10     28    0.00329 TGTGCCAGC… CASSWV… TRBV6… TRBD1  TRBJ2…    12      20    25
## # … with 5 more variables: J.start &lt;int&gt;, VJ.ins &lt;dbl&gt;, VD.ins &lt;dbl&gt;,
## #   DJ.ins &lt;dbl&gt;, Sequence &lt;lgl&gt;</code></pre>
</div>
<div id="filter-functional-non-functional-in-frame-out-of-frame-clonotypes" class="section level2">
<h2 class="hasAnchor">
<a href="#filter-functional-non-functional-in-frame-out-of-frame-clonotypes" class="anchor"></a>Filter functional / non-functional / in-frame / out-of-frame clonotypes</h2>
<p>Conveniently, functions are vectorised over the list of data frames; and <code><a href="../reference/coding.html">coding(immdata$data)</a></code> in the example below returns a list of data frames with coding sequences:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/coding.html">coding</a></span><span class="op">(</span><span class="va">immdata</span><span class="op">$</span><span class="va">data</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<p>The next one operates in a similar fashion:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/coding.html">noncoding</a></span><span class="op">(</span><span class="va">immdata</span><span class="op">$</span><span class="va">data</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<p>Now, the computation of the number of filtered sequences is straightforward:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="fu"><a href="../reference/coding.html">inframes</a></span><span class="op">(</span><span class="va">immdata</span><span class="op">$</span><span class="va">data</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>And for the out-of-frame clonotypes:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="fu"><a href="../reference/coding.html">outofframes</a></span><span class="op">(</span><span class="va">immdata</span><span class="op">$</span><span class="va">data</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="get-subset-of-clonotypes-with-a-specific-v-gene" class="section level2">
<h2 class="hasAnchor">
<a href="#get-subset-of-clonotypes-with-a-specific-v-gene" class="anchor"></a>Get subset of clonotypes with a specific V gene</h2>
<p>It is simple to subset data frame according to labels in the specified index. In the example the resulting data frame contains only records with ‘TRBV10-1’ V gene:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">immdata</span><span class="op">$</span><span class="va">data</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="va">V.name</span> <span class="op">==</span> <span class="st">'TRBV10-1'</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 24 x 15
##    Clones Proportion CDR3.nt    CDR3.aa V.name D.name J.name V.end D.start D.end
##     &lt;dbl&gt;      &lt;dbl&gt; &lt;chr&gt;      &lt;chr&gt;   &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;  &lt;int&gt;   &lt;int&gt; &lt;int&gt;
##  1      2   0.000235 TGCGCCAGC… CASSES… TRBV1… TRBD2  TRBJ2…    16      20    25
##  2      2   0.000235 TGCGCCAGC… CASSDG… TRBV1… TRBD1  TRBJ2…    13      15    22
##  3      1   0.000118 TGCGCCAGC… CASSGD… TRBV1… TRBD2  TRBJ2…     8      10    15
##  4      1   0.000118 TGCGCCACC… CATLRS… TRBV1… TRBD1  TRBJ2…     6       7     9
##  5      1   0.000118 TGCGCCAGC… CASSES… TRBV1… TRBD2  TRBJ2…    16      20    22
##  6      1   0.000118 TGCGCCAGC… CASSES… TRBV1… TRBD2  TRBJ2…    16      17    21
##  7      1   0.000118 TGCGCCAGC… CASRAS… TRBV1… TRBD2  TRBJ2…    10      13    21
##  8      1   0.000118 TGCGCCAGC… CASRRD… TRBV1… TRBD1  TRBJ2…     8      13    19
##  9      1   0.000118 TGCGCCAGC… CASSEV… TRBV1… TRBD1  TRBJ2…    14      19    24
## 10      1   0.000118 TGCGCCAGC… CASSEG… TRBV1… TRBD2  TRBJ2…    13      19    27
## # … with 14 more rows, and 5 more variables: J.start &lt;int&gt;, VJ.ins &lt;dbl&gt;,
## #   VD.ins &lt;dbl&gt;, DJ.ins &lt;dbl&gt;, Sequence &lt;lgl&gt;</code></pre>
</div>
</div>
<div id="downsampling" class="section level1">
<h1 class="hasAnchor">
<a href="#downsampling" class="anchor"></a>Downsampling</h1>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ds</span> <span class="op">=</span> <span class="fu"><a href="../reference/repSample.html">repSample</a></span><span class="op">(</span><span class="va">immdata</span><span class="op">$</span><span class="va">data</span>, <span class="st">"downsample"</span>, <span class="fl">100</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="va">ds</span>, <span class="va">nrow</span><span class="op">)</span></code></pre></div>
<pre><code>## A2-i129 A2-i131 A2-i133 A2-i132 A4-i191 A4-i192     MS1     MS2     MS3     MS4 
##      96      99      98     100      89      93      87     100      89     100 
##     MS5     MS6 
##      86      99</code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ds</span> <span class="op">=</span> <span class="fu"><a href="../reference/repSample.html">repSample</a></span><span class="op">(</span><span class="va">immdata</span><span class="op">$</span><span class="va">data</span>, <span class="st">"sample"</span>, .n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="va">ds</span>, <span class="va">nrow</span><span class="op">)</span></code></pre></div>
<pre><code>## A2-i129 A2-i131 A2-i133 A2-i132 A4-i191 A4-i192     MS1     MS2     MS3     MS4 
##      10      10      10      10      10      10      10      10      10      10 
##     MS5     MS6 
##      10      10</code></pre>
</div>
<div id="immunarch-data-format" class="section level1">
<h1 class="hasAnchor">
<a href="#immunarch-data-format" class="anchor"></a>Immunarch data format</h1>
<p><code>immunarch</code> comes with its own data format, including tab-delimited columns that can be specified as follows:</p>
<ul>
<li><p>“Clones” - count or number of barcodes (events, UMIs) or reads;</p></li>
<li><p>“Proportion” - proportion of barcodes (events, UMIs) or reads;</p></li>
<li><p>“CDR3.nt” - CDR3 nucleotide sequence;</p></li>
<li><p>“CDR3.aa” - CDR3 amino acid sequence;</p></li>
<li><p>“V.name” - names of aligned Variable gene segments;</p></li>
<li><p>“D.name” - names of aligned Diversity gene segments or NA;</p></li>
<li><p>“J.name” - names of aligned Joining gene segments;</p></li>
<li><p>“V.end” - last positions of aligned V gene segments (1-based);</p></li>
<li><p>“D.start” - positions of D’5 end of aligned D gene segments (1-based);</p></li>
<li><p>“D.end” - positions of D’3 end of aligned D gene segments (1-based);</p></li>
<li><p>“J.start” - first positions of aligned J gene segments (1-based);</p></li>
<li><p>“VJ.ins” - number of inserted nucleotides (N-nucleotides) at V-J junction (-1 for receptors with VDJ recombination);</p></li>
<li><p>“VD.ins” - number of inserted nucleotides (N-nucleotides) at V-D junction (-1 for receptors with VJ recombination);</p></li>
<li><p>“DJ.ins” - number of inserted nucleotides (N-nucleotides) at D-J junction (-1 for receptors with VJ recombination);</p></li>
<li><p>“Sequence” - full nucleotide sequence.</p></li>
</ul>
<!---
## Additional functions to help with dplyr

- `todf` - convert the input to R data frame.

- `todt` - convert the input to R data table.

- `db_add` - add a new column to the existing database.

- `db_rem` - remove a specific column from the existing database.

- `db_write` - write / rewrite a database table.

With these functions you can write the subset of your repertoires to new tables with `db_write`, after you have filtered out the noncoding sequences, for instance.

-->
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://www.linkedin.com/in/vdnaz">Vadim I. Nazarov</a>, <a href="https://www.linkedin.com/in/vasily-tsvetkov-227218ab">Vasily O. Tsvetkov</a>, Eugene Rumynskiy, Aleksandr A. Popov, Ivan Balashov, <a href="https://immunomind.io">ImmunoMind</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '94207562f34b455c0796790d104e3549',
    indexName: 'immunarch',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
